{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(GenomicRanges)\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "library(ggplot2)\n",
    "library(viridis)\n",
    "library(readxl)\n",
    "library(pheatmap)\n",
    "library(RColorBrewer)\n",
    "library(tidyr)\n",
    "library(gridExtra)\n",
    "library(corrplot)\n",
    "library(ggfortify)\n",
    "library(ggdendro)\n",
    "library(umap)\n",
    "library(DESeq2)\n",
    "library(rjson)\n",
    "library(ggrepel)\n",
    "library(ggpubr)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "rPath = \"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/\"\n",
    "figPath = \"CART_CUTRUN_Project/results/paper_figure/RNAseq_QC/\"\n",
    "## Read in data\n",
    "cellList <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hdList <- paste0(\"HD\", c(1:3, 5:7))\n",
    "exprList = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "\n",
    "# alignD = c()\n",
    "# for(cell in cellList){\n",
    "#   for(expr in exprList){\n",
    "#     for(hd in hdList){\n",
    "\n",
    "#       rAlignTmp = fread(paste0(rPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".summary\"))\n",
    "#       rAlignExon = rAlignTmp[nrow(rAlignTmp)]%>% as.numeric\n",
    "#       rAlign = fread(paste0(rPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".stat/RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".cnt\"), fill = TRUE, header = FALSE)[1:3, ]\n",
    "\n",
    "#       alignD = data.frame(cell = cell, expr = expr, hd = hd,\n",
    "#                           seqDepth = as.numeric(rAlign$V4[1]),\n",
    "#                           alignRate = rAlign$V2[1]/as.numeric(rAlign$V4[1]) * 100,\n",
    "#                           nonIntronicAlign = rAlign$V2[1]/rAlign$V2[1] * 100,\n",
    "#                           codingExonAlign = rAlignExon/rAlign$V2[1] * 100,\n",
    "#                           uniExon = as.numeric(rAlign$V1[2])/rAlign$V2[1] * 100,\n",
    "#                           multiExon = rAlign$V2[2]/rAlign$V2[1] * 100,\n",
    "#                           type = \"HD\") %>% rbind(alignD, .)\n",
    "\n",
    "#     }\n",
    "#   }\n",
    "\n",
    "# }\n",
    "\n",
    "## patient\n",
    "## Patient data infusion product of CM-derived cells \n",
    "patientRaw = read_xlsx(\"CART_CUTRUN_Project/report/CART_CUTRUN/analysis/20201028_NHL_CUTandRUN_PatientData.xlsx\")\n",
    "cm_ind = which(patientRaw$`CD8+ T cell isolation` == \"CD8TCM\")\n",
    "patient_id_list = patientRaw$Xnumber[cm_ind] %>% tolower\n",
    "patient_id_list\n",
    "\n",
    "rPath = \"CART_CUTRUN_Project/results/RNAseq/process/NHL_RSEM_gencode_v21/\"\n",
    "\n",
    "\n",
    "# for(sample in paste0(\"Sample_\", patient_id_list, \"_IP_CD8-CAR\")){\n",
    "#   if(file.exists(paste0(rPath, sample, \"/RSEM.summary\"))){\n",
    "#     print(sample)\n",
    "#     patient_id = sample %>% gsub(\"_IP_CD8-CAR\", \"\", .) %>% gsub(\"Sample_\", \"\", .)\n",
    "#     rAlignTmp = fread(paste0(rPath, sample, \"/RSEM.summary\"))\n",
    "#     rAlignExon = rAlignTmp[nrow(rAlignTmp)]%>% as.numeric\n",
    "#     rAlign = fread(paste0(rPath, sample, \"/RSEM.stat/RSEM.cnt\"), fill = TRUE, header = FALSE)[1:3, ]\n",
    "   \n",
    "#     alignD = data.frame(cell = \"Patient\", expr = \"Product\", hd = patient_id,\n",
    "#                         seqDepth = as.numeric(rAlign$V4[1]),\n",
    "#                         alignRate = rAlign$V2[1]/as.numeric(rAlign$V4[1]) * 100,\n",
    "#                         nonIntronicAlign = rAlign$V2[1]/rAlign$V2[1] * 100,\n",
    "#                         codingExonAlign = rAlignExon/rAlign$V2[1] * 100,\n",
    "#                         uniExon = as.numeric(rAlign$V1[2])/rAlign$V2[1] * 100,\n",
    "#                         multiExon = rAlign$V2[2]/rAlign$V2[1] * 100,\n",
    "#                         type = \"Patient\") %>% rbind(alignD, .)\n",
    "\n",
    "#   }\n",
    "\n",
    "# }\n",
    " \n",
    "\n",
    "# # hdList <- paste0(\"HD\", 4:7)\n",
    "# # exprList = c(\"Input\")\n",
    "# # for(cell in cellList){\n",
    "# #   for(expr in exprList){\n",
    "# #     for(hd in hdList){\n",
    "# # \n",
    "# #       kAlign = fromJSON(file = paste0(kPath,  \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \"/run_info.json\"))\n",
    "# #       rAlignExon = fread(paste0(rPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".summary\")) %>% as.numeric\n",
    "# #       rAlign = fread(paste0(rPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".stat/RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".cnt\"), fill = TRUE, header = FALSE)[1:3, ]\n",
    "# # \n",
    "# #       alignD = data.frame(cell = cell, expr = expr, hd = hd,\n",
    "# #                           seqDepth = as.numeric(kAlign$n_processed),\n",
    "# #                           alignRate = rAlign$V4[1]/as.numeric(kAlign$n_processed) * 100,\n",
    "# #                           nonIntronicAlign = rAlign$V2[1]/rAlign$V4[1] * 100,\n",
    "# #                           codingExonAlign = rAlignExon/rAlign$V4[1] * 100,\n",
    "# #                           uniExon = as.numeric(rAlign$V1[2])/rAlign$V2[1] * 100,\n",
    "# #                           multiExon = rAlign$V2[2]/rAlign$V2[1] * 100,\n",
    "# #                           type = \"RSEM\") %>% rbind(alignD, .)\n",
    "# # \n",
    "# #     }\n",
    "# #   }\n",
    "# # \n",
    "# # }\n",
    "# alignD$cell = factor(alignD$cell, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\"))\n",
    "\n",
    "\n",
    "# save(alignD, file = paste0(outPath, \"RData/RSEM_alignSummary_HD1-3_5-7_N_CM_EM_EMRA_Patients.RData\"))\n",
    "load(file = paste0(outPath, \"RData/RSEM_alignSummary_HD1-3_5-7_N_CM_EM_EMRA_Patients.RData\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "palette1 <- RColorBrewer::brewer.pal(8, \"Set1\")\n",
    "palette2 <- RColorBrewer::brewer.pal(8, \"Set2\")\n",
    "# Combine them to create a 16-color palette\n",
    "palette_col <- c(palette1, palette2)\n",
    "palette_col\n",
    "figPath = \"CART_CUTRUN_Project/results/paper_figure/RNAseq_QC/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Sequencing depth\n",
    "alignD %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>%\n",
    "ggplot(aes(x = cell, y = seqDepth, fill = cell)) +\n",
    "  geom_boxplot(outlier.shape = NA) +\n",
    "  geom_jitter(aes(color = hd), position = position_jitter(0.15)) + \n",
    "  facet_grid(~expr, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, option = \"magma\", alpha = 0.5) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Sequencing Depth\") +\n",
    "  xlab(\"\") +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "\n",
    "ggsave(paste0(figPath, \"RNAseq_SequencingDepth.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Sequencing depth - HD\n",
    "alignD %>% filter(hd %in% hdList) %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>%\n",
    "ggplot(aes(x = cell, y = seqDepth, fill = cell)) +\n",
    "  geom_boxplot(outlier.shape = NA) +\n",
    "  geom_jitter(aes(color = hd), position = position_jitter(0.15)) + \n",
    "  facet_grid(~expr, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, end = 0.75, option = \"magma\", alpha = 0.5) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Sequencing Depth\") +\n",
    "  xlab(\"\") +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "\n",
    "ggsave(paste0(figPath, \"RNAseq_SequencingDepth_HD.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Sequencing depth - PT\n",
    "alignD %>% filter(hd %in% patient_id_list) %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>%\n",
    "ggplot(aes(x = cell, y = seqDepth, fill = cell)) +\n",
    "  geom_boxplot(outlier.shape = NA) +\n",
    "  geom_jitter(aes(color = hd), position = position_jitter(0.15)) + \n",
    "  facet_grid(~expr, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 1, option = \"magma\", alpha = 0.5) +\n",
    "  scale_color_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Sequencing Depth\") +\n",
    "  xlab(\"\") +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "ggsave(paste0(figPath, \"RNAseq_SequencingDepth_PT.pdf\"), width = 8, height = 8)\n",
    "\n",
    "alignD %>% filter(hd %in% patient_id_list) %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>%\n",
    "ggplot(aes(x = hd, y = seqDepth, fill = hd)) +\n",
    "  geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "  # scale_fill_viridis(discrete = TRUE, begin = 0.1, option = \"magma\", alpha = 0.8) +\n",
    "  scale_fill_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Sequencing Depth\") +\n",
    "  xlab(\"Patient\") +\n",
    "  ggpubr::rremove(\"legend\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "\n",
    "ggsave(paste0(figPath, \"RNAseq_SequencingDepth_PT_bar.pdf\"), width = 8, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Alignment rate\n",
    "alignD  %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>% ggplot(aes(x = cell, y = alignRate, fill = cell)) +\n",
    "  geom_boxplot(outlier.shape = NA) +\n",
    "  geom_jitter(aes(color = hd), position = position_jitter(0.15)) + \n",
    "  facet_grid(~expr, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, option = \"magma\", alpha = 0.5) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Alignment Rate\") +\n",
    "  xlab(\"\") +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "\n",
    "ggsave(paste0(figPath, \"RNAseq_AlignmentRate.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Alignment rate - HD\n",
    "alignD %>% filter(hd %in% hdList) %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>% ggplot(aes(x = cell, y = alignRate, fill = cell)) +\n",
    "  geom_boxplot(outlier.shape = NA) +\n",
    "  geom_jitter(aes(color = hd), position = position_jitter(0.15)) + \n",
    "  facet_grid(~expr, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, end = 0.75, option = \"magma\", alpha = 0.5) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Alignment Rate\") +\n",
    "  xlab(\"\") +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "\n",
    "ggsave(paste0(figPath, \"RNAseq_AlignmentRate_HD.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Alignment rate - PT\n",
    "alignD  %>% filter(hd %in% patient_id_list) %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>% ggplot(aes(x = cell, y = alignRate, fill = cell)) +\n",
    "  geom_boxplot(outlier.shape = NA) +\n",
    "  geom_jitter(aes(color = hd), position = position_jitter(0.15)) + \n",
    "  facet_grid(~expr, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 1, option = \"magma\", alpha = 0.5) +\n",
    "  scale_color_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Alignment Rate\") +\n",
    "  xlab(\"\") +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "ggsave(paste0(figPath, \"RNAseq_AlignmentRate_PT.pdf\"), width = 8, height = 8)\n",
    "# Alignment rate - PT\n",
    "alignD  %>% filter(hd %in% patient_id_list) %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>% ggplot(aes(x = hd, y = alignRate, fill = hd)) +\n",
    "  geom_bar(stat = \"identity\") +\n",
    "  scale_fill_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Alignment Rate\") +\n",
    "  xlab(\"Patient\") +\n",
    "  ggpubr::rremove(\"legend\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  ylim(0, 100)\n",
    "\n",
    "ggsave(paste0(figPath, \"RNAseq_AlignmentRate_PT_bar.pdf\"), width = 8, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Coding exon rate\n",
    "alignD  %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>% ggplot(aes(x = cell, y = codingExonAlign, fill = cell)) +\n",
    "  geom_boxplot(outlier.shape = NA) +\n",
    "  geom_jitter(aes(color = hd), position = position_jitter(0.15)) + \n",
    "  facet_grid(~expr, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, option = \"magma\", alpha = 0.5) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Coding Exon Rate\") +\n",
    "  xlab(\"\") +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "ggsave(paste0(figPath, \"RNAseq_ExonRate.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Coding exon rate - HD\n",
    "alignD %>% filter(hd %in% hdList) %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>% ggplot(aes(x = cell, y = codingExonAlign, fill = cell)) +\n",
    "  geom_boxplot(outlier.shape = NA) +\n",
    "  geom_jitter(aes(color = hd), position = position_jitter(0.15)) + \n",
    "  facet_grid(~expr, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, end = 0.75, option = \"magma\", alpha = 0.5) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Coding Exon Rate\") +\n",
    "  xlab(\"\") +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "ggsave(paste0(figPath, \"RNAseq_ExonRate_HD.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Coding exon rate - PT\n",
    "alignD  %>% filter(hd %in% patient_id_list) %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>% ggplot(aes(x = cell, y = codingExonAlign, fill = cell)) +\n",
    "  geom_boxplot(outlier.shape = NA) +\n",
    "  geom_jitter(aes(color = hd), position = position_jitter(0.15)) + \n",
    "  facet_grid(~expr, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 1, option = \"magma\", alpha = 0.5) +\n",
    "  scale_color_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Coding Exon Rate\") +\n",
    "  xlab(\"\") +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "ggsave(paste0(figPath, \"RNAseq_ExonRate_PT.pdf\"), width = 8, height = 8)\n",
    "alignD  %>% filter(hd %in% patient_id_list) %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>% ggplot(aes(x = hd, y = codingExonAlign, fill = hd)) +\n",
    "  geom_bar(stat = \"identity\") +\n",
    "  scale_fill_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Coding Exon Rate\") +\n",
    "  xlab(\"Patient\") +\n",
    "  ggpubr::rremove(\"legend\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  ylim(0, 100)\n",
    "ggsave(paste0(figPath, \"RNAseq_ExonRate_PT_bar.pdf\"), width = 8, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Uni-mapping and Multi-mapping rate \n",
    "alignD  %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>% ggplot(aes(x = cell, y = uniExon, fill = cell)) +\n",
    "  geom_boxplot(outlier.shape = NA) +\n",
    "  geom_jitter(aes(color = hd), position = position_jitter(0.15)) + \n",
    "  facet_grid(~expr, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1, option = \"magma\", alpha = 0.5) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Uniquely Alignment Rate\") +\n",
    "  xlab(\"\") +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "ggsave(paste0(figPath, \"RNAseq_UniquelyAlignmentRate.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Uni-mapping and Multi-mapping rate - HD\n",
    "alignD %>% filter(hd %in% hdList) %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>% ggplot(aes(x = cell, y = uniExon, fill = cell)) +\n",
    "  geom_boxplot(outlier.shape = NA) +\n",
    "  geom_jitter(aes(color = hd), position = position_jitter(0.15)) + \n",
    "  facet_grid(~expr, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 0.1,  end = 0.75, option = \"magma\", alpha = 0.5) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Uniquely Alignment Rate\") +\n",
    "  xlab(\"\") +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "ggsave(paste0(figPath, \"RNAseq_UniquelyAlignmentRate_HD.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Uni-mapping and Multi-mapping rate - PT\n",
    "alignD  %>% filter(hd %in% patient_id_list) %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>% ggplot(aes(x = cell, y = uniExon, fill = cell)) +\n",
    "  geom_boxplot(outlier.shape = NA) +\n",
    "  geom_jitter(aes(color = hd), position = position_jitter(0.15)) + \n",
    "  facet_grid(~expr, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, begin = 1, option = \"magma\", alpha = 0.5) +\n",
    "  scale_color_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Uniquely Alignment Rate\") +\n",
    "  xlab(\"\") +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0)\n",
    "ggsave(paste0(figPath, \"RNAseq_UniquelyAlignmentRate_PT.pdf\"), width = 8, height = 8)\n",
    "alignD  %>% filter(hd %in% patient_id_list) %>% filter((expr == \"Input\" & cell != \"EMRA\") | (expr == \"Product\" & cell %in% c(\"CM\", \"EM\", \"Patient\"))) %>% ggplot(aes(x = hd, y = uniExon, fill = hd)) +\n",
    "  geom_bar(stat = \"identity\") +\n",
    "  scale_fill_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ylab(\"Uniquely Alignment Rate\") +\n",
    "  xlab(\"Patient\") +\n",
    "  ggpubr::rremove(\"legend\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  ylim(0, 100)\n",
    "ggsave(paste0(figPath, \"RNAseq_UniquelyAlignmentRate_PT_bar.pdf\"), width = 8, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Patient data infusion product of CM-derived cells \n",
    "patientRaw = read_xlsx(\"CART_CUTRUN_Project/report/CART_CUTRUN/analysis/20201028_NHL_CUTandRUN_PatientData.xlsx\")\n",
    "cm_ind = which(patientRaw$`CD8+ T cell isolation` == \"CD8TCM\")\n",
    "patient_id_list = patientRaw$Xnumber[cm_ind] %>% tolower\n",
    "patient_id_list\n",
    "\n",
    "## Patient data infusion product of CM-derived cells \n",
    "inPath = \"CART_CUTRUN_Project/results/RNAseq/process/NHL_RSEM_gencode_v21/\"\n",
    "\n",
    "pt_expCount = c()\n",
    "pt_name_tmp_list = c()\n",
    "for(sample in paste0(\"Sample_\", patient_id_list, \"_IP_CD8-CAR\")){\n",
    "  if(file.exists(paste0(inPath, sample, \"/RSEM.genes.results\"))){\n",
    "    print(sample)\n",
    "    patient_id = sample %>% gsub(\"Sample_\", \"\", .) %>% gsub(\"_IP_CD8-CAR\", \"\", .)\n",
    "    tmp = fread(paste0(inPath, sample, \"/RSEM.genes.results\")) %>% dplyr::select(expected_count)\n",
    "    pt_expCount = cbind(pt_expCount, round(tmp))\n",
    "    pt_name_tmp_list = c(pt_name_tmp_list, paste0(\"Product_Patient_\", patient_id))\n",
    "  }\n",
    "}\n",
    "pt_expCount = data.frame(pt_expCount)\n",
    "colnames(pt_expCount) = pt_name_tmp_list #paste0(patient_id_list, \"_PT\")\n",
    "\n",
    "# geneID = fread(paste0(inPath, sample, \"/RSEM.genes.results\"))$gene_id\n",
    "# pt_expCount$gene_name = geneID\n",
    "\n",
    "head(pt_expCount)\n",
    "dim(pt_expCount)\n",
    "\n",
    "## healthy donor data\n",
    "## Read in RNA-seq data\n",
    "cellList <- c(\"CM\", \"EM\") #c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hdList <- paste0(\"HD\", c(1:3, 5:7))\n",
    "exprList = \"Product\" #c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "\n",
    "inPath <- \"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\" #\"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "\n",
    "hd_expCount = c()\n",
    "colnames_tmp = c()\n",
    "for(cell in cellList){\n",
    "  for(expr in exprList){\n",
    "    for(hd in hdList){\n",
    "      # print(paste(expr, cell, hd, sep = \"_\"))\n",
    "      sample = paste0(\"RNA_CD8_\", cell, \"_\", expr, \"_\", hd)\n",
    "      if(file.exists(paste0(inPath, sample, \".genes.results\"))){\n",
    "        print(sample)\n",
    "        dataTmp = fread(paste0(inPath, sample, \".genes.results\")) %>% dplyr::select(expected_count)\n",
    "        # dataTmp <- fread(paste0(inPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".genes.results\")) %>% dplyr::select(expected_count)\n",
    "        hd_expCount <- cbind(hd_expCount, round(dataTmp))\n",
    "        colnames_tmp = c(colnames_tmp, paste0(expr, \"_\", cell, \"_\", hd))\n",
    "      }\n",
    "\n",
    "    }\n",
    "  }\n",
    "\n",
    "}\n",
    "\n",
    "cellList <- c(\"N\", \"CM\", \"EM\") #c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hdList <- paste0(\"HD\", c(1:3, 5:7))\n",
    "exprList = \"Input\" #c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "\n",
    "for(cell in cellList){\n",
    "  for(expr in exprList){\n",
    "    for(hd in hdList){\n",
    "      # print(paste(expr, cell, hd, sep = \"_\"))\n",
    "      sample = paste0(\"RNA_CD8_\", cell, \"_\", expr, \"_\", hd)\n",
    "      if(file.exists(paste0(inPath, sample, \".genes.results\"))){\n",
    "        print(sample)\n",
    "        dataTmp = fread(paste0(inPath, sample, \".genes.results\")) %>% dplyr::select(expected_count)\n",
    "        # dataTmp <- fread(paste0(inPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".genes.results\")) %>% dplyr::select(expected_count)\n",
    "        hd_expCount <- cbind(hd_expCount, round(dataTmp))\n",
    "        colnames_tmp = c(colnames_tmp, paste0(expr, \"_\", cell, \"_\", hd))\n",
    "      }\n",
    "\n",
    "    }\n",
    "  }\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "hd_expCount = data.frame(hd_expCount)\n",
    "colnames(hd_expCount) = colnames_tmp\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Distribution of Gene Expression.\n",
    "pt_data = stack(data.frame(pt_expCount)) %>% mutate(label = ind) %>% separate(ind, into = c(\"Experiment\", \"CellType\", \"HD\"), sep = \"_\")\n",
    "head(pt_data)\n",
    "\n",
    "hd_data = stack(data.frame(hd_expCount)) %>% mutate(label = ind) %>% separate(ind, into = c(\"Experiment\", \"CellType\", \"HD\"), sep = \"_\")\n",
    "head(hd_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Number of genes that have non-zero expression\n",
    "rbind(hd_data, pt_data) %>% filter(values > 0) %>% group_by(Experiment, CellType, HD) %>% summarize(geneN = n()) %>% ggplot(aes(x = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")), y = geneN, fill = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")))) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = HD), position = position_jitter(0.15)) + \n",
    "  facet_grid(~Experiment, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, option = \"magma\", alpha = 0.8, begin = 0.2) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  xlab(\"\") +\n",
    "  ylab(\"Number of Non-zero Expressed Genes\")\n",
    "ggsave(paste0(figPath, \"RNAseq_NonZeroGeneNum.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Number of genes that have non-zero expression - HD\n",
    "rbind(hd_data, pt_data) %>% filter(values > 0) %>% group_by(Experiment, CellType, HD) %>% summarize(geneN = n())  %>% filter(HD %in% hdList) %>% \n",
    "ggplot(aes(x = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")), y = geneN, fill = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")))) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = HD), position = position_jitter(0.15)) + \n",
    "  facet_grid(~Experiment, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, option = \"magma\", alpha = 0.8, begin = 0.2, end = 0.8) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  xlab(\"\") +\n",
    "  ylab(\"Number of Non-zero Expressed Genes\")\n",
    "ggsave(paste0(figPath, \"RNAseq_NonZeroGeneNum_HD.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Number of genes that have non-zero expression - PT\n",
    "rbind(hd_data, pt_data) %>% filter(values > 0) %>% group_by(Experiment, CellType, HD) %>% summarize(geneN = n())  %>% filter(HD %in% patient_id_list) %>% ggplot(aes(x = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")), y = geneN, fill = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")))) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = HD), position = position_jitter(0.15)) + \n",
    "  facet_grid(~Experiment, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, option = \"magma\", alpha = 0.8, begin = 1) +\n",
    "  scale_color_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  xlab(\"\") +\n",
    "  ylab(\"Number of Non-zero Expressed Genes\")\n",
    "ggsave(paste0(figPath, \"RNAseq_NonZeroGeneNum_PT.pdf\"), width = 8, height = 8)\n",
    "# Number of genes that have non-zero expression - PT\n",
    "rbind(hd_data, pt_data) %>% filter(values > 0) %>% group_by(Experiment, CellType, HD) %>% summarize(geneN = n())  %>% filter(HD %in% patient_id_list) %>% ggplot(aes(x = HD, y = geneN, fill = HD)) +\n",
    "  geom_bar(stat = \"identity\") +\n",
    "  scale_fill_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rremove(\"legend\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  xlab(\"Patient\") +\n",
    "  ylab(\"Number of Non-zero Expressed Genes\")\n",
    "ggsave(paste0(figPath, \"RNAseq_NonZeroGeneNum_PT_bar.pdf\"), width = 8, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "   Min. 1st Qu.  Median    Mean 3rd Qu.    Max. \n",
       "      1      10     106    1702     771 3749910 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "rbind(hd_data, pt_data) %>% filter(values > 0) %$% values %>% summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Percentage of genes that have relatively low expression.\n",
    "# 10 is the first quantile of overall non-zero gene expression.\n",
    "\n",
    "rbind(hd_data, pt_data) %>% filter(values < 10) %>% group_by(Experiment, CellType, HD) %>% summarize(LowExpressionNum = n()/nrow(pt_expCount) * 100) %>% ggplot(aes(x = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")), y = LowExpressionNum, fill = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")))) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = HD), position = position_jitter(0.15)) + \n",
    "  facet_grid(~Experiment, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, option = \"magma\", alpha = 0.8, begin = 0.2) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  xlab(\"\") +\n",
    "  ylab(\"% of Low Expression Genes\") +\n",
    "  ggtitle(\"GEX <= 1st Quantile of NonZero Genes: 10\")\n",
    "ggsave(paste0(figPath, \"RNAseq_LowExpressedGene.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Percentage of genes that have relatively low expression.\n",
    "# 10 is the first quantile of overall non-zero gene expression.\n",
    "# HD\n",
    "rbind(hd_data, pt_data) %>% filter(values < 10) %>% group_by(Experiment, CellType, HD) %>% summarize(LowExpressionNum = n()/nrow(pt_expCount) * 100)  %>% filter(HD %in% hdList) %>% ggplot(aes(x = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")), y = LowExpressionNum, fill = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")))) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = HD), position = position_jitter(0.15)) + \n",
    "  facet_grid(~Experiment, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, option = \"magma\", alpha = 0.8, begin = 0.2, end = 0.8) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  xlab(\"\") +\n",
    "  ylab(\"% of Low Expression Genes\") +\n",
    "  ggtitle(\"GEX <= 1st Quantile of NonZero Genes: 10\")\n",
    "ggsave(paste0(figPath, \"RNAseq_LowExpressedGene_HD.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Percentage of genes that have relatively low expression.\n",
    "# 10 is the first quantile of overall non-zero gene expression.\n",
    "# PT\n",
    "rbind(hd_data, pt_data) %>% filter(values < 10) %>% group_by(Experiment, CellType, HD) %>% summarize(LowExpressionNum = n()/nrow(pt_expCount) * 100) %>% filter(HD %in% patient_id_list) %>% ggplot(aes(x = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")), y = LowExpressionNum, fill = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")))) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = HD), position = position_jitter(0.15)) + \n",
    "  facet_grid(~Experiment, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, option = \"magma\", alpha = 0.8, begin = 1) +\n",
    "  scale_color_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  xlab(\"\") +\n",
    "  ylab(\"% of Low Expression Genes\") +\n",
    "  ggtitle(\"GEX <= 1st Quantile of NonZero Genes: 10\")\n",
    "ggsave(paste0(figPath, \"RNAseq_LowExpressedGene_PT.pdf\"), width = 8, height = 8)\n",
    "rbind(hd_data, pt_data) %>% filter(values < 10) %>% group_by(Experiment, CellType, HD) %>% summarize(LowExpressionNum = n()/nrow(pt_expCount) * 100) %>% filter(HD %in% patient_id_list) %>% ggplot(aes(x = HD, y = LowExpressionNum, fill = HD)) +\n",
    "  geom_bar(stat = \"identity\") +\n",
    "  scale_fill_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rremove(\"legend\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  xlab(\"Patient\") +\n",
    "  ylab(\"% of Low Expression Genes\") +\n",
    "  ggtitle(\"GEX <= 1st Quantile of NonZero Genes: 10\")\n",
    "\n",
    "ggsave(paste0(figPath, \"RNAseq_LowExpressedGene_PT_bar.pdf\"), width = 8, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# Percentage of genes that have relatively high expression.\n",
    "# 771 is the third quantile of overall non-zero gene expression.\n",
    "\n",
    "rbind(hd_data, pt_data) %>% filter(values >= 771) %>% group_by(Experiment, CellType, HD) %>% summarize(LowExpressionNum = n()/nrow(pt_expCount) * 100) %>% ggplot(aes(x = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")), y = LowExpressionNum, fill = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")))) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = HD), position = position_jitter(0.15)) + \n",
    "  facet_grid(~Experiment, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, option = \"magma\", alpha = 0.8, begin = 0.2) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  xlab(\"\") +\n",
    "  ylab(\"% of High Expression Genes\") +\n",
    "  ggtitle(\"GEX >= 3rd Quantile of Nonzero Genes: 771\")\n",
    "ggsave(paste0(figPath, \"RNAseq_HighExpressedGene.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Percentage of genes that have relatively high expression.\n",
    "# 771 is the third quantile of overall non-zero gene expression.\n",
    "# HD\n",
    "rbind(hd_data, pt_data) %>% filter(values >= 771) %>% group_by(Experiment, CellType, HD) %>% summarize(LowExpressionNum = n()/nrow(pt_expCount) * 100)  %>% filter(HD %in% hdList) %>% ggplot(aes(x = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")), y = LowExpressionNum, fill = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")))) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = HD), position = position_jitter(0.15)) + \n",
    "  facet_grid(~Experiment, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, option = \"magma\", alpha = 0.8, begin = 0.2, end = 0.8) +\n",
    "  scale_color_manual(values = palette_col) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  xlab(\"\") +\n",
    "  ylab(\"% of High Expression Genes\") +\n",
    "  ggtitle(\"GEX >= 3rd Quantile of Nonzero Genes: 771\")\n",
    "ggsave(paste0(figPath, \"RNAseq_HighExpressedGene_HD.pdf\"), width = 8, height = 8)\n",
    "\n",
    "# Percentage of genes that have relatively high expression.\n",
    "# 771 is the third quantile of overall non-zero gene expression.\n",
    "# PT\n",
    "rbind(hd_data, pt_data) %>% filter(values >= 771) %>% group_by(Experiment, CellType, HD) %>% summarize(LowExpressionNum = n()/nrow(pt_expCount) * 100) %>% filter(HD %in% patient_id_list) %>% ggplot(aes(x = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")), y = LowExpressionNum, fill = factor(CellType, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\", \"Patient\")))) +\n",
    "  geom_boxplot() +\n",
    "  geom_jitter(aes(color = HD), position = position_jitter(0.15)) + \n",
    "  facet_grid(~Experiment, scales = \"free\", space = \"free_x\") +\n",
    "  scale_fill_viridis(discrete = TRUE, option = \"magma\", alpha = 0.8, begin = 1) +\n",
    "  scale_color_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rremove(\"legend.title\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  xlab(\"\") +\n",
    "  ylab(\"% of High Expression Genes\") +\n",
    "  ggtitle(\"GEX >= 3rd Quantile of Nonzero Genes: 771\")\n",
    "ggsave(paste0(figPath, \"RNAseq_HighExpressedGene_PT.pdf\"), width = 8, height = 8)\n",
    "rbind(hd_data, pt_data) %>% filter(values >= 771) %>% group_by(Experiment, CellType, HD) %>% summarize(LowExpressionNum = n()/nrow(pt_expCount) * 100) %>% filter(HD %in% patient_id_list) %>% ggplot(aes(x = HD, y = LowExpressionNum, fill = HD)) +\n",
    "  geom_bar(stat = \"identity\") +\n",
    "  scale_fill_manual(values = palette_col[7:16]) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rremove(\"legend\") +\n",
    "  ggpubr::rotate_x_text(angle = 0) +\n",
    "  xlab(\"Patient\") +\n",
    "  ylab(\"% of High Expression Genes\") +\n",
    "  ggtitle(\"GEX >= 3rd Quantile of Nonzero Genes: 771\")\n",
    "ggsave(paste0(figPath, \"RNAseq_HighExpressedGene_PT_bar.pdf\"), width = 8, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Patient data infusion product of CM-derived cells \n",
    "patientRaw = read_xlsx(\"CART_CUTRUN_Project/report/CART_CUTRUN/analysis/20201028_NHL_CUTandRUN_PatientData.xlsx\")\n",
    "# cm_ind = which(patientRaw$`CD8+ T cell isolation` == \"CD8TCM\")\n",
    "patient_id_list = patientRaw$Xnumber %>% tolower\n",
    "patient_id_list\n",
    "\n",
    "## Patient data infusion product of CM-derived cells \n",
    "inPath = \"CART_CUTRUN_Project/results/RNAseq/process/NHL_RSEM_gencode_v21/\"\n",
    "\n",
    "pt_expCount = c()\n",
    "pt_name_tmp_list = c()\n",
    "for(sample in paste0(\"Sample_\", patient_id_list, \"_IP_CD8-CAR\")){\n",
    "  if(file.exists(paste0(inPath, sample, \"/RSEM.genes.results\"))){\n",
    "    print(sample)\n",
    "    patient_id = sample %>% gsub(\"Sample_\", \"\", .) %>% gsub(\"_IP_CD8-CAR\", \"\", .)\n",
    "    tmp = fread(paste0(inPath, sample, \"/RSEM.genes.results\")) %>% dplyr::select(expected_count)\n",
    "    pt_expCount = cbind(pt_expCount, round(tmp))\n",
    "    pt_name_tmp_list = c(pt_name_tmp_list, paste0(\"Product_Patient_\", patient_id))\n",
    "  }\n",
    "}\n",
    "pt_expCount = data.frame(pt_expCount)\n",
    "colnames(pt_expCount) = pt_name_tmp_list #paste0(patient_id_list, \"_PT\")\n",
    "\n",
    "geneID = fread(paste0(inPath, sample, \"/RSEM.genes.results\"))$gene_id\n",
    "rownames(pt_expCount) = geneID\n",
    "\n",
    "head(pt_expCount)\n",
    "dim(pt_expCount)\n",
    "\n",
    "## healthy donor data\n",
    "## Read in RNA-seq data\n",
    "cellList <- c(\"N\", \"CM\", \"EM\") #c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hdList <- paste0(\"HD\", c(1:3, 5:7))\n",
    "exprList = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "\n",
    "inPath <- \"CART_CUTRUN_Project/results/RNAseq/process/HD_RSEM_gencode_v21/\" #\"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "\n",
    "hd_expCount = c()\n",
    "colnames_tmp = c()\n",
    "for(cell in cellList){\n",
    "  for(expr in exprList){\n",
    "    for(hd in hdList){\n",
    "      # print(paste(expr, cell, hd, sep = \"_\"))\n",
    "      sample = paste0(\"RNA_CD8_\", cell, \"_\", expr, \"_\", hd)\n",
    "      if(file.exists(paste0(inPath, sample, \"/RSEM.genes.results\"))){\n",
    "        print(sample)\n",
    "        dataTmp = fread(paste0(inPath, sample, \"/RSEM.genes.results\")) %>% dplyr::select(expected_count)\n",
    "        # dataTmp <- fread(paste0(inPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".genes.results\")) %>% dplyr::select(expected_count)\n",
    "        hd_expCount <- cbind(hd_expCount, round(dataTmp))\n",
    "        colnames_tmp = c(colnames_tmp, paste0(expr, \"_\", cell, \"_\", hd))\n",
    "      }\n",
    "\n",
    "    }\n",
    "  }\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "hd_expCount = data.frame(hd_expCount)\n",
    "colnames(hd_expCount) = colnames_tmp\n",
    "\n",
    "geneID = fread(paste0(inPath, sample, \"/RSEM.genes.results\"))$gene_id\n",
    "rownames(hd_expCount) = geneID\n",
    "\n",
    "head(hd_expCount)\n",
    "dim(hd_expCount)\n",
    "\n",
    "dim(pt_expCount)\n",
    "dim(hd_expCount)\n",
    "# countMat = inner_join(pt_expCount, hd_expCount, by = \"gene_name\")\n",
    "# rownames(countMat) = countMat$gene_name\n",
    "# countMat = countMat %>% dplyr::select(-gene_name)\n",
    "# head(countMat)\n",
    "# dim(countMat)\n",
    "\n",
    "saveRDS(list(patientRaw, pt_expCount, hd_expCount), file = \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/RData/RNAseq_patientMeta_patientRawMatrix_hdRawMatrix.rds\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
