{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## CUTRUN DE and assign to nearest genes\n",
    "\n",
    "library(pacman)\n",
    "p_load(data.table, dplyr, ggplot2, viridis, magrittr, VennDiagram, ggpubr, limma, edgeR, tidyr, GenomicRanges, RColorBrewer, pheatmap, Seurat, fgsea, GSEABase, limma, TxDb.Hsapiens.UCSC.hg38.knownGene, org.Hs.eg.db)\n",
    "\n",
    "cell_list <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hd_list <- paste0(\"HD\", c(1:3, 5:7))\n",
    "expr_list = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "cell_comp_list = c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "hist_list <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "in_path <- \"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "out_path <- \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/\"\n",
    "fig_path <- \"CART_CUTRUN_Project/results/paper_figure/\"\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## PCA for RNA and CUTRUN\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\")) ## 'cell_comp_list''cell_list''countMat''designInfo''expr_list''fig_path''hd_list''hist_list''in_path''inPath''out_path''outPath'\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/RData/HD1-3_5-7_normLimma_perExprCondition.RData\") ## data, selectR, dataS, voomDDS, results, \n",
    "\n",
    "geneID = fread(paste0(\"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\", \"RNA_CD8_N_Input_HD1.genes.results\"))$gene_id\n",
    "data = data.frame(data)\n",
    "select_exp = \"Input\"\n",
    "\n",
    "rna_m = data[, colnames(data) %>% stringr::str_detect(select_exp)]\n",
    "k27_m = countMat[[\"H3K27me3\"]][, colnames(countMat[[\"H3K27me3\"]]) %>% stringr::str_detect(select_exp)]\n",
    "k4_m = countMat[[\"H3K4me2\"]][, colnames(countMat[[\"H3K4me2\"]]) %>% stringr::str_detect(select_exp)]\n",
    "\n",
    "rownames(rna_m) = geneID\n",
    "rownames(k27_m) = mPeak[[\"H3K27me3\"]] %>% data.frame %>% mutate(region = paste0(seqnames, \":\", start, \"-\", end)) %$% region\n",
    "rownames(k4_m) = mPeak[[\"H3K4me2\"]] %>% data.frame %>% mutate(region = paste0(seqnames, \":\", start, \"-\", end)) %$% region\n",
    "\n",
    "dim(rna_m)\n",
    "dim(k27_m)\n",
    "dim(k4_m)\n",
    "head(rna_m)\n",
    "head(k27_m)\n",
    "head(k4_m)\n",
    "\n",
    "rna_voom = voomDDS$E[, colnames(voomDDS$E) %>% stringr::str_detect(select_exp)]\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\"))\n",
    "k27_voom = voomDDS[[\"H3K27me3\"]]$E[, colnames(voomDDS[[\"H3K27me3\"]]$E) %>% stringr::str_detect(select_exp)]\n",
    "k4_voom = voomDDS[[\"H3K4me2\"]]$E[, colnames(voomDDS[[\"H3K4me2\"]]$E) %>% stringr::str_detect(select_exp)]\n",
    "rownames(rna_voom) = 1:nrow(rna_voom)\n",
    "rownames(k27_voom) = paste0(\"K27me3_\", 1:nrow(k27_voom))\n",
    "rownames(k4_voom) = paste0(\"K4me2_\", 1:nrow(k4_voom))\n",
    "\n",
    "dim(rna_voom)\n",
    "dim(k27_voom)\n",
    "dim(k4_voom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "rna_obj = CreateSeuratObject(counts = rna_m, project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 2000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = 10000000) %>% \n",
    "ScaleData()\n",
    "rna_obj = RunPCA(rna_obj, features = VariableFeatures(object = rna_obj), npcs = 20)\n",
    "\n",
    "k27_obj = CreateSeuratObject(counts = k27_m, project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = round(nrow(k27_m)/10*8)) %>% \n",
    "ScaleData()\n",
    "k27_obj = RunPCA(k27_obj, features = VariableFeatures(object = k27_obj), npcs = 15)\n",
    "\n",
    "k4_obj = CreateSeuratObject(counts = k4_m, project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = round(nrow(k4_m)/10*8)) %>% \n",
    "ScaleData()\n",
    "k4_obj = RunPCA(k4_obj, features = VariableFeatures(object = k4_obj), npcs = 15)\n",
    "\n",
    "cr_obj = CreateSeuratObject(counts = rbind(k27_m, k4_m), project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = round((nrow(k27_m) + nrow(k4_m))/10*8)) %>% \n",
    "ScaleData()\n",
    "cr_obj = RunPCA(cr_obj, features = VariableFeatures(object = cr_obj), npcs = 15)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=5)\n",
    "# DimPlot(rna_obj, reduction = \"pca\")\n",
    "ct_list = c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "p1 = Embeddings(rna_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 6), levels = ct_list)) %>% filter(celltype != \"EMRA\") %>%\n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"RNA-seq\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p2 = Embeddings(k27_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 4), levels = ct_list)) %>% filter(celltype != \"EMRA\") %>%\n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p3 = Embeddings(k4_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 4), levels = ct_list)) %>% filter(celltype != \"EMRA\") %>%\n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p4 = Embeddings(cr_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 4), levels = ct_list)) %>% filter(celltype != \"EMRA\") %>%\n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3+H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "ggarrange(p1, p2, p3, p4, nrow = 1, ncol = 4, common.legend = TRUE)\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig1/PCA_seurat_procedure.pdf\", width = 20, height = 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "anosim(x = t(rna_m[, 1:18]), grouping = factor(rep(ct_list[1:3],      6), levels = ct_list[1:3])) \n",
       "Dissimilarity: bray \n",
       "\n",
       "ANOSIM statistic R: 0.1432 \n",
       "      Significance: 0.04 \n",
       "\n",
       "Permutation: free\n",
       "Number of permutations: 999\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "anosim(x = t(k27_m[, 1:12]), grouping = factor(rep(ct_list[1:3],      4), levels = ct_list[1:3])) \n",
       "Dissimilarity: bray \n",
       "\n",
       "ANOSIM statistic R: -0.1852 \n",
       "      Significance: 0.898 \n",
       "\n",
       "Permutation: free\n",
       "Number of permutations: 999\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "anosim(x = t(k4_m[, 1:12]), grouping = factor(rep(ct_list[1:3],      4), levels = ct_list[1:3])) \n",
       "Dissimilarity: bray \n",
       "\n",
       "ANOSIM statistic R: -0.169 \n",
       "      Significance: 0.891 \n",
       "\n",
       "Permutation: free\n",
       "Number of permutations: 999\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "anosim(x = t(rbind(k27_m, k4_m)[, 1:12]), grouping = factor(rep(ct_list[1:3],      4), levels = ct_list[1:3])) \n",
       "Dissimilarity: bray \n",
       "\n",
       "ANOSIM statistic R: -0.2083 \n",
       "      Significance: 0.944 \n",
       "\n",
       "Permutation: free\n",
       "Number of permutations: 999\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "anosim(t(rna_m[, 1:18]), grouping = factor(rep(ct_list[1:3], 6), levels = ct_list[1:3]))\n",
    "anosim(t(k27_m[, 1:12]), grouping = factor(rep(ct_list[1:3], 4), levels = ct_list[1:3]))\n",
    "anosim(t(k4_m[, 1:12]), grouping = factor(rep(ct_list[1:3], 4), levels = ct_list[1:3]))\n",
    "anosim(t(rbind(k27_m, k4_m)[, 1:12]), grouping = factor(rep(ct_list[1:3], 4), levels = ct_list[1:3]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in vegdist(x, method = distance):\n",
      "“results may be meaningless because data have negative entries in method “bray””\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "anosim(x = Embeddings(rna_obj, reduction = \"pca\")[1:18, 1:2] %>%      data.frame, grouping = factor(rep(ct_list[1:3], 6), levels = ct_list[1:3])) \n",
       "Dissimilarity: bray \n",
       "\n",
       "ANOSIM statistic R: -0.0858 \n",
       "      Significance: 0.858 \n",
       "\n",
       "Permutation: free\n",
       "Number of permutations: 999\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in vegdist(x, method = distance):\n",
      "“results may be meaningless because data have negative entries in method “bray””\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "anosim(x = Embeddings(k27_obj, reduction = \"pca\")[1:12, 1:2] %>%      data.frame, grouping = factor(rep(ct_list[1:3], 4), levels = ct_list[1:3])) \n",
       "Dissimilarity: bray \n",
       "\n",
       "ANOSIM statistic R:    -0 \n",
       "      Significance: 0.463 \n",
       "\n",
       "Permutation: free\n",
       "Number of permutations: 999\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in vegdist(x, method = distance):\n",
      "“results may be meaningless because data have negative entries in method “bray””\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "anosim(x = Embeddings(k4_obj, reduction = \"pca\")[1:12, 1:2] %>%      data.frame, grouping = factor(rep(ct_list[1:3], 4), levels = ct_list[1:3])) \n",
       "Dissimilarity: bray \n",
       "\n",
       "ANOSIM statistic R: -0.1667 \n",
       "      Significance: 0.847 \n",
       "\n",
       "Permutation: free\n",
       "Number of permutations: 999\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in vegdist(x, method = distance):\n",
      "“results may be meaningless because data have negative entries in method “bray””\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "\n",
       "Call:\n",
       "anosim(x = Embeddings(cr_obj, reduction = \"pca\")[1:12, 1:2] %>%      data.frame, grouping = factor(rep(ct_list[1:3], 4), levels = ct_list[1:3])) \n",
       "Dissimilarity: bray \n",
       "\n",
       "ANOSIM statistic R: -0.1273 \n",
       "      Significance: 0.86 \n",
       "\n",
       "Permutation: free\n",
       "Number of permutations: 999\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "anosim(Embeddings(rna_obj, reduction = \"pca\")[1:18, 1:2] %>% data.frame, grouping = factor(rep(ct_list[1:3], 6), levels = ct_list[1:3]))\n",
    "anosim(Embeddings(k27_obj, reduction = \"pca\")[1:12, 1:2] %>% data.frame, grouping = factor(rep(ct_list[1:3], 4), levels = ct_list[1:3]))\n",
    "anosim(Embeddings(k4_obj, reduction = \"pca\")[1:12, 1:2] %>% data.frame, grouping = factor(rep(ct_list[1:3], 4), levels = ct_list[1:3]))\n",
    "anosim(Embeddings(cr_obj, reduction = \"pca\")[1:12, 1:2] %>% data.frame, grouping = factor(rep(ct_list[1:3], 4), levels = ct_list[1:3]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#!/bin/bash\n",
    "ml SAMtools/1.11-GCC-10.2.0 deepTools/3.3.1-foss-2019b-Python-3.7.4 libpciaccess/0.14-GCCcore-8.3.0\n",
    "outPath=\"CART_CUTRUN_Project/results/CUTANDRUN/bigwig\"\n",
    "genePath=\"CART_CUTRUN_Project/results/RNAseq/analysis/TSV\"\n",
    "analysisOut=\"CART_CUTRUN_Project/results/paper_figure/fig1/Tornado\"\n",
    "# for hist in H3K27me3 #H3K4me2\n",
    "# do\n",
    "#     for ct in N #CM EM EMRA\n",
    "#     do\n",
    "#         echo $hist\n",
    "#         echo $ct\n",
    "#     \t  file1=\"${hist}_CD8_${ct}_Input_HD4_rep1\"\n",
    "#         file2=\"${hist}_CD8_${ct}_Input_HD5_rep1\"\n",
    "#         file3=\"${hist}_CD8_${ct}_Input_HD6_rep1\"\n",
    "#         file4=\"${hist}_CD8_${ct}_Input_HD7_rep1\"\n",
    "#         inFile1=\"CART_CUTRUN_Project/results/CUTANDRUN/process/${file1}/alignment/bowtie2_align.bam\"\n",
    "#         inFile2=\"CART_CUTRUN_Project/results/CUTANDRUN/process/${file2}/alignment/bowtie2_align.bam\"\n",
    "#         inFile3=\"CART_CUTRUN_Project/results/CUTANDRUN/process/${file3}/alignment/bowtie2_align.bam\"\n",
    "#         inFile4=\"CART_CUTRUN_Project/results/CUTANDRUN/process/${file4}/alignment/bowtie2_align.bam\"\n",
    "\n",
    "#         samtools merge $outPath/${hist}_${ct}_bowtie2_align.bam $inFile1 $inFile2 $inFile3 $inFile4\n",
    "#         samtools sort -o $outPath/${hist}_${ct}_bowtie2_align.sorted.bam $outPath/${hist}_${ct}_bowtie2_align.bam\n",
    "#         samtools index $outPath/${hist}_${ct}_bowtie2_align.sorted.bam\n",
    "#         bamCoverage -b $outPath/${hist}_${ct}_bowtie2_align.sorted.bam -o $outPath/${hist}_${ct}_bowtie2_align.bw\n",
    "#     done\n",
    "# done\n",
    "cores=6\n",
    "for hist in H3K4me2 H3K27me3\n",
    "do\n",
    "    for ct in N CM EM # EMRA\n",
    "    do\n",
    "        echo $hist\n",
    "        echo $ct\n",
    "        computeMatrix reference-point -S $outPath/${hist}_${ct}_bowtie2_align.bw \\\n",
    "                              -R $genePath/Input_${ct}_ordered_gene_region.tsv\\\n",
    "                              --skipZeros -o $analysisOut/${hist}_${ct}.mat.gz -p $cores -a 1000 -b 1000 --referencePoint TSS --samplesLabel ${hist}_${ct}\n",
    "\n",
    "        plotHeatmap -m $analysisOut/${hist}_${ct}.mat.gz -out $analysisOut/${hist}_${ct}_sort_gex.png --sortRegions keep --regionsLabel \"\" --colorMap Reds\n",
    "        plotHeatmap -m $analysisOut/${hist}_${ct}.mat.gz -out $analysisOut/${hist}_${ct}_sort_peak_signal.png --sortUsing sum --regionsLabel \"\" --colorMap Reds\n",
    "    done\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(pacman)\n",
    "p_load(data.table, dplyr, ggplot2, viridis, magrittr, VennDiagram, ggpubr, limma, edgeR, tidyr, GenomicRanges, RColorBrewer, pheatmap, Seurat, fgsea, GSEABase, limma, TxDb.Hsapiens.UCSC.hg38.knownGene, org.Hs.eg.db, ggrepel)\n",
    "fig_path <- \"CART_CUTRUN_Project/results/paper_figure_paper1/\"\n",
    "\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "\n",
    "## Read in RNA-seq data\n",
    "cellList <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hdList <- paste0(\"HD\", c(1:3, 5:7))\n",
    "exprList = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "\n",
    "\n",
    "inPath <- \"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "outPath <- \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/\"\n",
    "\n",
    "rna_expCount = c()\n",
    "rna_TPM = c()\n",
    "rna_FPKM = c()\n",
    "for(cell in cellList){\n",
    "  for(expr in exprList){\n",
    "    for(hd in hdList){\n",
    "      # print(paste(expr, cell, hd, sep = \"_\"))\n",
    "      dataTmp <- fread(paste0(inPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".genes.results\")) %>% dplyr::select(expected_count)\n",
    "      rna_expCount <- cbind(rna_expCount, round(dataTmp))\n",
    "      dataTmp <- fread(paste0(inPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".genes.results\")) %>% dplyr::select(TPM)\n",
    "      rna_TPM <- cbind(rna_TPM, round(dataTmp))\n",
    "      dataTmp <- fread(paste0(inPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".genes.results\")) %>% dplyr::select(FPKM)\n",
    "      rna_FPKM <- cbind(rna_FPKM, round(dataTmp))\n",
    "\n",
    "    }\n",
    "  }\n",
    "\n",
    "}\n",
    "geneID = fread(paste0(inPath, \"RNA_CD8_N_Input_HD1.genes.results\"))$gene_id\n",
    "rownames(rna_expCount) <- geneID\n",
    "colnames(rna_expCount) <- paste(rep(rep(exprList, each = length(hdList)), length(cellList)), rep(cellList, each = length(hdList)*length(exprList)), rep(hdList, length(exprList)*length(cellList)), sep = \"_\")\n",
    "rownames(rna_TPM) <- geneID\n",
    "colnames(rna_TPM) <- paste(rep(rep(exprList, each = length(hdList)), length(cellList)), rep(cellList, each = length(hdList)*length(exprList)), rep(hdList, length(exprList)*length(cellList)), sep = \"_\")\n",
    "rownames(rna_FPKM) <- geneID\n",
    "colnames(rna_FPKM) <- paste(rep(rep(exprList, each = length(hdList)), length(cellList)), rep(cellList, each = length(hdList)*length(exprList)), rep(hdList, length(exprList)*length(cellList)), sep = \"_\")\n",
    "\n",
    "library(chromVAR)\n",
    "library(SummarizedExperiment)\n",
    "library(Matrix)\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "\n",
    "countMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_countMat.rds\"))\n",
    "normMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_normMat.rds\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "target_gene_list = c(\"Tim3\", \"LAG\", \"CCL3\", \"CCL4\") "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## mean across healthy donors\n",
    "fig_path = \"CART_CUTRUN_Project/results/paper_figure/fig1/\"\n",
    "pdf(paste0(fig_path, \"/K4_VS_K27_pointColorSizeRNA_TRgenes_nolabel.pdf\"), width = 20, height = 20)\n",
    "\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    rna_select_cond = paste0(\"Input_\", cT)\n",
    "    cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "    rna_select_data = rna_expCount\n",
    "    cr_select_data = normMat$H3K27me3\n",
    "\n",
    "    cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "    rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "    k27_data = log2(normMat$H3K27me3[, cr_col_index] + 1) %>% rowMeans %>% t %>% t %>% data.frame\n",
    "    colnames(k27_data) = c(\"K27\")\n",
    "    k27_data$gene_name = rownames(k27_data) %>% tolower\n",
    "\n",
    "    k4_data = log2(normMat$H3K4me2[, cr_col_index] + 1) %>% rowMeans %>% t %>%  t %>% data.frame\n",
    "    colnames(k4_data) = c(\"K4\")\n",
    "    k4_data$gene_name = rownames(k4_data) %>% tolower\n",
    "\n",
    "    rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "    RNA_data = log2(data.frame(rna_select_data)[, rna_col_index] + 1) %>% rowMeans %>% t %>%  t %>% data.frame\n",
    "    colnames(RNA_data) = c(\"RNA\")\n",
    "    RNA_data$gene_name = rna_gene\n",
    "\n",
    "    cr_data = cbind(k27_data, K4 = k4_data$K4)\n",
    "    cr_rna_full = full_join(RNA_data, cr_data, by = \"gene_name\") %>% dplyr::select(gene_name, K4, K27, RNA)\n",
    "    cr_rna_full[is.na(cr_rna_full)] = 0\n",
    "\n",
    "\n",
    "\n",
    "    print(cr_rna_full %>% ggplot() +\n",
    "        geom_point(aes(x = K4, y = K27, color = RNA, size = RNA), alpha = 0.5) +\n",
    "        theme_bw(base_size = 50) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_viridis() +\n",
    "        ggtitle(cT) +\n",
    "        geom_vline(xintercept = 4.5, linetype = \"dashed\") +\n",
    "        geom_hline(yintercept = 2.5, linetype = \"dashed\"))\n",
    "\n",
    "\n",
    "\n",
    "}\n",
    "dev.off()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## mean across healthy donors\n",
    "fig_path = \"CART_CUTRUN_Project/results/fig1/\"\n",
    "pdf(paste0(fig_path, \"/K4_VS_K27_pointColorSizeRNA_TRgenes_withlabel.pdf\"), width = 20, height = 20)\n",
    "\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    rna_select_cond = paste0(\"Input_\", cT)\n",
    "    cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "    rna_select_data = rna_expCount\n",
    "    cr_select_data = normMat$H3K27me3\n",
    "\n",
    "    cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "    rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "    k27_data = log2(normMat$H3K27me3[, cr_col_index] + 1) %>% rowMeans %>% t %>% t %>% data.frame\n",
    "    colnames(k27_data) = c(\"K27\")\n",
    "    k27_data$gene_name = rownames(k27_data) %>% tolower\n",
    "\n",
    "    k4_data = log2(normMat$H3K4me2[, cr_col_index] + 1) %>% rowMeans %>% t %>%  t %>% data.frame\n",
    "    colnames(k4_data) = c(\"K4\")\n",
    "    k4_data$gene_name = rownames(k4_data) %>% tolower\n",
    "\n",
    "    rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "    RNA_data = log2(data.frame(rna_select_data)[, rna_col_index] + 1) %>% rowMeans %>% t %>%  t %>% data.frame\n",
    "    colnames(RNA_data) = c(\"RNA\")\n",
    "    RNA_data$gene_name = rna_gene\n",
    "\n",
    "    cr_data = cbind(k27_data, K4 = k4_data$K4)\n",
    "    cr_rna_full = full_join(RNA_data, cr_data, by = \"gene_name\") %>% dplyr::select(gene_name, K4, K27, RNA)\n",
    "    cr_rna_full[is.na(cr_rna_full)] = 0\n",
    "\n",
    "\n",
    "\n",
    "    print(cr_rna_full %>% ggplot() +\n",
    "        geom_point(aes(x = K4, y = K27, color = RNA, size = RNA), alpha = 0.5) +\n",
    "        geom_point(aes(x = K4, y = K27, size = RNA), data = cr_rna_full %>% dplyr::filter(gene_name %in% tolower(target_gene_list)), color = \"black\", size = 8) +\n",
    "        geom_text_repel(aes(x = K4, y = K27, label = gene_name), data = cr_rna_full %>% dplyr::filter(gene_name %in% tolower(target_gene_list)), color = \"black\", size = 15) +\n",
    "        theme_bw(base_size = 50) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_viridis() +\n",
    "        ggtitle(cT) +\n",
    "        geom_vline(xintercept = 4.5, linetype = \"dashed\") +\n",
    "        geom_hline(yintercept = 2.5, linetype = \"dashed\"))\n",
    "\n",
    "\n",
    "\n",
    "}\n",
    "dev.off()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## PCA of cordant or discordant genes and see how well the cell types are separated\n",
    "cr_rna_full$K4_group = \"-\"\n",
    "cr_rna_full$K4_group[which(cr_rna_full$K4 > 4.5)] = \"+\"\n",
    "cr_rna_full$K27_group = \"-\"\n",
    "cr_rna_full$K27_group[which(cr_rna_full$K27 > 2.5)] = \"+\"\n",
    "cr_rna_full$RNA_group = \"-\"\n",
    "cr_rna_full$RNA_group[which(cr_rna_full$RNA > 2.5)] = \"+\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## full gene\n",
    "rna_col_index = which(rna_expCount %>% colnames %>% stringr::str_detect(\"Input\"))\n",
    "cr_col_index = which(normMat$H3K27me %>% colnames %>% stringr::str_detect(\"Input\"))\n",
    "\n",
    "rna_m = rna_expCount %>% data.frame\n",
    "rownames(rna_m) = rownames(rna_expCount)\n",
    "rna_m = rna_m[, rna_col_index]\n",
    "\n",
    "rna_obj = CreateSeuratObject(counts = rna_m[, 1:18], project = \"cart\", min.cells = 0, min.features = 50)  %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = 1000000) %>% \n",
    "ScaleData()\n",
    "rna_obj = RunPCA(rna_obj, features = VariableFeatures(object = rna_obj), npcs = 10)\n",
    "\n",
    "k27_obj = CreateSeuratObject(counts = normMat$H3K27me[, cr_col_index][, 1:12], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = round(nrow(normMat$H3K27me)/10*8)) %>% \n",
    "ScaleData()\n",
    "k27_obj = RunPCA(k27_obj, features = VariableFeatures(object = k27_obj), npcs = 10)\n",
    "\n",
    "k4_obj = CreateSeuratObject(counts = normMat$H3K4me2[, cr_col_index][, 1:12], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = round(nrow(normMat$H3K4me2)/10*8)) %>% \n",
    "ScaleData()\n",
    "k4_obj = RunPCA(k4_obj, features = VariableFeatures(object = k4_obj), npcs = 10)\n",
    "\n",
    "cr_obj = CreateSeuratObject(counts = rbind(normMat$H3K27me, normMat$H3K4me2)[, cr_col_index][, 1:12], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = round((nrow(normMat$H3K27me) + nrow(normMat$H3K4me2))/10*8)) %>% \n",
    "ScaleData()\n",
    "cr_obj = RunPCA(cr_obj, features = VariableFeatures(object = cr_obj), npcs = 10)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=5)\n",
    "# DimPlot(rna_obj, reduction = \"pca\")\n",
    "ct_list = c(\"N\", \"CM\", \"EM\") #, \"EMRA\")\n",
    "p1 = Embeddings(rna_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 6), levels = ct_list)) %>% filter(celltype != \"EMRA\") %>%\n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"RNA-seq\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p2 = Embeddings(k27_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 4), levels = ct_list)) %>% filter(celltype != \"EMRA\") %>%\n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p3 = Embeddings(k4_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 4), levels = ct_list)) %>% filter(celltype != \"EMRA\") %>%\n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p4 = Embeddings(cr_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 4), levels = ct_list)) %>% filter(celltype != \"EMRA\") %>%\n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3+H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "ggarrange(p1, p2, p3, p4, nrow = 1, ncol = 4, common.legend = TRUE)\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig1/PCA_seurat_procedure_usingReadsAroundGenes.pdf\", width = 20, height = 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "rna_col_index = which(rna_expCount %>% colnames %>% stringr::str_detect(\"Input\"))\n",
    "cr_col_index = which(normMat$H3K27me %>% colnames %>% stringr::str_detect(\"Input\"))\n",
    "rna_m = rna_expCount %>% data.frame\n",
    "rownames(rna_m) = rownames(rna_expCount)\n",
    "\n",
    "\n",
    "rna_obj = CreateSeuratObject(counts = rna_m[, rna_col_index][, 1:18], project = \"cart\", min.cells = 0, min.features = 50)  %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = nrow(rna_m)) %>% \n",
    "ScaleData()\n",
    "\n",
    "k27_obj = CreateSeuratObject(counts = normMat$H3K27me[, cr_col_index][, 1:12], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = nrow(normMat$H3K27me)) %>% \n",
    "ScaleData()\n",
    "\n",
    "k4_obj = CreateSeuratObject(counts = normMat$H3K4me2[, cr_col_index][, 1:12], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = nrow(normMat$H3K4me2)) %>% \n",
    "ScaleData()\n",
    "\n",
    "cr_obj = CreateSeuratObject(counts = rbind(normMat$H3K27me, normMat$H3K4me2)[, cr_col_index][, 1:12], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = nrow(normMat$H3K27me)*2) %>% \n",
    "ScaleData()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# gene_category = \"RNA+K4+K27-_CE\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"+\", K27_group == \"-\", RNA_group == \"+\") %$% gene_name\n",
    "\n",
    "gene_category = \"RNA-K4+K27-_DCNE\"\n",
    "target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"+\", K27_group == \"-\", RNA_group == \"-\") %$% gene_name\n",
    "\n",
    "# gene_category = \"RNA+K4-K27+_DE\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"-\", K27_group == \"+\", RNA_group == \"+\") %$% gene_name\n",
    "\n",
    "# gene_category = \"RNA-K4-K27+_CNE\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"-\", K27_group == \"+\", RNA_group == \"-\") %$% gene_name\n",
    "\n",
    "# gene_category = \"RNA+K4+K27+_BE\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"+\", K27_group == \"+\", RNA_group == \"+\") %$% gene_name\n",
    "\n",
    "# gene_category = \"RNA-K4+K27+_BNE\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"+\", K27_group == \"+\", RNA_group == \"-\") %$% gene_name\n",
    "\n",
    "# gene_category = \"RNA+K4-K27-_NME\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"-\", K27_group == \"-\", RNA_group == \"+\") %$% gene_name\n",
    "\n",
    "# gene_category = \"RNA-K4-K27-_NMNE\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"-\", K27_group == \"-\", RNA_group == \"-\") %$% gene_name\n",
    "\n",
    "rna_gene_name_list = rownames(rna_m) %>% gsub(\".*_\", \"\", .) %>% tolower \n",
    "rna_row_index = which(rna_gene_name_list %in% target_gene)\n",
    "\n",
    "cr_gene_name_list = rownames(normMat$H3K27me3) %>% tolower\n",
    "cr_row_index = which(cr_gene_name_list %in% target_gene)\n",
    "\n",
    "rna_obj_subset = rna_obj[rna_row_index, ]\n",
    "rna_obj_subset = RunPCA(rna_obj_subset, features = VariableFeatures(object = rna_obj_subset), npcs = 10)\n",
    "\n",
    "k27_obj_subset = k27_obj[cr_row_index, ]\n",
    "k27_obj_subset = RunPCA(k27_obj_subset, features = VariableFeatures(object = k27_obj_subset), npcs = 10)\n",
    "\n",
    "k4_obj_subset = k4_obj[cr_row_index, ]\n",
    "k4_obj_subset = RunPCA(k4_obj_subset, features = VariableFeatures(object = k4_obj_subset), npcs = 10)\n",
    "\n",
    "cr_obj_subset = cr_obj[c(cr_row_index, nrow(normMat$H3K27me) + cr_row_index), ]\n",
    "cr_obj_subset = RunPCA(cr_obj_subset, features = VariableFeatures(object = cr_obj_subset), npcs = 10)\n",
    "\n",
    "options(repr.plot.width=15, repr.plot.height=5)\n",
    "# DimPlot(rna_obj, reduction = \"pca\")\n",
    "ct_list = c(\"N\", \"CM\", \"EM\") #, \"EMRA\")\n",
    "p12 = Embeddings(rna_obj_subset, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 6), levels = ct_list)) %>% filter(celltype != \"EMRA\") %>%\n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"RNA-seq\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p22 = Embeddings(k27_obj_subset, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 4), levels = ct_list)) %>% filter(celltype != \"EMRA\") %>%\n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p32 = Embeddings(k4_obj_subset, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 4), levels = ct_list)) %>% filter(celltype != \"EMRA\") %>%\n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p42 = Embeddings(cr_obj_subset, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 4), levels = ct_list)) %>% filter(celltype != \"EMRA\") %>%\n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3+H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "ggarrange(p12, p22, p32, p42, nrow = 1, ncol = 4, common.legend = TRUE)\n",
    "ggsave(filename = paste0(\"CART_CUTRUN_Project/results/paper_figure/fig1/PCA_seurat_procedure_\", gene_category, \".pdf\"), width = 20, height = 6)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## patient related\n",
    "\n",
    "## Patient data infusion product of CM-derived cells \n",
    "patientRaw = read_xlsx(\"CART_CUTRUN_Project/report/CART_CUTRUN/analysis/20201028_NHL_CUTandRUN_PatientData.xlsx\")\n",
    "cm_ind = which(patientRaw$`CD8+ T cell isolation` == \"CD8TCM\")\n",
    "patient_id_list = patientRaw$Xnumber[cm_ind] %>% tolower\n",
    "patient_id_list\n",
    "\n",
    "## Patient data infusion product of CM-derived cells \n",
    "inPath = \"CART_CUTRUN_Project/results/RNAseq/process/NHL_RSEM_gencode_v21/\"\n",
    "\n",
    "pt_expCount = c()\n",
    "pt_name_tmp_list = c()\n",
    "for(sample in paste0(\"Sample_\", patient_id_list, \"_IP_CD8-CAR\")){\n",
    "  if(file.exists(paste0(inPath, sample, \"/RSEM.genes.results\"))){\n",
    "    print(sample)\n",
    "    tmp = fread(paste0(inPath, sample, \"/RSEM.genes.results\")) %>% dplyr::select(expected_count)\n",
    "    pt_expCount = cbind(pt_expCount, round(tmp))\n",
    "    pt_name_tmp_list = c(pt_name_tmp_list, sample)\n",
    "  }\n",
    "}\n",
    "pt_expCount = data.frame(pt_expCount)\n",
    "colnames(pt_expCount) = paste0(pt_name_tmp_list, \"_PT\") #paste0(patient_id_list, \"_PT\")\n",
    "\n",
    "geneID = fread(paste0(inPath, sample, \"/RSEM.genes.results\"))$gene_id\n",
    "pt_expCount$gene_name = geneID\n",
    "\n",
    "head(pt_expCount)\n",
    "dim(pt_expCount)\n",
    "\n",
    "## healthy donor data\n",
    "## Read in RNA-seq data\n",
    "cellList <- \"CM\" #c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hdList <- paste0(\"HD\", c(1:3, 5:7))\n",
    "exprList = \"Product\" #c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "\n",
    "inPath <- \"CART_CUTRUN_Project/results/RNAseq/process/HD_RSEM_gencode_v21/\" #\"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "\n",
    "hd_expCount = c()\n",
    "for(cell in cellList){\n",
    "  for(expr in exprList){\n",
    "    for(hd in hdList){\n",
    "      # print(paste(expr, cell, hd, sep = \"_\"))\n",
    "      sample = paste0(\"RNA_CD8_\", cell, \"_\", expr, \"_\", hd)\n",
    "      if(file.exists(paste0(inPath, sample, \"/RSEM.genes.results\"))){\n",
    "        print(sample)\n",
    "        dataTmp = fread(paste0(inPath, sample, \"/RSEM.genes.results\")) %>% dplyr::select(expected_count)\n",
    "        # dataTmp <- fread(paste0(inPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".genes.results\")) %>% dplyr::select(expected_count)\n",
    "        hd_expCount <- cbind(hd_expCount, round(dataTmp))\n",
    "      }\n",
    "\n",
    "    }\n",
    "  }\n",
    "\n",
    "}\n",
    "hd_expCount = data.frame(hd_expCount)\n",
    "colnames(hd_expCount) <- paste(rep(rep(exprList, each = length(hdList)), length(cellList)), rep(cellList, each = length(hdList)*length(exprList)), rep(hdList, length(exprList)*length(cellList)), sep = \"_\")\n",
    "\n",
    "geneID = fread(paste0(inPath, sample, \"/RSEM.genes.results\"))$gene_id\n",
    "hd_expCount$gene_name <- geneID\n",
    "head(hd_expCount)\n",
    "dim(hd_expCount)\n",
    "\n",
    "dim(pt_expCount)\n",
    "dim(hd_expCount)\n",
    "countMat = inner_join(pt_expCount, hd_expCount, by = \"gene_name\")\n",
    "rownames(countMat) = countMat$gene_name\n",
    "countMat = countMat %>% dplyr::select(-gene_name)\n",
    "head(countMat)\n",
    "dim(countMat)\n",
    "\n",
    "countMat_rna = countMat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=7, repr.plot.height=6)\n",
    "selectR = which(rowSums(countMat_rna) > 10) ## remove low count genes\n",
    "length(selectR)\n",
    "\n",
    "rna_obj = CreateSeuratObject(counts = countMat_rna[selectR, ], project = \"cart\", min.cells = 0, min.features = 10) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR)) %>%  # length(selectR)\n",
    "ScaleData()\n",
    "\n",
    "rna_obj = RunPCA(rna_obj, features = VariableFeatures(object = rna_obj), npcs = 10)\n",
    "rna_plot = Embeddings(rna_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", ncol(pt_expCount) - 1), rep(\"HD\", ncol(hd_expCount) - 1))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"RNA-seq\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "pdf(\"CART_CUTRUN_Project/results/paper_figure/fig1/CMpatient_CMhealthyDonor_PCA_on_RNA-seq_allfeatures.pdf\", width = 7, height = 6)\n",
    "rna_plot\n",
    "dev.off()\n",
    "rna_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "patientRaw = read_xlsx(\"CART_CUTRUN_Project/report/CART_CUTRUN/analysis/20201028_NHL_CUTandRUN_PatientData.xlsx\")\n",
    "cm_ind = which(patientRaw$`CD8+ T cell isolation` == \"CD8TCM\")\n",
    "patientRaw$Xnumber[cm_ind]\n",
    "\n",
    "## 1. TSS neighbor regions\n",
    "## use TSS region for differential detection\n",
    "## Get the TSS location and +/-1kb region around TSS\n",
    "gtf <- rtracklayer::import('/fh/fast/gottardo_r/yezheng_working/SupplementaryData/hg38/transcriptomeHumanReference/gencode.v33.annotation.gtf') #rtracklayer::import('/fh/fast/gottardo_r/yezheng_working/SupplementaryData/hg38/GENCODE/gencode.v21.annotation.gtf')\n",
    "gtf_df=as.data.frame(gtf)\n",
    "\n",
    "\n",
    "## get the TSS location and extend by 1kb up and downstream\n",
    "tss_df = rbind(\n",
    "    gtf_df %>% data.frame %>% dplyr::filter(type == \"gene\", strand == \"+\") %>% dplyr::select(seqnames, TSS = start, strand, source, gene_id, gene_name, gene_name, gene_type) ,\n",
    "    gtf_df %>% data.frame %>% dplyr::filter(type == \"gene\", strand == \"-\") %>% dplyr::select(seqnames, TSS = end, strand, source, gene_id, gene_name, gene_name, gene_type) \n",
    ")\n",
    "tss_filter_df = tss_df %>% dplyr::filter(!(seqnames %in% c(\"chrM\", \"chrX\", \"chrY\")))\n",
    "\n",
    "tss_gr = GRanges(seqnames = tss_filter_df$seqnames, ranges = IRanges(start = tss_filter_df$TSS - 1000, end = tss_filter_df$TSS + 1000), strand = tss_filter_df$strand, gene_id = tss_filter_df$gene_id, gene_name = tss_filter_df$gene_name, gene_type = tss_filter_df$gene_type)\n",
    "tss_gr \n",
    "\n",
    "## patient data\n",
    "\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "histList = c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "countMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_NHLpatient_noChrXYM_countMat.rds\"))\n",
    "normMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_NHLpatient_noChrXYM_normMat.rds\"))\n",
    "designInfo = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_NHLpatient_noChrXYM_designInfo.rds\"))\n",
    "\n",
    "pt_countMat = countMat\n",
    "pt_countMat$H3K27me3 = countMat$H3K27me3[, cm_ind]\n",
    "pt_countMat$H3K4me2 = countMat$H3K4me2[, cm_ind]\n",
    "colnames(pt_countMat$H3K27me3) = paste0(colnames(pt_countMat$H3K27me3), \"_PT\")\n",
    "colnames(pt_countMat$H3K4me2) = paste0(colnames(pt_countMat$H3K4me2), \"_PT\")\n",
    "\n",
    "pt_designInfo = designInfo %>% dplyr::filter(exps %in% patientRaw$Xnumber[cm_ind])\n",
    "head(pt_countMat$H3K27me3)\n",
    "pt_designInfo %>% arrange(hist, exps)\n",
    "\n",
    "## healthy donor data\n",
    "countMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_countMat.rds\"))\n",
    "normMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_normMat.rds\"))\n",
    "designInfo = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_designInfo.rds\"))\n",
    "\n",
    "selectC = colnames(countMat$H3K27me3) %>% stringr::str_detect(\"CM_Product\")\n",
    "head(countMat$H3K4me2[, selectC])\n",
    "\n",
    "hd_countMat = countMat\n",
    "hd_countMat$H3K27me3 = countMat$H3K27me3[, selectC]\n",
    "hd_countMat$H3K4me2 = countMat$H3K4me2[, selectC]\n",
    "\n",
    "hd_designInfo = designInfo[which(stringr::str_detect(designInfo$exps, \"CM_Product\")), ]\n",
    "hd_designInfo\n",
    "\n",
    "for(hist in histList){\n",
    "    tmp = c()\n",
    "\n",
    "    for(j in 1:ncol(pt_countMat[[hist]])){\n",
    "        pt_designInfo_tmp = pt_designInfo %>% dplyr::filter(hist == hist)\n",
    "        tmp = cbind(tmp, round(pt_countMat[[hist]][, j]/pt_designInfo_tmp$depth[j] * 13240732)) ## remove sequencing depth effect\n",
    "    }\n",
    "    for(j in 1:ncol(hd_countMat[[hist]])){\n",
    "        hd_designInfo_tmp = hd_designInfo %>% dplyr::filter(hist == hist)\n",
    "        tmp = cbind(tmp, round(hd_countMat[[hist]][, j]/hd_designInfo_tmp$depth[j] * 13240732)) ## remove sequencing depth effect\n",
    "    }\n",
    "    countMat[[hist]] = tmp\n",
    "    colnames(countMat[[hist]]) = c(colnames(pt_countMat[[hist]]), colnames(hd_countMat[[hist]]))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=22, repr.plot.height=10)\n",
    "hist = \"H3K27me3\"\n",
    "selectR = which(rowSums(countMat[[hist]]) > 50) ## remove low count genes\n",
    "\n",
    "tmp = countMat[[hist]][selectR, ]\n",
    "\n",
    "k27_obj = CreateSeuratObject(counts = countMat[[hist]][selectR, ], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR)) %>% \n",
    "ScaleData()\n",
    "k27_obj = RunPCA(k27_obj, features = VariableFeatures(object = k27_obj), npcs = 15)\n",
    "k27_plot = Embeddings(k27_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "hist = \"H3K4me2\"\n",
    "selectR = which(rowSums(countMat[[hist]]) > 1000) ## remove low count genes\n",
    "length(selectR)\n",
    "k4_obj = CreateSeuratObject(counts = countMat[[hist]][selectR, ], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR)) %>% \n",
    "ScaleData()\n",
    "k4_obj = RunPCA(k4_obj, features = VariableFeatures(object = k4_obj), npcs = 15)\n",
    "\n",
    "k4_plot = Embeddings(k4_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "cr_obj = CreateSeuratObject(counts = rbind(tmp, countMat[[hist]][selectR, ]), project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR) + nrow(tmp)) %>% \n",
    "ScaleData()\n",
    "cr_obj = RunPCA(cr_obj, features = VariableFeatures(object = cr_obj), npcs = 15)\n",
    "cr_plot = Embeddings(cr_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3 + H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "\n",
    "pdf(\"CART_CUTRUN_Project/results/paper_figure/fig1/CMpatient_CMhealthyDonor_PCA_on_all_TSSneighbors.pdf\", width = 15, height = 6)\n",
    "ggarrange(k27_plot, k4_plot, cr_plot, common.legend = TRUE, ncol = 3, nrow = 1)\n",
    "dev.off()\n",
    "ggarrange(k27_plot, k4_plot, cr_plot, common.legend = TRUE, ncol = 3, nrow = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "pdf(\"CART_CUTRUN_Project/results/paper_figure/fig1/CMpatient_CMhealthyDonor_PCA_on_all_TSSneighbors_RNA_K27_K4.pdf\", width = 20, height = 6)\n",
    "ggarrange(rna_plot, k27_plot, k4_plot, cr_plot, common.legend = TRUE, ncol = 4, nrow = 1)\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=21, repr.plot.height=6)\n",
    "## select cordant and discordant genes\n",
    "\n",
    "# gene_category = \"RNA+K4+K27-_CE\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"+\", K27_group == \"-\", RNA_group == \"+\") %$% gene_name\n",
    "\n",
    "# gene_category = \"RNA-K4+K27-_DCNE\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"+\", K27_group == \"-\", RNA_group == \"-\") %$% gene_name\n",
    "\n",
    "# gene_category = \"RNA+K4-K27+_DE\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"-\", K27_group == \"+\", RNA_group == \"+\") %$% gene_name\n",
    "\n",
    "# gene_category = \"RNA-K4-K27+_CNE\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"-\", K27_group == \"+\", RNA_group == \"-\") %$% gene_name\n",
    "\n",
    "# gene_category = \"RNA+K4+K27+_BE\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"+\", K27_group == \"+\", RNA_group == \"+\") %$% gene_name\n",
    "\n",
    "# gene_category = \"RNA-K4+K27+_BNE\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"+\", K27_group == \"+\", RNA_group == \"-\") %$% gene_name\n",
    "\n",
    "# gene_category = \"RNA+K4-K27-_NME\"\n",
    "# target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"-\", K27_group == \"-\", RNA_group == \"+\") %$% gene_name\n",
    "\n",
    "gene_category = \"RNA-K4-K27-_NMNE\"\n",
    "target_gene = cr_rna_full %>% dplyr::filter(K4_group == \"-\", K27_group == \"-\", RNA_group == \"-\") %$% gene_name\n",
    "\n",
    "rna_gene_name_list = rownames(countMat_rna) %>% gsub(\".*_\", \"\", .) %>% tolower \n",
    "rna_row_index = which(rna_gene_name_list %in% target_gene)\n",
    "\n",
    "cr_gene_name_list = rownames(countMat[[\"H3K4me2\"]]) %>% tolower\n",
    "cr_row_index = which(cr_gene_name_list %in% target_gene)\n",
    "\n",
    "\n",
    "selectR = rna_row_index\n",
    "length(selectR)\n",
    "\n",
    "rna_obj = CreateSeuratObject(counts = countMat_rna[selectR, ], project = \"cart\", min.cells = 0, min.features = 10) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR)) %>%  # length(selectR)\n",
    "ScaleData()\n",
    "\n",
    "rna_obj = RunPCA(rna_obj, features = VariableFeatures(object = rna_obj), npcs = 10)\n",
    "rna_plot = Embeddings(rna_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", ncol(pt_expCount) - 1), rep(\"HD\", ncol(hd_expCount) - 1))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"RNA-seq\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "\n",
    "hist = \"H3K27me3\"\n",
    "selectR = cr_row_index\n",
    "\n",
    "tmp = countMat[[hist]][selectR, ]\n",
    "\n",
    "k27_obj = CreateSeuratObject(counts = countMat[[hist]][selectR, ], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR)) %>% \n",
    "ScaleData()\n",
    "k27_obj = RunPCA(k27_obj, features = VariableFeatures(object = k27_obj), npcs = 15)\n",
    "k27_plot = Embeddings(k27_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "hist = \"H3K4me2\"\n",
    "selectR = cr_row_index\n",
    "length(selectR)\n",
    "k4_obj = CreateSeuratObject(counts = countMat[[hist]][selectR, ], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR)) %>% \n",
    "ScaleData()\n",
    "k4_obj = RunPCA(k4_obj, features = VariableFeatures(object = k4_obj), npcs = 15)\n",
    "\n",
    "k4_plot = Embeddings(k4_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "cr_obj = CreateSeuratObject(counts = rbind(tmp, countMat[[hist]][selectR, ]), project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR) + nrow(tmp)) %>% \n",
    "ScaleData()\n",
    "cr_obj = RunPCA(cr_obj, features = VariableFeatures(object = cr_obj), npcs = 15)\n",
    "cr_plot = Embeddings(cr_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3 + H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "\n",
    "ggarrange(rna_plot, k27_plot, k4_plot, cr_plot, common.legend = TRUE, ncol = 4, nrow = 1)\n",
    "ggsave(paste0(\"CART_CUTRUN_Project/results/paper_figure/fig1/CMpatient_CMhealthyDonor_PCA_on_all_TSSneighbors_RNA_K27_K4_\", gene_category, \".pdf\"), width = 20, height = 6)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Bivalent genes\n",
    "## CUTRUN DE and assign to nearest genes\n",
    "\n",
    "library(pacman)\n",
    "p_load(data.table, dplyr, ggplot2, viridis, magrittr, VennDiagram, ggpubr, limma, edgeR, tidyr, GenomicRanges, RColorBrewer, pheatmap, Seurat, fgsea, GSEABase, limma, TxDb.Hsapiens.UCSC.hg38.knownGene, org.Hs.eg.db)\n",
    "\n",
    "cell_list <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hd_list <- paste0(\"HD\", c(1:3, 5:7))\n",
    "expr_list = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "cell_comp_list = c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "hist_list <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "in_path <- \"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "out_path <- \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/\"\n",
    "fig_path <- \"CART_CUTRUN_Project/results/paper_figure/\"\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "\n",
    "## Bivalent methylation marks\n",
    "\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/IDR_master_peak_list_chromVar_count_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\"))\n",
    "\n",
    "names(peakAll)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "## overlap K4me2 and K27me3 ensuring at least 50bp overlap, count number of regions after merging\n",
    "fig_path = \"CART_CUTRUN_Project/results/paper_figure/fig1\"\n",
    "bi_path <- \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/BivalentGenes/\"\n",
    "gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")\n",
    "\n",
    "get_nearest_gene <- function(gene_gr, region_gr){\n",
    "    region_assign_gene <- distanceToNearest(region_gr, gene_gr, select = \"all\")\n",
    "    # dist_tmp <- region_assign_gene %>% data.frame %$% distance\n",
    "    # region_assign_gene <- region_assign_gene[dist_tmp < 1000]\n",
    "    region_match_gene <- cbind(data.frame(region_gr)[region_assign_gene@from, ], data.frame(peak_id = region_assign_gene@from, gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[region_assign_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) )\n",
    "    return(region_match_gene)\n",
    "\n",
    "}\n",
    "bi_num <- c()\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    for(pt in paste0(\"HD\", 4:7)){\n",
    "        bi_olp <- findOverlaps(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]], minoverlap = 50)\n",
    "        bi_peak.gr <- reduce(append(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]][bi_olp@from], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]][bi_olp@to]))\n",
    "        bi_gene <- get_nearest_gene(gene_gr, bi_peak.gr)\n",
    "        write.table(unique(bi_gene$gene_name), paste0(bi_path, \"Bivalent_peaks_nearest_genes_\", cT, \"_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "\n",
    "        total_peak.gr <- append(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]]) %>% reduce\n",
    "        total_gene <- get_nearest_gene(gene_gr, total_peak.gr)\n",
    "\n",
    "        bi_num <- data.frame(\n",
    "            bi_peak_num = length(bi_peak.gr),\n",
    "            bi_gene_num = length(unique(bi_gene$gene_name)),\n",
    "            total_peak_num = length(total_peak.gr),\n",
    "            total_gene_num = length(unique(total_gene$gene_name)),\n",
    "            cT = cT,\n",
    "            pt = pt\n",
    "        ) %>% rbind(bi_num, .)\n",
    "    }\n",
    "}\n",
    "bi_num$cT <- factor(bi_num$cT, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\"))\n",
    "\n",
    "# p1 = bi_num %>% ggplot(aes(x = pt, y = bi_peak_num, fill = cT)) + \n",
    "# geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "# scale_fill_viridis(discrete = TRUE) +\n",
    "# theme_bw(base_size = 20) +\n",
    "# xlab(\"\") +\n",
    "# ylab(\"# of Peak Regions\") +\n",
    "# rremove(\"legend.title\")\n",
    "\n",
    "# p2 = bi_num %>% mutate(bi_peak_prop = bi_peak_num/total_peak_num * 100) %>% ggplot(aes(x = pt, y = bi_peak_prop, fill = cT)) + \n",
    "# geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "# scale_fill_viridis(discrete = TRUE) +\n",
    "# theme_bw(base_size = 20) +\n",
    "# xlab(\"\") +\n",
    "# ylab(\"% of Peak Regions\") +\n",
    "# rremove(\"legend.title\")\n",
    "\n",
    "# p3 = bi_num %>% ggplot(aes(x = pt, y = bi_gene_num, fill = cT)) + \n",
    "# geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "# scale_fill_viridis(discrete = TRUE) +\n",
    "# theme_bw(base_size = 20) +\n",
    "# xlab(\"\") +\n",
    "# ylab(\"# of Genes\") +\n",
    "# rremove(\"legend.title\")\n",
    "\n",
    "# p4 = bi_num %>% mutate(bi_gene_prop = bi_gene_num/total_gene_num * 100) %>% ggplot(aes(x = pt, y = bi_gene_prop, fill = cT)) + \n",
    "# geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "# scale_fill_viridis(discrete = TRUE) +\n",
    "# theme_bw(base_size = 20) +\n",
    "# xlab(\"\") +\n",
    "# ylab(\"% of Genes\") +\n",
    "# rremove(\"legend.title\")\n",
    "\n",
    "# ggarrange(p1, p2, p3, p4, ncol = 2, nrow = 2, common.legend = TRUE) \n",
    "\n",
    "\n",
    "b1 = bi_num %>% filter(cT != \"EMRA\") %>% ggplot(aes(x = cT, y = bi_peak_num, fill = cT)) + \n",
    "geom_boxplot() +\n",
    "scale_fill_viridis(discrete = TRUE, begin = 0, end = 0.75, alpha = 0.8) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Peak Regions\") +\n",
    "rremove(\"legend\")\n",
    "\n",
    "b2 = bi_num %>% filter(cT != \"EMRA\") %>% mutate(bi_peak_prop = bi_peak_num/total_peak_num * 100) %>% ggplot(aes(x = cT, y = bi_peak_prop, fill = cT)) + \n",
    "geom_boxplot() +\n",
    "scale_fill_viridis(discrete = TRUE, begin = 0, end = 0.75, alpha = 0.8) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Peak Regions\") +\n",
    "rremove(\"legend\")\n",
    "\n",
    "b3 = bi_num %>% filter(cT != \"EMRA\") %>% ggplot() + \n",
    "geom_boxplot(aes(x = cT, y = bi_gene_num, fill = cT)) +\n",
    "scale_fill_viridis(discrete = TRUE, begin = 0, end = 0.75, alpha = 0.8) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Genes\") +\n",
    "rremove(\"legend\")\n",
    "\n",
    "\n",
    "b4 = bi_num %>% filter(cT != \"EMRA\") %>% mutate(bi_gene_prop = bi_gene_num/total_gene_num * 100) %>% ggplot() + \n",
    "geom_boxplot(aes(x = cT, y = bi_gene_prop, fill = cT)) +\n",
    "scale_fill_viridis(discrete = TRUE, begin = 0, end = 0.75, alpha = 0.8) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rremove(\"legend\") \n",
    "\n",
    "ggarrange(b1, b2, b3, b4, ncol = 4, nrow = 1)\n",
    "ggsave(paste0(fig_path, \"/bivalent_peak_gene_num_prop_boxplot.pdf\"), width = 28, height = 7)\n",
    "\n",
    "\n",
    "# pdf(paste0(fig_path, \"/bivalent_peak_gene_num_prop.pdf\"), width = 10, height = 10)\n",
    "# ggarrange(p1, p2, p3, p4, ncol = 2, nrow = 2, common.legend = TRUE) \n",
    "# dev.off()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "load(file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition.RData\"))\n",
    "rna_results <- results\n",
    "\n",
    "bi_change <- c()\n",
    "bi_change_gene_lfc <- c()\n",
    "ct_list <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "pt_list <- paste0(\"HD\", 4:7)\n",
    "for(i in 1:length(ct_list)){\n",
    "    cT <- ct_list[i]\n",
    "    cT1 <- cT\n",
    "    for(j in 1:length(ct_list)){\n",
    "        if(i != j){\n",
    "            cT2 <- ct_list[j]\n",
    "            for(pt in pt_list){\n",
    "                bi_olp <- findOverlaps(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]], minoverlap = 50)\n",
    "                bi_olp2 <- findOverlaps(peakAll[[paste0(\"H3K27me3_\", cT2, \"_Input_\", pt)]], peakAll[[paste0(\"H3K4me2_\", cT2, \"_Input_\", pt)]], minoverlap = 50)\n",
    "\n",
    "                bi_peak.gr <- reduce(append(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]][bi_olp@from], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]][bi_olp@to]))\n",
    "                bi_peak.gr2 <- reduce(append(peakAll[[paste0(\"H3K27me3_\", cT2, \"_Input_\", pt)]][bi_olp2@from], peakAll[[paste0(\"H3K4me2_\", cT2, \"_Input_\", pt)]][bi_olp2@to]))\n",
    "\n",
    "                bi_again <- findOverlaps(bi_peak.gr, bi_peak.gr2, minoverlap = 50)@from %>% unique\n",
    "                bi_k27 <- findOverlaps(bi_peak.gr, peakAll[[paste0(\"H3K27me3_\", cT2, \"_Input_\", pt)]], minoverlap = 50)@from %>% unique\n",
    "                bi_k4 <- findOverlaps(bi_peak.gr, peakAll[[paste0(\"H3K4me2_\", cT2, \"_Input_\", pt)]], minoverlap = 50)@from %>% unique\n",
    "                bi_k27 <- setdiff(setdiff(bi_k27, bi_k4), bi_again)\n",
    "                bi_k4 <- setdiff(setdiff(bi_k4, bi_k27), bi_again)\n",
    "                bi_none <- setdiff(1:length(bi_peak.gr), unique(c(bi_again, bi_k27, bi_k4)) %>% sort)\n",
    "\n",
    "                bi_gene <- get_nearest_gene(gene_gr, bi_peak.gr)\n",
    "                bi_again_gene <- bi_gene %>% filter(peak_id %in% bi_again) %$% gene_name %>% unique\n",
    "                bi_k27_gene <- bi_gene %>% filter(peak_id %in% bi_k27) %$% gene_name %>% unique\n",
    "                bi_k4_gene <- bi_gene %>% filter(peak_id %in% bi_k4) %$% gene_name %>% unique\n",
    "                bi_none_gene <- setdiff(unique(bi_gene$gene_name), c(bi_again_gene, bi_k27_gene, bi_k4_gene)) %>% unique\n",
    "                write.table(bi_again_gene, paste0(bi_path, \"BivalentPeaksOf\", cT1, \"_BivalentPeakAgainIn\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                write.table(bi_k27_gene, paste0(bi_path, \"BivalentPeaksOf\", cT1, \"_H3K27me3In\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                write.table(bi_k4_gene, paste0(bi_path, \"BivalentPeaksOf\", cT1, \"_H3K4me2In\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                write.table(bi_none_gene, paste0(bi_path, \"BivalentPeaksOf\", cT1, \"_NoPeakIn\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                \n",
    "                if(i < j){\n",
    "                    bi_again_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% bi_again_gene)\n",
    "                    bi_k27_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% bi_k27_gene)\n",
    "                    bi_k4_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% bi_k4_gene)\n",
    "                    bi_none_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% bi_none_gene)\n",
    "                }else{\n",
    "                    bi_again_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$logFC) %>% filter(gene_name %in% bi_again_gene)\n",
    "                    bi_k27_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$logFC) %>% filter(gene_name %in% bi_k27_gene)\n",
    "                    bi_k4_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$logFC) %>% filter(gene_name %in% bi_k4_gene)\n",
    "                    bi_none_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$logFC) %>% filter(gene_name %in% bi_none_gene)\n",
    "\n",
    "                }\n",
    "                \n",
    "                bi_change <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    bi_original_num = length(bi_peak.gr), bi_again_num = length(bi_again), bi_k27_num = length(bi_k27), bi_k4_num = length(bi_k4), bi_none_num = length(bi_none),\n",
    "                    bi_original_gene_num = length(unique(bi_gene$gene_name)), bi_again_gene_num = length(bi_again_gene), bi_k27_gene_num = length(bi_k27_gene), bi_k4_gene_num = length(bi_k4_gene), bi_none_gene_num = length(bi_none_gene)) %>% \n",
    "                    mutate(bi_again_prop = bi_again_num/bi_original_num * 100, bi_k27_prop = bi_k27_num/bi_original_num * 100, bi_k4_prop = bi_k4_num/bi_original_num * 100, bi_none_prop = bi_none_num/bi_original_num * 100) %>% \n",
    "                    mutate(bi_again_gene_prop = bi_again_gene_num/bi_original_gene_num * 100, bi_k27_gene_prop = bi_k27_gene_num/bi_original_gene_num * 100, bi_k4_gene_prop = bi_k4_gene_num/bi_original_gene_num * 100, bi_none_gene_prop = bi_none_gene_num/bi_original_gene_num * 100) %>% \n",
    "                    rbind(bi_change, .)\n",
    "                \n",
    "                bi_change_gene_lfc <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    logFC = bi_again_gene_lfc$GEX, gene_name = bi_again_gene_lfc$gene_name, type = \"bi_again\"\n",
    "                ) %>% rbind(bi_change_gene_lfc, .)\n",
    "                bi_change_gene_lfc <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    logFC = bi_k27_gene_lfc$GEX, gene_name = bi_k27_gene_lfc$gene_name, type = \"bi_k27\"\n",
    "                ) %>% rbind(bi_change_gene_lfc, .)\n",
    "                bi_change_gene_lfc <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    logFC = bi_k4_gene_lfc$GEX, gene_name = bi_k4_gene_lfc$gene_name, type = \"bi_k4\"\n",
    "                ) %>% rbind(bi_change_gene_lfc, .)\n",
    "                if(nrow(bi_none_gene_lfc) > 0){\n",
    "                    bi_change_gene_lfc <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    logFC = bi_none_gene_lfc$GEX, gene_name = bi_none_gene_lfc$gene_name, type = \"bi_none\"\n",
    "                ) %>% rbind(bi_change_gene_lfc, .)\n",
    "                }\n",
    "                \n",
    "            }\n",
    "        }\n",
    "        \n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "bi_change %>% head\n",
    "bi_change_gene_lfc %>% head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(tidyr)\n",
    "bi_change$cT <- factor(bi_change$cT, levels = ct_list)\n",
    "bi_change$cT_compare <- factor(bi_change$cT_compare, levels = ct_list)\n",
    "\n",
    "\n",
    "p1 = bi_change %>% pivot_longer(., cols = c(\"bi_again_prop\", \"bi_k27_prop\", \"bi_k4_prop\", \"bi_none_prop\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%\n",
    "filter(cT != \"EMRA\", cT_compare != \"EMRA\", name != \"bi_none_prop\") %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = factor(name, levels = c(\"bi_again_prop\", \"bi_k27_prop\", \"bi_k4_prop\"), labels = c(\"Bivalent->again\", \"Bivalent->H3K27me3\", \"Bivalent->H3K4me2\")), label = round(value, 1))) +\n",
    "geom_boxplot(position = position_dodge(width = 0.9)) +\n",
    "stat_summary(\n",
    "    fun.y = median,\n",
    "    geom = 'line',\n",
    "    aes(group = name, colour = name),\n",
    "    position = position_dodge(width = 0.9) #this has to be added\n",
    "  ) +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "scale_color_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Peaks\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "p12 = bi_change %>% pivot_longer(., cols = c(\"bi_again_prop\", \"bi_k27_prop\", \"bi_k4_prop\", \"bi_none_prop\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%\n",
    "filter(cT != \"EMRA\", cT_compare != \"EMRA\", name != \"bi_none_prop\") %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = factor(name, levels = c(\"bi_again_prop\", \"bi_k27_prop\", \"bi_k4_prop\"), labels = c(\"Bivalent->again\", \"Bivalent->H3K27me3\", \"Bivalent->H3K4me2\")), label = round(value, 1))) +\n",
    "geom_boxplot(position = position_dodge(width = 0.9)) +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "scale_color_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Peaks\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "p2 = bi_change %>% pivot_longer(., cols = c(\"bi_again_gene_prop\", \"bi_k27_gene_prop\", \"bi_k4_gene_prop\", \"bi_none_gene_prop\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%\n",
    "filter(cT != \"EMRA\", cT_compare != \"EMRA\", name != \"bi_none_gene_prop\") %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = factor(name, levels = c(\"bi_again_gene_prop\", \"bi_k27_gene_prop\", \"bi_k4_gene_prop\"), labels = c(\"Bivalent->again\", \"Bivalent->H3K27me3\", \"Bivalent->H3K4me2\")), label = round(value, 1))) +\n",
    "geom_boxplot(position = position_dodge(width = 0.9)) +\n",
    "stat_summary(\n",
    "    fun.y = median,\n",
    "    geom = 'line',\n",
    "    aes(group = name, colour = name),\n",
    "    position = position_dodge(width = 0.9) #this has to be added\n",
    "  ) +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "scale_color_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "p22 = bi_change %>% pivot_longer(., cols = c(\"bi_again_gene_prop\", \"bi_k27_gene_prop\", \"bi_k4_gene_prop\", \"bi_none_gene_prop\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%\n",
    "filter(cT != \"EMRA\", cT_compare != \"EMRA\", name != \"bi_none_gene_prop\") %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = factor(name, levels = c(\"bi_again_gene_prop\", \"bi_k27_gene_prop\", \"bi_k4_gene_prop\"), labels = c(\"Bivalent->again\", \"Bivalent->H3K27me3\", \"Bivalent->H3K4me2\")), label = round(value, 1))) +\n",
    "geom_boxplot(position = position_dodge(width = 0.9)) +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "scale_color_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "pdf(paste0(fig_path, \"/bivalent_peak_gene_change_between_celltypes.pdf\"), width = 10, height = 8)\n",
    "p1\n",
    "p12\n",
    "p2\n",
    "p22\n",
    "dev.off()\n",
    "\n",
    "ggarrange(p1, p2, nrow = 2, ncol = 1, common.legend = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "bi_change_gene_lfc %>% head\n",
    "bi_change_gene_lfc$cT <- factor(bi_change_gene_lfc$cT, levels = ct_list)\n",
    "bi_change_gene_lfc$cT_compare <- factor(bi_change_gene_lfc$cT_compare, levels = ct_list)\n",
    "\n",
    "bi_change_gene_lfc %>% filter(cT != \"EMRA\", cT_compare != \"EMRA\", type != \"bi_none\") %>%\n",
    "ggplot(aes(cT_compare, y = logFC, fill = factor(type, levels = c(\"bi_again\", \"bi_k27\", \"bi_k4\"), labels = c(\"Bivalent->again\", \"Bivalent->H3K27me3\", \"Bivalent->H3K4me2\")))) +\n",
    "geom_boxplot() +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "xlab(\"\") +\n",
    "ylab(\"logFC of GEX\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "\n",
    "ggsave(paste0(fig_path, \"/bivalent_peak_gene_change_between_celltypes_gex_logfc.pdf\"), width = 10, height = 8)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cordant_gene_df = c()\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    rna_select_cond = paste0(\"Input_\", cT)\n",
    "    cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "    rna_select_data = rna_expCount\n",
    "    cr_select_data = normMat$H3K27me3\n",
    "\n",
    "    cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "    rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "    k27_data = log2(normMat$H3K27me3[, cr_col_index] + 1) %>% rowMeans %>% t %>% t %>% data.frame\n",
    "    colnames(k27_data) = c(\"K27\")\n",
    "    k27_data$gene_name = rownames(k27_data) %>% tolower\n",
    "\n",
    "    k4_data = log2(normMat$H3K4me2[, cr_col_index] + 1) %>% rowMeans %>% t %>%  t %>% data.frame\n",
    "    colnames(k4_data) = c(\"K4\")\n",
    "    k4_data$gene_name = rownames(k4_data) %>% tolower\n",
    "\n",
    "    rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "    RNA_data = log2(data.frame(rna_select_data)[, rna_col_index] + 1) %>% rowMeans %>% t %>%  t %>% data.frame\n",
    "    colnames(RNA_data) = c(\"RNA\")\n",
    "    RNA_data$gene_name = rna_gene\n",
    "\n",
    "    cr_data = cbind(k27_data, K4 = k4_data$K4)\n",
    "    cr_rna_full = full_join(RNA_data, cr_data, by = \"gene_name\") %>% dplyr::select(gene_name, K4, K27, RNA)\n",
    "    cr_rna_full[is.na(cr_rna_full)] = 0\n",
    "\n",
    "    ## PCA of cordant or discordant genes and see how well the cell types are separated\n",
    "    cr_rna_full$K4_group = \"-\"\n",
    "    cr_rna_full$K4_group[which(cr_rna_full$K4 > 4.5)] = \"+\"\n",
    "    cr_rna_full$K27_group = \"-\"\n",
    "    cr_rna_full$K27_group[which(cr_rna_full$K27 > 2.5)] = \"+\"\n",
    "    cr_rna_full$RNA_group = \"-\"\n",
    "    cr_rna_full$RNA_group[which(cr_rna_full$RNA > 2.5)] = \"+\"\n",
    "\n",
    "    cr_rna_full = cr_rna_full %>% mutate(gene_type = paste0(\"RNA\", RNA_group, \"K4\", K4_group, \"K27\", K27_group)) %>% data.frame\n",
    "        \n",
    "\n",
    "    tmp = cr_rna_full %>% group_by(gene_type) %>% summarize(N = n()) %>% \n",
    "        mutate(cell_type = cT)\n",
    "\n",
    "\n",
    "    cordant_gene_df = rbind(cordant_gene_df, left_join(cr_rna_full %>% mutate(gene_name_index = paste0(gene_name, \"_\", 1:nrow(cr_rna_full))), tmp, by = \"gene_type\"))\n",
    "}\n",
    "cordant_gene_df$gene_type_name = factor(cordant_gene_df$gene_type, levels = c(\"RNA+K4-K27-\", \"RNA-K4-K27-\", \"RNA+K4+K27-\", \"RNA-K4+K27+\", \n",
    "\"RNA+K4-K27+\", \"RNA-K4+K27-\", \"RNA-K4-K27+\", \"RNA+K4+K27+\"), labels = c(\"RNA+K4-K27-(NME)\", \"RNA-K4-K27-(NMNE)\", \"RNA+K4+K27-(CE)\", \"RNA-K4+K27+(BNE)\", \n",
    "\"RNA+K4-K27+(DE)\", \"RNA-K4+K27-(DCNE)\", \"RNA-K4-K27+(CNE)\", \"RNA+K4+K27+(BE)\"))\n",
    "\n",
    "cordant_gene_df$cell_type = factor(cordant_gene_df$cell_type, levels = c(\"N\", \"CM\", \"EM\"))\n",
    "cordant_gene_df$gene_num = 1\n",
    "head(cordant_gene_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cordant_gene_df %>% head\n",
    "cordant_gene_df %>% write.table(\"CART_CUTRUN_Project/results/paper_figure/fig1/concordant_discordant_summary_N_CM_EM_EMRA.csv\", quote = F, row.names = F, sep = \",\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "target_start_gene = cordant_gene_df %>% dplyr::filter(cell_type == \"N\", gene_type %in% c(\"RNA-K4+K27+\", \"RNA+K4+K27+\")) %$% gene_name_index\n",
    "head(target_start_gene)\n",
    "length(target_start_gene)\n",
    "length(unique(target_start_gene))\n",
    "target_start_gene_cm = cordant_gene_df %>% dplyr::filter(cell_type == \"CM\", gene_type %in% c(\"RNA-K4+K27+\", \"RNA+K4+K27+\")) %$% gene_name_index\n",
    "head(target_start_gene_cm)\n",
    "length(target_start_gene_cm)\n",
    "length(unique(target_start_gene_cm))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "target_start_gene = cordant_gene_df %>% dplyr::filter(cell_type == \"N\", gene_type %in% c(\"RNA-K4+K27+\", \"RNA+K4+K27+\")) %$% gene_name\n",
    "head(target_start_gene)\n",
    "length(target_start_gene)\n",
    "length(unique(target_start_gene))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=5)\n",
    "pdf(paste0(fig_path, \"/FlowChart_allGenes.pdf\"), width =11, height = 11)\n",
    "ggplot(cordant_gene_df %>% filter(cell_type != \"EMRA\"),\n",
    "       aes(x = cell_type, stratum = gene_type_name, alluvium = gene_name_index,\n",
    "           y = gene_num,\n",
    "           fill = gene_type_name, label = N)) +\n",
    "  scale_x_discrete(expand = c(.1, .1)) +\n",
    "  geom_flow() +\n",
    "  geom_stratum(alpha = .5) +\n",
    "  geom_fit_text(stat = \"stratum\", size = 10, min.size = 6, formatter = comma) +\n",
    "  theme(legend.position = \"bottom\") +\n",
    "  ggtitle(\"\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=5)\n",
    "pdf(paste0(fig_path, \"/FlowChart_BE_BNE.pdf\"), width =9, height = 9)\n",
    "ggplot(cordant_gene_df %>% filter(cell_type != \"EMRA\", gene_name_index %in% target_start_gene),\n",
    "       aes(x = cell_type, stratum = gene_type_name, alluvium = gene_name_index,\n",
    "           y = gene_num,\n",
    "           fill = gene_type_name, label = N)) +\n",
    "  scale_x_discrete(expand = c(.1, .1)) +\n",
    "  geom_flow() +\n",
    "  geom_stratum(alpha = .5) +\n",
    "  geom_fit_text(stat = \"stratum\", size = 10, min.size = 6, formatter = comma) +\n",
    "  theme(legend.position = \"bottom\") +\n",
    "  ggtitle(\"\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## target gene for N\n",
    "\n",
    "cordant_gene_df = c()\n",
    "for(cT in c(\"N\", \"CM\", \"EM\")){\n",
    "    rna_select_cond = paste0(\"Input_\", cT)\n",
    "    cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "    rna_select_data = rna_expCount\n",
    "    cr_select_data = normMat$H3K27me3\n",
    "\n",
    "    cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "    rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "    k27_data = log2(normMat$H3K27me3[, cr_col_index] + 1) %>% rowMeans %>% t %>% t %>% data.frame\n",
    "    colnames(k27_data) = c(\"K27\")\n",
    "    k27_data$gene_name = rownames(k27_data) %>% tolower\n",
    "\n",
    "    k4_data = log2(normMat$H3K4me2[, cr_col_index] + 1) %>% rowMeans %>% t %>%  t %>% data.frame\n",
    "    colnames(k4_data) = c(\"K4\")\n",
    "    k4_data$gene_name = rownames(k4_data) %>% tolower\n",
    "\n",
    "    rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "    RNA_data = log2(data.frame(rna_select_data)[, rna_col_index] + 1) %>% rowMeans %>% t %>%  t %>% data.frame\n",
    "    colnames(RNA_data) = c(\"RNA\")\n",
    "    RNA_data$gene_name = rna_gene\n",
    "\n",
    "    cr_data = cbind(k27_data, K4 = k4_data$K4)\n",
    "    cr_rna_full = full_join(RNA_data, cr_data, by = \"gene_name\") %>% dplyr::select(gene_name, K4, K27, RNA)\n",
    "    cr_rna_full[is.na(cr_rna_full)] = 0\n",
    "\n",
    "    ## PCA of cordant or discordant genes and see how well the cell types are separated\n",
    "    cr_rna_full$K4_group = \"-\"\n",
    "    cr_rna_full$K4_group[which(cr_rna_full$K4 > 4.5)] = \"+\"\n",
    "    cr_rna_full$K27_group = \"-\"\n",
    "    cr_rna_full$K27_group[which(cr_rna_full$K27 > 2.5)] = \"+\"\n",
    "    cr_rna_full$RNA_group = \"-\"\n",
    "    cr_rna_full$RNA_group[which(cr_rna_full$RNA > 2.5)] = \"+\"\n",
    "\n",
    "    cr_rna_full = cr_rna_full %>% mutate(gene_type = paste0(\"RNA\", RNA_group, \"K4\", K4_group, \"K27\", K27_group)) %>% \n",
    "    mutate(gene_name_index = paste0(gene_name, \"_\", 1:nrow(cr_rna_full))) %>% data.frame %>% dplyr::filter(gene_name_index %in% target_start_gene)\n",
    "    print(dim(cr_rna_full))\n",
    "    tmp = cr_rna_full %>% group_by(gene_type) %>% summarize(N = n()) %>% \n",
    "        mutate(cell_type = cT)\n",
    "    print(dim(tmp))\n",
    "    print(tmp$N %>% sum)\n",
    "    cordant_gene_df = rbind(cordant_gene_df, left_join(cr_rna_full, tmp, by = \"gene_type\"))\n",
    "    print(dim(left_join(cr_rna_full, tmp, by = \"gene_type\")))\n",
    "}\n",
    "cordant_gene_df$gene_type_name = factor(cordant_gene_df$gene_type, levels = c(\"RNA+K4-K27-\", \"RNA-K4-K27-\", \"RNA+K4+K27-\", \"RNA-K4+K27+\", \n",
    "\"RNA+K4-K27+\", \"RNA-K4+K27-\", \"RNA-K4-K27+\", \"RNA+K4+K27+\"), labels = c(\"RNA+K4-K27-(NME)\", \"RNA-K4-K27-(NMNE)\", \"RNA+K4+K27-(CE)\", \"RNA-K4+K27+(BNE)\", \n",
    "\"RNA+K4-K27+(DE)\", \"RNA-K4+K27-(DCNE)\", \"RNA-K4-K27+(CNE)\", \"RNA+K4+K27+(BE)\"))\n",
    "\n",
    "cordant_gene_df$cell_type = factor(cordant_gene_df$cell_type, levels = c(\"N\", \"CM\", \"EM\"))\n",
    "cordant_gene_df$gene_num = 1\n",
    "head(cordant_gene_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=13, repr.plot.height=10)\n",
    "pdf(paste0(fig_path, \"/FlowChart_BE_BNE_targetgeneonly_for_N.pdf\"), width =9, height = 9)\n",
    "\n",
    "ggplot(cordant_gene_df, # %>% filter(cell_type != \"EMRA\", gene_name_index %in% target_start_gene),\n",
    "       aes(x = cell_type, stratum = gene_type_name, alluvium = gene_name_index,\n",
    "           y = gene_num,\n",
    "           fill = gene_type_name, label = N)) +\n",
    "  scale_x_discrete(expand = c(.1, .1)) +\n",
    "  geom_flow() +\n",
    "  geom_stratum(alpha = .5) +\n",
    "  geom_fit_text(stat = \"stratum\", size = 10, min.size = 6, formatter = comma) +\n",
    "  theme(legend.position = \"bottom\") +\n",
    "  ggtitle(\"\") +\n",
    "  ylab(\"Number of Genes\")\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## target gene for CM\n",
    "cordant_gene_df = c()\n",
    "for(cT in c(\"N\", \"CM\", \"EM\")){\n",
    "    rna_select_cond = paste0(\"Input_\", cT)\n",
    "    cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "    rna_select_data = rna_expCount\n",
    "    cr_select_data = normMat$H3K27me3\n",
    "\n",
    "    cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "    rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "    k27_data = log2(normMat$H3K27me3[, cr_col_index] + 1) %>% rowMeans %>% t %>% t %>% data.frame\n",
    "    colnames(k27_data) = c(\"K27\")\n",
    "    k27_data$gene_name = rownames(k27_data) %>% tolower\n",
    "\n",
    "    k4_data = log2(normMat$H3K4me2[, cr_col_index] + 1) %>% rowMeans %>% t %>%  t %>% data.frame\n",
    "    colnames(k4_data) = c(\"K4\")\n",
    "    k4_data$gene_name = rownames(k4_data) %>% tolower\n",
    "\n",
    "    rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "    RNA_data = log2(data.frame(rna_select_data)[, rna_col_index] + 1) %>% rowMeans %>% t %>%  t %>% data.frame\n",
    "    colnames(RNA_data) = c(\"RNA\")\n",
    "    RNA_data$gene_name = rna_gene\n",
    "\n",
    "    cr_data = cbind(k27_data, K4 = k4_data$K4)\n",
    "    cr_rna_full = full_join(RNA_data, cr_data, by = \"gene_name\") %>% dplyr::select(gene_name, K4, K27, RNA)\n",
    "    cr_rna_full[is.na(cr_rna_full)] = 0\n",
    "\n",
    "    ## PCA of cordant or discordant genes and see how well the cell types are separated\n",
    "    cr_rna_full$K4_group = \"-\"\n",
    "    cr_rna_full$K4_group[which(cr_rna_full$K4 > 4.5)] = \"+\"\n",
    "    cr_rna_full$K27_group = \"-\"\n",
    "    cr_rna_full$K27_group[which(cr_rna_full$K27 > 2.5)] = \"+\"\n",
    "    cr_rna_full$RNA_group = \"-\"\n",
    "    cr_rna_full$RNA_group[which(cr_rna_full$RNA > 2.5)] = \"+\"\n",
    "\n",
    "    cr_rna_full = cr_rna_full %>% mutate(gene_type = paste0(\"RNA\", RNA_group, \"K4\", K4_group, \"K27\", K27_group)) %>% \n",
    "    mutate(gene_name_index = paste0(gene_name, \"_\", 1:nrow(cr_rna_full))) %>% data.frame %>% dplyr::filter(gene_name_index %in% target_start_gene_cm)\n",
    "    print(dim(cr_rna_full))\n",
    "    tmp = cr_rna_full %>% group_by(gene_type) %>% summarize(N = n()) %>% \n",
    "        mutate(cell_type = cT)\n",
    "    print(dim(tmp))\n",
    "    print(tmp$N %>% sum)\n",
    "    cordant_gene_df = rbind(cordant_gene_df, left_join(cr_rna_full, tmp, by = \"gene_type\"))\n",
    "    print(dim(left_join(cr_rna_full, tmp, by = \"gene_type\")))\n",
    "}\n",
    "cordant_gene_df$gene_type_name = factor(cordant_gene_df$gene_type, levels = c(\"RNA+K4-K27-\", \"RNA-K4-K27-\", \"RNA+K4+K27-\", \"RNA-K4+K27+\", \n",
    "\"RNA+K4-K27+\", \"RNA-K4+K27-\", \"RNA-K4-K27+\", \"RNA+K4+K27+\"), labels = c(\"RNA+K4-K27-(NME)\", \"RNA-K4-K27-(NMNE)\", \"RNA+K4+K27-(CE)\", \"RNA-K4+K27+(BNE)\", \n",
    "\"RNA+K4-K27+(DE)\", \"RNA-K4+K27-(DCNE)\", \"RNA-K4-K27+(CNE)\", \"RNA+K4+K27+(BE)\"))\n",
    "\n",
    "cordant_gene_df$cell_type = factor(cordant_gene_df$cell_type, levels = c(\"CM\", \"N\", \"EM\"))\n",
    "cordant_gene_df$gene_num = 1\n",
    "head(cordant_gene_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=13, repr.plot.height=10)\n",
    "# pdf(paste0(fig_path, \"/FlowChart_BE_BNE_targetgeneonly_forCM.pdf\"), width =9, height = 9)\n",
    "\n",
    "ggplot(cordant_gene_df %>% filter(cell_type != \"N\"),\n",
    "       aes(x = cell_type, stratum = gene_type_name, alluvium = gene_name_index,\n",
    "           y = gene_num,\n",
    "           fill = gene_type_name, label = N)) +\n",
    "  scale_x_discrete(expand = c(.1, .1)) +\n",
    "  geom_flow() +\n",
    "  geom_stratum(alpha = .5) +\n",
    "  geom_fit_text(stat = \"stratum\", size = 10, min.size = 6, formatter = comma) +\n",
    "  theme(legend.position = \"bottom\") +\n",
    "  ggtitle(\"\") +\n",
    "  ylab(\"Number of Genes\")\n",
    "# dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "45e2e54819d2e10ea1ece5ec6a1f6b7f379f315a337a1110cb8530c4d6073b39"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
