{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(GenomicRanges)\n",
    "library(dplyr)\n",
    "library(data.table)\n",
    "library(ggplot2)\n",
    "library(viridis)\n",
    "library(readxl)\n",
    "library(pheatmap)\n",
    "library(RColorBrewer)\n",
    "library(tidyr)\n",
    "library(gridExtra)\n",
    "library(corrplot)\n",
    "library(tibble)\n",
    "library(ggrepel)\n",
    "library(ggpubr)\n",
    "library(DEFormats)\n",
    "library(BiocParallel)\n",
    "register(MulticoreParam(4))\n",
    "library(chromVAR)\n",
    "library(GenomicFeatures)\n",
    "library(BSgenome.Hsapiens.UCSC.hg38)\n",
    "library(ChIPseeker)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(org.Hs.eg.db)\n",
    "library(pheatmap)\n",
    "library(RColorBrewer)\n",
    "library(tidyr)\n",
    "library(gridExtra)\n",
    "library(ggrepel)\n",
    "library(limma)\n",
    "library(ggdendro)\n",
    "library(ggfortify)\n",
    "library(ggpubr)\n",
    "library(magrittr)\n",
    "library(Seurat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "patientRaw = read_xlsx(\"CART_CUTRUN_Project/report/CART_CUTRUN/analysis/20201028_NHL_CUTandRUN_PatientData.xlsx\")\n",
    "cm_ind = which(patientRaw$`CD8+ T cell isolation` == \"CD8TCM\")\n",
    "patientRaw$Xnumber[cm_ind]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## 1. TSS neighbor regions\n",
    "## use TSS region for differential detection\n",
    "## Get the TSS location and +/-1kb region around TSS\n",
    "gtf <- rtracklayer::import('SupplementaryData/hg38/transcriptomeHumanReference/gencode.v33.annotation.gtf') \n",
    "gtf_df=as.data.frame(gtf)\n",
    "\n",
    "\n",
    "## get the TSS location and extend by 1kb up and downstream\n",
    "tss_df = rbind(\n",
    "    gtf_df %>% data.frame %>% dplyr::filter(type == \"gene\", strand == \"+\") %>% dplyr::select(seqnames, TSS = start, strand, source, gene_id, gene_name, gene_name, gene_type) ,\n",
    "    gtf_df %>% data.frame %>% dplyr::filter(type == \"gene\", strand == \"-\") %>% dplyr::select(seqnames, TSS = end, strand, source, gene_id, gene_name, gene_name, gene_type) \n",
    ")\n",
    "tss_filter_df = tss_df %>% dplyr::filter(!(seqnames %in% c(\"chrM\", \"chrX\", \"chrY\")))\n",
    "\n",
    "tss_gr = GRanges(seqnames = tss_filter_df$seqnames, ranges = IRanges(start = tss_filter_df$TSS - 1000, end = tss_filter_df$TSS + 1000), strand = tss_filter_df$strand, gene_id = tss_filter_df$gene_id, gene_name = tss_filter_df$gene_name, gene_type = tss_filter_df$gene_type)\n",
    "tss_gr \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## patient data\n",
    "\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "histList = c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "countMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_NHLpatient_noChrXYM_countMat.rds\"))\n",
    "normMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_NHLpatient_noChrXYM_normMat.rds\"))\n",
    "designInfo = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_NHLpatient_noChrXYM_designInfo.rds\"))\n",
    "\n",
    "pt_countMat = countMat\n",
    "pt_countMat$H3K27me3 = countMat$H3K27me3[, cm_ind]\n",
    "pt_countMat$H3K4me2 = countMat$H3K4me2[, cm_ind]\n",
    "colnames(pt_countMat$H3K27me3) = paste0(colnames(pt_countMat$H3K27me3), \"_PT\")\n",
    "colnames(pt_countMat$H3K4me2) = paste0(colnames(pt_countMat$H3K4me2), \"_PT\")\n",
    "\n",
    "pt_designInfo = designInfo %>% dplyr::filter(exps %in% patientRaw$Xnumber[cm_ind])\n",
    "head(pt_countMat$H3K27me3)\n",
    "pt_designInfo %>% arrange(hist, exps)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## healthy donor data\n",
    "countMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_countMat.rds\"))\n",
    "normMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_normMat.rds\"))\n",
    "designInfo = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_designInfo.rds\"))\n",
    "\n",
    "selectC = colnames(countMat$H3K27me3) %>% stringr::str_detect(\"CM_Product\")\n",
    "head(countMat$H3K4me2[, selectC])\n",
    "\n",
    "hd_countMat = countMat\n",
    "hd_countMat$H3K27me3 = countMat$H3K27me3[, selectC]\n",
    "hd_countMat$H3K4me2 = countMat$H3K4me2[, selectC]\n",
    "\n",
    "hd_designInfo = designInfo[which(stringr::str_detect(designInfo$exps, \"CM_Product\")), ]\n",
    "hd_designInfo\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "for(hist in histList){\n",
    "    tmp = c()\n",
    "\n",
    "    for(j in 1:ncol(pt_countMat[[hist]])){\n",
    "        pt_designInfo_tmp = pt_designInfo %>% dplyr::filter(hist == hist)\n",
    "        tmp = cbind(tmp, round(pt_countMat[[hist]][, j]/pt_designInfo_tmp$depth[j] * 13240732)) ## remove sequencing depth effect\n",
    "    }\n",
    "    for(j in 1:ncol(hd_countMat[[hist]])){\n",
    "        hd_designInfo_tmp = hd_designInfo %>% dplyr::filter(hist == hist)\n",
    "        tmp = cbind(tmp, round(hd_countMat[[hist]][, j]/hd_designInfo_tmp$depth[j] * 13240732)) ## remove sequencing depth effect\n",
    "    }\n",
    "    countMat[[hist]] = tmp\n",
    "    colnames(countMat[[hist]]) = c(colnames(pt_countMat[[hist]]), colnames(hd_countMat[[hist]]))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "groupList = c(\"PT\", \"HD\")\n",
    "# hdList <- infoData$H3K4me2$Xnumber\n",
    "histList <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "target = data.frame(\n",
    "    group = c(\n",
    "        rep(\"PT\", ncol(pt_countMat$H3K4me2)), \n",
    "        rep(\"HD\", ncol(hd_countMat$H3K4me2))\n",
    "    )\n",
    ")\n",
    "\n",
    "target$group = factor(target$group, levels = groupList)\n",
    "\n",
    "## Experimental design\n",
    "treat <- factor(target$group)\n",
    "design <- model.matrix(~0 + treat)\n",
    "colnames(design) <- levels(treat)\n",
    "contrast <- makeContrasts(PT_HD = PT - HD, levels = design)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "min_dist_peak_to_gene = 5000\n",
    "results = vector(\"list\", length(histList))\n",
    "results_fullInfo = vector(\"list\", length(histList))\n",
    "voomDDS = vector(\"list\", length(histList))\n",
    "for(hist in histList){\n",
    "  ## Filter and delete low expressed genes\n",
    "    selectR = which(rowSums(countMat[[hist]]) > 200) ## remove low count genes\n",
    "    \n",
    "    dataS = countMat[[hist]][selectR, ]\n",
    "    peakRegionList = tss_gr[selectR] %>% data.frame %>% mutate(region = paste0(seqnames, \":\", start, \"-\", end, \"(\", gene_name, \")\")) %$% region\n",
    "    rownames(dataS) = peakRegionList\n",
    "    voomDDS[[hist]] = voom(counts = dataS, design = design, normalize.method = \"cyclicloess\", plot = FALSE)\n",
    "    ## option 1 using voomDDS option2 using normDDS as normalized input.\n",
    "    inputDDS = voomDDS[[hist]]\n",
    "    ## corfit$consensus\n",
    "    fit = lmFit(inputDDS, design)\n",
    "    fitContrast = contrasts.fit(fit, contrast)\n",
    "    fitBayes = eBayes(fitContrast, robust = TRUE)\n",
    "\n",
    "\n",
    "    results[[hist]] = list()\n",
    "    results_fullInfo[[hist]] = list()\n",
    "    for(i in 1:ncol(contrast)){\n",
    "      ## Results\n",
    "      res = topTable(fit = fitBayes, adjust.method = 'fdr', coef = i, number = nrow(inputDDS), sort = 'P') %>% data.table\n",
    "\n",
    "      # res = data.table(PeakRegion = rownames(res), res)\n",
    "      res[, Significance := ifelse((adj.P.Val <= 0.1 & sign(logFC) == 1 & abs(logFC) >= 1), 'Up',\n",
    "                              ifelse((adj.P.Val <= 0.1 & sign(logFC) == -1 & abs(logFC) >= 1), 'Down', 'notDE'))]\n",
    "      results[[hist]][[i]] = left_join(res, dataS %>% data.frame %>% dplyr::mutate(ID = rownames(dataS)), by = \"ID\")\n",
    "\n",
    "      ## overlap with genes\n",
    "      # results_peakRegion = results[[hist]][[i]] %$% PeakRegion\n",
    "      # results_chrom = gsub(\":.*\", \"\", results_peakRegion)\n",
    "      # results_start = gsub(\".*:\", \"\", results_peakRegion) %>% gsub(\"-.*\", \"\", .) %>% as.numeric\n",
    "      # results_end = gsub(\".*:\", \"\", results_peakRegion) %>% gsub(\".*-\", \"\", .) %>% as.numeric\n",
    "      # results_peakRegion_gr = GRanges(seqnames = results_chrom, IRanges(start = results_start, end = results_end), strand = \"*\")\n",
    "      \n",
    "      # dp_match_gene = distanceToNearest(results_peakRegion_gr, gene_gr, select = \"all\") ## get the nearest gene for each peak\n",
    "      # dp_match_gene = dp_match_gene[which(data.frame(dp_match_gene)$distance <= min_dist_peak_to_gene)]\n",
    "\n",
    "      # results_fullInfo[[hist]][[i]] = cbind(\n",
    "      #           results[[hist]][[i]][dp_match_gene@from, ], \n",
    "      #           gene_gr[dp_match_gene@to] %>% data.frame %>% dplyr::select(gene_name, gene_id, gene_type)\n",
    "      # )\n",
    "      # results_fullInfo[[hist]][[i]]$gene_region = paste0(data.frame(gene_gr[dp_match_gene@to])$seqnames, \":\", data.frame(gene_gr[dp_match_gene@to])$start, \"-\", data.frame(gene_gr[dp_match_gene@to])$end)\n",
    "      # results_fullInfo[[hist]][[i]]$peak_distance_to_gene = data.frame(dp_match_gene)$distance\n",
    "      # ## Output\n",
    "      write.table(results[[hist]][[i]], file = paste0(outPath, '/CUTRUN_limma_tables/DE_', hist, '_TSSregions_cmPatient_vs_cmHealthyDonor_InfusionProduct_adj0.1_logFC2.csv'), quote = FALSE, row.names = FALSE, sep = \",\")\n",
    "      # write.table(results_fullInfo[[hist]][[i]], file = paste0(outPath, '/CUTRUN_limma_tables/DE_', hist, '_patient_TSSregions_CR_VS_PD_adj0.1_logFC2_link_nearest_genes.csv'), quote = FALSE, row.names = FALSE, sep = \",\")\n",
    "    }\n",
    "    names(results[[hist]]) <- colnames(contrast)\n",
    "    # names(results_fullInfo[[hist]]) <- colnames(contrast)\n",
    "\n",
    "\n",
    "}\n",
    "save(results, voomDDS, file = paste0(outPath, \"/RData/results_histList_noChrXYM_TSSregions_cmPatient_vs_cmHealthyDonor_InfusionProduct.RData\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=22, repr.plot.height=10)\n",
    "hist = \"H3K27me3\"\n",
    "selectR = which(rowSums(countMat[[hist]]) > 200) ## remove low count genes\n",
    "\n",
    "tmp = countMat[[hist]][selectR, ]\n",
    "\n",
    "k27_obj = CreateSeuratObject(counts = countMat[[hist]][selectR, ], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR)) %>% \n",
    "ScaleData()\n",
    "k27_obj = RunPCA(k27_obj, features = VariableFeatures(object = k27_obj), npcs = 15)\n",
    "k27_plot = Embeddings(k27_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point() +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "hist = \"H3K4me2\"\n",
    "selectR = which(rowSums(countMat[[hist]]) > 200) ## remove low count genes\n",
    "k4_obj = CreateSeuratObject(counts = countMat[[hist]][selectR, ], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR)) %>% \n",
    "ScaleData()\n",
    "k4_obj = RunPCA(k4_obj, features = VariableFeatures(object = k4_obj), npcs = 15)\n",
    "\n",
    "k4_plot = Embeddings(k4_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point() +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "cr_obj = CreateSeuratObject(counts = rbind(tmp, countMat[[hist]][selectR, ]), project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR) + nrow(tmp)) %>% \n",
    "ScaleData()\n",
    "cr_obj = RunPCA(cr_obj, features = VariableFeatures(object = cr_obj), npcs = 15)\n",
    "cr_plot = Embeddings(cr_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point() +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3 + H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "pdf(\"CART_CUTRUN_Project/results/paper_figure/patient-related/CMpatient_CMhealthyDonor_PCA_on_all_TSSneighbors.pdf\", width = 15, height = 6)\n",
    "ggarrange(k27_plot, k4_plot, cr_plot, common.legend = TRUE, ncol = 3, nrow = 1)\n",
    "dev.off()\n",
    "ggarrange(k27_plot, k4_plot, cr_plot, common.legend = TRUE, ncol = 3, nrow = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## call peaks and associated with nearest genes\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "histList = c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "topN = 20\n",
    "load(file = paste0(outPath, \"/RData/PatientData_peakOverlap_withDuplicates_seacrOnly_top0_\", topN, \".RData\"))\n",
    "peakAll_pt = peakAll\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "peakAll_hd = peakAll\n",
    "\n",
    "mPeak = vector(\"list\", 2)\n",
    "for(hist in histList){\n",
    "    tmp = GRanges()\n",
    "    for(pt in patientRaw$Xnumber[cm_ind]){\n",
    "        tmp = append(tmp, peakAll_pt[[hist]][[pt]])\n",
    "    }\n",
    "\n",
    "    selectC = names(peakAll_hd)[names(peakAll_hd) %>% stringr::str_detect(\"CM_Product\") & names(peakAll_hd) %>% stringr::str_detect(hist)]\n",
    "    for(hd in selectC){\n",
    "        tmp = append(tmp, peakAll_hd[[hd]])\n",
    "    }\n",
    "    mPeak[[hist]] = reduce(tmp)\n",
    "}\n",
    "mPeak$H3K4me2 %>% length\n",
    "mPeak$H3K27me3 %>% length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## overlap with bam file to get count -- !!! Run it in terminal directly\n",
    "# library(doMC)\n",
    "# coreN <- 6\n",
    "# registerDoMC(cores = coreN)\n",
    "\n",
    "# outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "# histList = c(\"H3K27me3\", \"H3K4me2\")\n",
    "# bamSelect = vector(\"list\", length(histList))\n",
    "# patient_ordered = vector(\"list\", length(histList))\n",
    "# fragment_tss_counts = vector(\"list\", length(histList))\n",
    "\n",
    "# patientRaw = read_xlsx(\"CART_CUTRUN_Project/report/CART_CUTRUN/analysis/20201028_NHL_CUTandRUN_PatientData.xlsx\")\n",
    "# cm_ind = which(patientRaw$`CD8+ T cell isolation` == \"CD8TCM\")\n",
    "# patientRaw$Xnumber[cm_ind]\n",
    "\n",
    "# ## dataset 1\n",
    "# inPath = \"/shared/ngs/illumina/sfiorenz/201014_D00300_1082_AHHYJJBCX3/analysis/\"\n",
    "# index = grep(\"IP\", list.files(inPath), fixed = TRUE)[!grep(\"IP\", list.files(inPath), fixed = TRUE) %in% grep(\"IgG\", list.files(inPath), fixed = TRUE)]\n",
    "# sampleL = list.files(inPath)[index]\n",
    "# sampleLtmp = sampleL %>% strsplit(\"_\") %>% unlist\n",
    "# patientL = sampleLtmp[seq(4, length(sampleLtmp), 7)]\n",
    "# histL = sampleLtmp[seq(6, length(sampleLtmp), 7)]\n",
    "\n",
    "# for(i in 1:length(sampleL)){\n",
    "#   sample = sampleL[i]\n",
    "#   hist = histL[i]\n",
    "#   pt = patientL[i]\n",
    "#   if(pt %in% patientRaw$Xnumber[cm_ind]){\n",
    "#     bamFile = paste0(inPath, sample, \"/bowtie2_align.bam\")\n",
    "#     if(file.exists(bamFile)){\n",
    "#         bamSelect[[hist]] = c(bamSelect[[hist]], bamFile)\n",
    "#         patient_ordered[[hist]] = c(patient_ordered[[hist]], pt)\n",
    "#     }else{\n",
    "#         print(paste0(bamFile, \" does not exist!\"))\n",
    "#     }\n",
    "#   }\n",
    "# }\n",
    "\n",
    "# ## dataset 2\n",
    "\n",
    "# inPath = \"/shared/ngs/illumina/tphi/200924_D00300_1063_AHHLGCBCX3/analysis/\"\n",
    "# index = grep(\"IP\", list.files(inPath), fixed = TRUE)[!grep(\"IP\", list.files(inPath), fixed = TRUE) %in% grep(\"IgG\", list.files(inPath), fixed = TRUE)]\n",
    "# sampleL = list.files(inPath)[index]\n",
    "# sampleLtmp = sampleL %>% strsplit(\"_\") %>% unlist\n",
    "# patientL = sampleLtmp[seq(3, length(sampleLtmp), 6)]\n",
    "# histL = sampleLtmp[seq(5, length(sampleLtmp), 6)]\n",
    "\n",
    "# for(i in 1:length(sampleL)){\n",
    "#   sample = sampleL[i]\n",
    "#   hist = histL[i]\n",
    "#   pt = patientL[i]\n",
    "#   if(pt %in% patientRaw$Xnumber[cm_ind]){\n",
    "#     bamFile = paste0(inPath, sample, \"/bowtie2_align.bam\")\n",
    "#     if(file.exists(bamFile)){\n",
    "#         bamSelect[[hist]] = c(bamSelect[[hist]], bamFile)\n",
    "#         patient_ordered[[hist]] = c(patient_ordered[[hist]], pt)\n",
    "#     }else{\n",
    "#         print(paste0(bamFile, \" does not exist!\"))\n",
    "#     }\n",
    "#   }\n",
    "# }\n",
    "\n",
    "# ## healthy donor\n",
    "# bamDir <- \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "# # fragment_tss_counts = vector(\"list\", length(histList))\n",
    "# for(hist in histList){\n",
    "#    for(cell in \"CM\"){\n",
    "#     for(expr in \"Product\"){\n",
    "#       hdL = paste0(\"HD\", 1:7)\n",
    "#       for(hd in hdL){\n",
    "#         ## Get the replicate number and peak calling type\n",
    "#         rep = \"rep1\"\n",
    "#         bamType = \"bowtie2_align.bam\"\n",
    "#         bamFile = paste0(bamDir, hist, \"_CD8_\", cell, \"_\", expr, \"_\", hd, \"_\", rep, \"/alignment/\", bamType)\n",
    "#         if(file.exists(bamFile)){\n",
    "#           bamSelect[[hist]] = c(bamSelect[[hist]], bamFile)\n",
    "#           patient_ordered[[hist]] = c(patient_ordered[[hist]], hd)\n",
    "#         }else{\n",
    "#           print(paste0(bamFile, \" does not exist!\"))\n",
    "#         }\n",
    "#       }\n",
    "#     }\n",
    "#    }\n",
    "# }\n",
    "\n",
    "\n",
    "# ## read in the reads\n",
    "# for(hist in histList){\n",
    "#   fragment_tss_counts[[hist]] <- mclapply(as.list(bamSelect[[hist]]), chromVAR::getCounts, mPeak[[hist]], paired = TRUE, by_rg = FALSE, format = \"bam\", mc.cores = coreN)\n",
    "# }\n",
    "\n",
    "# saveRDS(fragment_tss_counts, file = paste0(outPath, \"/RData/peakRegions_chromVar_count_CMpatients_CMhealthydonor_noChrXYM.rds\"))\n",
    "# saveRDS(list(bamSelect, patient_ordered), file = paste0(outPath, \"/RData/peakRegions_chromVar_count_CMpatients_CMhealthydonor_noChrXYM_file_pt_info.rds\"))\n",
    "\n",
    "\n",
    "\n",
    "fragment_tss_counts = readRDS(file = paste0(outPath, \"/RData/peakRegions_chromVar_count_CMpatients_CMhealthydonor_noChrXYM.rds\"))\n",
    "tmp = readRDS(file = paste0(outPath, \"/RData/peakRegions_chromVar_count_CMpatients_CMhealthydonor_noChrXYM_file_pt_info.rds\"))\n",
    "bamSelect = tmp[[1]]\n",
    "patient_ordered = tmp[[2]]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(chromVAR)\n",
    "library(SummarizedExperiment)\n",
    "library(Matrix)\n",
    "\n",
    "## count matrix column name\n",
    "countMat = vector(\"list\", length(histList))\n",
    "normMat = vector(\"list\", length(histList))\n",
    "seqDepth = vector(\"list\", length(histList))\n",
    "designInfo = c()\n",
    "for(hist in histList){\n",
    "\n",
    "  countMat[[hist]] <- matrix(NA, length(mPeak[[hist]]), length(bamSelect[[hist]]))\n",
    "  normMat[[hist]] <- matrix(NA, length(mPeak[[hist]]), length(bamSelect[[hist]]))\n",
    "  colnames(countMat[[hist]]) <- patient_ordered[[hist]]\n",
    "  colnames(normMat[[hist]]) <- patient_ordered[[hist]]\n",
    "  seqDepth[[hist]] <- NULL\n",
    "  for (k in 1:length(bamSelect[[hist]])){\n",
    "    countMat[[hist]][, k] <- counts(fragment_tss_counts[[hist]][[k]])[,1]\n",
    "    seqDepth[[hist]][k] <- fragment_tss_counts[[hist]][[k]]@colData[1,1]\n",
    "    normMat[[hist]][, k] <- countMat[[hist]][, k]/seqDepth[[hist]][k] * 13240732 ## divide by the seqDepth and multiply by the largest seqDepth\n",
    "  }\n",
    "  designInfo <- data.frame(exps = patient_ordered[[hist]], depth = seqDepth[[hist]], hist = hist) %>% rbind(designInfo, .)\n",
    "}\n",
    "rownames(countMat$H3K4me2) = mPeak$H3K4me2 %>% data.frame %>% mutate(region = paste0(seqnames, \":\", start, \"-\", end)) %$% region\n",
    "rownames(countMat$H3K27me3) = mPeak$H3K27me3 %>% data.frame %>% mutate(region = paste0(seqnames, \":\", start, \"-\", end)) %$% region\n",
    "rownames(normMat$H3K4me2) = mPeak$H3K4me2 %>% data.frame %>% mutate(region = paste0(seqnames, \":\", start, \"-\", end)) %$% region\n",
    "rownames(normMat$H3K27me3) = mPeak$H3K27me3 %>% data.frame %>% mutate(region = paste0(seqnames, \":\", start, \"-\", end)) %$% region\n",
    "\n",
    "# ## read in patient info\n",
    "# patientRaw = read_xlsx(\"CART_CUTRUN_Project/report/CART_CUTRUN/analysis/20201028_NHL_CUTandRUN_PatientData.xlsx\")\n",
    "# info = patientRaw %>% data.frame %>% dplyr::select(Xnumber, candr_day, age, sex, bestresp) \n",
    "\n",
    "# countMat$H3K27me3 = countMat$H3K27me3[, patientRaw$Xnumber]\n",
    "# countMat$H3K4me2 = countMat$H3K4me2[, patientRaw$Xnumber]\n",
    "\n",
    "# infoData = list(H3K27me3 = data.frame(Xnumber = colnames(countMat[[histList[1]]])) %>% left_join(., info, by = \"Xnumber\") %>% mutate(candr_day = factor(candr_day), age = factor(age), sex = factor(sex), bestresp = factor(bestresp, levels = c(\"CR\", \"PR\", \"SD\", \"PD\"))), \n",
    "#                 H3K4me2 = data.frame(Xnumber = colnames(countMat[[histList[2]]])) %>% left_join(., info, by = \"Xnumber\") %>% mutate(candr_day = factor(candr_day), age = factor(age), sex = factor(sex), bestresp = factor(bestresp, levels = c(\"CR\", \"PR\", \"SD\", \"PD\"))))\n",
    "\n",
    "# colnames(countMat$H3K27me3) = paste0(infoData$H3K27me3$Xnumber, \"_\", infoData$H3K27me3$bestresp)\n",
    "# colnames(countMat$H3K4me2) = paste0(infoData$H3K4me2$Xnumber, \"_\", infoData$H3K4me2$bestresp)\n",
    "\n",
    "\n",
    "saveRDS(countMat, file = paste0(outPath, \"/RData/peakRegion_chromVar_count_CMpatient_CMhealthydonor_noChrXYM_countMat.rds\"))\n",
    "saveRDS(normMat, file = paste0(outPath, \"/RData/peakRegion_chromVar_count_CMpatient_CMhealthydonor_noChrXYM_normMat.rds\"))\n",
    "saveRDS(designInfo, file = paste0(outPath, \"/RData/peakRegion_chromVar_count_CMpatient_CMhealthydonor_noChrXYM_designInfo.rds\"))\n",
    "# normMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_normMat.rds\"))\n",
    "# designInfo = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_designInfo.rds\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "groupList = c(\"PT\", \"HD\")\n",
    "# hdList <- infoData$H3K4me2$Xnumber\n",
    "histList <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "target = data.frame(\n",
    "    group = c(\n",
    "        rep(\"PT\", 15), \n",
    "        rep(\"HD\", 7)\n",
    "    )\n",
    ")\n",
    "\n",
    "target$group = factor(target$group, levels = groupList)\n",
    "\n",
    "## Experimental design\n",
    "treat <- factor(target$group)\n",
    "design <- model.matrix(~0 + treat)\n",
    "colnames(design) <- levels(treat)\n",
    "contrast <- makeContrasts(PT_HD = PT - HD, levels = design)\n",
    "\n",
    "target\n",
    "design\n",
    "contrast\n",
    "\n",
    "\n",
    "## read in genes location\n",
    "gtf_filter_df = rtracklayer::import('SupplementaryData/hg38/GENCODE/gencode.v21.annotation.gtf') %>% \n",
    "    as.data.frame %>% \n",
    "    dplyr::filter(type == \"gene\", !(seqnames %in% c(\"chrM\", \"chrX\", \"chrY\")))\n",
    "gene_gr = GRanges(seqnames = gtf_filter_df$seqnames, ranges = IRanges(start = gtf_filter_df$start, end = gtf_filter_df$end), strand = gtf_filter_df$strand, gene_id = gtf_filter_df$gene_id, gene_name = gtf_filter_df$gene_name, gene_type = gtf_filter_df$gene_type)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "min_dist_peak_to_gene = 5000\n",
    "results = vector(\"list\", length(histList))\n",
    "results_fullInfo = vector(\"list\", length(histList))\n",
    "voomDDS = vector(\"list\", length(histList))\n",
    "for(hist in histList){\n",
    "  ## Filter and delete low expressed genes\n",
    "    selectR = which(rowSums(countMat[[hist]]) > 200) ## remove low count genes\n",
    "    \n",
    "    dataS = countMat[[hist]][selectR, ]\n",
    "    # rownames(dataS) = peakRegionList\n",
    "    voomDDS[[hist]] = voom(counts = dataS, design = design, normalize.method = \"cyclicloess\", plot = FALSE)\n",
    "    ## option 1 using voomDDS option2 using normDDS as normalized input.\n",
    "    inputDDS = voomDDS[[hist]]\n",
    "    ## corfit$consensus\n",
    "    fit = lmFit(inputDDS, design)\n",
    "    fitContrast = contrasts.fit(fit, contrast)\n",
    "    fitBayes = eBayes(fitContrast, robust = TRUE)\n",
    "\n",
    "\n",
    "    results[[hist]] = list()\n",
    "    results_fullInfo[[hist]] = list()\n",
    "    for(i in 1:ncol(contrast)){\n",
    "      ## Results\n",
    "      res = topTable(fit = fitBayes, adjust.method = 'fdr', coef = i, number = nrow(inputDDS), sort = 'P') \n",
    "\n",
    "      res = data.table(PeakRegion = rownames(res), res)\n",
    "      res[, Significance := ifelse((adj.P.Val <= 0.1 & sign(logFC) == 1 & abs(logFC) >= 1), 'Up',\n",
    "                              ifelse((adj.P.Val <= 0.1 & sign(logFC) == -1 & abs(logFC) >= 1), 'Down', 'notDE'))]\n",
    "      results[[hist]][[i]] = left_join(res, dataS %>% data.frame %>% dplyr::mutate(PeakRegion = rownames(dataS)), by = \"PeakRegion\")\n",
    "\n",
    "      ## overlap with genes\n",
    "      results_peakRegion = results[[hist]][[i]] %$% PeakRegion\n",
    "      results_chrom = gsub(\":.*\", \"\", results_peakRegion)\n",
    "      results_start = gsub(\".*:\", \"\", results_peakRegion) %>% gsub(\"-.*\", \"\", .) %>% as.numeric\n",
    "      results_end = gsub(\".*:\", \"\", results_peakRegion) %>% gsub(\".*-\", \"\", .) %>% as.numeric\n",
    "      results_peakRegion_gr = GRanges(seqnames = results_chrom, IRanges(start = results_start, end = results_end), strand = \"*\")\n",
    "      \n",
    "      dp_match_gene = distanceToNearest(results_peakRegion_gr, gene_gr, select = \"all\") ## get the nearest gene for each peak\n",
    "      dp_match_gene = dp_match_gene[which(data.frame(dp_match_gene)$distance <= min_dist_peak_to_gene)]\n",
    "\n",
    "      results_fullInfo[[hist]][[i]] = cbind(\n",
    "                results[[hist]][[i]][dp_match_gene@from, ], \n",
    "                gene_gr[dp_match_gene@to] %>% data.frame %>% dplyr::select(gene_name, gene_id, gene_type)\n",
    "      )\n",
    "      results_fullInfo[[hist]][[i]]$gene_region = paste0(data.frame(gene_gr[dp_match_gene@to])$seqnames, \":\", data.frame(gene_gr[dp_match_gene@to])$start, \"-\", data.frame(gene_gr[dp_match_gene@to])$end)\n",
    "      results_fullInfo[[hist]][[i]]$peak_distance_to_gene = data.frame(dp_match_gene)$distance\n",
    "      ## Output\n",
    "      write.table(results[[hist]][[i]], file = paste0(outPath, '/CUTRUN_limma_tables/DE_', hist, '_peakRegions_cmPatient_vs_cmHealthyDonor_InfusionProduct_adj0.1_logFC2.csv'), quote = FALSE, row.names = FALSE, sep = \",\")\n",
    "      write.table(results_fullInfo[[hist]][[i]], file = paste0(outPath, '/CUTRUN_limma_tables/DE_', hist, '_peakRegions_cmPatient_vs_cmHealthyDonor_InfusionProduct_adj0.1_logFC2_link_nearest_genes.csv'), quote = FALSE, row.names = FALSE, sep = \",\")\n",
    "    }\n",
    "    names(results[[hist]]) <- colnames(contrast)\n",
    "    names(results_fullInfo[[hist]]) <- colnames(contrast)\n",
    "\n",
    "\n",
    "}\n",
    "save(results, results_fullInfo, voomDDS, file = paste0(outPath, \"/RData/results_histList_noChrXYM_peakRegions_cmPatient_vs_cmHealthyDonor_InfusionProduct.RData\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=22, repr.plot.height=10)\n",
    "hist = \"H3K27me3\"\n",
    "selectR = which(rowSums(countMat[[hist]]) > 200) ## remove low count genes\n",
    "\n",
    "tmp = countMat[[hist]][selectR, ]\n",
    "\n",
    "k27_obj = CreateSeuratObject(counts = countMat[[hist]][selectR, ], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR)) %>% \n",
    "ScaleData()\n",
    "k27_obj = RunPCA(k27_obj, features = VariableFeatures(object = k27_obj), npcs = 15)\n",
    "k27_plot = Embeddings(k27_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point() +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "hist = \"H3K4me2\"\n",
    "selectR = which(rowSums(countMat[[hist]]) > 200) ## remove low count genes\n",
    "k4_obj = CreateSeuratObject(counts = countMat[[hist]][selectR, ], project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR)) %>% \n",
    "ScaleData()\n",
    "k4_obj = RunPCA(k4_obj, features = VariableFeatures(object = k4_obj), npcs = 15)\n",
    "\n",
    "k4_plot = Embeddings(k4_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point() +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "cr_obj = CreateSeuratObject(counts = rbind(tmp, countMat[[hist]][selectR, ]), project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = length(selectR) + nrow(tmp)) %>% \n",
    "ScaleData()\n",
    "cr_obj = RunPCA(cr_obj, features = VariableFeatures(object = cr_obj), npcs = 15)\n",
    "cr_plot = Embeddings(cr_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point() +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3 + H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "pdf(\"CART_CUTRUN_Project/results/paper_figure/patient-related/CMpatient_CMhealthyDonor_PCA_on_all_peakRegions.pdf\", width = 15, height = 6)\n",
    "ggarrange(k27_plot, k4_plot, cr_plot, common.legend = TRUE, ncol = 3, nrow = 1)\n",
    "dev.off()\n",
    "ggarrange(k27_plot, k4_plot, cr_plot, common.legend = TRUE, ncol = 3, nrow = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
