{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## CUTRUN DE and assign to nearest genes\n",
    "\n",
    "library(pacman)\n",
    "p_load(data.table, dplyr, ggplot2, viridis, magrittr, VennDiagram, ggpubr, limma, edgeR, tidyr, GenomicRanges, RColorBrewer, pheatmap, Seurat, fgsea, GSEABase, limma, TxDb.Hsapiens.UCSC.hg38.knownGene, org.Hs.eg.db)\n",
    "\n",
    "cell_list <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hd_list <- paste0(\"HD\", c(1:3, 5:7))\n",
    "expr_list = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "cell_comp_list = c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "hist_list <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "in_path <- \"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "out_path <- \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/\"\n",
    "fig_path <- \"CART_CUTRUN_Project/results/paper_figure/\"\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## PCA for RNA and CUTRUN\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\")) ## 'cell_comp_list''cell_list''countMat''designInfo''expr_list''fig_path''hd_list''hist_list''in_path''inPath''out_path''outPath'\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/RData/HD1-3_5-7_normLimma_perExprCondition.RData\") ## data, selectR, dataS, voomDDS, results, \n",
    "\n",
    "geneID = fread(paste0(\"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\", \"RNA_CD8_N_Input_HD1.genes.results\"))$gene_id\n",
    "data = data.frame(data)\n",
    "select_exp = \"Input\"\n",
    "\n",
    "rna_m = data[, (colnames(data) %>% stringr::str_detect(select_exp)) & !(colnames(data) %>% stringr::str_detect(\"EMRA\"))]\n",
    "k27_m = countMat[[\"H3K27me3\"]][, (colnames(countMat[[\"H3K27me3\"]]) %>% stringr::str_detect(select_exp)) & !(colnames(countMat[[\"H3K27me3\"]]) %>% stringr::str_detect(\"EMRA\"))]\n",
    "k4_m = countMat[[\"H3K4me2\"]][, (colnames(countMat[[\"H3K4me2\"]]) %>% stringr::str_detect(select_exp)) & !(colnames(countMat[[\"H3K4me2\"]]) %>% stringr::str_detect(\"EMRA\"))]\n",
    "\n",
    "rownames(rna_m) = geneID\n",
    "rownames(k27_m) = mPeak[[\"H3K27me3\"]] %>% data.frame %>% mutate(region = paste0(seqnames, \":\", start, \"-\", end)) %$% region\n",
    "rownames(k4_m) = mPeak[[\"H3K4me2\"]] %>% data.frame %>% mutate(region = paste0(seqnames, \":\", start, \"-\", end)) %$% region\n",
    "\n",
    "dim(rna_m)\n",
    "dim(k27_m)\n",
    "dim(k4_m)\n",
    "head(rna_m)\n",
    "head(k27_m)\n",
    "head(k4_m)\n",
    "\n",
    "rna_voom = voomDDS$E[, colnames(voomDDS$E) %>% stringr::str_detect(select_exp)]\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\"))\n",
    "k27_voom = voomDDS[[\"H3K27me3\"]]$E[, colnames(voomDDS[[\"H3K27me3\"]]$E) %>% stringr::str_detect(select_exp)]\n",
    "k4_voom = voomDDS[[\"H3K4me2\"]]$E[, colnames(voomDDS[[\"H3K4me2\"]]$E) %>% stringr::str_detect(select_exp)]\n",
    "rownames(rna_voom) = 1:nrow(rna_voom)\n",
    "rownames(k27_voom) = paste0(\"K27me3_\", 1:nrow(k27_voom))\n",
    "rownames(k4_voom) = paste0(\"K4me2_\", 1:nrow(k4_voom))\n",
    "\n",
    "dim(rna_voom)\n",
    "dim(k27_voom)\n",
    "dim(k4_voom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "rna_obj = CreateSeuratObject(counts = rna_m, project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 2000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = 2000) %>% \n",
    "ScaleData()\n",
    "rna_obj = RunPCA(rna_obj, features = VariableFeatures(object = rna_obj), npcs = 15)\n",
    "\n",
    "# k27_obj = CreateSeuratObject(counts = k27_m, project = \"cart\", min.cells = 0, min.features = 10) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "# FindVariableFeatures(., selection.method = \"vst\", nfeatures = 1000) %>% \n",
    "# ScaleData()\n",
    "# # FindVariableFeatures(., selection.method = \"vst\", nfeatures = round(nrow(k27_m)/10*8)) %>% \n",
    "\n",
    "# k27_obj = RunPCA(k27_obj, features = VariableFeatures(object = k27_obj), npcs = 10)\n",
    "\n",
    "# k4_obj = CreateSeuratObject(counts = k4_m, project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "# FindVariableFeatures(., selection.method = \"vst\", nfeatures = 1000) %>%  #\n",
    "# ScaleData()\n",
    "# k4_obj = RunPCA(k4_obj, features = VariableFeatures(object = k4_obj), npcs = 10)\n",
    "\n",
    "\n",
    "# cr_obj = CreateSeuratObject(counts = rbind(k27_m, k4_m), project = \"cart\", min.cells = 0, min.features = 50) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "# FindVariableFeatures(., selection.method = \"vst\", nfeatures = 1000 ) %>%  #round((nrow(k27_m) + nrow(k4_m))/10*8)\n",
    "# ScaleData()\n",
    "# cr_obj = RunPCA(cr_obj, features = VariableFeatures(object = cr_obj), npcs = 10)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=5)\n",
    "# DimPlot(rna_obj, reduction = \"pca\")\n",
    "ct_list = c(\"N\", \"CM\", \"EM\")\n",
    "p1 = Embeddings(rna_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 6), levels = ct_list)) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"RNA-seq\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p2 = Embeddings(k27_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 4), levels = ct_list)) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p3 = Embeddings(k4_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 4), levels = ct_list)) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p4 = Embeddings(cr_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 4), levels = ct_list)) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3+H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "ggarrange(p1, p2, p3, p4, nrow = 1, ncol = 4, common.legend = TRUE)\n",
    "# ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig1/PCA_seurat_procedure_N_CM_EM.pdf\", width = 20, height = 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(pacman)\n",
    "p_load(data.table, dplyr, ggplot2, viridis, magrittr, VennDiagram, ggpubr, limma, edgeR, tidyr, GenomicRanges, RColorBrewer, pheatmap, Seurat, fgsea, GSEABase, limma, TxDb.Hsapiens.UCSC.hg38.knownGene, org.Hs.eg.db, ggrepel)\n",
    "fig_path <- \"CART_CUTRUN_Project/results/paper_figure_paper1/\"\n",
    "\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "\n",
    "x_thres = 4.5\n",
    "y_thres = 2.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cT = \"N\"\n",
    "cr_rna_match = readRDS(file = paste0(outPath, \"/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "p_n = cr_rna_match %>% ggplot(aes(x = H3K4me2, y = H3K27me3, color = RNA)) +\n",
    "        geom_point(size = 1) +\n",
    "        theme_bw(base_size = 30) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_viridis() +\n",
    "        geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10', size = 1.5) +\n",
    "        geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\", size = 1.5) +\n",
    "         labs(color = \"RNA: log2(expected count + 1)\") +  guides(color = guide_colourbar(barwidth = 10, barheight = 1))+\n",
    "        ggtitle(\"Naive\")\n",
    "\n",
    "cT = \"CM\"\n",
    "cr_rna_match = readRDS(file = paste0(outPath, \"/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "p_cm = cr_rna_match %>% ggplot(aes(x = H3K4me2, y = H3K27me3, color = RNA)) +\n",
    "        geom_point(size = 1) +\n",
    "        theme_bw(base_size = 30) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_viridis() +\n",
    "        geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10', size = 1.5) +\n",
    "        geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\", size = 1.5) +\n",
    "         labs(color = \"RNA: log2(expected count + 1)\") +  guides(color = guide_colourbar(barwidth = 10, barheight = 1))+\n",
    "        ggtitle(\"CM\")\n",
    "\n",
    "cT = \"EM\"\n",
    "cr_rna_match = readRDS(file = paste0(outPath, \"/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "p_em = cr_rna_match %>% ggplot(aes(x = H3K4me2, y = H3K27me3, color = RNA)) +\n",
    "        geom_point(size = 1) +\n",
    "        theme_bw(base_size = 30) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_viridis() +\n",
    "        geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10', size = 1.5) +\n",
    "        geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\", size = 1.5) +\n",
    "        labs(color = \"RNA: log2(expected count + 1)\") +  guides(color = guide_colourbar(barwidth = 10, barheight = 1))+\n",
    "        ggtitle(\"EM\") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=25, repr.plot.height=8)\n",
    "ggarrange(p_n, p_cm, p_em, nrow = 1, common.legend = TRUE) \n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig2/K27_vs_K4_colorbyRNA_N_CM_EM.pdf\", width = 21, height = 9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cT = \"N\"\n",
    "cr_rna_match = readRDS(file = paste0(outPath, \"/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "p_n = cr_rna_match %>% ggplot(aes(x = H3K4me2, y = H3K27me3, color = RNA)) +\n",
    "        geom_point(size = 1, shape = 19, alpha = 0.2) +\n",
    "        theme_bw(base_size = 30) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_viridis() +\n",
    "        geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10', size = 1.5) +\n",
    "        geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\", size = 1.5) +\n",
    "         labs(color = \"RNA: log2(expected count + 1)\") +   guides(color = guide_colourbar(barwidth = 15, barheight = 1))+\n",
    "        ggtitle(\"Naive\")\n",
    "\n",
    "cT = \"CM\"\n",
    "cr_rna_match = readRDS(file = paste0(outPath, \"/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "p_cm = cr_rna_match %>% ggplot(aes(x = H3K4me2, y = H3K27me3, color = RNA)) +\n",
    "        geom_point(size = 1, shape = 19, alpha = 0.2) +\n",
    "        theme_bw(base_size = 30) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_viridis() +\n",
    "        geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10', size = 1.5) +\n",
    "        geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\", size = 1.5) +\n",
    "        labs(color = \"RNA: log2(expected count + 1)\") +  guides(color = guide_colourbar(barwidth = 15, barheight = 1))+\n",
    "        ggtitle(\"CM\")\n",
    "\n",
    "cT = \"EM\"\n",
    "cr_rna_match = readRDS(file = paste0(outPath, \"/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "p_em = cr_rna_match %>% ggplot(aes(x = H3K4me2, y = H3K27me3, color = RNA)) +\n",
    "        geom_point(size = 1, shape = 19, alpha = 0.2) +\n",
    "        theme_bw(base_size = 30) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_viridis() +\n",
    "        geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10', size = 1.5) +\n",
    "        geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\", size = 1.5) +\n",
    "        labs(color = \"RNA: log2(expected count + 1)\") +  guides(color = guide_colourbar(barwidth = 15, barheight = 1)) + \n",
    "        ggtitle(\"EM\") \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=25, repr.plot.height=8)\n",
    "ggarrange(p_n, p_cm, p_em, nrow = 1, common.legend = TRUE) \n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig2/K27_vs_K4_colorbyRNA_N_CM_EM_opacity0.2.pdf\", width = 21, height = 9)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Bivalent genes\n",
    "## CUTRUN DE and assign to nearest genes\n",
    "\n",
    "library(pacman)\n",
    "p_load(data.table, dplyr, ggplot2, viridis, magrittr, VennDiagram, ggpubr, limma, edgeR, tidyr, GenomicRanges, RColorBrewer, pheatmap, Seurat, fgsea, GSEABase, limma, TxDb.Hsapiens.UCSC.hg38.knownGene, org.Hs.eg.db)\n",
    "\n",
    "cell_list <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hd_list <- paste0(\"HD\", c(1:3, 5:7))\n",
    "expr_list = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "cell_comp_list = c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "hist_list <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "in_path <- \"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "out_path <- \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/\"\n",
    "fig_path <- \"CART_CUTRUN_Project/results/paper_figure/\"\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "\n",
    "## Bivalent methylation marks\n",
    "\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/IDR_master_peak_list_chromVar_count_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\"))\n",
    "\n",
    "names(peakAll)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "## overlap K4me2 and K27me3 ensuring at least 50bp overlap, count number of regions after merging\n",
    "fig_path = \"CART_CUTRUN_Project/results/paper_figure/fig1\"\n",
    "bi_path <- \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/BivalentGenes/\"\n",
    "gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")\n",
    "\n",
    "get_nearest_gene <- function(gene_gr, region_gr){\n",
    "    region_assign_gene <- distanceToNearest(region_gr, gene_gr, select = \"all\")\n",
    "    # dist_tmp <- region_assign_gene %>% data.frame %$% distance\n",
    "    # region_assign_gene <- region_assign_gene[dist_tmp < 1000]\n",
    "    region_match_gene <- cbind(data.frame(region_gr)[region_assign_gene@from, ], data.frame(peak_id = region_assign_gene@from, gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[region_assign_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) )\n",
    "    return(region_match_gene)\n",
    "\n",
    "}\n",
    "bi_num <- c()\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    for(pt in paste0(\"HD\", 4:7)){\n",
    "        bi_olp <- findOverlaps(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]], minoverlap = 50)\n",
    "        bi_peak.gr <- reduce(append(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]][bi_olp@from], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]][bi_olp@to]))\n",
    "        bi_gene <- get_nearest_gene(gene_gr, bi_peak.gr)\n",
    "        write.table(unique(bi_gene$gene_name), paste0(bi_path, \"Bivalent_peaks_nearest_genes_\", cT, \"_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "\n",
    "        total_peak.gr <- append(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]]) %>% reduce\n",
    "        total_gene <- get_nearest_gene(gene_gr, total_peak.gr)\n",
    "\n",
    "        bi_num <- data.frame(\n",
    "            bi_peak_num = length(bi_peak.gr),\n",
    "            bi_gene_num = length(unique(bi_gene$gene_name)),\n",
    "            total_peak_num = length(total_peak.gr),\n",
    "            total_gene_num = length(unique(total_gene$gene_name)),\n",
    "            cT = cT,\n",
    "            pt = pt\n",
    "        ) %>% rbind(bi_num, .)\n",
    "    }\n",
    "}\n",
    "bi_num$cT <- factor(bi_num$cT, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\"))\n",
    "\n",
    "# p1 = bi_num %>% ggplot(aes(x = pt, y = bi_peak_num, fill = cT)) + \n",
    "# geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "# scale_fill_viridis(discrete = TRUE) +\n",
    "# theme_bw(base_size = 20) +\n",
    "# xlab(\"\") +\n",
    "# ylab(\"# of Peak Regions\") +\n",
    "# rremove(\"legend.title\")\n",
    "\n",
    "# p2 = bi_num %>% mutate(bi_peak_prop = bi_peak_num/total_peak_num * 100) %>% ggplot(aes(x = pt, y = bi_peak_prop, fill = cT)) + \n",
    "# geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "# scale_fill_viridis(discrete = TRUE) +\n",
    "# theme_bw(base_size = 20) +\n",
    "# xlab(\"\") +\n",
    "# ylab(\"% of Peak Regions\") +\n",
    "# rremove(\"legend.title\")\n",
    "\n",
    "# p3 = bi_num %>% ggplot(aes(x = pt, y = bi_gene_num, fill = cT)) + \n",
    "# geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "# scale_fill_viridis(discrete = TRUE) +\n",
    "# theme_bw(base_size = 20) +\n",
    "# xlab(\"\") +\n",
    "# ylab(\"# of Genes\") +\n",
    "# rremove(\"legend.title\")\n",
    "\n",
    "# p4 = bi_num %>% mutate(bi_gene_prop = bi_gene_num/total_gene_num * 100) %>% ggplot(aes(x = pt, y = bi_gene_prop, fill = cT)) + \n",
    "# geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "# scale_fill_viridis(discrete = TRUE) +\n",
    "# theme_bw(base_size = 20) +\n",
    "# xlab(\"\") +\n",
    "# ylab(\"% of Genes\") +\n",
    "# rremove(\"legend.title\")\n",
    "\n",
    "# ggarrange(p1, p2, p3, p4, ncol = 2, nrow = 2, common.legend = TRUE) \n",
    "\n",
    "\n",
    "b1 = bi_num %>% filter(cT != \"EMRA\") %>% ggplot(aes(x = cT, y = bi_peak_num, fill = cT)) + \n",
    "geom_boxplot() +\n",
    "scale_fill_viridis(discrete = TRUE, begin = 0, end = 0.75, alpha = 0.8) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Peak Regions\") +\n",
    "rremove(\"legend\")\n",
    "\n",
    "b2 = bi_num %>% filter(cT != \"EMRA\") %>% mutate(bi_peak_prop = bi_peak_num/total_peak_num * 100) %>% ggplot(aes(x = cT, y = bi_peak_prop, fill = cT)) + \n",
    "geom_boxplot() +\n",
    "scale_fill_viridis(discrete = TRUE, begin = 0, end = 0.75, alpha = 0.8) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Peak Regions\") +\n",
    "rremove(\"legend\")\n",
    "\n",
    "b3 = bi_num %>% filter(cT != \"EMRA\") %>% ggplot() + \n",
    "geom_boxplot(aes(x = cT, y = bi_gene_num, fill = cT)) +\n",
    "scale_fill_viridis(discrete = TRUE, begin = 0, end = 0.75, alpha = 0.8) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Genes\") +\n",
    "rremove(\"legend\")\n",
    "\n",
    "\n",
    "b4 = bi_num %>% filter(cT != \"EMRA\") %>% mutate(bi_gene_prop = bi_gene_num/total_gene_num * 100) %>% ggplot() + \n",
    "geom_boxplot(aes(x = cT, y = bi_gene_prop, fill = cT)) +\n",
    "scale_fill_viridis(discrete = TRUE, begin = 0, end = 0.75, alpha = 0.8) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rremove(\"legend\") \n",
    "\n",
    "ggarrange(b1, b2, b3, b4, ncol = 4, nrow = 1)\n",
    "# ggsave(paste0(fig_path, \"/bivalent_peak_gene_num_prop_boxplot.pdf\"), width = 28, height = 7)\n",
    "\n",
    "\n",
    "# pdf(paste0(fig_path, \"/bivalent_peak_gene_num_prop.pdf\"), width = 10, height = 10)\n",
    "# ggarrange(p1, p2, p3, p4, ncol = 2, nrow = 2, common.legend = TRUE) \n",
    "# dev.off()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "load(file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition.RData\"))\n",
    "rna_results <- results\n",
    "\n",
    "bi_change <- c()\n",
    "bi_change_gene_lfc <- c()\n",
    "ct_list <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "pt_list <- paste0(\"HD\", 4:7)\n",
    "for(i in 1:length(ct_list)){\n",
    "    cT <- ct_list[i]\n",
    "    cT1 <- cT\n",
    "    for(j in 1:length(ct_list)){\n",
    "        if(i != j){\n",
    "            cT2 <- ct_list[j]\n",
    "            for(pt in pt_list){\n",
    "                bi_olp <- findOverlaps(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]], minoverlap = 50)\n",
    "                bi_olp2 <- findOverlaps(peakAll[[paste0(\"H3K27me3_\", cT2, \"_Input_\", pt)]], peakAll[[paste0(\"H3K4me2_\", cT2, \"_Input_\", pt)]], minoverlap = 50)\n",
    "\n",
    "                bi_peak.gr <- reduce(append(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]][bi_olp@from], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]][bi_olp@to]))\n",
    "                bi_peak.gr2 <- reduce(append(peakAll[[paste0(\"H3K27me3_\", cT2, \"_Input_\", pt)]][bi_olp2@from], peakAll[[paste0(\"H3K4me2_\", cT2, \"_Input_\", pt)]][bi_olp2@to]))\n",
    "\n",
    "                bi_again <- findOverlaps(bi_peak.gr, bi_peak.gr2, minoverlap = 50)@from %>% unique\n",
    "                bi_k27 <- findOverlaps(bi_peak.gr, peakAll[[paste0(\"H3K27me3_\", cT2, \"_Input_\", pt)]], minoverlap = 50)@from %>% unique\n",
    "                bi_k4 <- findOverlaps(bi_peak.gr, peakAll[[paste0(\"H3K4me2_\", cT2, \"_Input_\", pt)]], minoverlap = 50)@from %>% unique\n",
    "                bi_k27 <- setdiff(setdiff(bi_k27, bi_k4), bi_again)\n",
    "                bi_k4 <- setdiff(setdiff(bi_k4, bi_k27), bi_again)\n",
    "                bi_none <- setdiff(1:length(bi_peak.gr), unique(c(bi_again, bi_k27, bi_k4)) %>% sort)\n",
    "\n",
    "                bi_gene <- get_nearest_gene(gene_gr, bi_peak.gr)\n",
    "                bi_again_gene <- bi_gene %>% filter(peak_id %in% bi_again) %$% gene_name %>% unique\n",
    "                bi_k27_gene <- bi_gene %>% filter(peak_id %in% bi_k27) %$% gene_name %>% unique\n",
    "                bi_k4_gene <- bi_gene %>% filter(peak_id %in% bi_k4) %$% gene_name %>% unique\n",
    "                bi_none_gene <- setdiff(unique(bi_gene$gene_name), c(bi_again_gene, bi_k27_gene, bi_k4_gene)) %>% unique\n",
    "                # write.table(bi_again_gene, paste0(bi_path, \"BivalentPeaksOf\", cT1, \"_BivalentPeakAgainIn\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                # write.table(bi_k27_gene, paste0(bi_path, \"BivalentPeaksOf\", cT1, \"_H3K27me3In\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                # write.table(bi_k4_gene, paste0(bi_path, \"BivalentPeaksOf\", cT1, \"_H3K4me2In\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                # write.table(bi_none_gene, paste0(bi_path, \"BivalentPeaksOf\", cT1, \"_NoPeakIn\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                \n",
    "                if(i < j){\n",
    "                    bi_again_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% bi_again_gene)\n",
    "                    bi_k27_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% bi_k27_gene)\n",
    "                    bi_k4_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% bi_k4_gene)\n",
    "                    bi_none_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% bi_none_gene)\n",
    "                }else{\n",
    "                    bi_again_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$logFC) %>% filter(gene_name %in% bi_again_gene)\n",
    "                    bi_k27_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$logFC) %>% filter(gene_name %in% bi_k27_gene)\n",
    "                    bi_k4_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$logFC) %>% filter(gene_name %in% bi_k4_gene)\n",
    "                    bi_none_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$logFC) %>% filter(gene_name %in% bi_none_gene)\n",
    "\n",
    "                }\n",
    "                \n",
    "                bi_change <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    bi_original_num = length(bi_peak.gr), bi_again_num = length(bi_again), bi_k27_num = length(bi_k27), bi_k4_num = length(bi_k4), bi_none_num = length(bi_none),\n",
    "                    bi_original_gene_num = length(unique(bi_gene$gene_name)), bi_again_gene_num = length(bi_again_gene), bi_k27_gene_num = length(bi_k27_gene), bi_k4_gene_num = length(bi_k4_gene), bi_none_gene_num = length(bi_none_gene)) %>% \n",
    "                    mutate(bi_again_prop = bi_again_num/bi_original_num * 100, bi_k27_prop = bi_k27_num/bi_original_num * 100, bi_k4_prop = bi_k4_num/bi_original_num * 100, bi_none_prop = bi_none_num/bi_original_num * 100) %>% \n",
    "                    mutate(bi_again_gene_prop = bi_again_gene_num/bi_original_gene_num * 100, bi_k27_gene_prop = bi_k27_gene_num/bi_original_gene_num * 100, bi_k4_gene_prop = bi_k4_gene_num/bi_original_gene_num * 100, bi_none_gene_prop = bi_none_gene_num/bi_original_gene_num * 100) %>% \n",
    "                    rbind(bi_change, .)\n",
    "                \n",
    "                bi_change_gene_lfc <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    logFC = bi_again_gene_lfc$GEX, gene_name = bi_again_gene_lfc$gene_name, type = \"bi_again\"\n",
    "                ) %>% rbind(bi_change_gene_lfc, .)\n",
    "                bi_change_gene_lfc <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    logFC = bi_k27_gene_lfc$GEX, gene_name = bi_k27_gene_lfc$gene_name, type = \"bi_k27\"\n",
    "                ) %>% rbind(bi_change_gene_lfc, .)\n",
    "                bi_change_gene_lfc <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    logFC = bi_k4_gene_lfc$GEX, gene_name = bi_k4_gene_lfc$gene_name, type = \"bi_k4\"\n",
    "                ) %>% rbind(bi_change_gene_lfc, .)\n",
    "                if(nrow(bi_none_gene_lfc) > 0){\n",
    "                    bi_change_gene_lfc <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    logFC = bi_none_gene_lfc$GEX, gene_name = bi_none_gene_lfc$gene_name, type = \"bi_none\"\n",
    "                ) %>% rbind(bi_change_gene_lfc, .)\n",
    "                }\n",
    "                \n",
    "            }\n",
    "        }\n",
    "        \n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "bi_change %>% head\n",
    "bi_change_gene_lfc %>% head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(tidyr)\n",
    "bi_change$cT <- factor(bi_change$cT, levels = ct_list)\n",
    "bi_change$cT_compare <- factor(bi_change$cT_compare, levels = ct_list)\n",
    "\n",
    "\n",
    "p1 = bi_change %>% pivot_longer(., cols = c(\"bi_again_prop\", \"bi_k27_prop\", \"bi_k4_prop\", \"bi_none_prop\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%\n",
    "filter(cT != \"EMRA\", cT_compare != \"EMRA\", name != \"bi_none_prop\") %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = factor(name, levels = c(\"bi_again_prop\", \"bi_k27_prop\", \"bi_k4_prop\"), labels = c(\"Bivalent->Bivalent\", \"Bivalent->H3K27me3\", \"Bivalent->H3K4me2\")), label = round(value, 1))) +\n",
    "geom_boxplot(position = position_dodge(width = 0.9)) +\n",
    "stat_summary(\n",
    "    fun.y = median,\n",
    "    geom = 'line',\n",
    "    aes(group = name, colour = name),\n",
    "    position = position_dodge(width = 0.9) #this has to be added\n",
    "  ) +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "scale_color_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Peaks\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "p12 = bi_change %>% pivot_longer(., cols = c(\"bi_again_prop\", \"bi_k27_prop\", \"bi_k4_prop\", \"bi_none_prop\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%\n",
    "filter(cT != \"EMRA\", cT_compare != \"EMRA\", name != \"bi_none_prop\") %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = factor(name, levels = c(\"bi_again_prop\", \"bi_k27_prop\", \"bi_k4_prop\"), labels = c(\"Bivalent->Bivalent\", \"Bivalent->H3K27me3\", \"Bivalent->H3K4me2\")), label = round(value, 1))) +\n",
    "geom_boxplot(position = position_dodge(width = 0.9)) +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "scale_color_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Peaks\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "p2 = bi_change %>% pivot_longer(., cols = c(\"bi_again_gene_prop\", \"bi_k27_gene_prop\", \"bi_k4_gene_prop\", \"bi_none_gene_prop\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%\n",
    "filter(cT != \"EMRA\", cT_compare != \"EMRA\", name != \"bi_none_gene_prop\") %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = factor(name, levels = c(\"bi_again_gene_prop\", \"bi_k27_gene_prop\", \"bi_k4_gene_prop\"), labels = c(\"Bivalent->Bivalent\", \"Bivalent->H3K27me3\", \"Bivalent->H3K4me2\")), label = round(value, 1))) +\n",
    "geom_boxplot(position = position_dodge(width = 0.9)) +\n",
    "stat_summary(\n",
    "    fun.y = median,\n",
    "    geom = 'line',\n",
    "    aes(group = name, colour = name),\n",
    "    position = position_dodge(width = 0.9) #this has to be added\n",
    "  ) +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "scale_color_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "p22 = bi_change %>% pivot_longer(., cols = c(\"bi_again_gene_prop\", \"bi_k27_gene_prop\", \"bi_k4_gene_prop\", \"bi_none_gene_prop\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%\n",
    "filter(cT != \"EMRA\", cT_compare != \"EMRA\", name != \"bi_none_gene_prop\") %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = factor(name, levels = c(\"bi_again_gene_prop\", \"bi_k27_gene_prop\", \"bi_k4_gene_prop\"), labels = c(\"Bivalent->Bivalent\", \"Bivalent->H3K27me3\", \"Bivalent->H3K4me2\")), label = round(value, 1))) +\n",
    "geom_boxplot(position = position_dodge(width = 0.9)) +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "scale_color_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "# pdf(paste0(fig_path, \"/bivalent_peak_gene_change_between_celltypes.pdf\"), width = 10, height = 8)\n",
    "# p1\n",
    "# p12\n",
    "# p2\n",
    "# p22\n",
    "# dev.off()\n",
    "\n",
    "ggarrange(p12, p22, nrow = 2, ncol = 1, common.legend = TRUE)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "bi_change %>% \n",
    "    pivot_longer(., cols = c(\"bi_again_prop\", \"bi_k27_prop\", \"bi_k4_prop\", \"bi_none_prop\")) %>%\n",
    "    colnames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "bi_change[c(\"bi_again_prop\", \"bi_k27_prop\", \"bi_k4_prop\", \"bi_none_prop\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "bi_change_gene_lfc %>% head\n",
    "bi_change_gene_lfc$cT <- factor(bi_change_gene_lfc$cT, levels = ct_list)\n",
    "bi_change_gene_lfc$cT_compare <- factor(bi_change_gene_lfc$cT_compare, levels = ct_list)\n",
    "\n",
    "bi_change_gene_lfc %>% filter(cT != \"EMRA\", cT_compare != \"EMRA\", type != \"bi_none\") %>%\n",
    "ggplot(aes(cT_compare, y = logFC, fill = factor(type, levels = c(\"bi_again\", \"bi_k27\", \"bi_k4\"), labels = c(\"Bivalent->Bivalent\", \"Bivalent->H3K27me3\", \"Bivalent->H3K4me2\")))) +\n",
    "geom_boxplot() +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "xlab(\"\") +\n",
    "ylab(\"logFC of GEX\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "# ggsave(paste0(fig_path, \"/bivalent_peak_gene_change_between_celltypes_gex_logfc.pdf\"), width = 10, height = 8)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(dplyr)\n",
    "library(ggplot2)\n",
    "library(rstatix)\n",
    "# tmp = bi_change_gene_lfc\n",
    "# bi_change_gene_lfc = tmp\n",
    "# bi_change_gene_lfc$type = factor(bi_change_gene_lfc$type, levels = c(\"bi_again\", \"bi_k27\", \"bi_k4\", \"bi_none\"), \n",
    "#              labels = c(\"Bivalent->Bivalent\", \"Bivalent->H3K27me3\", \"Bivalent->H3K4me2\", \"Bivalent->None\"))\n",
    "\n",
    "bi_change_gene_lfc %>% \n",
    "  filter(cT != \"EMRA\", cT_compare != \"EMRA\", type != \"Bivalent->None\") %>%\n",
    "  ggplot(aes(cT_compare, y = logFC, fill = type)) +\n",
    "  geom_boxplot() +\n",
    "  facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "  stat_compare_means(method = \"anova\", label = \"p.format\", \n",
    "                     symnum.args = list(cutpoints = c(0, 0.001, 0.01, 0.05, 1), \n",
    "                                        symbols = c(\"***\", \"**\", \"*\", \"ns\"))) + \n",
    "  theme_bw(base_size = 20) +\n",
    "  scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "  xlab(\"\") +\n",
    "  ylab(\"logFC of GEX\") +\n",
    "  rremove(\"legend.title\") +\n",
    "  theme(legend.position = \"top\")\n",
    "\n",
    "ggsave(paste0(fig_path, \"/bivalent_peak_gene_change_between_celltypes_gex_logfc_withANOVAtestPValues.pdf\"), width = 10, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "lfc_thres = 1\n",
    "bi_change_gene_lfc$logFC_group = \"nonDE\"\n",
    "bi_change_gene_lfc$logFC_group[bi_change_gene_lfc$logFC < -lfc_thres] = \"down_regulated\"\n",
    "bi_change_gene_lfc$logFC_group[bi_change_gene_lfc$logFC > lfc_thres] = \"up_regulated\"\n",
    "tmp1 = bi_change_gene_lfc %>% group_by(cT, cT_compare, pt, type, logFC_group) %>% summarize(gene_num = n())\n",
    "tmp2 = bi_change_gene_lfc %>% group_by(cT, cT_compare, pt, type) %>% summarize(gene_total_num = n())\n",
    "bi_change_gene_lfc_group = left_join(tmp1, tmp2, by = c(\"cT\", \"cT_compare\", \"pt\", \"type\")) %>% mutate(gene_prop = gene_num/gene_total_num * 100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=25, repr.plot.height=8)\n",
    "fig_path = \"CART_CUTRUN_Project/results/paper_figure/fig5\"\n",
    "\n",
    "p_gene_prop = bi_change_gene_lfc_group %>% filter(cT != \"EMRA\", cT_compare != \"EMRA\", type != \"bi_none\") %>% ggplot(aes(x = cT_compare, y = gene_prop, fill = type)) +\n",
    "geom_boxplot(aes(alpha = logFC_group)) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.5) +\n",
    "# scale_color_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 1) +\n",
    "theme_bw(base_size = 30) +\n",
    "facet_grid(~cT, scales = \"free\") +\n",
    "ylab(\"% of Genes\") +\n",
    "scale_alpha_discrete(range = c(0.3, 1)) +\n",
    "xlab(\"\") +\n",
    "rremove(\"legend\")\n",
    "\n",
    "p_gene_prop\n",
    "ggsave(paste0(fig_path, \"/bivalent_gene_expression_change_proportion_logFC_thres1.pdf\"), width = 9, height = 8)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "p22\n",
    "ggsave(paste0(fig_path, \"/bivalent_peak_gene_change_between_celltypes_geneprop.pdf\"), width = 9, height = 8)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "lfc_thres = 2\n",
    "bi_change_gene_lfc$logFC_group = \"nonDE\"\n",
    "bi_change_gene_lfc$logFC_group[bi_change_gene_lfc$logFC < -lfc_thres] = \"DE\"\n",
    "bi_change_gene_lfc$logFC_group[bi_change_gene_lfc$logFC > lfc_thres] = \"DE\"\n",
    "tmp1 = bi_change_gene_lfc %>% group_by(cT, cT_compare, pt, type, logFC_group) %>% summarize(gene_num = n())\n",
    "tmp2 = bi_change_gene_lfc %>% group_by(cT, cT_compare, pt) %>% summarize(gene_total_num = n())\n",
    "bi_change_gene_lfc_group = left_join(tmp1, tmp2, by = c(\"cT\", \"cT_compare\", \"pt\")) %>% mutate(gene_prop = gene_num/gene_total_num * 100)\n",
    "bi_change_gene_lfc_group$type = factor(bi_change_gene_lfc_group$type, levels = c(\"bi_again\", \"bi_k27\", \"bi_k4\"), labels = c(\"Bivalent->Bivalent\", \"Bivalent->H3K27me3\", \"Bivalent->H3K4me2\"))\n",
    "\n",
    "bi_change_gene_lfc_group %>% head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "p_gene_prop = bi_change_gene_lfc_group %>% filter(cT != \"EMRA\", cT_compare != \"EMRA\", type != \"bi_none\", logFC_group == \"DE\") %>% \n",
    "ggplot(aes(x = cT_compare, y = gene_prop, fill = type)) +\n",
    "geom_boxplot() +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "# scale_color_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 1) +\n",
    "theme_bw(base_size = 20) +\n",
    "facet_grid(~cT, scales = \"free\") +\n",
    "ylab(\"% of Genes\") +\n",
    "scale_alpha_discrete(range = c(0.3, 1)) +\n",
    "xlab(\"\") +\n",
    "theme(legend.position = \"top\") +\n",
    "ggpubr::rremove(\"legend.title\") +\n",
    "ylim(0, 60)\n",
    "\n",
    "p_gene_prop\n",
    "ggsave(paste0(fig_path, \"/bivalent_gene_DE_proportion_logFC_thres2.pdf\"), width = 9, height = 9)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## PCA for RNA and CUTRUN\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\")) ## 'cell_comp_list''cell_list''countMat''designInfo''expr_list''fig_path''hd_list''hist_list''in_path''inPath''out_path''outPath'\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/RData/HD1-3_5-7_normLimma_perExprCondition.RData\") ## data, selectR, dataS, voomDDS, results, \n",
    "\n",
    "geneID = fread(paste0(\"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\", \"RNA_CD8_N_Input_HD1.genes.results\"))$gene_id\n",
    "data = data.frame(data)\n",
    "select_exp = \"Product\"\n",
    "\n",
    "rna_m = data[, (colnames(data) %>% stringr::str_detect(select_exp)) & !(colnames(data) %>% stringr::str_detect(\"EMRA\"))  & !(colnames(data) %>% stringr::str_detect(\"N\"))]\n",
    "k27_m = countMat[[\"H3K27me3\"]][, (colnames(countMat[[\"H3K27me3\"]]) %>% stringr::str_detect(select_exp)) & !(colnames(countMat[[\"H3K27me3\"]]) %>% stringr::str_detect(\"EMRA\")) & !(colnames(countMat[[\"H3K27me3\"]]) %>% stringr::str_detect(\"N\"))]\n",
    "k4_m = countMat[[\"H3K4me2\"]][, (colnames(countMat[[\"H3K4me2\"]]) %>% stringr::str_detect(select_exp)) & !(colnames(countMat[[\"H3K4me2\"]]) %>% stringr::str_detect(\"EMRA\")) & !(colnames(countMat[[\"H3K4me2\"]]) %>% stringr::str_detect(\"N\"))]\n",
    "\n",
    "rownames(rna_m) = geneID\n",
    "rownames(k27_m) = mPeak[[\"H3K27me3\"]] %>% data.frame %>% mutate(region = paste0(seqnames, \":\", start, \"-\", end)) %$% region\n",
    "rownames(k4_m) = mPeak[[\"H3K4me2\"]] %>% data.frame %>% mutate(region = paste0(seqnames, \":\", start, \"-\", end)) %$% region\n",
    "\n",
    "dim(rna_m)\n",
    "dim(k27_m)\n",
    "dim(k4_m)\n",
    "head(rna_m)\n",
    "head(k27_m)\n",
    "head(k4_m)\n",
    "\n",
    "rna_voom = voomDDS$E[, colnames(voomDDS$E) %>% stringr::str_detect(select_exp)]\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\"))\n",
    "k27_voom = voomDDS[[\"H3K27me3\"]]$E[, colnames(voomDDS[[\"H3K27me3\"]]$E) %>% stringr::str_detect(select_exp)]\n",
    "k4_voom = voomDDS[[\"H3K4me2\"]]$E[, colnames(voomDDS[[\"H3K4me2\"]]$E) %>% stringr::str_detect(select_exp)]\n",
    "rownames(rna_voom) = 1:nrow(rna_voom)\n",
    "rownames(k27_voom) = paste0(\"K27me3_\", 1:nrow(k27_voom))\n",
    "rownames(k4_voom) = paste0(\"K4me2_\", 1:nrow(k4_voom))\n",
    "\n",
    "dim(rna_voom)\n",
    "dim(k27_voom)\n",
    "dim(k4_voom)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "rna_obj_tmp = CreateSeuratObject(counts = rna_m, project = \"cart\", min.cells = 0, min.features = 10) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) \n",
    "\n",
    "rna_obj = rna_obj_tmp %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = 1000) %>% \n",
    "ScaleData()\n",
    "rna_obj = RunPCA(rna_obj, features = VariableFeatures(object = rna_obj), npcs = 10)\n",
    "\n",
    "k27_obj_tmp = CreateSeuratObject(counts = k27_m, project = \"cart\", min.cells = 0, min.features = 10) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000)\n",
    "\n",
    "k27_obj = k27_obj_tmp %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = 10000) %>%  #\n",
    "ScaleData()\n",
    "\n",
    "k27_obj = RunPCA(k27_obj, features = VariableFeatures(object = k27_obj), npcs = 12)\n",
    "\n",
    "k4_obj_tmp = CreateSeuratObject(counts = k4_m, project = \"cart\", min.cells = 0, min.features = 10) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) \n",
    "k4_obj = k4_obj_tmp %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = 1000) %>%  #round(nrow(k4_m)/10*8)\n",
    "ScaleData()\n",
    "k4_obj = RunPCA(k4_obj, features = VariableFeatures(object = k4_obj), npcs = 12)\n",
    "\n",
    "# cr_obj = CreateSeuratObject(counts = rbind(k27_m, k4_m), project = \"cart\", min.cells = 0, min.features = 10) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "# FindVariableFeatures(., selection.method = \"vst\", nfeatures = 500) %>%  #round((nrow(k27_m) + nrow(k4_m))/10*8)\n",
    "# ScaleData()\n",
    "# cr_obj = RunPCA(cr_obj, features = VariableFeatures(object = cr_obj), npcs = 12)\n",
    "\n",
    "\n",
    "cr_obj = CreateSeuratObject(counts = rbind(k27_obj_tmp@assays$RNA@data, k4_obj_tmp@assays$RNA@data), project = \"cart\", min.cells = 0, min.features = 10) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = 1000) %>%  #round((nrow(k27_m) + nrow(k4_m))/10*8)\n",
    "ScaleData()\n",
    "cr_obj = RunPCA(cr_obj, features = VariableFeatures(object = cr_obj), npcs = 12)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=7)\n",
    "# DimPlot(rna_obj, reduction = \"pca\")\n",
    "ct_list = c(\"CM\", \"EM\")\n",
    "p1 = Embeddings(rna_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(hd = factor(rep(paste0(\"HD\", c(1, 2, 3, 5, 6, 7)), 2))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = hd)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_brewer(palette = \"Paired\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"RNA-seq\") +\n",
    "rremove(\"legend.title\")+ theme(legend.position = \"top\")\n",
    "\n",
    "p2 = Embeddings(k27_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(hd = factor(rep(paste0(\"HD\", 1:7), 2))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = hd)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_brewer(palette = \"Paired\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3\") +\n",
    "rremove(\"legend.title\")+ theme(legend.position = \"top\")\n",
    "\n",
    "p3 = Embeddings(k4_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(hd = factor(rep(paste0(\"HD\", 1:7), 2))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = hd)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_brewer(palette = \"Paired\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K4me2\") +\n",
    "rremove(\"legend.title\")+ theme(legend.position = \"top\")\n",
    "\n",
    "p4 = Embeddings(cr_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(hd = factor(rep(paste0(\"HD\", 1:7), 2))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = hd)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_brewer(palette = \"Paired\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3+H3K4me2\") +\n",
    "rremove(\"legend.title\")+ theme(legend.position = \"top\")\n",
    "\n",
    "ggarrange(p1, p2, p3, p4, nrow = 1, ncol = 4, common.legend = FALSE) \n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig6/PCA_seurat_procedure_CAR-T_Product_CM_EM_colorbyHD.pdf\", width = 20, height = 7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=15, repr.plot.height=5)\n",
    "# DimPlot(rna_obj, reduction = \"pca\")\n",
    "ct_list = c(\"CM\", \"EM\")\n",
    "p1 = Embeddings(rna_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 6), levels = ct_list)) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"RNA-seq\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p2 = Embeddings(k27_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 7), levels = ct_list)) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p3 = Embeddings(k4_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 7), levels = ct_list)) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p4 = Embeddings(cr_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(rep(ct_list, each = 7), levels = ct_list)) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.75) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3+H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "ggarrange(p1, p2, p3, p4, nrow = 1, ncol = 4, common.legend = TRUE)\n",
    "# ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig6/PCA_seurat_procedure_CAR-T_Product_CM_EM.pdf\", width = 20, height = 6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## patient related\n",
    "\n",
    "## Patient data infusion product of CM-derived cells \n",
    "patientRaw = read_xlsx(\"CART_CUTRUN_Project/report/CART_CUTRUN/analysis/20201028_NHL_CUTandRUN_PatientData.xlsx\")\n",
    "cm_ind = which(patientRaw$`CD8+ T cell isolation` == \"CD8TCM\")\n",
    "patient_id_list = patientRaw$Xnumber[cm_ind] %>% tolower\n",
    "patient_id_list\n",
    "\n",
    "## Patient data infusion product of CM-derived cells \n",
    "inPath = \"CART_CUTRUN_Project/results/RNAseq/process/NHL_RSEM_gencode_v21/\"\n",
    "\n",
    "pt_expCount = c()\n",
    "pt_name_tmp_list = c()\n",
    "for(sample in paste0(\"Sample_\", patient_id_list, \"_IP_CD8-CAR\")){\n",
    "  if(file.exists(paste0(inPath, sample, \"/RSEM.genes.results\"))){\n",
    "    print(sample)\n",
    "    tmp = fread(paste0(inPath, sample, \"/RSEM.genes.results\")) %>% dplyr::select(expected_count)\n",
    "    pt_expCount = cbind(pt_expCount, round(tmp))\n",
    "    pt_name_tmp_list = c(pt_name_tmp_list, sample)\n",
    "  }\n",
    "}\n",
    "pt_expCount = data.frame(pt_expCount)\n",
    "colnames(pt_expCount) = paste0(pt_name_tmp_list, \"_PT\") #paste0(patient_id_list, \"_PT\")\n",
    "\n",
    "geneID = fread(paste0(inPath, sample, \"/RSEM.genes.results\"))$gene_id\n",
    "pt_expCount$gene_name = geneID\n",
    "\n",
    "head(pt_expCount)\n",
    "dim(pt_expCount)\n",
    "\n",
    "## healthy donor data\n",
    "## Read in RNA-seq data\n",
    "cellList <- \"CM\" #c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hdList <- paste0(\"HD\", c(1:3, 5:7))\n",
    "exprList = \"Product\" #c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "\n",
    "inPath <- \"CART_CUTRUN_Project/results/RNAseq/process/HD_RSEM_gencode_v21/\" #\"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "\n",
    "hd_expCount = c()\n",
    "for(cell in cellList){\n",
    "  for(expr in exprList){\n",
    "    for(hd in hdList){\n",
    "      # print(paste(expr, cell, hd, sep = \"_\"))\n",
    "      sample = paste0(\"RNA_CD8_\", cell, \"_\", expr, \"_\", hd)\n",
    "      if(file.exists(paste0(inPath, sample, \"/RSEM.genes.results\"))){\n",
    "        print(sample)\n",
    "        dataTmp = fread(paste0(inPath, sample, \"/RSEM.genes.results\")) %>% dplyr::select(expected_count)\n",
    "        # dataTmp <- fread(paste0(inPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".genes.results\")) %>% dplyr::select(expected_count)\n",
    "        hd_expCount <- cbind(hd_expCount, round(dataTmp))\n",
    "      }\n",
    "\n",
    "    }\n",
    "  }\n",
    "\n",
    "}\n",
    "hd_expCount = data.frame(hd_expCount)\n",
    "colnames(hd_expCount) <- paste(rep(rep(exprList, each = length(hdList)), length(cellList)), rep(cellList, each = length(hdList)*length(exprList)), rep(hdList, length(exprList)*length(cellList)), sep = \"_\")\n",
    "\n",
    "geneID = fread(paste0(inPath, sample, \"/RSEM.genes.results\"))$gene_id\n",
    "hd_expCount$gene_name <- geneID\n",
    "head(hd_expCount)\n",
    "dim(hd_expCount)\n",
    "\n",
    "dim(pt_expCount)\n",
    "dim(hd_expCount)\n",
    "countMat = inner_join(pt_expCount, hd_expCount, by = \"gene_name\")\n",
    "rownames(countMat) = countMat$gene_name\n",
    "countMat = countMat %>% dplyr::select(-gene_name)\n",
    "head(countMat)\n",
    "dim(countMat)\n",
    "\n",
    "countMat_rna = countMat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=7, repr.plot.height=6)\n",
    "selectR = which(rowSums(countMat_rna) > 10) ## remove low count genes\n",
    "length(selectR)\n",
    "\n",
    "rna_obj = CreateSeuratObject(counts = countMat_rna[selectR, ], project = \"cart\", min.cells = 0, min.features = 10) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = 50000) %>%  # length(selectR)\n",
    "ScaleData()\n",
    "\n",
    "rna_obj = RunPCA(rna_obj, features = VariableFeatures(object = rna_obj), npcs = 10)\n",
    "rna_plot = Embeddings(rna_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", ncol(pt_expCount) - 1), rep(\"HD\", ncol(hd_expCount) - 1))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"RNA-seq\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "# pdf(\"CART_CUTRUN_Project/results/paper_figure/fig1/CMpatient_CMhealthyDonor_PCA_on_RNA-seq_allfeatures.pdf\", width = 7, height = 6)\n",
    "# rna_plot\n",
    "# dev.off()\n",
    "rna_plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "patientRaw = read_xlsx(\"CART_CUTRUN_Project/report/CART_CUTRUN/analysis/20201028_NHL_CUTandRUN_PatientData.xlsx\")\n",
    "cm_ind = which(patientRaw$`CD8+ T cell isolation` == \"CD8TCM\")\n",
    "patientRaw$Xnumber[cm_ind]\n",
    "\n",
    "## 1. TSS neighbor regions\n",
    "## use TSS region for differential detection\n",
    "## Get the TSS location and +/-1kb region around TSS\n",
    "gtf <- rtracklayer::import('/fh/fast/gottardo_r/yezheng_working/SupplementaryData/hg38/transcriptomeHumanReference/gencode.v33.annotation.gtf') #rtracklayer::import('/fh/fast/gottardo_r/yezheng_working/SupplementaryData/hg38/GENCODE/gencode.v21.annotation.gtf')\n",
    "gtf_df=as.data.frame(gtf)\n",
    "\n",
    "\n",
    "## get the TSS location and extend by 1kb up and downstream\n",
    "tss_df = rbind(\n",
    "    gtf_df %>% data.frame %>% dplyr::filter(type == \"gene\", strand == \"+\") %>% dplyr::select(seqnames, TSS = start, strand, source, gene_id, gene_name, gene_name, gene_type) ,\n",
    "    gtf_df %>% data.frame %>% dplyr::filter(type == \"gene\", strand == \"-\") %>% dplyr::select(seqnames, TSS = end, strand, source, gene_id, gene_name, gene_name, gene_type) \n",
    ")\n",
    "tss_filter_df = tss_df %>% dplyr::filter(!(seqnames %in% c(\"chrM\", \"chrX\", \"chrY\")))\n",
    "\n",
    "tss_gr = GRanges(seqnames = tss_filter_df$seqnames, ranges = IRanges(start = tss_filter_df$TSS - 1000, end = tss_filter_df$TSS + 1000), strand = tss_filter_df$strand, gene_id = tss_filter_df$gene_id, gene_name = tss_filter_df$gene_name, gene_type = tss_filter_df$gene_type)\n",
    "tss_gr \n",
    "\n",
    "## patient data\n",
    "\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "histList = c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "countMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_NHLpatient_noChrXYM_countMat.rds\"))\n",
    "normMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_NHLpatient_noChrXYM_normMat.rds\"))\n",
    "designInfo = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_NHLpatient_noChrXYM_designInfo.rds\"))\n",
    "\n",
    "pt_countMat = countMat\n",
    "pt_countMat$H3K27me3 = countMat$H3K27me3[, cm_ind]\n",
    "pt_countMat$H3K4me2 = countMat$H3K4me2[, cm_ind]\n",
    "colnames(pt_countMat$H3K27me3) = paste0(colnames(pt_countMat$H3K27me3), \"_PT\")\n",
    "colnames(pt_countMat$H3K4me2) = paste0(colnames(pt_countMat$H3K4me2), \"_PT\")\n",
    "\n",
    "pt_designInfo = designInfo %>% dplyr::filter(exps %in% patientRaw$Xnumber[cm_ind])\n",
    "head(pt_countMat$H3K27me3)\n",
    "pt_designInfo %>% arrange(hist, exps)\n",
    "\n",
    "## healthy donor data\n",
    "countMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_countMat.rds\"))\n",
    "normMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_normMat.rds\"))\n",
    "designInfo = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_designInfo.rds\"))\n",
    "\n",
    "selectC = colnames(countMat$H3K27me3) %>% stringr::str_detect(\"CM_Product\")\n",
    "head(countMat$H3K4me2[, selectC])\n",
    "\n",
    "hd_countMat = countMat\n",
    "hd_countMat$H3K27me3 = countMat$H3K27me3[, selectC]\n",
    "hd_countMat$H3K4me2 = countMat$H3K4me2[, selectC]\n",
    "\n",
    "hd_designInfo = designInfo[which(stringr::str_detect(designInfo$exps, \"CM_Product\")), ]\n",
    "hd_designInfo\n",
    "\n",
    "for(hist in histList){\n",
    "    tmp = c()\n",
    "\n",
    "    for(j in 1:ncol(pt_countMat[[hist]])){\n",
    "        pt_designInfo_tmp = pt_designInfo %>% dplyr::filter(hist == hist)\n",
    "        tmp = cbind(tmp, round(pt_countMat[[hist]][, j]/pt_designInfo_tmp$depth[j] * 13240732)) ## remove sequencing depth effect\n",
    "    }\n",
    "    for(j in 1:ncol(hd_countMat[[hist]])){\n",
    "        hd_designInfo_tmp = hd_designInfo %>% dplyr::filter(hist == hist)\n",
    "        tmp = cbind(tmp, round(hd_countMat[[hist]][, j]/hd_designInfo_tmp$depth[j] * 13240732)) ## remove sequencing depth effect\n",
    "    }\n",
    "    countMat[[hist]] = tmp\n",
    "    colnames(countMat[[hist]]) = c(colnames(pt_countMat[[hist]]), colnames(hd_countMat[[hist]]))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=22, repr.plot.height=10)\n",
    "hist = \"H3K27me3\"\n",
    "selectR = which(rowSums(countMat[[hist]]) > 1000) ## remove low count genes\n",
    "\n",
    "tmp = countMat[[hist]][selectR, ]\n",
    "\n",
    "k27_obj = CreateSeuratObject(counts = countMat[[hist]][selectR, ], project = \"cart\", min.cells = 0, min.features = 10) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = round(length(selectR)/2)) %>% \n",
    "ScaleData()\n",
    "k27_obj = RunPCA(k27_obj, features = VariableFeatures(object = k27_obj), npcs = 15)\n",
    "k27_plot = Embeddings(k27_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "hist = \"H3K4me2\"\n",
    "selectR = which(rowSums(countMat[[hist]]) > 1000) ## remove low count genes\n",
    "length(selectR)\n",
    "k4_obj = CreateSeuratObject(counts = countMat[[hist]][selectR, ], project = \"cart\", min.cells = 0, min.features = 10) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = round(length(selectR)/2)) %>% \n",
    "ScaleData()\n",
    "k4_obj = RunPCA(k4_obj, features = VariableFeatures(object = k4_obj), npcs = 15)\n",
    "\n",
    "k4_plot = Embeddings(k4_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "cr_obj = CreateSeuratObject(counts = rbind(tmp, countMat[[hist]][selectR, ]), project = \"cart\", min.cells = 0, min.features = 10) %>%  \n",
    "# NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 1000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"vst\", nfeatures = round((length(selectR) + nrow(tmp))/1)) %>% \n",
    "ScaleData()\n",
    "cr_obj = RunPCA(cr_obj, features = VariableFeatures(object = cr_obj), npcs = 15)\n",
    "cr_plot = Embeddings(cr_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(group = c(rep(\"PT\", 15), rep(\"HD\", 7))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = group)) +\n",
    "geom_point(size = 5) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE, begin = 0.2, end = 0.8, option = \"magma\") +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3 + H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "\n",
    "pdf(\"CART_CUTRUN_Project/results/paper_figure/fig6/CMpatient_CMhealthyDonor_PCA_on_all_TSSneighbors.pdf\", width = 15, height = 6)\n",
    "ggarrange(k27_plot, k4_plot, cr_plot, common.legend = TRUE, ncol = 3, nrow = 1)\n",
    "dev.off()\n",
    "ggarrange(k27_plot, k4_plot, cr_plot, common.legend = TRUE, ncol = 3, nrow = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "metadata_path = \"CART_CUTRUN_Project/meta/\"\n",
    "clinical_data = fread(paste0(metadata_path, \"NHL_DL2_FluCy_Propscore_linear_model_CandR_only_pfs.csv\"), sep = \",\", header = TRUE) %>% data.frame\n",
    "head(clinical_data)\n",
    "\n",
    "clinical_data %>% colnames\n",
    "clinical_select = c(\"xnumber\", \"cd8_max\", \"cd4_max\", \"log_cd8_max\", \"propscore_cd8_max\", \"propscore\", \"median_cd8_max\", \"preld_ldh\", \"pfscode\",\t\"pfsdaysafterfirstcartcell\", \"pfs_late_code\", \"log_pfs_days\")\n",
    "clinical_data_select = clinical_data[, clinical_select]\n",
    "clinical_data_select"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "marker_gene_table = fread(paste0(metadata_path, \"Table_S3.csv\"), sep = \",\", header = TRUE)\n",
    "marker_gene = marker_gene_table$GeneNameClean\n",
    "k27_TF_gene = marker_gene_table %>% filter(Analysis == \"K27\") %$% GeneNameClean\n",
    "k4_TF_gene = marker_gene_table %>% filter(Analysis == \"K4\") %$% GeneNameClean\n",
    "rna_TF_gene = marker_gene_table %>% filter(Analysis == \"RNA\") %$% GeneNameClean\n",
    "\n",
    "marker_gene_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "patientRaw = read_xlsx(\"CART_CUTRUN_Project/report/CART_CUTRUN/analysis/20201028_NHL_CUTandRUN_PatientData.xlsx\")\n",
    "cm_ind = which(patientRaw$`CD8+ T cell isolation` == \"CD8TCM\")\n",
    "patientRaw$Xnumber[cm_ind]\n",
    "\n",
    "## patient data\n",
    "\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "histList = c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "countMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_NHLpatient_noChrXYM_countMat.rds\"))\n",
    "normMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_NHLpatient_noChrXYM_normMat.rds\")) ## RC\n",
    "designInfo = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_NHLpatient_noChrXYM_designInfo.rds\"))\n",
    "\n",
    "# pt_countMat = countMat\n",
    "# pt_countMat$H3K27me3 = countMat$H3K27me3[, cm_ind]\n",
    "# pt_countMat$H3K4me2 = countMat$H3K4me2[, cm_ind]\n",
    "# colnames(pt_countMat$H3K27me3) = paste0(colnames(pt_countMat$H3K27me3), \"_PT\")\n",
    "# colnames(pt_countMat$H3K4me2) = paste0(colnames(pt_countMat$H3K4me2), \"_PT\")\n",
    "\n",
    "pt_normMat = normMat\n",
    "pt_normMat$H3K27me3 = log2(normMat$H3K27me3[, patientRaw$Xnumber[cm_ind]]+1)\n",
    "pt_normMat$H3K4me2 = log2(normMat$H3K4me2[, patientRaw$Xnumber[cm_ind]]+1)\n",
    "# colnames(pt_normMat$H3K27me3) = paste0(colnames(pt_normMat$H3K27me3), \"_PT\")\n",
    "# colnames(pt_normMat$H3K4me2) = paste0(colnames(pt_normMat$H3K4me2), \"_PT\")\n",
    "\n",
    "\n",
    "pt_designInfo = designInfo %>% dplyr::filter(exps %in% patientRaw$Xnumber[cm_ind])\n",
    "head(pt_normMat$H3K27me3)\n",
    "pt_designInfo %>% arrange(hist, exps)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=25, repr.plot.height=20)\n",
    "\n",
    "pt_select_tmp = pt_normMat$H3K27me3[marker_gene, ] %>% data.frame\n",
    "pt_select = pt_select_tmp %>% mutate(gene_num = rownames(pt_select_tmp)) %>% pivot_longer(cols = colnames(pt_select_tmp), names_to = \"xnumber\")\n",
    "head(pt_select, 20)\n",
    "pt_select %>% dim\n",
    "\n",
    "pt_clinical = left_join(left_join(pt_select, clinical_data_select, by = \"xnumber\"), marker_gene_table[, c(\"Analysis\", \"GeneNameClean\")], by = c(\"gene_num\" = \"GeneNameClean\"))\n",
    "pt_clinical %>% head\n",
    "\n",
    "pt_clinical %>% ggplot(aes(x = as.factor(propscore_cd8_max), y = value)) +\n",
    "geom_boxplot(aes(fill = Analysis)) +\n",
    "theme_bw(base_size = 25) +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 10, scales = \"free\") + \n",
    "stat_compare_means(vjust = 1) +\n",
    "scale_fill_brewer(palette = \"Dark2\") +\n",
    "xlab(\"Propensity Score CD8 Max\") +\n",
    "ylab(\"H3K27me3: log2(Scaled Count + 1)\")\n",
    "\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K27me3_signal_patient_propscore_cd8_max.pdf\", width = 25, height = 18)\n",
    "\n",
    "pt_clinical %>% filter(gene_num %in% k27_TF_gene) %>% ggplot(aes(x = as.factor(propscore_cd8_max), y = value)) +\n",
    "geom_boxplot(aes(fill = as.factor(propscore_cd8_max))) +\n",
    "theme_bw(base_size = 25) +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 5, scales = \"free\") + \n",
    "stat_compare_means(vjust = 1, size = 6.5) +\n",
    "scale_fill_brewer(palette = \"Paired\") +\n",
    "xlab(\"Propensity Score CD8 Max\") +\n",
    "ylab(\"H3K27me3: log2(Scaled Count + 1)\") +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K27me3_signal_patient_propscore_cd8_max_k27DEgenes.pdf\", width = 18, height = 18)\n",
    "\n",
    "pt_clinical %>% ggplot(aes(x = log_cd8_max, y = value)) +\n",
    "geom_point(aes(color = Analysis)) +\n",
    "geom_smooth(aes(color = Analysis), method = \"lm\") +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_color_brewer(palette = \"Dark2\") +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 10, scales = \"free\") +\n",
    "xlab(\"Log CD8 Max\") +\n",
    "ylab(\"H3K27me3: log2(Scaled Count + 1)\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K27me3_signal_patient_log_cd8_max.pdf\", width = 25, height = 18)\n",
    "\n",
    "pt_clinical %>% filter(gene_num %in% k27_TF_gene) %>% ggplot(aes(x = log_cd8_max, y = value)) +\n",
    "geom_point(aes(color = log_cd8_max)) +\n",
    "geom_smooth(color = \"grey\", method = \"lm\", se = FALSE) +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_color_viridis() +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 5, scales = \"free\") +\n",
    "xlab(\"Log CD8 Max\") +\n",
    "ylab(\"H3K27me3: log2(Scaled Count + 1)\") +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K27me3_signal_patient_log_cd8_max_k27DEgenes.pdf\", width = 20, height = 14)\n",
    "\n",
    "## \"pfscode\",\t\"pfsdaysafterfirstcartcell\", \"pfs_late_code\", \"log_pfs_days\"\n",
    "\n",
    "options(repr.plot.width=25, repr.plot.height=20)\n",
    "\n",
    "pt_clinical %>% ggplot(aes(x = as.factor(pfs_late_code), y = value)) +\n",
    "geom_boxplot(aes(fill = Analysis)) +\n",
    "theme_bw(base_size = 25) +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 10, scales = \"free\") + \n",
    "stat_compare_means(vjust = 1) +\n",
    "scale_fill_brewer(palette = \"Dark2\") +\n",
    "xlab(\"PFS Late Code\") +\n",
    "ylab(\"H3K27me3: log2(Scaled Count + 1)\")\n",
    "\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K27me3_signal_patient_PFS_late_code.pdf\", width = 25, height = 18)\n",
    "\n",
    "pt_clinical %>% filter(gene_num %in% k27_TF_gene) %>% ggplot(aes(x = as.factor(pfs_late_code), y = value)) +\n",
    "geom_boxplot(aes(fill = as.factor(pfs_late_code))) +\n",
    "theme_bw(base_size = 25) +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 5, scales = \"free\") + \n",
    "stat_compare_means(vjust = 1, size = 6.5) +\n",
    "scale_fill_brewer(palette = \"Paired\") +\n",
    "xlab(\"PFS Late Code\") +\n",
    "ylab(\"H3K27me3: log2(Scaled Count + 1)\") +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K27me3_signal_patient_PFS_late_code_k27DEgenes.pdf\", width = 18, height = 18)\n",
    "\n",
    "pt_clinical %>% ggplot(aes(x = log_pfs_days, y = value)) +\n",
    "geom_point(aes(color = Analysis)) +\n",
    "geom_smooth(aes(color = Analysis), method = \"lm\") +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_color_brewer(palette = \"Dark2\") +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 10, scales = \"free\") +\n",
    "xlab(\"Log PFS Days\") +\n",
    "ylab(\"H3K27me3: log2(Scaled Count + 1)\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K27me3_signal_patient_log_pfs_days.pdf\", width = 25, height = 18)\n",
    "\n",
    "pt_clinical %>% filter(gene_num %in% k27_TF_gene) %>% ggplot(aes(x = log_pfs_days, y = value)) +\n",
    "geom_point(aes(color = log_pfs_days)) +\n",
    "geom_smooth(color = \"grey\", method = \"lm\", se = FALSE) +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_color_viridis() +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 5, scales = \"free\") +\n",
    "xlab(\"Log PFS Days\") +\n",
    "ylab(\"H3K27me3: log2(Scaled Count + 1)\") +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K27me3_signal_patient_log_pfs_days_k27DEgenes.pdf\", width = 20, height = 14)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=25, repr.plot.height=20)\n",
    "\n",
    "pt_select_tmp = pt_normMat$H3K4me2[marker_gene, ] %>% data.frame\n",
    "pt_select = pt_select_tmp %>% mutate(gene_num = rownames(pt_select_tmp)) %>% pivot_longer(cols = colnames(pt_select_tmp), names_to = \"xnumber\")\n",
    "head(pt_select, 20)\n",
    "pt_select %>% dim\n",
    "\n",
    "pt_clinical = left_join(left_join(pt_select, clinical_data_select, by = \"xnumber\"), marker_gene_table[, c(\"Analysis\", \"GeneNameClean\")], by = c(\"gene_num\" = \"GeneNameClean\"))\n",
    "pt_clinical %>% head\n",
    "\n",
    "pt_clinical %>% ggplot(aes(x = as.factor(propscore_cd8_max), y = value)) +\n",
    "geom_boxplot(aes(fill = Analysis)) +\n",
    "theme_bw(base_size = 25) +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 10, scales = \"free\") + \n",
    "stat_compare_means(vjust = 1) +\n",
    "scale_fill_brewer(palette = \"Dark2\") +\n",
    "xlab(\"Propensity Score CD8 Max\") +\n",
    "ylab(\"H3K4me2: log2(Scaled Count + 1)\")\n",
    "\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K4me2_signal_patient_propscore_cd8_max.pdf\", width = 25, height = 18)\n",
    "\n",
    "pt_clinical %>% filter(gene_num %in% k4_TF_gene) %>% ggplot(aes(x = as.factor(propscore_cd8_max), y = value)) +\n",
    "geom_boxplot(aes(fill = as.factor(propscore_cd8_max))) +\n",
    "theme_bw(base_size = 25) +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 5, scales = \"free\") + \n",
    "stat_compare_means(vjust = 1, size = 6.5) +\n",
    "scale_fill_brewer(palette = \"Paired\") +\n",
    "xlab(\"Propensity Score CD8 Max\") +\n",
    "ylab(\"H3K4me2: log2(Scaled Count + 1)\") +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K4me2_signal_patient_propscore_cd8_max_k4DEgenes.pdf\", width = 18, height = 18)\n",
    "\n",
    "pt_clinical %>% ggplot(aes(x = log_cd8_max, y = value)) +\n",
    "geom_point(aes(color = Analysis)) +\n",
    "geom_smooth(aes(color = Analysis), method = \"lm\") +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_color_brewer(palette = \"Dark2\") +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 10, scales = \"free\") +\n",
    "xlab(\"Log CD8 Max\") +\n",
    "ylab(\"H3K4me2: log2(Scaled Count + 1)\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K4me2_signal_patient_log_cd8_max.pdf\", width = 25, height = 18)\n",
    "\n",
    "pt_clinical %>% filter(gene_num %in% k4_TF_gene) %>% ggplot(aes(x = log_cd8_max, y = value)) +\n",
    "geom_point(aes(color = log_cd8_max)) +\n",
    "geom_smooth(color = \"grey\", method = \"lm\", se = FALSE) +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_color_viridis() +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 5, scales = \"free\") +\n",
    "xlab(\"Log CD8 Max\") +\n",
    "ylab(\"H3K4me2: log2(Scaled Count + 1)\") +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K4me2_signal_patient_log_cd8_max_k4DEgenes.pdf\", width = 20, height = 10)\n",
    "\n",
    "## \"pfscode\",\t\"pfsdaysafterfirstcartcell\", \"pfs_late_code\", \"log_pfs_days\"\n",
    "\n",
    "options(repr.plot.width=25, repr.plot.height=20)\n",
    "\n",
    "pt_clinical %>% ggplot(aes(x = as.factor(pfs_late_code), y = value)) +\n",
    "geom_boxplot(aes(fill = Analysis)) +\n",
    "theme_bw(base_size = 25) +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 10, scales = \"free\") + \n",
    "stat_compare_means(vjust = 1) +\n",
    "scale_fill_brewer(palette = \"Dark2\") +\n",
    "xlab(\"PFS Late Code\") +\n",
    "ylab(\"H3K4me2: log2(Scaled Count + 1)\")\n",
    "\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K4me2_signal_patient_PFS_late_code.pdf\", width = 25, height = 18)\n",
    "\n",
    "pt_clinical %>% filter(gene_num %in% k4_TF_gene) %>% ggplot(aes(x = as.factor(pfs_late_code), y = value)) +\n",
    "geom_boxplot(aes(fill = as.factor(pfs_late_code))) +\n",
    "theme_bw(base_size = 25) +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 5, scales = \"free\") + \n",
    "stat_compare_means(vjust = 1, size = 6.5) +\n",
    "scale_fill_brewer(palette = \"Paired\") +\n",
    "xlab(\"PFS Late Code\") +\n",
    "ylab(\"H3K4me2: log2(Scaled Count + 1)\") +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K4me2_signal_patient_PFS_late_code_k4DEgenes.pdf\", width = 18, height = 18)\n",
    "\n",
    "pt_clinical %>% ggplot(aes(x = log_pfs_days, y = value)) +\n",
    "geom_point(aes(color = Analysis)) +\n",
    "geom_smooth(aes(color = Analysis), method = \"lm\") +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_color_brewer(palette = \"Dark2\") +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 10, scales = \"free\") +\n",
    "xlab(\"Log PFS Days\") +\n",
    "ylab(\"H3K4me2: log2(Scaled Count + 1)\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K4me2_signal_patient_log_pfs_days.pdf\", width = 25, height = 18)\n",
    "\n",
    "pt_clinical %>% filter(gene_num %in% k4_TF_gene) %>% ggplot(aes(x = log_pfs_days, y = value)) +\n",
    "geom_point(aes(color = log_pfs_days)) +\n",
    "geom_smooth(color = \"grey\", method = \"lm\", se = FALSE) +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_color_viridis() +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 5, scales = \"free\") +\n",
    "xlab(\"Log PFS Days\") +\n",
    "ylab(\"H3K4me2: log2(Scaled Count + 1)\") +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/H3K4me2_signal_patient_log_pfs_days_k4DEgenes.pdf\", width = 20, height = 10)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## RNA\n",
    "row_ind = which((rownames(rna_obj@assays$RNA@data) %>% gsub(\"COMMD3-BMI1\", \"COMMD3BMI1\", .) %>% gsub(\".*-\", \"\", .) %>% toupper) %in% marker_gene)\n",
    "rna_normMat = rna_obj@assays$RNA@data[row_ind, 1:10] %>% data.frame\n",
    "colnames(rna_normMat) = colnames(rna_normMat) %>% gsub(\"Sample_\", \"\", .) %>% gsub(\"_.*\", \"\", .) %>% toupper\n",
    "rownames(rna_normMat) = rownames(rna_normMat) %>% as.vector %>% as.character %>% gsub(\".*-\", \"\", .) %>% toupper\n",
    "rna_normMat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "options(repr.plot.width=25, repr.plot.height=20)\n",
    "\n",
    "pt_select_tmp = rna_normMat %>% data.frame\n",
    "pt_select = pt_select_tmp %>% mutate(gene_num = rownames(pt_select_tmp)) %>% pivot_longer(cols = colnames(pt_select_tmp), names_to = \"xnumber\")\n",
    "head(pt_select, 20)\n",
    "pt_select %>% dim\n",
    "\n",
    "pt_clinical = left_join(left_join(pt_select, clinical_data_select, by = \"xnumber\"), marker_gene_table[, c(\"Analysis\", \"GeneNameClean\")], by = c(\"gene_num\" = \"GeneNameClean\"))\n",
    "pt_clinical %>% head\n",
    "\n",
    "pt_clinical %>% ggplot(aes(x = as.factor(propscore_cd8_max), y = value)) +\n",
    "geom_boxplot(aes(fill = Analysis)) +\n",
    "theme_bw(base_size = 25) +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 10, scales = \"free\") + \n",
    "stat_compare_means(vjust = 1) +\n",
    "scale_fill_brewer(palette = \"Dark2\") +\n",
    "xlab(\"Propensity Score CD8 Max\") +\n",
    "ylab(\"RNA-seq: log2(Scaled Count + 1)\")\n",
    "\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/RNA_signal_patient_propscore_cd8_max.pdf\", width = 25, height = 18)\n",
    "\n",
    "pt_clinical %>% filter(gene_num %in% rna_TF_gene) %>% ggplot(aes(x = as.factor(propscore_cd8_max), y = value)) +\n",
    "geom_boxplot(aes(fill = as.factor(propscore_cd8_max))) +\n",
    "theme_bw(base_size = 25) +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 5, scales = \"free\") + \n",
    "stat_compare_means(vjust = 1, size = 6.5) +\n",
    "scale_fill_brewer(palette = \"Paired\") +\n",
    "xlab(\"Propensity Score CD8 Max\") +\n",
    "ylab(\"RNA-seq: log2(Scaled Count + 1)\") +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/RNA_signal_patient_propscore_cd8_max_rnaDEgenes.pdf\", width = 18, height = 9)\n",
    "\n",
    "pt_clinical %>% ggplot(aes(x = log_cd8_max, y = value)) +\n",
    "geom_point(aes(color = Analysis)) +\n",
    "geom_smooth(aes(color = Analysis), method = \"lm\") +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_color_brewer(palette = \"Dark2\") +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 10, scales = \"free\") +\n",
    "xlab(\"Log CD8 Max\") +\n",
    "ylab(\"RNA-seq: log2(Scaled Count + 1)\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/RNA_signal_patient_log_cd8_max.pdf\", width = 25, height = 18)\n",
    "\n",
    "pt_clinical %>% filter(gene_num %in% rna_TF_gene) %>% ggplot(aes(x = log_cd8_max, y = value)) +\n",
    "geom_point(aes(color = log_cd8_max)) +\n",
    "geom_smooth(color = \"grey\", method = \"lm\", se = FALSE) +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_color_viridis() +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 5, scales = \"free\") +\n",
    "xlab(\"Log CD8 Max\") +\n",
    "ylab(\"RNA-seq: log2(Scaled Count + 1)\") +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/RNA_signal_patient_log_cd8_max_rnaDEgenes.pdf\", width = 20, height = 7)\n",
    "\n",
    "## \"pfscode\",\t\"pfsdaysafterfirstcartcell\", \"pfs_late_code\", \"log_pfs_days\"\n",
    "\n",
    "options(repr.plot.width=25, repr.plot.height=20)\n",
    "\n",
    "pt_clinical %>% ggplot(aes(x = as.factor(pfs_late_code), y = value)) +\n",
    "geom_boxplot(aes(fill = Analysis)) +\n",
    "theme_bw(base_size = 25) +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 10, scales = \"free\") + \n",
    "stat_compare_means(vjust = 1) +\n",
    "scale_fill_brewer(palette = \"Dark2\") +\n",
    "xlab(\"PFS Late Code\") +\n",
    "ylab(\"RNA-seq: log2(Scaled Count + 1)\")\n",
    "\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/RNA_signal_patient_PFS_late_code.pdf\", width = 25, height = 18)\n",
    "\n",
    "pt_clinical %>% filter(gene_num %in% rna_TF_gene) %>% ggplot(aes(x = as.factor(pfs_late_code), y = value)) +\n",
    "geom_boxplot(aes(fill = as.factor(pfs_late_code))) +\n",
    "theme_bw(base_size = 25) +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 5, scales = \"free\") + \n",
    "stat_compare_means(vjust = 1, size = 6.5) +\n",
    "scale_fill_brewer(palette = \"Paired\") +\n",
    "xlab(\"PFS Late Code\") +\n",
    "ylab(\"RNA-seq: log2(Scaled Count + 1)\") +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/RNA_signal_patient_PFS_late_code_rnaDEgenes.pdf\", width = 18, height = 9)\n",
    "\n",
    "pt_clinical %>% ggplot(aes(x = log_pfs_days, y = value)) +\n",
    "geom_point(aes(color = Analysis)) +\n",
    "geom_smooth(aes(color = Analysis), method = \"lm\") +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_color_brewer(palette = \"Dark2\") +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 10, scales = \"free\") +\n",
    "xlab(\"Log PFS Days\") +\n",
    "ylab(\"RNA-seq: log2(Scaled Count + 1)\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/RNA_signal_patient_log_pfs_days.pdf\", width = 25, height = 18)\n",
    "\n",
    "pt_clinical %>% filter(gene_num %in% rna_TF_gene) %>% ggplot(aes(x = log_pfs_days, y = value)) +\n",
    "geom_point(aes(color = log_pfs_days)) +\n",
    "geom_smooth(color = \"grey\", method = \"lm\", se = FALSE) +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_color_viridis() +\n",
    "facet_wrap(~factor(gene_num, levels = marker_gene), ncol = 5, scales = \"free\") +\n",
    "xlab(\"Log PFS Days\") +\n",
    "ylab(\"RNA-seq: log2(Scaled Count + 1)\") +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = \"CART_CUTRUN_Project/results/paper_figure/fig7/RNA_signal_patient_log_pfs_days_rnaDEgenes.pdf\", width = 20, height = 7)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
