{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## CUTRUN DE and assign to nearest genes\n",
    "\n",
    "library(pacman)\n",
    "p_load(data.table, dplyr, ggplot2, viridis, magrittr, VennDiagram, ggpubr, limma, edgeR, tidyr, GenomicRanges, RColorBrewer, pheatmap, Seurat, fgsea, GSEABase, limma, TxDb.Hsapiens.UCSC.hg38.knownGene, org.Hs.eg.db)\n",
    "\n",
    "cell_list <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hd_list <- paste0(\"HD\", c(1:3, 5:7))\n",
    "expr_list = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "cell_comp_list = c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "hist_list <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "in_path <- \"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "out_path <- \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/\"\n",
    "fig_path <- \"CART_CUTRUN_Project/results/paper_figure/\"\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## CUTRUN pairwise DP\n",
    "## 1. pairwise\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/IDR_master_peak_list_chromVar_count_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "\n",
    "cellList <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hdList <- paste0(\"HD\", 1:7)\n",
    "exprList = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "cellCompList = c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "histList <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "checkRep = function(hist, cell, expr, hd){\n",
    "  rep = \"rep1\"\n",
    "  if(paste(hist, cell, expr, hd, sep = \"_\") == \"H3K27me3_CM_Input_HD1\"){\n",
    "    rep = \"rep2\"\n",
    "  }\n",
    "  if(paste(hist, cell, expr, hd, sep = \"_\") == \"H3K27me3_N_Input_HD2\"){\n",
    "    rep = \"rep2\"\n",
    "  }\n",
    "\n",
    "  return(rep)\n",
    "\n",
    "}\n",
    "\n",
    "checkDup = function(expr, hd){\n",
    "\n",
    "  dupType = \"normLibSize\" #\"norm\"\n",
    "  if(expr == \"Input\" && hd %in% paste0(\"HD\", 1:3)){\n",
    "    dupType = \"stringent\"\n",
    "  }\n",
    "  return(dupType)\n",
    "\n",
    "}\n",
    "\n",
    "checkBam = function(expr, hd){\n",
    "\n",
    "  bamType = \"bowtie2_align.bam\"\n",
    "  if(expr == \"Input\" && hd %in% paste0(\"HD\", 1:3)){\n",
    "    bamType = \"bowtie2_align.sorted.rmDup.sortName.bam\"\n",
    "  }\n",
    "  return(bamType)\n",
    "\n",
    "}\n",
    "\n",
    "checkHD = function(expr){\n",
    "\n",
    "  if(expr == \"Input\"){\n",
    "    return(paste0(\"HD\", 4:7))\n",
    "  }else{\n",
    "    return(paste0(\"HD\", 1:7))\n",
    "  }\n",
    "}\n",
    "\n",
    "target = c()\n",
    "for(cell in cellList){\n",
    "  for(expr in exprList){\n",
    "    hdL = checkHD(expr)\n",
    "    for(hd in hdL){\n",
    "      if( !((cell == \"EMRA\") & (expr == \"Input\") && (hd == \"HD2\"))){\n",
    "        target = rbind(target, data.frame(cell = cell, expr = expr, humanDonor = hd))\n",
    "      }\n",
    "\n",
    "    }\n",
    "  }\n",
    "}\n",
    "target\n",
    "\n",
    "## Consider human donor as random variable.\n",
    "## Experimental design\n",
    "treat <- factor(paste(target$cell, target$expr, sep=\".\"))\n",
    "design <- model.matrix(~0 + treat)\n",
    "colnames(design) <- levels(treat)\n",
    "contrast <- makeContrasts(Input_N_CM = CM.Input - N.Input,\n",
    "                    Input_N_EM = EM.Input - N.Input,\n",
    "                    Input_N_EMRA = EMRA.Input - N.Input,\n",
    "                    Input_CM_EM = EM.Input - CM.Input,\n",
    "                    Input_CM_EMRA = EMRA.Input - CM.Input,\n",
    "                    Input_EM_EMRA = EMRA.Input - EM.Input,\n",
    "                    Product_N_CM = CM.Product - N.Product,\n",
    "                    Product_N_EM = EM.Product - N.Product,\n",
    "                    Product_N_EMRA = EMRA.Product - N.Product,\n",
    "                    Product_CM_EM = EM.Product - CM.Product,\n",
    "                    Product_CM_EMRA = EMRA.Product - CM.Product,\n",
    "                    Product_EM_EMRA = EMRA.Product - EM.Product,\n",
    "                    Stim1_N_CM = CM.Stim1 - N.Stim1,\n",
    "                    Stim1_N_EM = EM.Stim1 - N.Stim1,\n",
    "                    Stim1_N_EMRA = EMRA.Stim1 - N.Stim1,\n",
    "                    Stim1_CM_EM = EM.Stim1 - CM.Stim1,\n",
    "                    Stim1_CM_EMRA = EMRA.Stim1 - CM.Stim1,\n",
    "                    Stim1_EM_EMRA = EMRA.Stim1 - EM.Stim1,\n",
    "                    Stim2_N_CM = CM.Stim2 - N.Stim2,\n",
    "                    Stim2_N_EM = EM.Stim2 - N.Stim2,\n",
    "                    Stim2_N_EMRA = EMRA.Stim2 - N.Stim2,\n",
    "                    Stim2_CM_EM = EM.Stim2 - CM.Stim2,\n",
    "                    Stim2_CM_EMRA = EMRA.Stim2 - CM.Stim2,\n",
    "                    Stim2_EM_EMRA = EMRA.Stim2 - EM.Stim2,\n",
    "                    Stim3_N_CM = CM.Stim3 - N.Stim3,\n",
    "                    Stim3_N_EM = EM.Stim3 - N.Stim3,\n",
    "                    Stim3_N_EMRA = EMRA.Stim3 - N.Stim3,\n",
    "                    Stim3_CM_EM = EM.Stim3 - CM.Stim3,\n",
    "                    Stim3_CM_EMRA = EMRA.Stim3 - CM.Stim3,\n",
    "                    Stim3_EM_EMRA = EMRA.Stim3 - EM.Stim3,\n",
    "                    levels = design)\n",
    "\n",
    "results = vector(\"list\", length(histList))\n",
    "voomDDS = vector(\"list\", length(histList))\n",
    "for(hist in histList){\n",
    "  ## Filter and delete low expressed genes\n",
    "    selectR <- which(rowSums(countMat[[hist]]) > 10) ## remove low count genes\n",
    "    dataS <- countMat[[hist]][selectR,]\n",
    "    geneNameList = rownames(data)[selectR]\n",
    "\n",
    "    voomDDS[[hist]] <- voom(counts = dataS, design = design, normalize.method = \"cyclicloess\", plot = FALSE)\n",
    "    ## option 1 using voomDDS option2 using normDDS as normalized input.\n",
    "    inputDDS <- voomDDS[[hist]]\n",
    "    corfit <- duplicateCorrelation(inputDDS, design, block = target$humanDonor)\n",
    "    ## corfit$consensus\n",
    "    fit <- lmFit(inputDDS, design, block = target$humanDonor, correlation = corfit$consensus)\n",
    "    fitContrast <- contrasts.fit(fit, contrast)\n",
    "    fitBayes <- eBayes(fitContrast, robust = TRUE)\n",
    "\n",
    "\n",
    "    results[[hist]] = list()\n",
    "    for(i in 1:ncol(contrast)){\n",
    "        ## Results\n",
    "        res <- topTable(fit = fitBayes, adjust.method = 'fdr', coef = i, number = nrow(inputDDS), sort = 'P')\n",
    "\n",
    "        res <- data.table(GeneName = geneNameList[as.numeric(rownames(res))], GeneIndex = rownames(res) %>% as.numeric, res)\n",
    "        res[, Significance := ifelse((adj.P.Val <= 0.05 & sign(logFC) == 1 & abs(logFC) >= 1), 'Up',\n",
    "                              ifelse((adj.P.Val <= 0.05 & sign(logFC) == -1 & abs(logFC) >= 1), 'Down', 'notDE'))]\n",
    "        results[[hist]][[i]] = res\n",
    "\n",
    "        ## Output\n",
    "        write.table(res, file = paste0(outPath, '/CUTRUN_limma_tables/DE_', colnames(contrast)[i], '_HD1-7_adj0.05_logFC1.csv'), quote = FALSE, row.names = FALSE, sep = \",\")\n",
    "    }\n",
    "    names(results[[hist]]) <- colnames(contrast)\n",
    "\n",
    "}\n",
    "save(results, voomDDS, file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\"))\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## heatmap\n",
    "## 1. pairwise\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/IDR_master_peak_list_chromVar_count_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\"))\n",
    "for(hist in histList){\n",
    "    norm_cr <- voomDDS[[hist]]$E\n",
    "\n",
    "    gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")\n",
    "    dp_match_gene <- distanceToNearest(mPeak[[hist]], gene_gr, select = \"all\")\n",
    "    dp_match_gene %>% data.frame %$% distance %>% summary\n",
    "\n",
    "    peak_match_gene <- data.frame(peak_id = dp_match_gene@from, gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[dp_match_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) %>% group_by(peak_id) %>% summarize(nearest_genes = paste(gene_name, collapse = \";\"))\n",
    "    peak_match_gene %>% head()\n",
    "\n",
    "    rownames(norm_cr) <- rep(\"notAssigned\", nrow(norm_cr))\n",
    "    rownames(norm_cr)[peak_match_gene$peak_id] <- peak_match_gene$nearest_genes\n",
    "    head(norm_cr)\n",
    "\n",
    "    de_gene_list <- c(\n",
    "    results[[hist]]$Input_N_CM %>% filter(Significance != \"notDE\") %$% GeneIndex,\n",
    "    results[[hist]]$Input_N_EM %>% filter(Significance != \"notDE\") %$% GeneIndex,\n",
    "    results[[hist]]$Input_N_EMRA %>% filter(Significance != \"notDE\") %$% GeneIndex,\n",
    "    results[[hist]]$Input_CM_EM %>% filter(Significance != \"notDE\") %$% GeneIndex,\n",
    "    results[[hist]]$Input_CM_EMRA %>% filter(Significance != \"notDE\") %$% GeneIndex,\n",
    "    results[[hist]]$Input_EM_EMRA %>% filter(Significance != \"notDE\") %$% GeneIndex\n",
    "    ) %>% unique\n",
    "    sample_select <- paste0(hist, \"_\", rep(cell_list, each = 4), rep(c(\"_Input_HD4\", \"_Input_HD5\", \"_Input_HD6\", \"_Input_HD7\"), length(cell_list)))\n",
    "\n",
    "    norm_cr[de_gene_list, ] %>% dim\n",
    "    norm_cr[de_gene_list, sample_select] %>% head\n",
    "\n",
    "    annotation_col <- data.frame(CellType = rep(c(\"N\", \"CM\", \"EM\", \"EMRA\"), each = 4))\n",
    "    rownames(annotation_col) <- sample_select\n",
    "    annotation_color <- list(CellType = c(\"N\" = \"#440154\", \"CM\" = \"#31688e\", \"EM\" = \"#35b779\", \"EMRA\" = \"#fde725\"))\n",
    "\n",
    "\n",
    "    test = pheatmap(norm_cr[de_gene_list, sample_select], cluster_rows = TRUE, cluster_cols = FALSE, show_rownames = FALSE, scale = \"none\", annotation_col = annotation_col, color = BlueAndRed(),\n",
    "    annotation_colors = annotation_color, filename = paste0(fig_path, \"CUTRUN_DP_\", hist, \"_pairwise_comparison_heatmap.pdf\"), width = 7, height = 7) \n",
    "\n",
    "    write.table(norm_cr[de_gene_list[test$tree_row$order], sample_select], file = paste0(fig_path, \"CUTRUN_DP_\", hist, \"_pairwise_comparison_heatmap.csv\"), quote = F, row.names = TRUE, col.names = TRUE, sep = \",\")\n",
    "\n",
    "}\n",
    "\n",
    "# 2. one vs rest\n",
    "\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\"))\n",
    "\n",
    "for(hist in histList){\n",
    "    norm_cr <- voomDDS[[hist]]$E\n",
    "\n",
    "    gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")\n",
    "    dp_match_gene <- distanceToNearest(mPeak[[hist]], gene_gr, select = \"all\")\n",
    "    dp_match_gene %>% data.frame %$% distance %>% summary\n",
    "\n",
    "    peak_match_gene <- data.frame(peak_id = dp_match_gene@from, gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[dp_match_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) %>% group_by(peak_id) %>% summarize(nearest_genes = paste(gene_name, collapse = \";\"))\n",
    "    peak_match_gene %>% head()\n",
    "\n",
    "    rownames(norm_cr) <- rep(\"notAssigned\", nrow(norm_cr))\n",
    "    rownames(norm_cr)[peak_match_gene$peak_id] <- peak_match_gene$nearest_genes\n",
    "\n",
    "    de_gene_list <- c()\n",
    "    for(target_cell in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "        load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM_one_vs_rest_\", target_cell, \".RData\"))\n",
    "        de_gene_list <- c(de_gene_list, results[[hist]][[paste0(\"Input_\", target_cell, \"_Others\")]] %>% filter(Significance != \"notDE\") %$% GeneIndex)\n",
    "    }\n",
    "    de_gene_list <- de_gene_list %>% unique\n",
    "\n",
    "    sample_select <- paste0(hist, \"_\", rep(cell_list, each = 4), rep(c(\"_Input_HD4\", \"_Input_HD5\", \"_Input_HD6\", \"_Input_HD7\"), length(cell_list)))\n",
    "\n",
    "    norm_cr[de_gene_list, ] %>% dim\n",
    "    norm_cr[de_gene_list, sample_select] %>% head\n",
    "\n",
    "    annotation_col <- data.frame(CellType = rep(c(\"N\", \"CM\", \"EM\", \"EMRA\"), each = 4))\n",
    "    rownames(annotation_col) <- sample_select\n",
    "    annotation_color <- list(CellType = c(\"N\" = \"#440154\", \"CM\" = \"#31688e\", \"EM\" = \"#35b779\", \"EMRA\" = \"#fde725\"))\n",
    "\n",
    "\n",
    "    test = pheatmap(norm_cr[de_gene_list, sample_select], cluster_rows = TRUE, cluster_cols = FALSE, show_rownames = FALSE, scale = \"none\", annotation_col = annotation_col, color = BlueAndRed(),\n",
    "    annotation_colors = annotation_color, filename = paste0(fig_path, \"CUTRUN_DP_\", hist, \"_one_vs_rest_heatmap.pdf\"), width = 7, height = 7) \n",
    "\n",
    "    write.table(norm_cr[de_gene_list[test$tree_row$order], sample_select], file = paste0(fig_path, \"CUTRUN_DP_\", hist, \"_one_vs_rest_heatmap.csv\"), quote = F, row.names = TRUE, col.names = TRUE, sep = \",\")\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## GSEA of CUTRUN\n",
    "## 1. pairwise\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "# load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "# load(file = paste0(outPath, \"/RData/IDR_master_peak_list_chromVar_count_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "# load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\"))\n",
    "pathway_list <- c(\"msigdb.v7.4.symbols.gmt\", \"h.all.v7.4.symbols.gmt\", \"c2.cp.v7.4.symbols.gmt\", \"c3.tft.v7.4.symbols.gmt\", \"c5.all.v7.4.entrez.gmt\", \"c7.all.v7.4.symbols.gmt\")\n",
    "pathway_name_list <- c(\"All_Gene_Sets\", \"Hallmark_All\", \"C2_noCGP\", \"C3_TFTonly\", \"C5_All\", \"C7_All\")\n",
    "cell_pair_list <- c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "\n",
    "get_abs_max <- function(x){\n",
    "    return(x[which(abs(x) == max(abs(x)))])\n",
    "}\n",
    "\n",
    "gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")\n",
    "\n",
    "for(hist in histList){\n",
    "    for(cell_pair in cell_pair_list){\n",
    "        dp_ind <- results[[hist]][[paste0(\"Input_\", cell_pair)]] %$% GeneIndex ## DiffPeak Peak ID\n",
    "        dp_match_gene <- distanceToNearest(mPeak[[hist]][dp_ind], gene_gr, select = \"all\") ## get the nearest gene for each peak\n",
    "        peak_match_gene <- data.frame(peak_id = dp_ind[dp_match_gene@from], gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[dp_match_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) ## get the peak id and matching gene name\n",
    "        peak_match_gene <- left_join(peak_match_gene, results[[hist]][[paste0(\"Input_\", cell_pair)]] %>% dplyr::select(GeneIndex, logFC), by = c(\"peak_id\" = \"GeneIndex\")) ## get the peak id, peak logFC, and matching gene name\n",
    "        gene_peak_logFC <- peak_match_gene %>% group_by(gene_name) %>% summarize(logFC_abs_max = get_abs_max(logFC)) ## for each gene get the abs max logFC peak\n",
    "\n",
    "        geneList <- gene_peak_logFC$logFC_abs_max\n",
    "        names(geneList) <- gene_peak_logFC$gene_name\n",
    "        geneList <- sort(geneList, decreasing = TRUE)\n",
    "\n",
    "        for(p in 1:length(pathway_list)){\n",
    "            pathway <- pathway_list[p]\n",
    "            pathway_name <- pathway_name_list[p]\n",
    "\n",
    "            module_file <- paste0(\"/fh/fast/gottardo_r/yezheng_working/SupplementaryData/GSEA/\", pathway)\n",
    "            gene_set <- getGmt(module_file)\n",
    "            gene_pathway <- geneIds(gene_set)\n",
    "\n",
    "            gsea_res <- fgsea(pathways = gene_pathway, stats = geneList, minSize = 10, maxSize = 500, eps = 0)\n",
    "            gsea_res <- gsea_res %>% arrange(padj, decreaing = FALSE)\n",
    "            fwrite(gsea_res, file = paste0(fig_path, \"/CUTRUN_\", hist, \"_nearest_gene_DP_pairwise_\", cell_pair, \"_\", pathway_name, \"_GSEA.csv\"), sep = \",\")\n",
    "\n",
    "            pdf(paste0(fig_path, \"/CUTRUN_\", hist, \"_nearest_gene_DP_\", cell_pair, \"_\", pathway_name, \"_GSEA_topEnrichmentScore.pdf\"), width = 15, height = 10)\n",
    "            topPathwaysUp <- gsea_res[ES > 0][head(order(pval), n=10), pathway]\n",
    "            topPathwaysDown <- gsea_res[ES < 0][head(order(pval), n=10), pathway]\n",
    "            topPathways <- c(topPathwaysUp, rev(topPathwaysDown)) %>% unique\n",
    "            plotGseaTable(\n",
    "                gene_pathway[topPathways], \n",
    "                geneList, \n",
    "                gsea_res,\n",
    "                gseaParam=0.5\n",
    "            )\n",
    "            dev.off()\n",
    "        }\n",
    "\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## GSEA of CUTRUN\n",
    "## 2. one vs rest\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "pathway_list <- c(\"msigdb.v7.4.symbols.gmt\", \"h.all.v7.4.symbols.gmt\", \"c2.cp.v7.4.symbols.gmt\", \"c3.tft.v7.4.symbols.gmt\", \"c5.all.v7.4.entrez.gmt\", \"c7.all.v7.4.symbols.gmt\")\n",
    "pathway_name_list <- c(\"All_Gene_Sets\", \"Hallmark_All\", \"C2_noCGP\", \"C3_TFTonly\", \"C5_All\", \"C7_All\")\n",
    "cell_pair_list <- c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "\n",
    "get_abs_max <- function(x){\n",
    "    return(x[which(abs(x) == max(abs(x)))])\n",
    "}\n",
    "\n",
    "gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")\n",
    "\n",
    "for(hist in histList){\n",
    "    for(target_cell in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "        load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM_one_vs_rest_\", target_cell, \".RData\"))\n",
    "        dp_ind <- results[[hist]][[paste0(\"Input_\", target_cell, \"_Others\")]] %$% GeneIndex ## DiffPeak Peak ID\n",
    "        dp_match_gene <- distanceToNearest(mPeak[[hist]][dp_ind], gene_gr, select = \"all\") ## get the nearest gene for each peak\n",
    "        peak_match_gene <- data.frame(peak_id = dp_ind[dp_match_gene@from], gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[dp_match_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) ## get the peak id and matching gene name\n",
    "        peak_match_gene <- left_join(peak_match_gene, results[[hist]][[paste0(\"Input_\", target_cell, \"_Others\")]] %>% dplyr::select(GeneIndex, logFC), by = c(\"peak_id\" = \"GeneIndex\")) ## get the peak id, peak logFC, and matching gene name\n",
    "        gene_peak_logFC <- peak_match_gene %>% group_by(gene_name) %>% summarize(logFC_abs_max = get_abs_max(logFC)) ## for each gene get the abs max logFC peak\n",
    "\n",
    "        geneList <- gene_peak_logFC$logFC_abs_max\n",
    "        names(geneList) <- gene_peak_logFC$gene_name\n",
    "        geneList <- sort(geneList, decreasing = TRUE)\n",
    "\n",
    "        for(p in 1:length(pathway_list)){\n",
    "            pathway <- pathway_list[p]\n",
    "            pathway_name <- pathway_name_list[p]\n",
    "\n",
    "            module_file <- paste0(\"/fh/fast/gottardo_r/yezheng_working/SupplementaryData/GSEA/\", pathway)\n",
    "            gene_set <- getGmt(module_file)\n",
    "            gene_pathway <- geneIds(gene_set)\n",
    "\n",
    "            gsea_res <- fgsea(pathways = gene_pathway, stats = geneList, minSize = 10, maxSize = 500, eps = 0)\n",
    "            gsea_res <- gsea_res %>% arrange(padj, decreaing = FALSE)\n",
    "            fwrite(gsea_res, file = paste0(fig_path, \"/CUTRUN_\", hist, \"_nearest_gene_DP_one_vs_rest_\", target_cell, \"_\", pathway_name, \"_GSEA.csv\"), sep = \",\")\n",
    "\n",
    "            pdf(paste0(fig_path, \"/CUTRUN_\", hist, \"_nearest_gene_DP_one_vs_rest_\", target_cell, \"_\", pathway_name, \"_GSEA_topEnrichmentScore.pdf\"), width = 15, height = 10)\n",
    "            topPathwaysUp <- gsea_res[ES > 0][head(order(pval), n=10), pathway]\n",
    "            topPathwaysDown <- gsea_res[ES < 0][head(order(pval), n=10), pathway]\n",
    "            topPathways <- c(topPathwaysUp, rev(topPathwaysDown)) %>% unique\n",
    "            plotGseaTable(\n",
    "                gene_pathway[topPathways], \n",
    "                geneList, \n",
    "                gsea_res,\n",
    "                gseaParam=0.5\n",
    "            )\n",
    "            dev.off()\n",
    "        }\n",
    "\n",
    "    }\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Figure 2 -\tDifferences between RNAseq and CUT&RUN identified gene sets\n",
    "# o\tDisplayed as scatter plot\n",
    "# \tRNAseq expression on Y vs log(CPM) for H3K4me2 on X for N, CM, EM and EMRA\n",
    "# o\tFull list of genes that lie in Q1 vs Q3\n",
    "# o\tGene ontology for genes that lie in Q1 bs Q3 \n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\")) ## voomDDS results\n",
    "cr_voomDDS <- voomDDS\n",
    "load(file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition.RData\"))\n",
    "rna_voomDDS <- voomDDS$E\n",
    "geneID <- fread(paste0(in_path, \"RNA_CD8_N_Input_HD1.genes.results\"))$gene_id\n",
    "rownames(rna_voomDDS) <- geneID[selectR] %>% gsub(\".*_\", \"\", .)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "select_gene <- list(\n",
    "    TF = read_excel(paste0(gene_path, \"2021_11_29_TFs.xlsx\")) %>% mutate(gene_name = `HGNC symbol`),\n",
    "    Tactive = read_excel(paste0(gene_path, \"2021_11_30_TCellActivationFunction.xlsx\")) %>% mutate(gene_name = `Official Symbol`),\n",
    "    CellCycle = read_excel(paste0(gene_path, \"2021_22_30_CellCycleAPoptosis.xlsx\")) %>% mutate(gene_name = `Symbol`),\n",
    "    Metabolism = read_excel(paste0(gene_path, \"2021_22_30_MetabolismGeneList_FinalList.xlsx\")) %>% mutate(gene_name = `Gene`)\n",
    ")\n",
    "\n",
    "for(gene_type in names(select_gene)){\n",
    "    pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_PeakSignal_scatterplot_\", gene_type, \"_maxCRsignal.pdf\"), width = 11, height =  11)\n",
    "   \n",
    "    for(hist in c(\"H3K27me3\", \"H3K4me2\")){\n",
    "        for(target_cell in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "            gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")\n",
    "            dp_match_gene <- distanceToNearest(mPeak[[hist]], gene_gr, select = \"all\")\n",
    "            dp_match_gene %>% data.frame %$% distance %>% summary\n",
    "            peak_match_gene <- data.frame(peak_id = dp_match_gene@from, gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[dp_match_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) #%>% group_by(peak_id) %>% summarize(nearest_genes = paste(gene_name, collapse = \";\"))\n",
    "\n",
    "            rna_gene_exp <- data.frame(gene_name = rownames(rna_voomDDS), GEX = rna_voomDDS[, paste0(rep(\"Input_\", 6), target_cell, c(\"_HD1\", \"_HD2\", \"_HD3\", \"_HD5\", \"_HD6\", \"_HD7\"))] %>% rowMeans )\n",
    "            rna_gene_exp %>% dim\n",
    "            cr_mean <- cr_voomDDS[[hist]]$E[, paste0(hist, \"_\", target_cell, c(\"_Input_HD4\", \"_Input_HD5\", \"_Input_HD6\", \"_Input_HD7\"))] %>% rowMeans\n",
    "            cr_gene_cr <- data.frame(gene_name = peak_match_gene$gene_name, CUTRUN_signal = cr_mean[peak_match_gene$peak_id]) \n",
    "            cr_gene_cr %>% dim\n",
    "            \n",
    "            if(hist == \"H3K4me2\"){\n",
    "                line_slope <- 1\n",
    "                line_intercept <- 0\n",
    "                line_x <- 5\n",
    "                line_y <- 5\n",
    "            }else{\n",
    "                line_slope <- -1\n",
    "                line_intercept <- 7\n",
    "                line_x <- 3.5\n",
    "                line_y <- 3.5\n",
    "            }\n",
    "            print(inner_join(cr_gene_cr, rna_gene_exp, by = \"gene_name\")  %>% filter(gene_name %in% select_gene[[gene_type]]$gene_name) %>% group_by(gene_name, GEX) %>% summarize(maxCR = max(CUTRUN_signal)) %>% \n",
    "            ## ggplot(aes(x = CUTRUN_signal, y = GEX)) +\n",
    "            ggplot(aes(x = maxCR, y = GEX)) +\n",
    "            geom_hex(bins = 200) +\n",
    "            ## geom_point(size = 0.05, color = \"grey\") +\n",
    "            theme_bw(base_size = 20) +\n",
    "            geom_abline(slope = line_slope, intercept = line_intercept, linetype = \"dashed\", color = \"red\", size = 1.2) +\n",
    "            geom_vline(xintercept = line_x, linetype = \"dashed\", color = \"red\", size = 1.2) +\n",
    "            geom_hline(yintercept = line_y, linetype = \"dashed\", color = \"red\", size = 1.2) +\n",
    "            xlab(paste0(hist, \" Normalized Peak Signals\")) +\n",
    "            ylab(\"Normalized Gene Expression\") +\n",
    "            ggtitle(paste0(target_cell, \" Cells \", gene_type)))\n",
    "\n",
    "        }\n",
    "    }\n",
    "    dev.off()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## hexbin plot comparing normalized gex and peak signal\n",
    "pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_PeakSignal_scatterplot.pdf\"), width = 11, height =  11)\n",
    "for(hist in c(\"H3K27me3\", \"H3K4me2\")){\n",
    "    for(target_cell in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "        gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")\n",
    "        dp_match_gene <- distanceToNearest(mPeak[[hist]], gene_gr, select = \"all\")\n",
    "        dp_match_gene %>% data.frame %$% distance %>% summary\n",
    "        peak_match_gene <- data.frame(peak_id = dp_match_gene@from, gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[dp_match_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) #%>% group_by(peak_id) %>% summarize(nearest_genes = paste(gene_name, collapse = \";\"))\n",
    "\n",
    "        rna_gene_exp <- data.frame(gene_name = rownames(rna_voomDDS), GEX = rna_voomDDS[, paste0(rep(\"Input_\", 6), target_cell, c(\"_HD1\", \"_HD2\", \"_HD3\", \"_HD5\", \"_HD6\", \"_HD7\"))] %>% rowMeans )\n",
    "        rna_gene_exp %>% dim\n",
    "        cr_mean <- cr_voomDDS[[hist]]$E[, paste0(hist, \"_\", target_cell, c(\"_Input_HD4\", \"_Input_HD5\", \"_Input_HD6\", \"_Input_HD7\"))] %>% rowMeans\n",
    "        cr_gene_cr <- data.frame(gene_name = peak_match_gene$gene_name, CUTRUN_signal = cr_mean[peak_match_gene$peak_id]) \n",
    "        cr_gene_cr %>% dim\n",
    "        \n",
    "        if(hist == \"H3K4me2\"){\n",
    "            line_slope <- 1\n",
    "            line_intercept <- 0\n",
    "            line_x <- 5\n",
    "            line_y <- 5\n",
    "        }else{\n",
    "            line_slope <- -1\n",
    "            line_intercept <- 7\n",
    "            line_x <- 3.5\n",
    "            line_y <- 3.5\n",
    "        }\n",
    "        print(inner_join(cr_gene_cr, rna_gene_exp, by = \"gene_name\")  %>% \n",
    "        ggplot(aes(x = CUTRUN_signal, y = GEX)) +\n",
    "        geom_hex(bins = 200) +\n",
    "        ## geom_point(size = 0.01, color = \"grey\") +\n",
    "        theme_bw(base_size = 20) +\n",
    "        geom_abline(slope = line_slope, intercept = line_intercept, linetype = \"dashed\", color = \"red\", size = 1.2) +\n",
    "        geom_vline(xintercept = line_x, linetype = \"dashed\", color = \"red\", size = 1.2) +\n",
    "        geom_hline(yintercept = line_y, linetype = \"dashed\", color = \"red\", size = 1.2) +\n",
    "        xlab(paste0(hist, \" Normalized Peak Signals\")) +\n",
    "        ylab(\"Normalized Gene Expression\") +\n",
    "        ggtitle(paste0(target_cell, \" Cells\")))\n",
    "\n",
    "    }\n",
    "}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## output DP with nearest genes\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/IDR_master_peak_list_chromVar_count_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\")) ## voomDDS results\n",
    "gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")\n",
    "\n",
    "\n",
    "for(hist in c(\"H3K27me3\", \"H3K4me2\")){\n",
    "    dp_match_gene <- distanceToNearest(mPeak[[hist]], gene_gr, select = \"all\")\n",
    "    dp_match_gene %>% data.frame %$% distance %>% summary\n",
    "    peak_match_gene <- data.frame(peak_id = dp_match_gene@from, gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[dp_match_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) %>% group_by(peak_id) %>% summarize(nearest_genes = paste(gene_name, collapse = \";\"))\n",
    "\n",
    "    for(sample in names(results[[hist]])){\n",
    "        info = sample %>% strsplit(\"_\") %>% unlist\n",
    "        limma_res <- results[[hist]][[sample]]\n",
    "        limma_res <- cbind(data.frame(mPeak[[hist]])[limma_res$GeneIndex, c(\"seqnames\", \"start\", \"end\")], limma_res)\n",
    "        dp_res <- right_join(peak_match_gene, limma_res, by = c(\"peak_id\" = \"GeneIndex\")) %>% arrange(adj.P.Val)\n",
    "\n",
    "        data_norm <- countMat[[hist]]/colSums(countMat[[hist]]) * 1000000\n",
    "        tmp = cbind(dp_res, \n",
    "            data_norm[\n",
    "                dp_res$peak_id, \n",
    "                c(\n",
    "                    paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", c(\"HD4\", \"HD5\", \"HD6\", \"HD7\")), \n",
    "                    paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", c(\"HD4\", \"HD5\", \"HD6\", \"HD7\"))\n",
    "                )]\n",
    "        ) \n",
    "        tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD4\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD4\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD4\")])\n",
    "        tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD5\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD5\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD5\")])\n",
    "        tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD6\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD6\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD6\")])\n",
    "        tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD7\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD7\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD7\")])\n",
    "\n",
    "        write.table(tmp, file = paste0(fig_path, \"CUTRUN_DP_\", hist, \"_\", sample, \"_pairwise_withNearestGenes.csv\"), sep = \",\", quote = F, row.names = F, col.names = T)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## hexbin plot comparing logFC of GEX and Peak signal\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\")) ## voomDDS results\n",
    "cr_results <- results\n",
    "load(file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition.RData\"))\n",
    "rna_results <- results\n",
    "\n",
    "pdf(paste0(fig_path, \"/RNAseq_CUTRUN_logFC_scatterplot.pdf\"), width = 11, height =  11)\n",
    "for(hist in c(\"H3K27me3\", \"H3K4me2\")){\n",
    "    for(cell_pair in c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")){\n",
    "\n",
    "        \n",
    "        gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")\n",
    "        dp_match_gene <- distanceToNearest(mPeak[[hist]], gene_gr, select = \"all\")\n",
    "        dp_match_gene %>% data.frame %$% distance %>% summary\n",
    "        peak_match_gene <- data.frame(peak_id = dp_match_gene@from, gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[dp_match_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) #%>% group_by(peak_id) %>% summarize(nearest_genes = paste(gene_name, collapse = \";\"))\n",
    "\n",
    "        rna_gene_exp <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cell_pair)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cell_pair)]]$logFC)\n",
    "        rna_gene_exp %>% dim\n",
    "        cr_gene_cr <- left_join(peak_match_gene, cr_results[[hist]][[paste0(\"Input_\", cell_pair)]], by = c(\"peak_id\" = \"GeneIndex\")) %>% group_by(gene_name) %>% summarize(logFC_abs_max = get_abs_max(logFC))\n",
    "        cr_gene_cr %>% dim\n",
    "        \n",
    "        if(hist == \"H3K4me2\"){\n",
    "            line_slope <- 1\n",
    "            line_intercept <- 0\n",
    "            line_x <- 5\n",
    "            line_y <- 5\n",
    "        }else{\n",
    "            line_slope <- -1\n",
    "            line_intercept <- 7\n",
    "            line_x <- 3.5\n",
    "            line_y <- 3.5\n",
    "        }\n",
    "        print(inner_join(cr_gene_cr, rna_gene_exp, by = \"gene_name\")  %>% \n",
    "        ggplot(aes(x = logFC_abs_max, y = GEX)) +\n",
    "        ## geom_hex(bins = 200) +\n",
    "        geom_point(size = 0.1, alpha = 0.3) +\n",
    "        theme_bw(base_size = 20) +\n",
    "        # geom_abline(slope = line_slope, intercept = line_intercept, linetype = \"dashed\", color = \"red\", size = 1.2) +\n",
    "        # geom_vline(xintercept = line_x, linetype = \"dashed\", color = \"red\", size = 1.2) +\n",
    "        # geom_hline(yintercept = line_y, linetype = \"dashed\", color = \"red\", size = 1.2) +\n",
    "        xlab(paste0(hist, \" logFC\")) +\n",
    "        ylab(\"Gene Expression logFC\") +\n",
    "        ggtitle(cell_pair))\n",
    "\n",
    "    }\n",
    "}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Bivalent methylation marks\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/IDR_master_peak_list_chromVar_count_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\"))\n",
    "\n",
    "names(peakAll)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## overlap K4me2 and K27me3 ensuring at least 50bp overlap, count number of regions after merging\n",
    "bi_path <- \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/BivalentGenes/\"\n",
    "gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")\n",
    "\n",
    "get_nearest_gene <- function(gene_gr, region_gr){\n",
    "    region_assign_gene <- distanceToNearest(region_gr, gene_gr, select = \"all\")\n",
    "    # dist_tmp <- region_assign_gene %>% data.frame %$% distance\n",
    "    # region_assign_gene <- region_assign_gene[dist_tmp < 1000]\n",
    "    region_match_gene <- cbind(data.frame(region_gr)[region_assign_gene@from, ], data.frame(peak_id = region_assign_gene@from, gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[region_assign_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) )\n",
    "    return(region_match_gene)\n",
    "\n",
    "}\n",
    "bi_num <- c()\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    for(pt in paste0(\"HD\", 4:7)){\n",
    "        bi_olp <- findOverlaps(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]], minoverlap = 50)\n",
    "        bi_peak.gr <- reduce(append(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]][bi_olp@from], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]][bi_olp@to]))\n",
    "        bi_gene <- get_nearest_gene(gene_gr, bi_peak.gr)\n",
    "        write.table(unique(bi_gene$gene_name), paste0(bi_path, \"Bivalent_peaks_nearest_genes_\", cT, \"_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "\n",
    "        total_peak.gr <- append(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]]) %>% reduce\n",
    "        total_gene <- get_nearest_gene(gene_gr, total_peak.gr)\n",
    "\n",
    "        bi_num <- data.frame(\n",
    "            bi_peak_num = length(bi_peak.gr),\n",
    "            bi_gene_num = length(unique(bi_gene$gene_name)),\n",
    "            total_peak_num = length(total_peak.gr),\n",
    "            total_gene_num = length(unique(total_gene$gene_name)),\n",
    "            cT = cT,\n",
    "            pt = pt\n",
    "        ) %>% rbind(bi_num, .)\n",
    "    }\n",
    "}\n",
    "bi_num$cT <- factor(bi_num$cT, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\"))\n",
    "\n",
    "p1 = bi_num %>% ggplot(aes(x = pt, y = bi_peak_num, fill = cT)) + \n",
    "geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Peak Regions\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p2 = bi_num %>% mutate(bi_peak_prop = bi_peak_num/total_peak_num * 100) %>% ggplot(aes(x = pt, y = bi_peak_prop, fill = cT)) + \n",
    "geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Peak Regions\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p3 = bi_num %>% ggplot(aes(x = pt, y = bi_gene_num, fill = cT)) + \n",
    "geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Genes\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p4 = bi_num %>% mutate(bi_gene_prop = bi_gene_num/total_gene_num * 100) %>% ggplot(aes(x = pt, y = bi_gene_prop, fill = cT)) + \n",
    "geom_bar(stat = \"identity\", position = position_dodge()) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "ggarrange(p1, p2, p3, p4, ncol = 2, nrow = 2, common.legend = TRUE) \n",
    "\n",
    "pdf(paste0(fig_path, \"/bivalent_peak_gene_num_prop_boxplot.pdf\"), width = 7, height = 7)\n",
    "bi_num %>% ggplot(aes(x = cT, y = bi_peak_num, fill = cT)) + \n",
    "geom_boxplot() +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Peak Regions\") +\n",
    "rremove(\"legend\")\n",
    "\n",
    "bi_num %>% mutate(bi_peak_prop = bi_peak_num/total_peak_num * 100) %>% ggplot(aes(x = cT, y = bi_peak_prop, fill = cT)) + \n",
    "geom_boxplot() +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Peak Regions\") +\n",
    "rremove(\"legend\")\n",
    "\n",
    "bi_num %>% ggplot() + \n",
    "geom_boxplot(aes(x = cT, y = bi_gene_num, fill = cT)) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Genes\") +\n",
    "rremove(\"legend\")\n",
    "\n",
    "\n",
    "bi_num %>% mutate(bi_gene_prop = bi_gene_num/total_gene_num * 100) %>% ggplot() + \n",
    "geom_boxplot(aes(x = cT, y = bi_gene_prop, fill = cT)) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rremove(\"legend\") \n",
    "dev.off()\n",
    "\n",
    "# pdf(paste0(fig_path, \"/bivalent_peak_gene_num_prop.pdf\"), width = 10, height = 10)\n",
    "# ggarrange(p1, p2, p3, p4, ncol = 2, nrow = 2, common.legend = TRUE) \n",
    "# dev.off()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "load(file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition.RData\"))\n",
    "rna_results <- results\n",
    "\n",
    "bi_change <- c()\n",
    "bi_change_gene_lfc <- c()\n",
    "ct_list <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "pt_list <- paste0(\"HD\", 4:7)\n",
    "for(i in 1:length(ct_list)){\n",
    "    cT <- ct_list[i]\n",
    "    cT1 <- cT\n",
    "    for(j in 1:length(ct_list)){\n",
    "        if(i != j){\n",
    "            cT2 <- ct_list[j]\n",
    "            for(pt in pt_list){\n",
    "                bi_olp <- findOverlaps(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]], minoverlap = 50)\n",
    "                bi_olp2 <- findOverlaps(peakAll[[paste0(\"H3K27me3_\", cT2, \"_Input_\", pt)]], peakAll[[paste0(\"H3K4me2_\", cT2, \"_Input_\", pt)]], minoverlap = 50)\n",
    "\n",
    "                bi_peak.gr <- reduce(append(peakAll[[paste0(\"H3K27me3_\", cT, \"_Input_\", pt)]][bi_olp@from], peakAll[[paste0(\"H3K4me2_\", cT, \"_Input_\", pt)]][bi_olp@to]))\n",
    "                bi_peak.gr2 <- reduce(append(peakAll[[paste0(\"H3K27me3_\", cT2, \"_Input_\", pt)]][bi_olp2@from], peakAll[[paste0(\"H3K4me2_\", cT2, \"_Input_\", pt)]][bi_olp2@to]))\n",
    "\n",
    "                bi_again <- findOverlaps(bi_peak.gr, bi_peak.gr2, minoverlap = 50)@from %>% unique\n",
    "                bi_k27 <- findOverlaps(bi_peak.gr, peakAll[[paste0(\"H3K27me3_\", cT2, \"_Input_\", pt)]], minoverlap = 50)@from %>% unique\n",
    "                bi_k4 <- findOverlaps(bi_peak.gr, peakAll[[paste0(\"H3K4me2_\", cT2, \"_Input_\", pt)]], minoverlap = 50)@from %>% unique\n",
    "                bi_k27 <- setdiff(setdiff(bi_k27, bi_k4), bi_again)\n",
    "                bi_k4 <- setdiff(setdiff(bi_k4, bi_k27), bi_again)\n",
    "                bi_none <- setdiff(1:length(bi_peak.gr), unique(c(bi_again, bi_k27, bi_k4)) %>% sort)\n",
    "\n",
    "                bi_gene <- get_nearest_gene(gene_gr, bi_peak.gr)\n",
    "                bi_again_gene <- bi_gene %>% filter(peak_id %in% bi_again) %$% gene_name %>% unique\n",
    "                bi_k27_gene <- bi_gene %>% filter(peak_id %in% bi_k27) %$% gene_name %>% unique\n",
    "                bi_k4_gene <- bi_gene %>% filter(peak_id %in% bi_k4) %$% gene_name %>% unique\n",
    "                bi_none_gene <- setdiff(unique(bi_gene$gene_name), c(bi_again_gene, bi_k27_gene, bi_k4_gene)) %>% unique\n",
    "                write.table(bi_again_gene, paste0(bi_path, \"BivalentPeaksOf\", cT1, \"_BivalentPeakAgainIn\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                write.table(bi_k27_gene, paste0(bi_path, \"BivalentPeaksOf\", cT1, \"_H3K27me3In\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                write.table(bi_k4_gene, paste0(bi_path, \"BivalentPeaksOf\", cT1, \"_H3K4me2In\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                write.table(bi_none_gene, paste0(bi_path, \"BivalentPeaksOf\", cT1, \"_NoPeakIn\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                \n",
    "                if(i < j){\n",
    "                    bi_again_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% bi_again_gene)\n",
    "                    bi_k27_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% bi_k27_gene)\n",
    "                    bi_k4_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% bi_k4_gene)\n",
    "                    bi_none_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% bi_none_gene)\n",
    "                }else{\n",
    "                    bi_again_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$logFC) %>% filter(gene_name %in% bi_again_gene)\n",
    "                    bi_k27_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$logFC) %>% filter(gene_name %in% bi_k27_gene)\n",
    "                    bi_k4_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$logFC) %>% filter(gene_name %in% bi_k4_gene)\n",
    "                    bi_none_gene_lfc <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT2, \"_\", cT1)]]$logFC) %>% filter(gene_name %in% bi_none_gene)\n",
    "\n",
    "                }\n",
    "                \n",
    "                bi_change <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    bi_original_num = length(bi_peak.gr), bi_again_num = length(bi_again), bi_k27_num = length(bi_k27), bi_k4_num = length(bi_k4), bi_none_num = length(bi_none),\n",
    "                    bi_original_gene_num = length(unique(bi_gene$gene_name)), bi_again_gene_num = length(bi_again_gene), bi_k27_gene_num = length(bi_k27_gene), bi_k4_gene_num = length(bi_k4_gene), bi_none_gene_num = length(bi_none_gene)) %>% \n",
    "                    mutate(bi_again_prop = bi_again_num/bi_original_num * 100, bi_k27_prop = bi_k27_num/bi_original_num * 100, bi_k4_prop = bi_k4_num/bi_original_num * 100, bi_none_prop = bi_none_num/bi_original_num * 100) %>% \n",
    "                    mutate(bi_again_gene_prop = bi_again_gene_num/bi_original_gene_num * 100, bi_k27_gene_prop = bi_k27_gene_num/bi_original_gene_num * 100, bi_k4_gene_prop = bi_k4_gene_num/bi_original_gene_num * 100, bi_none_gene_prop = bi_none_gene_num/bi_original_gene_num * 100) %>% \n",
    "                    rbind(bi_change, .)\n",
    "                \n",
    "                bi_change_gene_lfc <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    logFC = bi_again_gene_lfc$GEX, gene_name = bi_again_gene_lfc$gene_name, type = \"bi_again\"\n",
    "                ) %>% rbind(bi_change_gene_lfc, .)\n",
    "                bi_change_gene_lfc <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    logFC = bi_k27_gene_lfc$GEX, gene_name = bi_k27_gene_lfc$gene_name, type = \"bi_k27\"\n",
    "                ) %>% rbind(bi_change_gene_lfc, .)\n",
    "                bi_change_gene_lfc <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    logFC = bi_k4_gene_lfc$GEX, gene_name = bi_k4_gene_lfc$gene_name, type = \"bi_k4\"\n",
    "                ) %>% rbind(bi_change_gene_lfc, .)\n",
    "                if(nrow(bi_none_gene_lfc) > 0){\n",
    "                    bi_change_gene_lfc <- data.frame(\n",
    "                    cT = cT, cT_compare = cT2, pt = pt, \n",
    "                    logFC = bi_none_gene_lfc$GEX, gene_name = bi_none_gene_lfc$gene_name, type = \"bi_none\"\n",
    "                ) %>% rbind(bi_change_gene_lfc, .)\n",
    "                }\n",
    "                \n",
    "            }\n",
    "        }\n",
    "        \n",
    "    }\n",
    "}\n",
    "\n",
    "\n",
    "bi_change %>% head\n",
    "bi_change_gene_lfc %>% head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(tidyr)\n",
    "bi_change$cT <- factor(bi_change$cT, levels = ct_list)\n",
    "bi_change$cT_compare <- factor(bi_change$cT_compare, levels = ct_list)\n",
    "\n",
    "p1 = bi_change %>% pivot_longer(., cols = c(\"bi_again_prop\", \"bi_k27_prop\", \"bi_k4_prop\", \"bi_none_prop\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%  group_by(cT, pt, cT_compare) %>% arrange(rev(name)) %>% mutate(label_y = cumsum(value)) %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = name, label = round(value, 1))) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "geom_text(aes(y = label_y), vjust = 1, stat = \"identity\") +\n",
    "facet_grid(pt~cT, scale = \"free\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Peaks\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p2 = bi_change %>% pivot_longer(., cols = c(\"bi_again_num\", \"bi_k27_num\", \"bi_k4_num\", \"bi_none_num\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%  group_by(cT, pt, cT_compare) %>% arrange(rev(name)) %>% mutate(label_y = cumsum(value)) %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = name, label = round(value, 1))) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "geom_text(aes(y = label_y), vjust = 1, stat = \"identity\") +\n",
    "facet_grid(pt~cT, scale = \"free\", space = \"free_x\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Peaks\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "p3 = bi_change %>% pivot_longer(., cols = c(\"bi_again_gene_prop\", \"bi_k27_gene_prop\", \"bi_k4_gene_prop\", \"bi_none_gene_prop\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%  group_by(cT, pt, cT_compare) %>% arrange(rev(name)) %>% mutate(label_y = cumsum(value)) %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = name, label = round(value, 1))) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "geom_text(aes(y = label_y), vjust = 1, stat = \"identity\") +\n",
    "facet_grid(pt~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rremove(\"legend.title\") \n",
    "\n",
    "p4 = bi_change %>% pivot_longer(., cols = c(\"bi_again_gene_num\", \"bi_k27_gene_num\", \"bi_k4_gene_num\", \"bi_none_gene_num\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%  group_by(cT, pt, cT_compare) %>% arrange(rev(name)) %>% mutate(label_y = cumsum(value)) %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = name, label = round(value, 1))) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "geom_text(aes(y = label_y), vjust = 1, stat = \"identity\") +\n",
    "facet_grid(pt~cT, scale = \"free\", space = \"free_x\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Genes\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "\n",
    "pdf(paste0(fig_path, \"bivalent_peak_gene_change_between_celltypes_perHD.pdf\"), width = 13, height = 13)\n",
    "p1\n",
    "p2\n",
    "p3\n",
    "p4\n",
    "dev.off()\n",
    "\n",
    "ggarrange(p1, p2, p3, p4, nrow = 2, ncol = 2, common.legend = TRUE)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "p1 = bi_change %>% pivot_longer(., cols = c(\"bi_again_prop\", \"bi_k27_prop\", \"bi_k4_prop\", \"bi_none_prop\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = name, label = round(value, 1))) +\n",
    "geom_boxplot() +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Peaks\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "p2 = bi_change %>% pivot_longer(., cols = c(\"bi_again_gene_prop\", \"bi_k27_gene_prop\", \"bi_k4_gene_prop\", \"bi_none_gene_prop\")) %>% dplyr::select(name, value, cT, cT_compare, pt) %>%\n",
    "ggplot(aes(cT_compare, y = value, fill = name, label = round(value, 1))) +\n",
    "geom_boxplot() +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "pdf(paste0(fig_path, \"bivalent_peak_gene_change_between_celltypes.pdf\"), width = 11, height = 17)\n",
    "ggarrange(p1, p2, nrow = 2, ncol = 1, common.legend = TRUE)\n",
    "dev.off()\n",
    "\n",
    "ggarrange(p1, p2, nrow = 2, ncol = 1, common.legend = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "bi_change_gene_lfc %>% head\n",
    "bi_change_gene_lfc$cT <- factor(bi_change_gene_lfc$cT, levels = ct_list)\n",
    "bi_change_gene_lfc$cT_compare <- factor(bi_change_gene_lfc$cT_compare, levels = ct_list)\n",
    "\n",
    "p1 = bi_change_gene_lfc %>% \n",
    "ggplot(aes(cT_compare, y = logFC, fill = type)) +\n",
    "geom_boxplot() +\n",
    "facet_grid(pt~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "xlab(\"\") +\n",
    "ylab(\"logFC of GEX\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "p2 = bi_change_gene_lfc %>% \n",
    "ggplot(aes(cT_compare, y = logFC, fill = type)) +\n",
    "geom_boxplot() +\n",
    "facet_grid(~cT, scale = \"free_x\", space = \"free\") +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE, option = \"magma\", begin = 0.2, end = 1, alpha = 0.8) +\n",
    "xlab(\"\") +\n",
    "ylab(\"logFC of GEX\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "pdf(paste0(fig_path, \"bivalent_peak_gene_change_between_celltypes_gex_logfc_perHD.pdf\"), width = 11, height = 20)\n",
    "p1\n",
    "dev.off()\n",
    "pdf(paste0(fig_path, \"bivalent_peak_gene_change_between_celltypes_gex_logfc.pdf\"), width = 11, height = 8)\n",
    "p2\n",
    "dev.off()\n",
    "p1\n",
    "p2\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## K4 of earlier cell type overlap with K27 of later cell type and no bivalent in each cell type\n",
    "load(file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition.RData\"))\n",
    "rna_results <- results\n",
    "\n",
    "gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")\n",
    "\n",
    "get_nearest_gene <- function(gene_gr, region_gr){\n",
    "    region_assign_gene <- distanceToNearest(region_gr, gene_gr, select = \"all\")\n",
    "    region_match_gene <- cbind(data.frame(region_gr)[region_assign_gene@from, ], data.frame(peak_id = region_assign_gene@from, gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[region_assign_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) )\n",
    "    return(region_match_gene)\n",
    "\n",
    "}\n",
    "\n",
    "switch_summary <- c()\n",
    "gex_switch_summary <- c()\n",
    "ct_list <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "pt_list <- paste0(\"HD\", 4:7)\n",
    "\n",
    "mark1 <- \"H3K4me2\"\n",
    "mark2 <- \"H3K27me3\"\n",
    "for(i in 1:(length(ct_list)-1)){\n",
    "    cT1 <- ct_list[i]\n",
    "    for(j in (i+1):length(ct_list)){\n",
    "        cT2 <- ct_list[j]\n",
    "        for(pt in pt_list){          \n",
    "            bi_olp <- findOverlaps(peakAll[[paste0(mark1, \"_\", cT1, \"_Input_\", pt)]], peakAll[[paste0(mark2, \"_\", cT1, \"_Input_\", pt)]], minoverlap = 50)\n",
    "            bi_olp2 <- findOverlaps(peakAll[[paste0(mark1, \"_\", cT2, \"_Input_\", pt)]], peakAll[[paste0(mark2, \"_\", cT2, \"_Input_\", pt)]], minoverlap = 50)\n",
    "\n",
    "            pure_mark1 <- peakAll[[paste0(mark1, \"_\", cT1, \"_Input_\", pt)]][-unique(bi_olp@from)]\n",
    "            pure_mark2 <- peakAll[[paste0(mark2, \"_\", cT2, \"_Input_\", pt)]][-unique(bi_olp2@to)]\n",
    "            pure_mark1_gene <- get_nearest_gene(gene_gr, pure_mark1)\n",
    "            \n",
    "            switch_mark <- findOverlaps(pure_mark1, pure_mark2, minoverlap = 50)\n",
    "            switch_region <- reduce(append(pure_mark1[switch_mark@from], pure_mark2[switch_mark@to]))\n",
    "            switch_gene <- get_nearest_gene(gene_gr, switch_region)\n",
    "            rna_gene_exp <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% unique(switch_gene$gene_name))\n",
    "            \n",
    "            write.table(unique(switch_gene$gene_name), paste0(bi_path, mark1, \"Of\", cT1, \"_\", mark2, \"Of\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "                \n",
    "\n",
    "            switch_summary <- data.frame(\n",
    "                cT1 = cT1, cT2 = cT2, mark1 = mark1, mark2 = mark2, pt = pt, \n",
    "                switch_region_num = length(switch_region), switch_gene_num = length(unique(switch_gene$gene_name)),\n",
    "                pure_region_num = length(pure_mark1), pure_gene_num = length(unique(pure_mark1_gene$gene_name))\n",
    "            ) %>% rbind(switch_summary, .)\n",
    "\n",
    "            gex_switch_summary <- data.frame(\n",
    "                cT1 = cT1, cT2 = cT2, mark1 = mark1, mark2 = mark2, pt = pt, \n",
    "                logFC = rna_gene_exp$GEX, gene_name = rna_gene_exp$gene_name\n",
    "            ) %>% rbind(gex_switch_summary, .)\n",
    "        }\n",
    "    }\n",
    "}\n",
    "mark1 <- \"H3K27me3\"\n",
    "mark2 <- \"H3K4me2\"\n",
    "for(i in 1:(length(ct_list)-1)){\n",
    "    cT1 <- ct_list[i]\n",
    "    for(j in (i+1):length(ct_list)){\n",
    "        cT2 <- ct_list[j]\n",
    "        for(pt in pt_list){          \n",
    "            bi_olp <- findOverlaps(peakAll[[paste0(mark1, \"_\", cT1, \"_Input_\", pt)]], peakAll[[paste0(mark2, \"_\", cT1, \"_Input_\", pt)]], minoverlap = 50)\n",
    "            bi_olp2 <- findOverlaps(peakAll[[paste0(mark1, \"_\", cT2, \"_Input_\", pt)]], peakAll[[paste0(mark2, \"_\", cT2, \"_Input_\", pt)]], minoverlap = 50)\n",
    "\n",
    "            pure_mark1 <- peakAll[[paste0(mark1, \"_\", cT1, \"_Input_\", pt)]][-unique(bi_olp@from)]\n",
    "            pure_mark2 <- peakAll[[paste0(mark2, \"_\", cT2, \"_Input_\", pt)]][-unique(bi_olp2@to)]\n",
    "            pure_mark1_gene <- get_nearest_gene(gene_gr, pure_mark1)\n",
    "            \n",
    "            switch_mark <- findOverlaps(pure_mark1, pure_mark2, minoverlap = 50)\n",
    "            switch_region <- reduce(append(pure_mark1[switch_mark@from], pure_mark2[switch_mark@to]))\n",
    "            switch_gene <- get_nearest_gene(gene_gr, switch_region)\n",
    "            rna_gene_exp <- data.frame(gene_name = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$GeneName %>% gsub(\".*_\", \"\", .), GEX = rna_results[[paste0(\"Input_\", cT1, \"_\", cT2)]]$logFC) %>% filter(gene_name %in% unique(switch_gene$gene_name))\n",
    "            write.table(unique(switch_gene$gene_name), paste0(bi_path, mark1, \"Of\", cT1, \"_\", mark2, \"_Of\", cT2, \"_nearest_genes_\", pt, \".csv\"), quote = F, row.names = F, col.names = F, sep = \",\")\n",
    "            \n",
    "            switch_summary <- data.frame(\n",
    "                cT1 = cT1, cT2 = cT2, mark1 = mark1, mark2 = mark2, pt = pt, \n",
    "                switch_region_num = length(switch_region), switch_gene_num = length(unique(switch_gene$gene_name)),\n",
    "                pure_region_num = length(pure_mark1), pure_gene_num = length(unique(pure_mark1_gene$gene_name))\n",
    "            ) %>% rbind(switch_summary, .)\n",
    "\n",
    "            gex_switch_summary <- data.frame(\n",
    "                cT1 = cT1, cT2 = cT2, mark1 = mark1, mark2 = mark2, pt = pt, \n",
    "                logFC = rna_gene_exp$GEX, gene_name = rna_gene_exp$gene_name\n",
    "            ) %>% rbind(gex_switch_summary, .)\n",
    "        }\n",
    "    }\n",
    "}\n",
    "head(switch_summary)\n",
    "head(gex_switch_summary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "switch_summary$cT1 <- factor(switch_summary$cT1, levels = ct_list)\n",
    "switch_summary$cT2 <- factor(switch_summary$cT2, levels = ct_list)\n",
    "\n",
    "p1 = switch_summary %>% filter(mark1 == \"H3K4me2\") %>% mutate(switch_region_prop = switch_region_num/pure_region_num * 100) %>%\n",
    "ggplot(aes(x = pt, y = switch_region_prop, fill = cT1)) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "facet_grid(cT2~cT1) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Peaks\") +\n",
    "rotate_x_text(angle = 90) +\n",
    "rremove(\"legend\") +\n",
    "ggtitle(paste0(\"H3K4me2 -> H3K27me3\"))\n",
    "\n",
    "p2 = switch_summary  %>% filter(mark1 == \"H3K4me2\") %>% \n",
    "ggplot(aes(x = pt, y = switch_region_num, fill = cT1)) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "facet_grid(cT2~cT1) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Peaks\") +\n",
    "rotate_x_text(angle = 90) +\n",
    "rremove(\"legend\") +\n",
    "ggtitle(paste0(\"H3K4me2 -> H3K27me3\"))\n",
    "\n",
    "p3 = switch_summary  %>% filter(mark1 == \"H3K4me2\") %>% mutate(switch_gene_prop = switch_gene_num/pure_gene_num * 100) %>%\n",
    "ggplot(aes(x = pt, y = switch_gene_prop, fill = cT1)) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "facet_grid(cT2~cT1) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rotate_x_text(angle = 90) +\n",
    "rremove(\"legend\") +\n",
    "ggtitle(paste0(\"H3K4me2 -> H3K27me3\"))\n",
    "\n",
    "\n",
    "p4 = switch_summary  %>% filter(mark1 == \"H3K4me2\") %>% \n",
    "ggplot(aes(x = pt, y = switch_gene_num, fill = cT1)) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "facet_grid(cT2~cT1) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Genes\") +\n",
    "rotate_x_text(angle = 90) +\n",
    "rremove(\"legend\") +\n",
    "ggtitle(paste0(\"H3K4me2 -> H3K27me3\"))\n",
    "\n",
    "pdf(paste0(fig_path, \"mark_switch_peak_gene_num_prop_K4ToK27.pdf\"), width = 13, height = 13)\n",
    "ggarrange(p1, p2, p3, p4, nrow = 2, ncol = 2)\n",
    "dev.off()\n",
    "\n",
    "ggarrange(p1, p2, p3, p4, nrow = 2, ncol = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "switch_summary$cT1 <- factor(switch_summary$cT1, levels = ct_list)\n",
    "switch_summary$cT2 <- factor(switch_summary$cT2, levels = ct_list)\n",
    "\n",
    "p1 = switch_summary %>% filter(mark1 == \"H3K27me3\") %>% mutate(switch_region_prop = switch_region_num/pure_region_num * 100) %>%\n",
    "ggplot(aes(x = pt, y = switch_region_prop, fill = cT1)) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "facet_grid(cT2~cT1) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Peaks\") +\n",
    "rotate_x_text(angle = 90) +\n",
    "rremove(\"legend\") +\n",
    "ggtitle(paste0(\"H3K27me3 -> H3K4me2\"))\n",
    "\n",
    "p2 = switch_summary  %>% filter(mark1 == \"H3K27me3\") %>% \n",
    "ggplot(aes(x = pt, y = switch_region_num, fill = cT1)) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "facet_grid(cT2~cT1) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Peaks\") +\n",
    "rotate_x_text(angle = 90) +\n",
    "rremove(\"legend\") +\n",
    "ggtitle(paste0(\"H3K27me3 -> H3K4me2\"))\n",
    "\n",
    "p3 = switch_summary  %>% filter(mark1 == \"H3K27me3\") %>% mutate(switch_gene_prop = switch_gene_num/pure_gene_num * 100) %>%\n",
    "ggplot(aes(x = pt, y = switch_gene_prop, fill = cT1)) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "facet_grid(cT2~cT1) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rotate_x_text(angle = 90) +\n",
    "rremove(\"legend\") +\n",
    "ggtitle(paste0(\"H3K27me3 -> H3K4me2\"))\n",
    "\n",
    "\n",
    "p4 = switch_summary  %>% filter(mark1 == \"H3K27me3\") %>% \n",
    "ggplot(aes(x = pt, y = switch_gene_num, fill = cT1)) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "facet_grid(cT2~cT1) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Genes\") +\n",
    "rotate_x_text(angle = 90) +\n",
    "rremove(\"legend\") +\n",
    "ggtitle(paste0(\"H3K27me3 -> H3K4me2\"))\n",
    "\n",
    "pdf(paste0(fig_path, \"mark_switch_peak_gene_num_prop_K27ToK4.pdf\"), width = 13, height = 13)\n",
    "ggarrange(p1, p2, p3, p4, nrow = 2, ncol = 2)\n",
    "dev.off()\n",
    "\n",
    "ggarrange(p1, p2, p3, p4, nrow = 2, ncol = 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "switch_summary$cT1 <- factor(switch_summary$cT1, levels = ct_list)\n",
    "switch_level <- c(\"N->CM\", \"N->EM\", \"N->EMRA\", \"CM->EM\", \"CM->EMRA\", \"EM->EMRA\")\n",
    "p1 = switch_summary %>% mutate(switch_gene_prop = switch_gene_num/pure_gene_num * 100) %>%\n",
    "ggplot(aes(x = factor(paste0(cT1, \"->\", cT2), switch_level), y = switch_gene_prop, fill = cT1)) +\n",
    "geom_boxplot() +\n",
    "facet_grid(~paste0(mark1, \"->\", mark2)) +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "xlab(\"\") +\n",
    "ylab(\"% of Genes\") +\n",
    "rotate_x_text(angle = 90) +\n",
    "rremove(\"legend\") +\n",
    "ylim(0, 7)\n",
    "\n",
    "# p2 = switch_summary  %>% filter(mark1 == \"H3K27me3\") %>% mutate(switch_gene_prop = switch_gene_num/pure_gene_num * 100) %>%\n",
    "# ggplot(aes(x = factor(paste0(cT1, \"->\", cT2), switch_level), y = switch_gene_prop, fill = cT1)) +\n",
    "# geom_boxplot() +\n",
    "# theme_bw(base_size = 20) +\n",
    "# scale_fill_viridis(discrete = TRUE) +\n",
    "# xlab(\"\") +\n",
    "# ylab(\"% of Genes\") +\n",
    "# rotate_x_text(angle = 90) +\n",
    "# rremove(\"legend\") +\n",
    "# ggtitle(paste0(\"H3K27me3 -> H3K4me2\")) +\n",
    "# ylim(0, 7)\n",
    "pdf(paste0(fig_path, \"mark_switch_gene_prop.pdf\"), width = 13, height = 8)\n",
    "# ggarrange(p1, p2, nrow = 1, ncol = 2)\n",
    "p1\n",
    "dev.off()\n",
    "\n",
    "p1\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "switch_level <- c(\"N->CM\", \"N->EM\", \"N->EMRA\", \"CM->EM\", \"CM->EMRA\", \"EM->EMRA\")\n",
    "gex_switch_summary$cT1 <- factor(gex_switch_summary$cT1, levels = ct_list)\n",
    "\n",
    "p1 = gex_switch_summary %>% ggplot(aes(x = factor(paste0(cT1, \"->\", cT2), levels = switch_level), y = logFC, fill = cT1)) +\n",
    "geom_boxplot() +\n",
    "facet_grid(~paste0(mark1, \"->\", mark2)) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"logFC of GEX\") +\n",
    "rremove(\"legend\") +\n",
    "rotate_x_text(angle = 90) +\n",
    "scale_fill_viridis(discrete = TRUE)\n",
    "\n",
    "p2 = gex_switch_summary %>% ggplot(aes(x = factor(paste0(cT1, \"->\", cT2), levels = switch_level), y = logFC, fill = cT1)) +\n",
    "geom_boxplot() +\n",
    "facet_grid(pt~paste0(mark1, \"->\", mark2)) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"logFC of GEX\") +\n",
    "rremove(\"legend\") +\n",
    "rotate_x_text(angle = 90) +\n",
    "scale_fill_viridis(discrete = TRUE)\n",
    "\n",
    "pdf(paste0(fig_path, \"mark_switch_gene_expression_change_logFC.pdf\"), width = 13, height = 8)\n",
    "p1\n",
    "dev.off()\n",
    "\n",
    "pdf(paste0(fig_path, \"mark_switch_gene_expression_change_logFC_perHD.pdf\"), width = 13, height = 20)\n",
    "p2\n",
    "dev.off()\n",
    "\n",
    "p1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## correlation between hd across cell types\n",
    "## bin genome by 500bp\n",
    "# library(corrplot)\n",
    "# reprod = c()\n",
    "# fragCount = NULL\n",
    "# projPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "# for(hist in c(\"H3K27me3\", \"H3K4me2\")){\n",
    "#     for(ct in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "#         for(hd in paste0(\"HD\", 4:7)){\n",
    "#             if(is.null(fragCount)){\n",
    "\n",
    "#                 fragCount = read.table(paste0(projPath, hist, \"_CD8_\", ct, \"_Input_\", hd, \"_rep1/alignment/bowtie2_align.fragmentsCount.bin1000.bed\"), header = FALSE) \n",
    "#                 colnames(fragCount) = c(\"chrom\", \"bin\", paste0(hist, \"_\", ct, \"_\", hd))\n",
    "\n",
    "#             }else{\n",
    "\n",
    "#                 fragCountTmp = read.table(paste0(projPath, hist, \"_CD8_\", ct, \"_Input_\", hd, \"_rep1/alignment/bowtie2_align.fragmentsCount.bin1000.bed\"), header = FALSE) \n",
    "#                 colnames(fragCountTmp) = c(\"chrom\", \"bin\", paste0(hist, \"_\", ct, \"_\", hd))\n",
    "#                 fragCount = full_join(fragCount, fragCountTmp, by = c(\"chrom\", \"bin\"))\n",
    "\n",
    "#             }\n",
    "#         }\n",
    "#     }  \n",
    "# }\n",
    "\n",
    "# M = cor(fragCount %>% dplyr::select(-c(\"chrom\", \"bin\")) %>% log2(), use = \"complete.obs\") \n",
    "# saveRDS(M, file = paste0(\"CART_CUTRUN_Project/results/CUTANDRUN/analysis/RDS/HD_bin1000_corr_M.rds\"))\n",
    "pdf(\"CART_CUTRUN_Project/results/paper_figure/SuppFig-QC/HD_bin1000_reprod_corr.pdf\", width = 20, height = 20)\n",
    "corrplot(M, method = \"color\", outline = T, addgrid.col = \"darkgray\", order=\"hclust\", hclust.method = 'ward.D2', addrect = 2, rect.col = \"black\", rect.lwd = 3,cl.pos = \"b\", tl.col = \"indianred4\", tl.cex = 1, cl.cex = 1, addCoef.col = \"black\", number.digits = 2, number.cex = 1, col = colorRampPalette(c(\"midnightblue\",\"white\",\"darkred\"))(100))\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(corrplot)\n",
    "reprod = c()\n",
    "fragCount = NULL\n",
    "projPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "for(hist in c(\"H3K27me3\", \"H3K4me2\")){\n",
    "    for(ct in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "        for(hd in paste0(\"HD\", 4:7)){\n",
    "            if(is.null(fragCount)){\n",
    "\n",
    "                fragCount = read.table(paste0(projPath, hist, \"_CD8_\", ct, \"_Input_\", hd, \"_rep1/alignment/bowtie2_align.fragmentsCount.bin5000.bed\"), header = FALSE) \n",
    "                colnames(fragCount) = c(\"chrom\", \"bin\", paste0(hist, \"_\", ct, \"_\", hd))\n",
    "\n",
    "            }else{\n",
    "\n",
    "                fragCountTmp = read.table(paste0(projPath, hist, \"_CD8_\", ct, \"_Input_\", hd, \"_rep1/alignment/bowtie2_align.fragmentsCount.bin5000.bed\"), header = FALSE) \n",
    "                colnames(fragCountTmp) = c(\"chrom\", \"bin\", paste0(hist, \"_\", ct, \"_\", hd))\n",
    "                fragCount = full_join(fragCount, fragCountTmp, by = c(\"chrom\", \"bin\"))\n",
    "\n",
    "            }\n",
    "        }\n",
    "    }  \n",
    "}\n",
    "saveRDS(fragCount, file = paste0(\"CART_CUTRUN_Project/results/CUTANDRUN/analysis/RDS/HD_bin5000_fragCount.rds\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "M = cor(fragCount %>% dplyr::select(-c(\"chrom\", \"bin\")) %>% log2(), use = \"complete.obs\")\n",
    "corrplot(M, method = \"color\", outline = T, addgrid.col = \"darkgray\", order=\"hclust\", hclust.method = 'ward.D2', addrect = 2, rect.col = \"black\", rect.lwd = 3,cl.pos = \"b\", tl.col = \"indianred4\", tl.cex = 1, cl.cex = 1, addCoef.col = \"black\", number.digits = 2, number.cex = 1, col = colorRampPalette(c(\"midnightblue\",\"white\",\"darkred\"))(100))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=20, repr.plot.height=20)\n",
    "\n",
    "fragCount[is.na(fragCount)] = 0\n",
    "# head(fragCount)\n",
    "fragCountFilter <- fragCount[rowSums(fragCount %>% dplyr::select(-c(\"chrom\", \"bin\"))) >= quantile(rowSums(fragCount%>% dplyr::select(-c(\"chrom\", \"bin\"))), 0.5), ]\n",
    "# head(fragCountFilter)\n",
    "\n",
    "# dim(fragCount)\n",
    "# dim(fragCountFilter)\n",
    "M = cor(fragCountFilter %>% dplyr::select(-c(\"chrom\", \"bin\")), use = \"complete.obs\") \n",
    "# pdf(\"CART_CUTRUN_Project/results/paper_figure/SuppFig-QC/HD_bin1000_reprod_corr_fillNA0_0.25.pdf\", width = 20, height = 20)\n",
    "\n",
    "corrplot(M, method = \"color\", outline = T, addgrid.col = \"darkgray\", order=\"hclust\", hclust.method = 'ward.D2', addrect = 2, rect.col = \"black\", rect.lwd = 3,cl.pos = \"b\", tl.col = \"indianred4\", tl.cex = 1, cl.cex = 1, addCoef.col = \"black\", number.digits = 2, number.cex = 1, col = colorRampPalette(c(\"midnightblue\",\"white\",\"darkred\"))(100))\n",
    "# dev.off()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
