{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(pacman)\n",
    "p_load(data.table, dplyr, ggplot2, viridis, magrittr, VennDiagram, ggpubr, limma, edgeR, tidyr, GenomicRanges, RColorBrewer, pheatmap, Seurat, fgsea, GSEABase)\n",
    "\n",
    "cell_list <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hd_list <- paste0(\"HD\", c(1:3, 5:7))\n",
    "expr_list = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "cell_comp_list = c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "hist_list <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "in_path <- \"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "out_path <- \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/\"\n",
    "fig_path <- \"CART_CUTRUN_Project/results/paper_figure_pools/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Master peak annotation\n",
    "library(ChIPseeker)\n",
    "load(file = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\")\n",
    "pdf(\"CART_CUTRUN_Project/results/paper_figure/SuppFig-QC/master_peak_annotation.pdf\")\n",
    "plotAnnoPie(mPeakAnno[[\"H3K27me3\"]], title = \"H3K27me3\")\n",
    "plotAnnoPie(mPeakAnno[[\"H3K4me2\"]], title = \"H3K4me2\")\n",
    "plotAnnoBar(mPeakAnno[[\"H3K27me3\"]], title = \"H3K27me3\")\n",
    "plotAnnoBar(mPeakAnno[[\"H3K4me2\"]], title = \"H3K4me2\")\n",
    "plotDistToTSS(mPeakAnno[[\"H3K27me3\"]], title = \"H3K27me3\")\n",
    "plotDistToTSS(mPeakAnno[[\"H3K4me2\"]], title = \"H3K4me2\")\n",
    "dev.off()\n",
    "\n",
    "library(ChIPseeker)\n",
    "library(TxDb.Hsapiens.UCSC.hg38.knownGene)\n",
    "library(org.Hs.eg.db)\n",
    "options(ChIPseeker.ignore_1st_exon = TRUE)\n",
    "options(ChIPseeker.ignore_1st_intron = TRUE)\n",
    "options(ChIPseeker.ignore_downstream = TRUE)\n",
    "options(ChIPseeker.ignore_promoter_subcategory = TRUE)\n",
    "\n",
    "txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene\n",
    "annodb <- \"org.Hs.eg.db\"\n",
    "for(hist in c(\"H3K27me3\", \"H3K4me2\")){\n",
    "    for(ct in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "        peak_gr <- GRanges()\n",
    "        for(hd in paste0(\"HD\", 4:7)){\n",
    "            name = paste(hist, ct, \"Input\", hd, sep = \"_\")\n",
    "            peak_gr = append(peak_gr, peakAll[[name]])\n",
    "        }\n",
    "        peak_ct_anno = annotatePeak(reduce(peak_gr), tssRegion = c(-2000, 0), addFlankGeneInfo = TRUE,  TxDb = txdb, annoDb = annodb,  level = \"gene\")\n",
    "        pdf(paste0(\"CART_CUTRUN_Project/results/paper_figure/SuppFig-QC/\", hist, \"_\", ct, \"_annotation.pdf\"))\n",
    "        print(plotAnnoPie(peak_ct_anno, title = paste0(hist, \"_\", ct)))\n",
    "        print(plotAnnoBar(peak_ct_anno, title = paste0(hist, \"_\", ct)))\n",
    "        print(plotDistToTSS(peak_ct_anno, title = paste0(hist, \"_\", ct)))\n",
    "        dev.off()\n",
    "   }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## -\tNumbers of differentially expressed genes (RNAseq) in transcription of input of N vs CM vs EM vs EMRA - Venn diagram\n",
    "\n",
    "## 1. Pairwise comparisons\n",
    "## VOOM normalized and use donor id as random effect to remove donor variation\n",
    "## save(data, selectR, dataS, voomDDS, results, file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition.RData\"))\n",
    "load(file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition.RData\"))\n",
    "\n",
    "de_num <- c()\n",
    "for(cell_comp in cell_comp_list){\n",
    "    de_num <- data.frame(cond = \"Input\", comparison = cell_comp, de_num = results[[paste0(\"Input_\", cell_comp)]] %>% filter(Significance != \"notDE\") %>% nrow) %>%\n",
    "    rbind(de_num)\n",
    "}\n",
    "de_num$comparison <- factor(de_num$comparison, levels = cell_comp_list)\n",
    "\n",
    "de_num %>% ggplot(aes(x = comparison, y = de_num, label = de_num, fill = comparison)) +\n",
    "geom_bar(stat = \"identity\") +\n",
    "geom_text(vjust = -0.1) +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Differentially Expressed Genes\") +\n",
    "ggtitle(\"Input (FDR <= 0.05 |logFC| >= 1)\") +\n",
    "scale_fill_viridis(discrete = TRUE) +\n",
    "rotate_x_text(angle = 90) +\n",
    "rremove(\"legend\")\n",
    "ggsave(filename = paste0(fig_path, \"RNAseq_pairwise_de_bar_plot.pdf\"), width = 8, height = 7)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "## 2. N vs rest, CM vs rest, ... for venn diagram\n",
    "get_contrast <- function(cell_type, design){\n",
    "\n",
    "  if(cell_type == \"N\"){\n",
    "    contrast <- makeContrasts(\n",
    "      Input_N_Others = Others.Input - N.Input,\n",
    "      Product_N_Others = Others.Product - N.Product,\n",
    "      Stim1_N_Others = Others.Stim1 - N.Stim1,\n",
    "      Stim2_N_Others = Others.Stim2 - N.Stim2,\n",
    "      Stim3_N_Others = Others.Stim3 - N.Stim3,\n",
    "      levels = design\n",
    "    )\n",
    "  }\n",
    "  if(cell_type == \"CM\"){\n",
    "    contrast <- makeContrasts(\n",
    "      Input_CM_Others = Others.Input - CM.Input,\n",
    "      Product_CM_Others = Others.Product - CM.Product,\n",
    "      Stim1_CM_Others = Others.Stim1 - CM.Stim1,\n",
    "      Stim2_CM_Others = Others.Stim2 - CM.Stim2,\n",
    "      Stim3_CM_Others = Others.Stim3 - CM.Stim3,\n",
    "      levels = design\n",
    "    )\n",
    "  }\n",
    "  if(cell_type == \"EM\"){\n",
    "    contrast <- makeContrasts(\n",
    "      Input_EM_Others = Others.Input - EM.Input,\n",
    "      Product_EM_Others = Others.Product - EM.Product,\n",
    "      Stim1_EM_Others = Others.Stim1 - EM.Stim1,\n",
    "      Stim2_EM_Others = Others.Stim2 - EM.Stim2,\n",
    "      Stim3_EM_Others = Others.Stim3 - EM.Stim3,\n",
    "      levels = design\n",
    "    )\n",
    "  }\n",
    "  if(cell_type == \"EMRA\"){\n",
    "    contrast <- makeContrasts(\n",
    "      Input_EMRA_Others = Others.Input - EMRA.Input,\n",
    "      Product_EMRA_Others = Others.Product - EMRA.Product,\n",
    "      Stim1_EMRA_Others = Others.Stim1 - EMRA.Stim1,\n",
    "      Stim2_EMRA_Others = Others.Stim2 - EMRA.Stim2,\n",
    "      Stim3_EMRA_Others = Others.Stim3 - EMRA.Stim3,\n",
    "      levels = design\n",
    "    )\n",
    "  }\n",
    "  return(contrast)\n",
    "}\n",
    "\n",
    "data <- c()\n",
    "col_name_list <- c()\n",
    "for (cell in cellList) {\n",
    "  for (expr in exprList) {\n",
    "    for (hd in hdList) {\n",
    "      ## print(paste(expr, cell, hd, sep = \"_\"))\n",
    "      dataTmp <- fread(paste0(in_path, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".genes.results\")) %>% dplyr::select(expected_count)\n",
    "      data <- cbind(data, round(dataTmp))\n",
    "      col_name_list <- c(col_name_list, paste0(expr, \"_\", cell, \"_\", hd))\n",
    "    }\n",
    "  }\n",
    "}\n",
    "geneID <- fread(paste0(in_path, \"RNA_CD8_N_Input_HD1.genes.results\"))$gene_id\n",
    "rownames(data) <- geneID\n",
    "colnames(data) <- col_name_list ## paste(rep(rep(expr_list, each = length(hd_list)), length(cell_list)), rep(cell_list, each = length(hd_list)*length(expr_list)), rep(hd_list, length(expr_list)*length(cell_list)), sep = \"_\")\n",
    "\n",
    "## Filter and delete low expressed genes\n",
    "selectR <- which(rowSums(data) > 10) ## remove low count genes\n",
    "dataS <- data[selectR, ]\n",
    "geneNameList <- rownames(data)[selectR]\n",
    "\n",
    "\n",
    "for (target_cell in c(\"CM\", \"EM\", \"EMRA\")) { ## \"N\"\n",
    "  target <- data.frame(humanDonor = rep(hdList, length(exprList) * length(cellList)), cell = rep(cellList, each = length(hdList) * length(exprList)), expr = rep(rep(exprList, each = length(hdList))))\n",
    "  target$cell[which(target$cell != target_cell)] <- \"Others\"\n",
    "\n",
    "\n",
    "  ## Experimental design\n",
    "  treat <- factor(paste(target$cell, target$expr, sep = \".\"))\n",
    "  design <- model.matrix(~ 0 + treat)\n",
    "  colnames(design) <- levels(treat)\n",
    "  design %>% head()\n",
    "  contrast <- get_contrast(target_cell, design)\n",
    "  contrast %>% head()\n",
    "\n",
    "  \n",
    "  voomDDS <- voom(counts = dataS, design = design, normalize.method = \"cyclicloess\", plot = TRUE)\n",
    "  ## option 1 using voomDDS option2 using normDDS as normalized input.\n",
    "  inputDDS <- voomDDS\n",
    "  corfit <- duplicateCorrelation(inputDDS, design, block = target$humanDonor)\n",
    "  # corfit$consensus\n",
    "  fit <- lmFit(inputDDS, design, block = target$humanDonor, correlation = corfit$consensus)\n",
    "  fitContrast <- contrasts.fit(fit, contrast)\n",
    "  fitBayes <- eBayes(fitContrast, robust = TRUE)\n",
    "\n",
    "  results <- list()\n",
    "  for (i in 1:ncol(contrast)) {\n",
    "    #- Results\n",
    "    res <- topTable(fit = fitBayes, adjust.method = \"fdr\", coef = i, number = nrow(inputDDS), sort = \"P\")\n",
    "\n",
    "    res <- data.table(GeneName = geneNameList[as.numeric(rownames(res))], GeneIndex = rownames(res) %>% as.numeric(), res)\n",
    "    res[, Significance := ifelse((adj.P.Val <= 0.05 & sign(logFC) == 1 & abs(logFC) >= 1), \"Up\",\n",
    "      ifelse((adj.P.Val <= 0.05 & sign(logFC) == -1 & abs(logFC) >= 1), \"Down\", \"notDE\")\n",
    "    )]\n",
    "    results[[i]] <- res\n",
    "    #- Output\n",
    "    write.table(res, file = paste0(out_path, \"/limma_tables_HD1-3_5-7/one_vs_rest/DE_\", colnames(contrast)[i], \"_adj0.05_logFC1.csv\"), quote = FALSE, row.names = FALSE, sep = \",\")\n",
    "  }\n",
    "  names(results) <- colnames(contrast)\n",
    "  save(data, selectR, dataS, voomDDS, results, file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition_one_vs_rest_\", target_cell, \".RData\"))\n",
    "}\n",
    "\n",
    "de_list <- list()\n",
    "for(target_cell in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    load(file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition_one_vs_rest_\", target_cell, \".RData\"))\n",
    "    de_list[[target_cell]] <- results[[paste0(\"Input_\", target_cell, \"_Others\")]] %>% filter(Significance != \"notDE\") %$% GeneName\n",
    "}\n",
    "# results\n",
    "pdf(paste0(fig_path, \"RNAseq_DE_Input_Venn_diagram.pdf\"))\n",
    "venn.diagram(\n",
    "    x = list(de_list[[\"N\"]], de_list[[\"CM\"]], de_list[[\"EM\"]], de_list[[\"EMRA\"]]),\n",
    "    category.names = c(\n",
    "        paste0(\"N (\", length(de_list[[\"N\"]]), \")\"), \n",
    "        paste0(\"CM (\", length(de_list[[\"CM\"]]), \")\"),\n",
    "        paste0(\"EM (\", length(de_list[[\"EM\"]]), \")\"),\n",
    "        paste0(\"EMRA (\", length(de_list[[\"EMRA\"]]), \")\")),\n",
    "    filename = NULL, #paste0(fig_path, \"RNAseq_DE_Input_Venn_diagram.pdf\"\n",
    "    col = c(\"#440154\", \"#31688e\", \"#35b779\", \"#fde725\"), \n",
    "    fill = c(alpha(\"#440154\", 0.3), alpha(\"#31688e\", 0.3), alpha(\"#35b779\", 0.3), alpha(\"#fde725\", 0.3)),\n",
    "    resolution = 1000, \n",
    "    disable.logging = TRUE\n",
    ") %>% grid.draw\n",
    "\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## -\tWe used CUT&RUN to mark H3K27me3 and H3K4me2 regions. We used an isotype control to control for non-specifically bound DNA\n",
    "## 1. diagram: refer to cut&tag tutorial diagram\n",
    "## 2. QC outputs\n",
    "\n",
    "## 2.1 Sequencing depth\n",
    "seqDepTmp = fread(\"CART_CUTRUN_Project/results/CUTANDRUN/analysis/sequenceDepth.csv\", header = FALSE)\n",
    "seqDep = separate(seqDepTmp, \"V1\", into = c(\"Histone\", \"CD\", \"CellType\", \"Experiment\", \"HD\", \"Replicate\"), sep = \"_\") %>% mutate(CellType = factor(toupper(CellType), levels = c(\"BULK\", \"N\", \"CM\", \"EM\", \"EMRA\")))\n",
    "\n",
    "## data = seqDep %>% filter(CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", CellType != \"BULK\", Experiment != \"PatientIP\") %>% filter(!(Experiment == \"Input\" & (HD == \"HD1\" | HD == \"HD2\" | HD == \"HD3\")))  #, Experiment != \"PatientIP\"\n",
    "data = seqDep %>% filter(Experiment == \"Input\", CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", CellType != \"BULK\", Experiment != \"PatientIP\") %>% filter(!(Experiment == \"Input\" & (HD == \"HD1\" | HD == \"HD2\" | HD == \"HD3\")))  #, Experiment != \"PatientIP\"\n",
    "\n",
    "data %>% ggplot(aes(x = CellType, y = as.numeric(V2)/1000000, fill = CellType)) +\n",
    "    geom_boxplot(outlier.shape = NA) +\n",
    "    ## geom_jitter(aes(color = Experiment), position = position_jitter(0.15)) + \n",
    "    ## geom_jitter(position = position_jitter(0.15)) + \n",
    "    # geom_label_repel(data = data %>% filter(Experiment == \"Input\", HD %in% paste0(\"HD\", c(1, 2, 3, 6))), aes(label = paste0(Experiment, \"_\", HD)), position = position_jitterdodge()) +\n",
    "    facet_grid(~Histone, scales = \"free\", space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"Sequencing Depth per Million\") +\n",
    "    xlab(\"\") \n",
    "ggsave(filename = paste0(fig_path, \"cutrun_seqdepth_before_alignment.pdf\"), width = 10, height = 7)\n",
    "\n",
    "## 2.2 Alignment Rate\n",
    "alignRateTmp = fread(\"CART_CUTRUN_Project/results/CUTANDRUN/analysis/alignmentRate.csv\", header = FALSE)\n",
    "alignRate = separate(alignRateTmp, \"V1\", into = c(\"Histone\", \"CD\", \"CellType\", \"Experiment\", \"HD\", \"Replicate\"), sep = \"_\") %>% mutate(CellType = factor(toupper(CellType), levels = c(\"BULK\", \"N\", \"CM\", \"EM\", \"EMRA\")))\n",
    "\n",
    "## dataAR = alignRate %>% filter(CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\")  %>% filter(!(Experiment == \"Input\" & (HD == \"HD1\" | HD == \"HD2\" | HD == \"HD3\"))) #, Experiment != \"PatientIP\") \n",
    "dataAR = alignRate %>% filter(Experiment == \"Input\", CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\")  %>% filter(!(Experiment == \"Input\" & (HD == \"HD1\" | HD == \"HD2\" | HD == \"HD3\"))) #, Experiment != \"PatientIP\") \n",
    "\n",
    "dataM = left_join(data, dataAR, by = c(\"Histone\", \"CD\", \"CellType\", \"Experiment\", \"HD\", \"Replicate\")) %>% mutate(mappN = V2.x * V2.y/100)\n",
    "\n",
    "dataM %>% ggplot(aes(x = CellType, y = as.numeric(mappN)/1000000, fill = CellType)) +\n",
    "    geom_boxplot(outlier.shape = NA) +\n",
    "    ## geom_jitter(aes(color = Experiment), position = position_jitter(0.15)) + \n",
    "    ## geom_jitter(position = position_jitter(0.15)) + \n",
    "    # geom_label_repel(data = data %>% filter(Experiment == \"Input\", HD %in% paste0(\"HD\", c(1, 2, 3, 6))), aes(label = paste0(Experiment, \"_\", HD)), position = position_jitterdodge()) +\n",
    "    facet_grid(~Histone, scales = \"free\", space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"# of Mappable Read-pairs per Million\") +\n",
    "    xlab(\"\")\n",
    "ggsave(filename = paste0(fig_path, \"cutrun_seqdepth_after_alignment.pdf\"), width = 10, height = 7)\n",
    "\n",
    "dataHR = alignRate %>% filter(Experiment == \"Input\", CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", CellType != \"BULK\", Experiment != \"PatientIP\") %>% filter(!(Experiment == \"Input\" & (HD == \"HD1\" | HD == \"HD2\" | HD == \"HD3\"))) \n",
    "\n",
    "dataHR %>% ggplot(aes(x = CellType, y = as.numeric(V2), fill = CellType)) +\n",
    "    geom_boxplot(outlier.shape = NA) +\n",
    "    ## geom_jitter(aes(color = Experiment), position = position_jitter(0.15)) +\n",
    "    geom_jitter(position = position_jitter(0.15)) +\n",
    "    facet_grid(~Histone, scales = \"free\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"Alignment Rate\") +\n",
    "    xlab(\"\") \n",
    "ggsave(filename = paste0(fig_path, \"cutrun_alignment_rate.pdf\"), width = 10, height = 7)\n",
    "\n",
    "## 2.3 Duplicate rate\n",
    "dupRateTmp = fread(\"CART_CUTRUN_Project/results/CUTANDRUN/analysis/validDupRmRate.csv\", header = FALSE)\n",
    "dupRate = separate(dupRateTmp, \"V1\", into = c(\"Histone\", \"CD\", \"CellType\", \"Experiment\", \"HD\", \"Replicate\"), sep = \"_\") %>% mutate(DupRate = 100 - V3/V2*100, CellType = factor(toupper(CellType), levels = c(\"BULK\", \"N\", \"CM\", \"EM\", \"EMRA\")))\n",
    "\n",
    "data = dupRate %>% filter(CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", CellType != \"BULK\", Experiment != \"PatientIP\") %>% filter(!(Experiment == \"Input\" & (HD == \"HD1\" | HD == \"HD2\" | HD == \"HD3\"))) \n",
    "\n",
    "data %>% ggplot(aes(x = CellType, y = as.numeric(DupRate), fill = CellType)) +\n",
    "    geom_boxplot(outlier.shape = NA) +\n",
    "    geom_jitter(aes(color = Experiment), position = position_jitter(0.15)) +\n",
    "    #geom_label_repel(data = data %>% filter(Experiment == \"Input\", HD %in% paste0(\"HD\", c(1, 2, 3, 6))), aes(label = paste0(Experiment, \"_\", HD)), position = position_jitterdodge()) +\n",
    "    facet_grid(~Histone, scales = \"free\", space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"Duplicates Rate\") +\n",
    "    xlab(\"\") +\n",
    "    ylim(0, 100)\n",
    "ggsave(filename = paste0(fig_path, \"cutrun_duplicates_rate.pdf\"), width = 10, height = 7)\n",
    "\n",
    "data = dupRate %>% filter(CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", CellType != \"BULK\", Experiment != \"PatientIP\") %>% filter(!(Experiment == \"Input\" & (HD == \"HD1\" | HD == \"HD2\" | HD == \"HD3\"))) \n",
    "\n",
    "data %>% ggplot(aes(x = CellType, y = as.numeric(V3)/1000000, fill = CellType)) +\n",
    "    geom_boxplot(outlier.shape = NA) +\n",
    "    geom_jitter(aes(color = Experiment), position = position_jitter(0.15)) +\n",
    "    # geom_label_repel(data = data %>% filter(Experiment == \"Input\", HD %in% paste0(\"HD\", c(1, 2, 3, 6))), aes(label = paste0(Experiment, \"_\", HD)), position = position_jitterdodge()) +\n",
    "    facet_grid(~Histone, scales = \"free\", space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"Estimated Unique Library Size per Million\") +\n",
    "    xlab(\"\") \n",
    "ggsave(filename = paste0(fig_path, \"cutrun_unique_library_size.pdf\"), width = 10, height = 7)\n",
    "\n",
    "## 2.4 Fragment length\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "fileList = fread(\"CART_CUTRUN_Project/meta/Histone_IgG_pairing.csv\")$histone\n",
    "fragL = c()\n",
    "for(file in fileList){\n",
    "    fragL = fread(paste0(inPath, file, \"/alignment/bowtie2_align.fragLen.freq\")) %>% mutate(info = file, Weight = V2/sum(V2)) %>% rbind(fragL, .)\n",
    "\n",
    "}\n",
    "\n",
    "fileList = fread(\"CART_CUTRUN_Project/meta/Histone_IgG_pairing.csv\")$IgG\n",
    "for(file in fileList){\n",
    "    fragL = fread(paste0(inPath, file, \"/alignment/bowtie2_align.fragLen.freq\")) %>% mutate(info = file, Weight = V2/sum(V2)) %>% rbind(fragL, .)\n",
    "\n",
    "}\n",
    "fragLenD = separate(fragL, \"info\", into = c(\"Histone\", \"CD\", \"CellType\", \"Experiment\", \"HD\", \"Replicate\"), sep = \"_\") %>% mutate(CellType = factor(toupper(CellType), levels = c(\"BULK\", \"N\", \"CM\", \"EM\", \"EMRA\"))) %>% rename(fragLen = V1, Freq = V2)\n",
    "\n",
    "fragData = fragLenD %>% filter(CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", CellType != \"BULK\", Experiment != \"PatientIP\") %>% filter(!(Experiment == \"Input\" & (HD == \"HD1\" | HD == \"HD2\" | HD == \"HD3\"))) \n",
    "\n",
    "fragData %>% ggplot(aes(x = CellType, y = as.numeric(fragLen), weight = Weight, fill = CellType)) +\n",
    "    geom_violin(bw = 1) +\n",
    "    # geom_jitter(aes(color = Experiment), position = position_jitter(0.15)) +\n",
    "    facet_grid(~Histone, scales = \"free\", space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"Fragment Length\") +\n",
    "    xlab(\"\") +\n",
    "    scale_y_continuous(breaks = seq(50, 700, 50)) +\n",
    "    rremove(\"legend\")\n",
    "\n",
    "ggsave(filename = paste0(fig_path, \"cutrun_fragment_length.pdf\"), width = 10, height = 7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## 3 Peak calling\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "fileList = fread(\"CART_CUTRUN_Project/meta/Histone_IgG_pairing.csv\")$histone\n",
    "\n",
    "convertGR = function(peak){\n",
    "    if(nrow(peak) > 0){\n",
    "        peak.gr = GRanges(seqname = peak$V1, IRanges(peak$V2, peak$V3), strand = \"*\")\n",
    "    }else{\n",
    "        peak.gr = GRanges()\n",
    "    }\n",
    "    return(peak.gr)\n",
    "}\n",
    "# peakN = c()\n",
    "# peakLen = c()\n",
    "# peakCov = c()\n",
    "# peakOverlap = c()\n",
    "# for(file in fileList){\n",
    "#     print(file)\n",
    "#     macsNarrow = fread(paste0(inPath, file, \"/peakCalling/normLibSize/MACS2/macs2_narrow_q_0.05_peaks.narrowPeak\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\")\n",
    "#     macsBroad = fread(paste0(inPath, file, \"/peakCalling/normLibSize/MACS2/macs2_broad_q_0.05_peaks.broadPeak\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\")\n",
    "#     seacrControl = fread(paste0(inPath, file, \"/peakCalling/normLibSize/SEACR/seacr_control_seacrNorm.peaks.stringent.bed\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\")\n",
    "#     seacrTop = fread(paste0(inPath, file, \"/peakCalling/normLibSize/SEACR/seacr_top0.1_spikeInNorm.peaks.stringent.bed\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\")\n",
    "    \n",
    "    \n",
    "#     peakN = data.frame(peakN = c(macsNarrow %>% nrow, macsBroad %>% nrow, seacrControl %>% nrow, seacrTop %>% nrow), type = c(\"MACS2_Narrow\", \"MACS2_Broad\", \"SEACR_Control\", \"SEACR_Top10%\"), info = file) %>% rbind(peakN, .)\n",
    "    \n",
    "#     peakLen = data.frame(peakLen = c(macsNarrow$V3 - macsNarrow$V2, macsBroad$V3 - macsBroad$V2, seacrControl$V3 - seacrControl$V2, seacrTop$V3 - seacrTop$V2), type = c(rep(\"MACS2_Narrow\", nrow(macsNarrow)), rep(\"MACS2_Broad\", nrow(macsBroad)), rep(\"SEACR_Control\", nrow(seacrControl)), rep(\"SEACR_Top10%\", nrow(seacrTop))), info = file) %>% rbind(peakLen, .)\n",
    "    \n",
    "#     peakCov = data.frame(peakLen = c(sum(macsNarrow$V3 - macsNarrow$V2), sum(macsBroad$V3 - macsBroad$V2),  sum(seacrControl$V3 - seacrControl$V2), sum(seacrTop$V3 - seacrTop$V2)), type = c(\"MACS2_Narrow\", \"MACS2_Broad\", \"SEACR_Control\", \"SEACR_Top10%\"), info = file) %>% rbind(peakCov, .)\n",
    "    \n",
    "#     macsNarrow.gr = convertGR(macsNarrow)\n",
    "#     macsBroad.gr = convertGR(macsBroad)\n",
    "#     seacrControl.gr = convertGR(seacrControl)\n",
    "#     seacrTop.gr = convertGR(seacrTop)\n",
    "    \n",
    "#     narrowControl = findOverlaps(macsNarrow.gr, seacrControl.gr)\n",
    "#     broadControl = findOverlaps(macsBroad.gr, seacrControl.gr)\n",
    "#     narrowTop = findOverlaps(macsNarrow.gr, seacrTop.gr)\n",
    "#     broadTop = findOverlaps(macsBroad.gr, seacrTop.gr)\n",
    "\n",
    "#     peakOverlap = data.frame(\n",
    "#         overlapN = c(\n",
    "#             narrowControl@from %>% unique %>% length, \n",
    "#             narrowControl@to %>% unique %>% length,\n",
    "#             broadControl@from %>% unique %>% length, \n",
    "#             broadControl@to %>% unique %>% length,\n",
    "#             narrowTop@from %>% unique %>% length,\n",
    "#             narrowTop@to %>% unique %>% length,\n",
    "#             broadTop@from %>% unique %>% length, \n",
    "#             broadTop@to %>% unique %>% length\n",
    "#         ),\n",
    "#         overlapProp = c(\n",
    "#             (narrowControl@from %>% unique %>% length)/length(macsNarrow.gr) * 100, \n",
    "#             (narrowControl@to %>% unique %>% length)/length(seacrControl.gr) * 100,\n",
    "#             (broadControl@from %>% unique %>% length)/length(macsBroad.gr) * 100, \n",
    "#             (broadControl@to %>% unique %>% length)/length(seacrControl.gr) * 100,\n",
    "#             (narrowTop@from %>% unique %>% length)/length(macsNarrow.gr) * 100,\n",
    "#             (narrowTop@to %>% unique %>% length)/length(seacrTop.gr) * 100,\n",
    "#             (broadTop@from %>% unique %>% length)/length(macsBroad.gr) * 100, \n",
    "#             (broadTop@to %>% unique %>% length)/length(seacrTop.gr) * 100\n",
    "#         ),\n",
    "#         propLabel = c(\n",
    "#             \"MACS2_Narrow vs SEACR_Control/MACS2_Narrow\", \n",
    "#             \"MACS2_Narrow vs SEACR_Control/SEACR_Control\", \n",
    "#             \"MACS2_Broad vs SEACR_Control/MACS2_Broad\", \n",
    "#             \"MACS2_Broad vs SEACR_Control/SEACR_Control\", \n",
    "#             \"MACS2_Narrow vs SEACR_Top/MACS2_Narrow\", \n",
    "#             \"MACS2_Narrow vs SEACR_Top/SEACR_Top\", \n",
    "#             \"MACS2_Broad vs SEACR_Top/MACS2_Broad\", \n",
    "#             \"MACS2_Broad vs SEACR_Top/SEACR_Top\"\n",
    "#         ), \n",
    "#         numLabel = c(\n",
    "#             \"MACS2_Narrow vs SEACR_Control: MACS2_Narrow\", \n",
    "#             \"MACS2_Narrow vs SEACR_Control: SEACR_Control\", \n",
    "#             \"MACS2_Broad vs SEACR_Control: MACS2_Broad\", \n",
    "#             \"MACS2_Broad vs SEACR_Control: SEACR_Control\", \n",
    "#             \"MACS2_Narrow vs SEACR_Top: MACS2_Narrow\", \n",
    "#             \"MACS2_Narrow vs SEACR_Top: SEACR_Top\", \n",
    "#             \"MACS2_Broad vs SEACR_Top: MACS2_Broad\", \n",
    "#             \"MACS2_Broad vs SEACR_Top: SEACR_Top\"\n",
    "#         ), \n",
    "#         info = file\n",
    "#     ) %>% rbind(peakOverlap, .)\n",
    "    \n",
    "# }\n",
    "# save(peakN, peakCov, peakLen, peakOverlap, file = paste0(outPath, \"RData/peakN_Cov_Len_summary_macs_narrow_broad_q_0.05_seacr_control_top10.RData\")) ##  peakOverlap,\n",
    "load(file = paste0(outPath, \"RData/peakN_Cov_Len_summary_macs_narrow_broad_q_0.05_seacr_control_top10.RData\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# 3 Comparison of MACS2 and SEACR\n",
    "# 3.1 Numbers of peaks\n",
    "peakNData = separate(peakN, \"info\", into = c(\"Histone\", \"CD\", \"CellType\", \"Experiment\", \"HD\", \"Replicate\"), sep = \"_\") %>% mutate(CellType = factor(toupper(CellType), levels = c(\"BULK\", \"N\", \"CM\", \"EM\", \"EMRA\")))\n",
    "peakNData$type[peakNData$type == \"SEACR_Top1%\"] <- \"SEACR_Top10%\"\n",
    "peakNData$type <- factor(\n",
    "    peakNData$type, \n",
    "    levels = c(\"MACS2_Narrow\", \"MACS2_Broad\", \"SEACR_Control\", \"SEACR_Top10%\"), \n",
    "    labels = c(\"MACS2 Narrow\", \"MACS2 Broad\", \"SEACR Control\", \"SEACR Top10%\")\n",
    ")\n",
    "\n",
    "peakNData %>% filter(CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", Experiment != \"PatientIP\") %>% ggplot(aes(x = CellType, y = as.numeric(peakN), fill = CellType)) +\n",
    "    geom_boxplot(outlier.shape = NA) +\n",
    "    geom_jitter(aes(color = Experiment), position = position_jitter(0.15)) +\n",
    "    facet_grid(Histone~type, scales = \"free\", space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"# of Peaks\") +\n",
    "    xlab(\"\") \n",
    "\n",
    "ggsave(filename = paste0(fig_path, \"cutrun_peak_number_macs_q_0.05_seacr_top10.pdf\"), width = 12, height = 7)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "\n",
    "# 3.2\tPeak width and coverage\n",
    "peakLenData = separate(peakLen, \"info\", into = c(\"Histone\", \"CD\", \"CellType\", \"Experiment\", \"HD\", \"Replicate\"), sep = \"_\") %>% \n",
    "mutate(CellType = factor(toupper(CellType), levels = c(\"N\", \"CM\", \"EM\", \"EMRA\")))\n",
    "\n",
    "save(peakLenData, file = paste0(outPath, \"RData/peakOverlap_peakLenData_macs_q_0.05_seacr_top10.RData\"))\n",
    "\n",
    "\n",
    "# load(file = paste0(outPath, \"RData/peakOverlap_peakLenData_macs_q_0.05_seacr_top10.RData\"))\n",
    "# dim(peakLenData)\n",
    "\n",
    "peakLenData$type[peakLenData$type == \"SEACR_Top1%\"] <- \"SEACR_Top10%\"\n",
    "peakLenData$type <- factor(\n",
    "    peakLenData$type, \n",
    "    levels = c(\"MACS2_Narrow\", \"MACS2_Broad\", \"SEACR_Control\", \"SEACR_Top10%\"), \n",
    "    labels = c(\"MACS2 Narrow\", \"MACS2 Broad\", \"SEACR Control\", \"SEACR Top10%\")\n",
    ")\n",
    "\n",
    "peakLenData %>% filter(CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", Experiment != \"PatientIP\") %>% ggplot(aes(x = CellType, y = as.numeric(peakLen), fill = CellType)) +\n",
    "    geom_violin() +\n",
    "    # geom_jitter(aes(color = Experiment), position = position_jitter(0.15)) +\n",
    "    facet_grid(Histone~type, scales = \"free\", space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"Length of Peaks (bp)\") +\n",
    "    xlab(\"\") \n",
    "ggsave(filename = paste0(fig_path, \"cutrun_peak_length_macs_q_0.05_seacr_top10_violinplot.pdf\"), width = 12, height = 7)\n",
    "\n",
    "peakLenData %>% filter(CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", Experiment != \"PatientIP\") %>% ggplot(aes(x = CellType, y = as.numeric(peakLen), fill = CellType)) +\n",
    "    geom_boxplot(outlier.shape = NA) +\n",
    "    facet_grid(Histone~type, scales = \"free\", space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"Length of Peaks (bp)\") +\n",
    "    xlab(\"\") +\n",
    "    coord_cartesian(ylim = c(0, 15000)) \n",
    "ggsave(filename = paste0(fig_path, \"cutrun_peak_length_macs_q_0.05_seacr_top10_boxplot.pdf\"), width = 12, height = 7)\n",
    "\n",
    "peakCovData = separate(peakCov, \"info\", into = c(\"Histone\", \"CD\", \"CellType\", \"Experiment\", \"HD\", \"Replicate\"), sep = \"_\") %>% mutate(CellType = factor(toupper(CellType), levels = c(\"BULK\", \"N\", \"CM\", \"EM\", \"EMRA\")))\n",
    "peakCovData$type[peakCovData$type == \"SEACR_Top1%\"] <- \"SEACR_Top10%\"\n",
    "peakCovData$type <- factor(\n",
    "    peakCovData$type, \n",
    "    levels = c(\"MACS2_Narrow\", \"MACS2_Broad\", \"SEACR_Control\", \"SEACR_Top10%\"), \n",
    "    labels = c(\"MACS2 Narrow\", \"MACS2 Broad\", \"SEACR Control\", \"SEACR Top10%\")\n",
    ")\n",
    "peakCovData %>% filter(CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", Experiment != \"PatientIP\") %>% ggplot(aes(x = CellType, y = as.numeric(peakLen), fill = CellType)) +\n",
    "    geom_boxplot(outlier.shape = NA) +\n",
    "    geom_jitter(aes(color = Experiment), position = position_jitter(0.15)) +\n",
    "    facet_grid(Histone~type, scales = \"free\", space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"Coverage of Peaks (bp)\") +\n",
    "    xlab(\"\") \n",
    "\n",
    "ggsave(filename = paste0(fig_path, \"cutrun_peak_coverage_macs_q_0.05_seacr_top10.pdf\"), width = 12, height = 7)\n",
    "\n",
    "\n",
    "# # 3.3\tMethod comparison\n",
    "\n",
    "peakOverlapData = separate(peakOverlap, \"info\", into = c(\"Histone\", \"CD\", \"CellType\", \"Experiment\", \"HD\", \"Replicate\"), sep = \"_\") %>% mutate(CellType = factor(toupper(CellType), levels = c(\"BULK\", \"N\", \"CM\", \"EM\", \"EMRA\")))\n",
    "peakOverlapData$propLabel <- factor(\n",
    "    peakOverlapData$propLabel, \n",
    "    levels = c(\"MACS2_Narrow vs SEACR_Control/MACS2_Narrow\", \n",
    "            \"MACS2_Narrow vs SEACR_Control/SEACR_Control\", \n",
    "            \"MACS2_Broad vs SEACR_Control/MACS2_Broad\", \n",
    "            \"MACS2_Broad vs SEACR_Control/SEACR_Control\", \n",
    "            \"MACS2_Narrow vs SEACR_Top/MACS2_Narrow\", \n",
    "            \"MACS2_Narrow vs SEACR_Top/SEACR_Top\", \n",
    "            \"MACS2_Broad vs SEACR_Top/MACS2_Broad\", \n",
    "            \"MACS2_Broad vs SEACR_Top/SEACR_Top\"), \n",
    "    labels = c(\"Narrow vs Control/Narrow\", \n",
    "            \"Narrow vs Control/Control\", \n",
    "            \"Broad vs Control/Broad\", \n",
    "            \"Broad vs Control/Control\", \n",
    "            \"Narrow vs Top/Narrow\", \n",
    "            \"Narrow vs Top/Top\", \n",
    "            \"Broad vs Top/Broad\", \n",
    "            \"Broad vs Top/Top\")\n",
    ")\n",
    "peakOverlapData %>% filter(CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", Experiment != \"PatientIP\") %>% ggplot(aes(x = CellType, y = as.numeric(overlapProp), fill = CellType)) +\n",
    "    geom_boxplot(outlier.shape = NA) +\n",
    "    geom_jitter(aes(color = Experiment), position = position_jitter(0.15)) +\n",
    "    facet_grid(propLabel~Histone, scales = \"free\", space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"% of Peaks Overlapped\") +\n",
    "    xlab(\"\") \n",
    "ggsave(filename = paste0(fig_path, \"cutrun_peak_overlap_prop_macs_q_0.05_seacr_top10.pdf\"), width = 10, height = 25)\n",
    "\n",
    "\n",
    "peakOverlapData %>% filter(CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", Experiment != \"PatientIP\") %>% ggplot(aes(x = CellType, y = as.numeric(overlapN), fill = CellType)) +\n",
    "    geom_boxplot(outlier.shape = NA) +\n",
    "    geom_jitter(aes(color = Experiment), position = position_jitter(0.15)) +\n",
    "    facet_grid(numLabel~Histone, scales = \"free\", space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"# of Peaks Overlapped\") +\n",
    "    xlab(\"\") \n",
    "ggsave(filename = paste0(fig_path, \"cutrun_peak_overlap_num_macs_q_0.05_seacr_top10.pdf\"), width = 10, height = 25)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# 3.4\tPrecision across donors\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "\n",
    "blackList = c(\"H3K27me3_CD8_CM_Input_HD1_rep1\", \"H3K27me3_CD8_CM_Input_HD3_rep2\", \"H3K27me3_CD8_N_Input_HD2_rep1\", \"H3K27me3_CD8_N_Input_HD3_rep2\")\n",
    "fileList = fread(\"CART_CUTRUN_Project/meta/Histone_IgG_pairing.csv\") %>% select(histone) %>% mutate(path = histone) %>% filter(! histone %in% blackList) %>% separate(\"histone\", into = c(\"Histone\", \"CD\", \"CellType\", \"Experiment\", \"HD\", \"Replicate\"), sep = \"_\") \n",
    "fileListUni = fileList %>% select(Histone, CD, CellType, Experiment) %>% unique\n",
    "convertGR = function(peak){\n",
    "    peak.gr = GRanges(seqname = peak$V1, IRanges(peak$V2, peak$V3), strand = \"*\")\n",
    "    return(peak.gr)\n",
    "}\n",
    "\n",
    "reprodD = c()\n",
    "for(i in 1:nrow(fileListUni)){\n",
    "  print(i)\n",
    "  hist = fileListUni[i, 1] %>% as.character\n",
    "  cd = fileListUni[i, 2] %>% as.character\n",
    "  cell = fileListUni[i, 3] %>% as.character\n",
    "  exp = fileListUni[i, 4] %>% as.character\n",
    "  hdList = fileList %>% filter(Histone == hist, CD == cd, CellType == cell, Experiment == exp)\n",
    "  for(p1 in 1:nrow(hdList)){\n",
    "      for(p2 in 1:nrow(hdList)){\n",
    "        if(p1 != p2){\n",
    "          \n",
    "          ## MACS2 Narrow\n",
    "          peak1.gr = fread(paste0(inPath, hdList$path[p1], \"/peakCalling/normLibSize/MACS2/macs2_narrow_q_0.05_peaks.narrowPeak\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\") %>% convertGR()\n",
    "          peak2.gr = fread(paste0(inPath, hdList$path[p2], \"/peakCalling/normLibSize/MACS2/macs2_narrow_q_0.05_peaks.narrowPeak\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\") %>% convertGR()\n",
    "          overlap = findOverlaps(peak1.gr, peak2.gr)\n",
    "          reprodD = data.frame(info = hdList$path[p1], overlapHD = hdList$HD[p2], overlapN = overlap@from %>% unique %>% length, overlapP = overlap@from %>% unique %>% length/length(peak1.gr) * 100, peakN = length(peak1.gr), caller = \"MACS2_Narrow\") %>% rbind(reprodD)\n",
    "          \n",
    "          ## MACS2 Broad\n",
    "          peak1.gr = fread(paste0(inPath, hdList$path[p1], \"/peakCalling/normLibSize/MACS2/macs2_broad_q_0.05_peaks.broadPeak\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\") %>% convertGR()\n",
    "          peak2.gr = fread(paste0(inPath, hdList$path[p2], \"/peakCalling/normLibSize/MACS2/macs2_broad_q_0.05_peaks.broadPeak\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\") %>% convertGR()\n",
    "          overlap = findOverlaps(peak1.gr, peak2.gr)\n",
    "          reprodD = data.frame(info = hdList$path[p1], overlapHD = hdList$HD[p2], overlapN = overlap@from %>% unique %>% length, overlapP = overlap@from %>% unique %>% length/length(peak1.gr) * 100, peakN = length(peak1.gr), caller = \"MACS2_Broad\") %>% rbind(reprodD)\n",
    "         \n",
    "           ## SEACR Control\n",
    "          peak1.gr = fread(paste0(inPath, hdList$path[p1], \"/peakCalling/normLibSize/SEACR/seacr_control_seacrNorm.peaks.stringent.bed\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\") %>% convertGR()\n",
    "          peak2.gr = fread(paste0(inPath, hdList$path[p2], \"/peakCalling/normLibSize/SEACR/seacr_control_seacrNorm.peaks.stringent.bed\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\") %>% convertGR()\n",
    "          overlap = findOverlaps(peak1.gr, peak2.gr)\n",
    "          reprodD = data.frame(info = hdList$path[p1], overlapHD = hdList$HD[p2], overlapN = overlap@from %>% unique %>% length, overlapP = overlap@from %>% unique %>% length/length(peak1.gr) * 100, peakN = length(peak1.gr), caller = \"SEACR_Control\") %>% rbind(reprodD)\n",
    "    \n",
    "          p1N = length(peak1.gr)\n",
    "          p2N = length(peak2.gr)\n",
    "          peak1Con.gr = peak1.gr\n",
    "          peak2Con.gr = peak2.gr\n",
    "          \n",
    "          ## SEACR Top\n",
    "          peak1.gr = fread(paste0(inPath, hdList$path[p1], \"/peakCalling/normLibSize/SEACR/seacr_top0.1_spikeInNorm.peaks.stringent.bed\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\") %>% convertGR()\n",
    "          peak2.gr = fread(paste0(inPath, hdList$path[p2], \"/peakCalling/normLibSize/SEACR/seacr_top0.1_spikeInNorm.peaks.stringent.bed\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\") %>% convertGR()\n",
    "          overlap = findOverlaps(peak1.gr, peak2.gr)\n",
    "          reprodD = data.frame(info = hdList$path[p1], overlapHD = hdList$HD[p2], overlapN = overlap@from %>% unique %>% length, overlapP = overlap@from %>% unique %>% length/length(peak1.gr) * 100, peakN = length(peak1.gr), caller = \"SEACR_Top\") %>% rbind(reprodD)\n",
    "          \n",
    "          # ## SEACR Control in Top\n",
    "          # peak1ConTop.gr = peak1Con.gr[findOverlaps(peak1Con.gr, peak1.gr)@from %>% unique]\n",
    "          # peak2ConTop.gr = peak2Con.gr[findOverlaps(peak2Con.gr, peak2.gr)@from %>% unique]\n",
    "          # overlap = findOverlaps(peak1ConTop.gr, peak2ConTop.gr)\n",
    "          # reprodD = data.frame(info = hdList$path[p1], overlapHD = hdList$HD[p2], overlapN = overlap@from %>% unique %>% length, overlapP = overlap@from %>% unique %>% length/length(peak1ConTop.gr) * 100, peakN = length(peak1ConTop.gr), caller = \"SEACR_ControlInTop\") %>% rbind(reprodD)\n",
    "          \n",
    "          # ## MACS2 - topN\n",
    "          # peak1.gr = fread(paste0(inPath, hdList$path[p1], \"/peakCalling/relax/MACS2/macs2_noControl_peaks.narrowPeak\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\") %>% top_n(p1N, wt = V9) %>% convertGR()\n",
    "          # peak2.gr = fread(paste0(inPath, hdList$path[p2], \"/peakCalling/relax/MACS2/macs2_noControl_peaks.narrowPeak\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\") %>% top_n(p2N, wt = V9) %>% convertGR()\n",
    "          # overlap = findOverlaps(peak1.gr, peak2.gr)\n",
    "          # reprodD = data.frame(info = hdList$path[p1], overlapHD = hdList$HD[p2], overlapN = overlap@from %>% unique %>% length, overlapP = overlap@from %>% unique %>% length/length(peak1.gr) * 100, peakN = length(peak1.gr), caller = \"MACS2(TopN=SEACR_Control)\") %>% rbind(reprodD)\n",
    "         }\n",
    "        \n",
    "      }\n",
    "    \n",
    "  }\n",
    "  \n",
    "}\n",
    "reprodD$caller = factor(reprodD$caller, levels = c(\"MACS2_Narrow\", \"MACS2_Broad\", \"SEACR_Control\", \"SEACR_Top\")) ##\"MACS2(TopN=SEACR_Control)\", \"SEACR_ControlInTop\"\n",
    "save(reprodD, file = paste0(outPath, \"RData/peak_comparisons_macs_narrow_broad_q_0.05_seacr_contro_top10.RData\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "reprodData = separate(reprodD, \"info\", into = c(\"Histone\", \"CD\", \"CellType\", \"Experiment\", \"HD\", \"Replicate\"), sep = \"_\") %>% mutate(CellType = factor(toupper(CellType), levels = c(\"BULK\", \"N\", \"CM\", \"EM\", \"EMRA\")))\n",
    "\n",
    "reprodData %>% filter(CD == \"CD8\", CellType != \"MAIT\", CellType != \"NA\", Experiment != \"PatientIP\") %>% ggplot(aes(x = CellType, y = as.numeric(overlapP), fill = CellType)) +\n",
    "    geom_boxplot() +\n",
    "    # geom_jitter(aes(color = Experiment), position = position_jitter(0.15)) +\n",
    "    facet_grid(Histone~caller, scales = \"free\", space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    ylab(\"Reproducibility (%)\") +\n",
    "    xlab(\"\") +\n",
    "    rremove(\"legend\")\n",
    "ggsave(filename = paste0(fig_path, \"cutrun_peak_reproducibility_macs_q_0.05_seacr_top10.pdf\"), width = 12, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "fileList = fread(\"CART_CUTRUN_Project/meta/Histone_IgG_pairing.csv\")$histone\n",
    "\n",
    "chrom_num <- c()\n",
    "for(file in fileList){\n",
    "    print(file)\n",
    "    macsNarrow = fread(paste0(inPath, file, \"/peakCalling/normLibSize/MACS2/macs2_narrow_q_0.05_peaks.narrowPeak\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\")\n",
    "    macsBroad = fread(paste0(inPath, file, \"/peakCalling/normLibSize/MACS2/macs2_broad_q_0.05_peaks.broadPeak\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\")\n",
    "    seacrControl = fread(paste0(inPath, file, \"/peakCalling/normLibSize/SEACR/seacr_control_seacrNorm.peaks.stringent.bed\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\")\n",
    "    seacrTop = fread(paste0(inPath, file, \"/peakCalling/normLibSize/SEACR/seacr_top0.1_spikeInNorm.peaks.stringent.bed\")) %>% filter(V1 != \"chrY\", V1 != \"chrM\")\n",
    "    chrom_num <- data.frame(\n",
    "        chrom_num = c(\n",
    "            macsNarrow$V1 %>% unique %>% length, \n",
    "            macsBroad$V1 %>% unique %>% length,\n",
    "            seacrControl$V1 %>% unique %>% length,\n",
    "            seacrTop$V1 %>% unique %>% length\n",
    "        ),\n",
    "        type = c(\"MACS2 Narrow\", \"MACS2 Broad\", \"SEACR Control\", \"SEACR Top\"),\n",
    "        file = file\n",
    "    ) %>% rbind(chrom_num, .)\n",
    "}\n",
    "saveRDS(chrom_num, file = paste0(outPath, \"RData/peak_count_chrom_num.rds\"))\n",
    "\n",
    "chrom_num$type = factor(chrom_num$type, levels = c(\"MACS2 Narrow\", \"MACS2 Broad\", \"SEACR Control\", \"SEACR Top\"))\n",
    "chrom_num %>% ggplot(aes(x = type, y = chrom_num, fill = type)) +\n",
    "geom_boxplot() +\n",
    "scale_fill_brewer(palette = \"Set1\") +\n",
    "theme_bw(base_size = 20) +\n",
    "xlab(\"\") +\n",
    "ylab(\"Number of Chromosome having Peaks\") +\n",
    "scale_y_continuous(breaks = 1:23) +\n",
    "rremove(\"legend\") +\n",
    "rotate_x_text(angle = 90)\n",
    "\n",
    "ggsave(filename = paste0(fig_path, \"cutrun_number_chromosome_with_peak_macs_q_0.05_seacr_top10.pdf\"), width = 10, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "##\t3.5 FriP\n",
    "# load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/IDR_master_peak_list_chromVar_count_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "inMasterPeak = c()\n",
    "for(hist in hist_list){\n",
    "  inMasterPeak = colSums(countMat[[hist]]) %>% c(inMasterPeak, .)\n",
    "}\n",
    "\n",
    "frPeakTmp = cbind(cbind(frPeak, seqDepth = designInfo$depth), inMasterPeak = inMasterPeak) %>% dplyr::mutate(inPeakProp =  inPeakN/seqDepth * 100, inMasterProp = inMasterPeak/seqDepth * 100)\n",
    "\n",
    "frPeakD = rbind(frPeakTmp %>% select(-inMasterPeak, -inMasterProp) %>% dplyr::mutate(type = \"Individual Peak\"), frPeakTmp %>% select(-inPeakN, -inPeakProp) %>% mutate(inPeakN = inMasterPeak, inPeakProp = inMasterProp, type = \"Master Peak\") %>% select(-inMasterPeak, -inMasterProp))\n",
    "\n",
    "frPeakD$cell = factor(frPeakD$cell, levels = cell_list)\n",
    "\n",
    "frPeakD %>% \n",
    "  ggplot(aes(x = cell, y = as.numeric(inPeakProp), fill = cell)) +\n",
    "    geom_boxplot(outlier.shape = NA) +\n",
    "    geom_jitter(aes(color = expr), position = position_jitter(0.15)) +\n",
    "    #geom_label_repel(data = frPeakD %>% filter(expr == \"Input\"), aes(label = paste0(expr, \"_\", hd)), position = position_jitterdodge()) +\n",
    "    facet_grid(hist~type, space = \"free_x\") +\n",
    "    scale_fill_viridis(discrete = TRUE, option = \"viridis\", alpha = 0.8) +\n",
    "    scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "    theme_bw(base_size = 20) +\n",
    "    scale_y_continuous(breaks = c(10, 20, 30, 40, 50, 60, 70, 80, 90, 100)) +\n",
    "    ylab(\"FRagment proportion in Peaks regions (FRiPs)\") +\n",
    "    xlab(\"\")\n",
    "ggsave(filename = paste0(fig_path, \"cutrun_FRiPs_macs_q_0.05_seacr_top10.pdf\"), width = 12, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## DE peaks\n",
    "# -\tNumbers of differentially enriched peak regions (CUT&RUN) in transcription of input of N vs CM vs EM vs EMRA\n",
    "# o\tVenn diagram for H3K27me3 and H3K4me2\n",
    "\n",
    "## 1. pairwise differential peaks\n",
    "\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/IDR_master_peak_list_chromVar_count_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/results_histList.RData\"))\n",
    "\n",
    "cellList <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hdList <- paste0(\"HD\", 1:7)\n",
    "exprList = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "cellCompList = c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "histList <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "inPath <- \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath <- \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "bamDir <- \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "\n",
    "\n",
    "deNum = c()\n",
    "dePeak = vector(\"list\", length(histList))\n",
    "dePeakIndex = vector(\"list\", length(histList))\n",
    "for(hist in histList){\n",
    "  dePeak[[hist]] = list()\n",
    "  dePeakIndex[[hist]] = list()\n",
    "\n",
    "  for(exprIter in c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")){\n",
    "    dePeak[[hist]][[exprIter]] = c()\n",
    "    for(cellComp in c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")){\n",
    "      resData = results[[hist]][[paste(exprIter, cellComp, sep = \"_\")]]\n",
    "      deNum = rbind(data.frame(deNum = nrow(resData[Significance != 'notDE']),\n",
    "                               expr = exprIter,\n",
    "                               cellComp = cellComp,\n",
    "                               hist = hist), deNum)\n",
    "      dePeak[[hist]][[exprIter]] = c(dePeak[[hist]][[exprIter]], resData[Significance != 'notDE']$GeneName)\n",
    "      dePeakIndex[[hist]][[exprIter]] = c(dePeakIndex[[hist]][[exprIter]], resData[Significance != 'notDE']$GeneIndex)\n",
    "\n",
    "    }\n",
    "    dePeak[[hist]][[exprIter]] = dePeak[[hist]][[exprIter]] %>% unique\n",
    "    dePeakIndex[[hist]][[exprIter]] = dePeakIndex[[hist]][[exprIter]] %>% unique\n",
    "  }\n",
    "}\n",
    "\n",
    "deNum$cellComp = factor(deNum$cellComp, levels = c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\"))\n",
    "\n",
    "deNum %>% filter(expr == \"Input\") %>% ggplot(aes(x = cellComp, y = deNum,  fill = cellComp, label = deNum)) +\n",
    "  geom_bar(stat = \"identity\") +\n",
    "  facet_grid(~hist) +\n",
    "  geom_text(vjust = -0.1) +\n",
    "  theme_bw(base_size = 20) +\n",
    "  ggpubr::rotate_x_text(angle = 90) +\n",
    "  scale_fill_viridis(discrete = TRUE, option = \"viridis\") +\n",
    "  scale_color_viridis(discrete = TRUE, option = \"magma\") +\n",
    "  ggpubr::rremove(\"legend\") +\n",
    "  xlab(\"\") +\n",
    "  ylab(\"# of Differential Peaks\") +\n",
    "  ggtitle(\"Input (FDR <= 0.05 |logFC| >=1)\")\n",
    "\n",
    "ggsave(filename = paste0(fig_path, \"cutrun_differential_peak_number_macs_q_0.05_seacr_top10.pdf\"), width = 12, height = 8)\n",
    "\n",
    "## 2. compare N vs rest ... for venn diagram\n",
    "\n",
    "checkRep = function(hist, cell, expr, hd){\n",
    "  rep = \"rep1\"\n",
    "  if(paste(hist, cell, expr, hd, sep = \"_\") == \"H3K27me3_CM_Input_HD1\"){\n",
    "    rep = \"rep2\"\n",
    "  }\n",
    "  if(paste(hist, cell, expr, hd, sep = \"_\") == \"H3K27me3_N_Input_HD2\"){\n",
    "    rep = \"rep2\"\n",
    "  }\n",
    "\n",
    "  return(rep)\n",
    "\n",
    "}\n",
    "\n",
    "checkDup = function(expr, hd){\n",
    "\n",
    "  dupType = \"normLibSize\" #\"norm\"\n",
    "  if(expr == \"Input\" && hd %in% paste0(\"HD\", 1:3)){\n",
    "    dupType = \"stringent\"\n",
    "  }\n",
    "  return(dupType)\n",
    "\n",
    "}\n",
    "\n",
    "checkBam = function(expr, hd){\n",
    "\n",
    "  bamType = \"bowtie2_align.bam\"\n",
    "  if(expr == \"Input\" && hd %in% paste0(\"HD\", 1:3)){\n",
    "    bamType = \"bowtie2_align.sorted.rmDup.sortName.bam\"\n",
    "  }\n",
    "  return(bamType)\n",
    "\n",
    "}\n",
    "\n",
    "checkHD = function(expr){\n",
    "\n",
    "  if(expr == \"Input\"){\n",
    "    return(paste0(\"HD\", 4:7))\n",
    "  }else{\n",
    "    return(paste0(\"HD\", 1:7))\n",
    "  }\n",
    "}\n",
    "get_contrast <- function(cell_type, design){\n",
    "\n",
    "  if(cell_type == \"N\"){\n",
    "    contrast <- makeContrasts(\n",
    "      Input_N_Others = Others.Input - N.Input,\n",
    "      Product_N_Others = Others.Product - N.Product,\n",
    "      Stim1_N_Others = Others.Stim1 - N.Stim1,\n",
    "      Stim2_N_Others = Others.Stim2 - N.Stim2,\n",
    "      Stim3_N_Others = Others.Stim3 - N.Stim3,\n",
    "      levels = design\n",
    "    )\n",
    "  }\n",
    "  if(cell_type == \"CM\"){\n",
    "    contrast <- makeContrasts(\n",
    "      Input_CM_Others = Others.Input - CM.Input,\n",
    "      Product_CM_Others = Others.Product - CM.Product,\n",
    "      Stim1_CM_Others = Others.Stim1 - CM.Stim1,\n",
    "      Stim2_CM_Others = Others.Stim2 - CM.Stim2,\n",
    "      Stim3_CM_Others = Others.Stim3 - CM.Stim3,\n",
    "      levels = design\n",
    "    )\n",
    "  }\n",
    "  if(cell_type == \"EM\"){\n",
    "    contrast <- makeContrasts(\n",
    "      Input_EM_Others = Others.Input - EM.Input,\n",
    "      Product_EM_Others = Others.Product - EM.Product,\n",
    "      Stim1_EM_Others = Others.Stim1 - EM.Stim1,\n",
    "      Stim2_EM_Others = Others.Stim2 - EM.Stim2,\n",
    "      Stim3_EM_Others = Others.Stim3 - EM.Stim3,\n",
    "      levels = design\n",
    "    )\n",
    "  }\n",
    "  if(cell_type == \"EMRA\"){\n",
    "    contrast <- makeContrasts(\n",
    "      Input_EMRA_Others = Others.Input - EMRA.Input,\n",
    "      Product_EMRA_Others = Others.Product - EMRA.Product,\n",
    "      Stim1_EMRA_Others = Others.Stim1 - EMRA.Stim1,\n",
    "      Stim2_EMRA_Others = Others.Stim2 - EMRA.Stim2,\n",
    "      Stim3_EMRA_Others = Others.Stim3 - EMRA.Stim3,\n",
    "      levels = design\n",
    "    )\n",
    "  }\n",
    "  return(contrast)\n",
    "}\n",
    "\n",
    "\n",
    "for(target_cell in c(\"CM\", \"EM\", \"EMRA\")){ ## \"N\"\n",
    "  target = c()\n",
    "  for(cell in cellList){\n",
    "    for(expr in exprList){\n",
    "      hdL = checkHD(expr)\n",
    "      for(hd in hdL){\n",
    "        if( !((cell == \"EMRA\") & (expr == \"Input\") && (hd == \"HD2\"))){\n",
    "          target = rbind(target, data.frame(cell = cell, expr = expr, humanDonor = hd))\n",
    "        }\n",
    "\n",
    "      }\n",
    "    }\n",
    "  }\n",
    "  target$cell[which(target$cell != target_cell)] <- \"Others\"\n",
    "\n",
    "\n",
    "  ## Consider human donor as random variable.\n",
    "  ## Experimental design\n",
    "  treat <- factor(paste(target$cell, target$expr, sep=\".\"))\n",
    "  design <- model.matrix(~0 + treat)\n",
    "  colnames(design) <- levels(treat)\n",
    "  contrast <- get_contrast(target_cell, design)\n",
    "  contrast\n",
    "\n",
    "\n",
    "  results = vector(\"list\", length(histList))\n",
    "  voomDDS = vector(\"list\", length(histList))\n",
    "  for(hist in histList){\n",
    "    ## Filter and delete low expressed genes\n",
    "      selectR <- which(rowSums(countMat[[hist]]) > 10) ## remove low count genes\n",
    "      dataS <- countMat[[hist]][selectR,]\n",
    "      geneNameList = rownames(data)[selectR]\n",
    "\n",
    "      voomDDS[[hist]] <- voom(counts = dataS, design = design, normalize.method = \"cyclicloess\", plot = FALSE)\n",
    "      ## option 1 using voomDDS option2 using normDDS as normalized input.\n",
    "      inputDDS <- voomDDS[[hist]]\n",
    "      corfit <- duplicateCorrelation(inputDDS, design, block = target$humanDonor)\n",
    "      ## corfit$consensus\n",
    "      fit <- lmFit(inputDDS, design, block = target$humanDonor, correlation = corfit$consensus)\n",
    "      fitContrast <- contrasts.fit(fit, contrast)\n",
    "      fitBayes <- eBayes(fitContrast, robust = TRUE)\n",
    "\n",
    "\n",
    "      results[[hist]] = list()\n",
    "      for(i in 1:ncol(contrast)){\n",
    "          ## Results\n",
    "          res <- topTable(fit = fitBayes, adjust.method = 'fdr', coef = i, number = nrow(inputDDS), sort = 'P')\n",
    "\n",
    "          res <- data.table(GeneName = geneNameList[as.numeric(rownames(res))], GeneIndex = rownames(res) %>% as.numeric, res)\n",
    "          res[, Significance := ifelse((adj.P.Val <= 0.05 & sign(logFC) == 1 & abs(logFC) >= 1), 'Up',\n",
    "                                ifelse((adj.P.Val <= 0.05 & sign(logFC) == -1 & abs(logFC) >= 1), 'Down', 'notDE'))]\n",
    "          results[[hist]][[i]] = res\n",
    "\n",
    "          ## Output\n",
    "          write.table(res, file = paste0(outPath, '/CUTRUN_limma_tables/DE_', colnames(contrast)[i], '_HD1-7_adj0.05_logFC1_one_vs_rest_', target_cell, '.csv'), quote = FALSE, row.names = FALSE, sep = \",\")\n",
    "      }\n",
    "      names(results[[hist]]) <- colnames(contrast)\n",
    "\n",
    "\n",
    "  }\n",
    "\n",
    "  save(results, voomDDS, file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM_one_vs_rest_\", target_cell, \".RData\"))\n",
    "} \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "hist <- \"H3K4me2\" ## \"H3K4me2\"\n",
    "de_list <- list()\n",
    "for(target_cell in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM_one_vs_rest_\", target_cell, \".RData\"))\n",
    "    de_list[[target_cell]] <- results[[hist]][[paste0(\"Input_\", target_cell, \"_Others\")]] %>% filter(Significance != \"notDE\") %$% GeneName\n",
    "}\n",
    "# results\n",
    "pdf(paste0(fig_path, \"cutrun_DP_Input_\", hist, \"_Venn_diagram.pdf\"))\n",
    "venn.diagram(\n",
    "    x = list(de_list[[\"N\"]], de_list[[\"CM\"]], de_list[[\"EM\"]], de_list[[\"EMRA\"]]),\n",
    "    category.names = c(\n",
    "        paste0(\"N (\", length(de_list[[\"N\"]]), \")\"), \n",
    "        paste0(\"CM (\", length(de_list[[\"CM\"]]), \")\"),\n",
    "        paste0(\"EM (\", length(de_list[[\"EM\"]]), \")\"),\n",
    "        paste0(\"EMRA (\", length(de_list[[\"EMRA\"]]), \")\")),\n",
    "    filename = NULL, #paste0(fig_path, \"RNAseq_DE_Input_Venn_diagram.pdf\"\n",
    "    col = c(\"#440154\", \"#31688e\", \"#35b779\", \"#fde725\"), \n",
    "    fill = c(alpha(\"#440154\", 0.3), alpha(\"#31688e\", 0.3), alpha(\"#35b779\", 0.3), alpha(\"#fde725\", 0.3)),\n",
    "    resolution = 1000, \n",
    "    disable.logging = TRUE\n",
    ") %>% grid.draw\n",
    "\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# -\tGene set enrichment analysis of differentially expressed genes  and differentially enriched genes of N vs CM vs EM vs EMRA\n",
    "# o\tFull list of gene set differences\n",
    "\n",
    "## 1. pairwise DE\n",
    "load(file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition.RData\"))\n",
    "names(results)\n",
    "geneID <- fread(paste0(in_path, \"RNA_CD8_N_Input_HD1.genes.results\"))$gene_id\n",
    "\n",
    "norm_rna <- voomDDS$E\n",
    "rownames(norm_rna) <- geneID[selectR]\n",
    "\n",
    "de_gene_list <- c(\n",
    "  results$Input_N_CM %>% filter(Significance != \"notDE\") %$% GeneName,\n",
    "  results$Input_N_EM %>% filter(Significance != \"notDE\") %$% GeneName,\n",
    "  results$Input_N_EMRA %>% filter(Significance != \"notDE\") %$% GeneName,\n",
    "  results$Input_CM_EM %>% filter(Significance != \"notDE\") %$% GeneName,\n",
    "  results$Input_CM_EMRA %>% filter(Significance != \"notDE\") %$% GeneName,\n",
    "  results$Input_EM_EMRA %>% filter(Significance != \"notDE\") %$% GeneName\n",
    ") %>% unique\n",
    "sample_select <- c(\"Input_N_HD1\", \"Input_N_HD2\", \"Input_N_HD3\", \"Input_N_HD5\", \n",
    "\"Input_N_HD6\", \"Input_N_HD7\", \"Input_CM_HD1\", \n",
    "\"Input_CM_HD2\", \"Input_CM_HD3\", \"Input_CM_HD5\", \"Input_CM_HD6\", \n",
    "\"Input_CM_HD7\", \"Input_EM_HD1\", \"Input_EM_HD2\", \"Input_EM_HD3\", \n",
    "\"Input_EM_HD5\", \"Input_EM_HD6\", \"Input_EM_HD7\", \"Input_EMRA_HD1\", \n",
    "\"Input_EMRA_HD2\", \"Input_EMRA_HD3\", \"Input_EMRA_HD5\", \"Input_EMRA_HD6\", \n",
    "\"Input_EMRA_HD7\")\n",
    "norm_rna[de_gene_list, ] %>% dim\n",
    "norm_rna[de_gene_list, sample_select] %>% head\n",
    "\n",
    "annotation_col <- data.frame(CellType = rep(c(\"N\", \"CM\", \"EM\", \"EMRA\"), each = 6))\n",
    "rownames(annotation_col) <- sample_select\n",
    "annotation_color <- list(CellType = c(\"N\" = \"#440154\", \"CM\" = \"#31688e\", \"EM\" = \"#35b779\", \"EMRA\" = \"#fde725\"))\n",
    "\n",
    "\n",
    "test = pheatmap(norm_rna[de_gene_list, sample_select], cluster_rows = TRUE, cluster_cols = FALSE, show_rownames = FALSE, scale = \"none\", annotation_col = annotation_col, color = PurpleAndYellow(),\n",
    "annotation_colors = annotation_color, filename = paste0(fig_path, \"RNAseq_DE_genes_pairwise_comparison_heatmap.pdf\"), width = 7, height = 7)\n",
    "\n",
    "write.table(norm_rna[de_gene_list[test$tree_row$order], sample_select], file = paste0(fig_path, \"RNAseq_DE_genes_pairwise_comparison_heatmap.csv\"), quote = F, row.names = TRUE, col.names = TRUE, sep = \",\")\n",
    "\n",
    "## 2. One vs rest\n",
    "de_gene_list <- c()\n",
    "for(target_cell in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    load(file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition_one_vs_rest_\", target_cell, \".RData\"))\n",
    "    de_gene_list <- results[[paste0(\"Input_\", target_cell, \"_Others\")]] %>% filter(Significance != \"notDE\") %$% GeneName %>% c(de_gene_list, .)\n",
    "}\n",
    "de_gene_list <- de_gene_list %>% unique\n",
    "\n",
    "sample_select <- c(\"Input_N_HD1\", \"Input_N_HD2\", \"Input_N_HD3\", \"Input_N_HD5\", \n",
    "\"Input_N_HD6\", \"Input_N_HD7\", \"Input_CM_HD1\", \n",
    "\"Input_CM_HD2\", \"Input_CM_HD3\", \"Input_CM_HD5\", \"Input_CM_HD6\", \n",
    "\"Input_CM_HD7\", \"Input_EM_HD1\", \"Input_EM_HD2\", \"Input_EM_HD3\", \n",
    "\"Input_EM_HD5\", \"Input_EM_HD6\", \"Input_EM_HD7\", \"Input_EMRA_HD1\", \n",
    "\"Input_EMRA_HD2\", \"Input_EMRA_HD3\", \"Input_EMRA_HD5\", \"Input_EMRA_HD6\", \n",
    "\"Input_EMRA_HD7\")\n",
    "norm_rna[de_gene_list, ] %>% dim\n",
    "norm_rna[de_gene_list, sample_select] %>% head\n",
    "\n",
    "annotation_col <- data.frame(CellType = rep(c(\"N\", \"CM\", \"EM\", \"EMRA\"), each = 6))\n",
    "rownames(annotation_col) <- sample_select\n",
    "annotation_color <- list(CellType = c(\"N\" = \"#440154\", \"CM\" = \"#31688e\", \"EM\" = \"#35b779\", \"EMRA\" = \"#fde725\"))\n",
    "test = pheatmap(norm_rna[de_gene_list, sample_select], cluster_rows = TRUE, cluster_cols = FALSE, show_rownames = FALSE, scale = \"none\", annotation_col = annotation_col, color = PurpleAndYellow(),\n",
    "annotation_colors = annotation_color, filename = paste0(fig_path, \"RNAseq_DE_genes_one_vs_rest_comparison_heatmap.pdf\"), width = 7, height = 7)\n",
    "\n",
    "write.table(norm_rna[de_gene_list[test$tree_row$order], sample_select], file = paste0(fig_path, \"RNAseq_DE_genes_one_vs_rest_comparison_heatmap.csv\"), quote = F, row.names = TRUE, col.names = TRUE, sep = \",\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# GSEA\n",
    "## 1. Pairwise DE\n",
    "load(file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition.RData\"))\n",
    "pathway_list <- c(\"msigdb.v7.4.symbols.gmt\", \"h.all.v7.4.symbols.gmt\", \"c2.cp.v7.4.symbols.gmt\", \"c3.tft.v7.4.symbols.gmt\", \"c5.all.v7.4.entrez.gmt\", \"c7.all.v7.4.symbols.gmt\")\n",
    "pathway_name_list <- c(\"All_Gene_Sets\", \"Hallmark_All\", \"C2_noCGP\", \"C3_TFTonly\", \"C5_All\", \"C7_All\")\n",
    "cell_pair_list <- c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "\n",
    "for(cell_pair in cell_pair_list){\n",
    "    print(cell_pair)\n",
    "    geneList <- results[[paste0(\"Input_\", cell_pair)]]$logFC\n",
    "    names(geneList) <- results[[paste0(\"Input_\", cell_pair)]]$GeneName %>% gsub(\".*_\", \"\", .)\n",
    "    geneList <- sort(geneList, decreasing = TRUE)\n",
    "\n",
    "    for(p in 1:length(pathway_list)){\n",
    "        pathway <- pathway_list[p]\n",
    "        pathway_name <- pathway_name_list[p]\n",
    "\n",
    "        module_file <- paste0(\"/fh/fast/gottardo_r/yezheng_working/SupplementaryData/GSEA/\", pathway)\n",
    "        gene_set <- getGmt(module_file)\n",
    "        gene_pathway <- geneIds(gene_set)\n",
    "\n",
    "        gsea_res <- fgsea(pathways = gene_pathway, stats = geneList, minSize = 10, maxSize = 500, eps = 0)\n",
    "        gsea_res <- gsea_res %>% arrange(padj, decreaing = FALSE)\n",
    "        fwrite(gsea_res, file = paste0(fig_path, \"/RNAseq_DE_gene_pairwise_\", cell_pair, \"_\", pathway_name, \"_GSEA.csv\"), sep = \",\")\n",
    "\n",
    "        pdf(paste0(fig_path, \"/RNAseq_DE_gene_pairwise_\", cell_pair, \"_\", pathway_name, \"_GSEA_topEnrichmentScore.pdf\"), width = 10, height = 10)\n",
    "        topPathwaysUp <- gsea_res[ES > 0][head(order(pval), n=10), pathway]\n",
    "        topPathwaysDown <- gsea_res[ES < 0][head(order(pval), n=10), pathway]\n",
    "        topPathways <- c(topPathwaysUp, rev(topPathwaysDown)) %>% unique\n",
    "        plotGseaTable(\n",
    "            gene_pathway[topPathways], \n",
    "            geneList, \n",
    "            gsea_res,\n",
    "            gseaParam=0.5\n",
    "        )\n",
    "        dev.off()\n",
    "\n",
    "        pdf(paste0(fig_path, \"/RNAseq_DE_gene_pairwise_\", cell_pair, \"_\", pathway_name, \"_GSEA_topPadj.pdf\"), width = 10, height = 10)\n",
    "        collapsedPathways <- collapsePathways(gsea_res[order(padj)][padj < 0.01], \n",
    "                                            gene_pathway, geneList)\n",
    "        mainPathways <- gsea_res[pathway %in% collapsedPathways$mainPathways][\n",
    "                                order(-NES), pathway]\n",
    "        plotGseaTable(\n",
    "            gene_pathway[mainPathways], \n",
    "            geneList, \n",
    "            gsea_res, \n",
    "            gseaParam = 0.5\n",
    "        )\n",
    "        dev.off()\n",
    "\n",
    "    }\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## 2. one vs rest DE\n",
    "pathway_list <- c(\"msigdb.v7.4.symbols.gmt\", \"h.all.v7.4.symbols.gmt\", \"c2.cp.v7.4.symbols.gmt\", \"c3.tft.v7.4.symbols.gmt\", \"c5.all.v7.4.entrez.gmt\", \"c7.all.v7.4.symbols.gmt\")\n",
    "pathway_name_list <- c(\"All_Gene_Sets\", \"Hallmark_All\", \"C2_noCGP\", \"C3_TFTonly\", \"C5_All\", \"C7_All\")\n",
    "target_cell_list <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "\n",
    "for(target_cell in target_cell_list){\n",
    "    load(file = paste0(out_path, \"/RData/HD1-3_5-7_normLimma_perExprCondition_one_vs_rest_\", target_cell, \".RData\"))\n",
    "    \n",
    "    geneList <- results[[paste0(\"Input_\", target_cell, \"_Others\")]]$logFC\n",
    "    names(geneList) <- results[[paste0(\"Input_\", target_cell, \"_Others\")]]$GeneName %>% gsub(\".*_\", \"\", .)\n",
    "    geneList <- sort(geneList, decreasing = TRUE)\n",
    "\n",
    "    for(p in 1:length(pathway_list)){\n",
    "        pathway <- pathway_list[p]\n",
    "        pathway_name <- pathway_name_list[p]\n",
    "\n",
    "        module_file <- paste0(\"/fh/fast/gottardo_r/yezheng_working/SupplementaryData/GSEA/\", pathway)\n",
    "        gene_set <- getGmt(module_file)\n",
    "        gene_pathway <- geneIds(gene_set)\n",
    "\n",
    "        gsea_res <- fgsea(pathways = gene_pathway, stats = geneList, minSize = 10, maxSize = 500, eps = 0)\n",
    "        gsea_res <- gsea_res %>% arrange(padj, decreaing = FALSE)\n",
    "        fwrite(gsea_res, file = paste0(fig_path, \"/RNAseq_DE_gene_one_vs_rest_\", target_cell, \"_\", pathway_name, \"_GSEA.csv\"), sep = \",\")\n",
    "\n",
    "        pdf(paste0(fig_path, \"/RNAseq_DE_gene_one_vs_rest_\", target_cell, \"_\", pathway_name, \"_GSEA_topEnrichmentScore.pdf\"), width = 10, height = 10)\n",
    "        topPathwaysUp <- gsea_res[ES > 0][head(order(pval), n=10), pathway]\n",
    "        topPathwaysDown <- gsea_res[ES < 0][head(order(pval), n=10), pathway]\n",
    "        topPathways <- c(topPathwaysUp, rev(topPathwaysDown)) %>% unique\n",
    "        plotGseaTable(\n",
    "            gene_pathway[topPathways], \n",
    "            geneList, \n",
    "            gsea_res,\n",
    "            gseaParam=0.5\n",
    "        )\n",
    "        dev.off()\n",
    "\n",
    "    }\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
