{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(pacman)\n",
    "p_load(data.table, dplyr, ggplot2, viridis, magrittr, VennDiagram, ggpubr, limma, edgeR, tidyr, GenomicRanges, RColorBrewer, pheatmap, Seurat, fgsea, GSEABase, limma, TxDb.Hsapiens.UCSC.hg38.knownGene, org.Hs.eg.db)\n",
    "in_path <- \"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "out_path <- \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/\"\n",
    "fig_path <- \"CART_CUTRUN_Project/results/paper_figure/\"\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### METHOD 1 Compare GEX with Histone Signals around TSS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Get the TSS location and +/-1kb region around TSS\n",
    "gtf <- rtracklayer::import('/fh/fast/newell_e/yezheng_working/SupplementaryData/hg38/GENCODE/gencode.v21.annotation.gtf')\n",
    "gtf_df=as.data.frame(gtf)\n",
    "head(gtf_df)\n",
    "\n",
    "## get the TSS location and extend by 1kb up and downstream\n",
    "tss_df = rbind(\n",
    "    gtf_df %>% data.frame %>% dplyr::filter(type == \"gene\", strand == \"+\") %>% dplyr::select(seqnames, TSS = start, strand, source, gene_id, gene_name, gene_name, gene_type) ,\n",
    "    gtf_df %>% data.frame %>% dplyr::filter(type == \"gene\", strand == \"-\") %>% dplyr::select(seqnames, TSS = end, strand, source, gene_id, gene_name, gene_name, gene_type) \n",
    ")\n",
    "tss_filter_df = tss_df %>% dplyr::filter(!(seqnames %in% c(\"chrM\", \"chrX\", \"chrY\")))\n",
    "gtf_filter_df = gtf_df %>% dplyr::filter(type == \"gene\", !(seqnames %in% c(\"chrM\", \"chrX\", \"chrY\")))\n",
    "\n",
    "tss_gr = GRanges(seqnames = tss_filter_df$seqnames, ranges = IRanges(start = tss_filter_df$TSS - 1000, end = tss_filter_df$TSS + 1000), strand = tss_filter_df$strand, gene_id = tss_filter_df$gene_id, gene_name = tss_filter_df$gene_name, gene_type = tss_filter_df$gene_type)\n",
    "tss_gr \n",
    "\n",
    "gene_gr = GRanges(seqnames = gtf_filter_df$seqnames, ranges = IRanges(start = gtf_filter_df$start, end = gtf_filter_df$end), strand = gtf_filter_df$strand, gene_id = gtf_filter_df$gene_id, gene_name = gtf_filter_df$gene_name, gene_type = gtf_filter_df$gene_type)\n",
    "gene_gr "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## overlap with bam file to get count -- !!! Run it in terminal directly\n",
    "bamDir <- \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "cellList <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hdList <- paste0(\"HD\", 1:7)\n",
    "exprList = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "cellCompList = c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "histList <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "\n",
    "checkRep = function(hist, cell, expr, hd){\n",
    "  rep = \"rep1\"\n",
    "  if(paste(hist, cell, expr, hd, sep = \"_\") == \"H3K27me3_CM_Input_HD1\"){\n",
    "    rep = \"rep2\"\n",
    "  }\n",
    "  if(paste(hist, cell, expr, hd, sep = \"_\") == \"H3K27me3_N_Input_HD2\"){\n",
    "    rep = \"rep2\"\n",
    "  }\n",
    "\n",
    "  return(rep)\n",
    "\n",
    "}\n",
    "\n",
    "checkDup = function(expr, hd){\n",
    "\n",
    "  dupType = \"norm\"\n",
    "  if(expr == \"Input\" && hd %in% paste0(\"HD\", 1:3)){\n",
    "    dupType = \"stringent\"\n",
    "  }\n",
    "  return(dupType)\n",
    "\n",
    "}\n",
    "\n",
    "checkBam = function(expr, hd){\n",
    "\n",
    "  bamType = \"bowtie2_align.bam\"\n",
    "  if(expr == \"Input\" && hd %in% paste0(\"HD\", 1:3)){\n",
    "    bamType = \"bowtie2_align.sorted.rmDup.sortName.bam\"\n",
    "  }\n",
    "  return(bamType)\n",
    "\n",
    "}\n",
    "\n",
    "checkHD = function(expr){\n",
    "\n",
    "  if(expr == \"Input\"){\n",
    "    return(paste0(\"HD\", 4:7))\n",
    "  }else{\n",
    "    return(paste0(\"HD\", 1:7))\n",
    "  }\n",
    "}\n",
    " \n",
    "# library(doMC)\n",
    "# coreN <- 6\n",
    "# registerDoMC(cores = coreN)\n",
    "\n",
    "bamDir <- \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "bamSelect = vector(\"list\", length(histList))\n",
    "# fragment_tss_counts = vector(\"list\", length(histList))\n",
    "for(hist in histList){\n",
    "  bamSelect[[hist]] = c()\n",
    "   for(cell in cellList){\n",
    "    for(expr in exprList){\n",
    "      hdL = checkHD(expr)\n",
    "      for(hd in hdL){\n",
    "        ## Get the replicate number and peak calling type\n",
    "        rep = checkRep(hist, cell, expr, hd)\n",
    "        bamType = checkBam(expr, hd)\n",
    "        bamFile = paste0(bamDir, hist, \"_CD8_\", cell, \"_\", expr, \"_\", hd, \"_\", rep, \"/alignment/\", bamType)\n",
    "        if(file.exists(bamFile)){\n",
    "          bamSelect[[hist]] = c(bamSelect[[hist]], bamFile)\n",
    "        }else{\n",
    "          print(paste0(bamFile, \" does not exist!\"))\n",
    "        }\n",
    "      }\n",
    "    }\n",
    "   }\n",
    "  # bamSelect[[hist]]\n",
    "  ##get count matrix in parallel\n",
    "  # fragment_tss_counts[[hist]] <- mclapply(as.list(bamSelect[[hist]]), chromVAR::getCounts, tss_gr, paired = TRUE, by_rg = FALSE, format = \"bam\", mc.cores = coreN)\n",
    "\n",
    "}\n",
    "\n",
    "\n",
    "# saveRDS(fragment_tss_counts, file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM.rds\"))\n",
    "\n",
    "fragment_tss_counts = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM.rds\"))\n",
    "head(fragment_tss_counts)       \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "library(chromVAR)\n",
    "library(SummarizedExperiment)\n",
    "library(Matrix)\n",
    "\n",
    "## count matrix column name\n",
    "# countMat = vector(\"list\", length(histList))\n",
    "# normMat = vector(\"list\", length(histList))\n",
    "# seqDepth = vector(\"list\", length(histList))\n",
    "# designInfo = c()\n",
    "# for(hist in histList){\n",
    "#   sampleName = c()\n",
    "#   for(cell in cellList){\n",
    "#     for(expr in exprList){\n",
    "#       hdL = checkHD(expr)\n",
    "#       for(hd in hdL){\n",
    "#         sampleName = c(sampleName, paste(hist, cell, expr, hd, sep = \"_\"))\n",
    "#       }\n",
    "#     }\n",
    "#   }\n",
    "\n",
    "#   ## Form the matrix\n",
    "#   rmIndex = which(sampleName == paste0(hist, \"_EMRA_Input_HD2\"))\n",
    "#   if(length(rmIndex) > 0){\n",
    "#     sampleName = sampleName[-rmIndex]\n",
    "#   }\n",
    "#   countMat[[hist]] <- matrix(NA, length(tss_gr), length(bamSelect[[hist]]))\n",
    "#   normMat[[hist]] <- matrix(NA, length(tss_gr), length(bamSelect[[hist]]))\n",
    "#   colnames(countMat[[hist]]) <- sampleName\n",
    "#   colnames(normMat[[hist]]) <- sampleName\n",
    "#   seqDepth[[hist]] <- NULL\n",
    "#   for (k in 1:length(bamSelect[[hist]])){\n",
    "#     countMat[[hist]][, k] <- counts(fragment_tss_counts[[hist]][[k]])[,1]\n",
    "#     seqDepth[[hist]][k] <- fragment_tss_counts[[hist]][[k]]@colData[1,1]\n",
    "#     normMat[[hist]][, k] <- countMat[[hist]][, k]/seqDepth[[hist]][k] * 16294327 ## divide by the seqDepth and multiply by the largest seqDepth\n",
    "#   }\n",
    "#   designInfo <- data.frame(exps = sampleName, depth = seqDepth[[hist]], hist = hist) %>% rbind(designInfo, .)\n",
    "# }\n",
    "# rownames(countMat$H3K4me2) = data.frame(tss_gr)$gene_name\n",
    "# rownames(countMat$H3K27me3) = data.frame(tss_gr)$gene_name\n",
    "# rownames(normMat$H3K4me2) = data.frame(tss_gr)$gene_name\n",
    "# rownames(normMat$H3K27me3) = data.frame(tss_gr)$gene_name\n",
    "\n",
    "\n",
    "# saveRDS(countMat, file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_countMat.rds\"))\n",
    "# saveRDS(normMat, file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_normMat.rds\"))\n",
    "# saveRDS(designInfo, file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_designInfo.rds\"))\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "\n",
    "countMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_countMat.rds\"))\n",
    "normMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_normMat.rds\"))\n",
    "# designInfo = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_designInfo.rds\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=22, repr.plot.height=10)\n",
    "hist = \"H3K27me3\"\n",
    "selectR = which(rowSums(countMat[[hist]]) > 200) ## remove low count genes\n",
    "selectC = which(colnames(countMat[[hist]]) %>% stringr::str_detect(\"Input\"))\n",
    "\n",
    "tmp = countMat[[hist]][selectR, selectC]\n",
    "\n",
    "k27_obj = CreateSeuratObject(counts = countMat[[hist]][selectR, selectC], project = \"cart\", min.cells = 0, min.features = 5) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 100000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"mean.var.plot\", nfeatures = 1000) %>% \n",
    "ScaleData()\n",
    "\n",
    "k27_obj = RunPCA(k27_obj, features = VariableFeatures(object = k27_obj), npcs = 15)\n",
    "k27_plot = Embeddings(k27_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(gsub(\"H3K27me3_\", \"\", colnames(k27_obj)) %>% gsub(\"_.*\", \"\", .), levels = c(\"N\", \"CM\", \"EM\", \"EMRA\"))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point() +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "hist = \"H3K4me2\"\n",
    "selectR = which(rowSums(countMat[[hist]]) > 200) ## remove low count genes\n",
    "selectC = which(colnames(countMat[[hist]]) %>% stringr::str_detect(\"Input\"))\n",
    "k4_obj = CreateSeuratObject(counts = countMat[[hist]][selectR, selectC], project = \"cart\", min.cells = 0, min.features = 5) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 100000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"mean.var.plot\", nfeatures = 1000) %>% \n",
    "ScaleData()\n",
    "k4_obj = RunPCA(k4_obj, features = VariableFeatures(object = k4_obj), npcs = 15)\n",
    "\n",
    "k4_plot = Embeddings(k4_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% mutate(celltype = factor(gsub(\"H3K4me2_\", \"\", colnames(k4_obj)) %>% gsub(\"_.*\", \"\", .), levels = c(\"N\", \"CM\", \"EM\", \"EMRA\"))) %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point() +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "\n",
    "cr_obj = CreateSeuratObject(counts = rbind(tmp, countMat[[hist]][selectR, selectC]), project = \"cart\", min.cells = 0, min.features = 5) %>%  \n",
    "NormalizeData(., normalization.method = \"LogNormalize\", scale.factor = 100000000) %>% \n",
    "FindVariableFeatures(., selection.method = \"mean.var.plot\", nfeatures = 1000) %>% \n",
    "ScaleData()\n",
    "cr_obj = RunPCA(cr_obj, features = VariableFeatures(object = cr_obj), npcs = 15)\n",
    "\n",
    "cr_plot = Embeddings(cr_obj, reduction = \"pca\")[, 1:2] %>% data.frame %>% \n",
    "mutate(celltype = factor(c(\n",
    "        gsub(\"H3K27me3_\", \"\", colnames(k27_obj)) %>% gsub(\"_.*\", \"\", .)), \n",
    "    levels = c(\"N\", \"CM\", \"EM\", \"EMRA\"))\n",
    ") %>% \n",
    "ggplot(aes(x = PC_1, y = PC_2, color = celltype)) +\n",
    "geom_point() +\n",
    "theme_bw(base_size = 20) +\n",
    "scale_color_viridis(discrete = TRUE) +\n",
    "xlab(\"PC 1\") +\n",
    "ylab(\"PC 2\") +\n",
    "ggtitle(\"H3K27me3 + H3K4me2\") +\n",
    "rremove(\"legend.title\")\n",
    "# dev.off()\n",
    "# pdf(\"CART_CUTRUN_Project/results/paper_figure/patient-related/Patient_CR_PD_PCA_on_all_peaks.pdf\", width = 15, height = 6)\n",
    "# ggarrange(k27_plot, k4_plot, cr_plot, common.legend = TRUE, ncol = 3, nrow = 1)\n",
    "# dev.off()\n",
    "ggarrange(k27_plot, k4_plot, cr_plot, common.legend = TRUE, ncol = 3, nrow = 1)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Read in RNA-seq data\n",
    "cellList <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hdList <- paste0(\"HD\", c(1:3, 5:7))\n",
    "exprList = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "\n",
    "\n",
    "inPath <- \"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "outPath <- \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/\"\n",
    "\n",
    "rna_expCount = c()\n",
    "rna_TPM = c()\n",
    "rna_FPKM = c()\n",
    "for(cell in cellList){\n",
    "  for(expr in exprList){\n",
    "    for(hd in hdList){\n",
    "      # print(paste(expr, cell, hd, sep = \"_\"))\n",
    "      dataTmp <- fread(paste0(inPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".genes.results\")) %>% dplyr::select(expected_count)\n",
    "      rna_expCount <- cbind(rna_expCount, round(dataTmp))\n",
    "      dataTmp <- fread(paste0(inPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".genes.results\")) %>% dplyr::select(TPM)\n",
    "      rna_TPM <- cbind(rna_TPM, round(dataTmp))\n",
    "      dataTmp <- fread(paste0(inPath, \"RNA_CD8_\", cell, \"_\", expr, \"_\", hd, \".genes.results\")) %>% dplyr::select(FPKM)\n",
    "      rna_FPKM <- cbind(rna_FPKM, round(dataTmp))\n",
    "\n",
    "    }\n",
    "  }\n",
    "\n",
    "}\n",
    "geneID = fread(paste0(inPath, \"RNA_CD8_N_Input_HD1.genes.results\"))$gene_id\n",
    "rownames(rna_expCount) <- geneID\n",
    "colnames(rna_expCount) <- paste(rep(rep(exprList, each = length(hdList)), length(cellList)), rep(cellList, each = length(hdList)*length(exprList)), rep(hdList, length(exprList)*length(cellList)), sep = \"_\")\n",
    "rownames(rna_TPM) <- geneID\n",
    "colnames(rna_TPM) <- paste(rep(rep(exprList, each = length(hdList)), length(cellList)), rep(cellList, each = length(hdList)*length(exprList)), rep(hdList, length(exprList)*length(cellList)), sep = \"_\")\n",
    "rownames(rna_FPKM) <- geneID\n",
    "colnames(rna_FPKM) <- paste(rep(rep(exprList, each = length(hdList)), length(cellList)), rep(cellList, each = length(hdList)*length(exprList)), rep(hdList, length(exprList)*length(cellList)), sep = \"_\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## add target genes\n",
    "library(readxl)\n",
    "library(ggrepel)\n",
    "gene_path <- \"CART_CUTRUN_Project/functionalGeneList/\"\n",
    "select_gene <- list(\n",
    "    TF = read_excel(paste0(gene_path, \"2021_11_29_TFs.xlsx\")) %>% mutate(gene_name = `HGNC symbol`),\n",
    "    Tactive = read_excel(paste0(gene_path, \"2021_11_30_TCellActivationFunction.xlsx\")) %>% mutate(gene_name = `Official Symbol`),\n",
    "    CellCycle = read_excel(paste0(gene_path, \"2021_22_30_CellCycleAPoptosis.xlsx\")) %>% mutate(gene_name = `Symbol`),\n",
    "    Metabolism = read_excel(paste0(gene_path, \"2021_22_30_MetabolismGeneList_FinalList.xlsx\")) %>% mutate(gene_name = `Gene`)\n",
    ")\n",
    "small_list = read_excel(paste0(gene_path, \"SmallCuratedGeneList.xlsx\")) \n",
    "head(small_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## mean across healthy donors\n",
    "for(gene_group in unique(small_list$Group)){\n",
    "    pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_scatterplot_SmallCuratedGeneList_\", gsub(\" \", \"_\", gene_group), \".pdf\"), width = 13, height =  11)\n",
    "    for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "        rna_select_cond = paste0(\"Input_\", cT)\n",
    "        cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "        rna_select_data = rna_expCount\n",
    "        cr_select_data = normMat$H3K27me3\n",
    "\n",
    "        rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "        cr_gene = rownames(cr_select_data) %>% tolower\n",
    "        rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "        rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "        cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "        rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "        cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "        \n",
    "        target_gene = small_list %>% dplyr::filter(Group == gene_group) %$% Gene %>% tolower\n",
    "\n",
    "        cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "        rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "        cr_rna_match = data.frame(\n",
    "            H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "            H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "            RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans,\n",
    "            gene_name = rna_gene[rna_row_index] %>% toupper\n",
    "        )\n",
    "\n",
    "        x_thres = 4.5\n",
    "        y_thres = 2.5\n",
    "        panel1 = cr_rna_match %>% dplyr::filter(H3K4me2 <= x_thres, RNA >= y_thres) %>% nrow()\n",
    "        panel2 = cr_rna_match %>% dplyr::filter(H3K4me2 > x_thres, RNA >= y_thres) %>% nrow()\n",
    "        panel3 = cr_rna_match %>% dplyr::filter(H3K4me2 <= x_thres, RNA < y_thres) %>% nrow()\n",
    "        panel4 = cr_rna_match %>% dplyr::filter(H3K4me2 > x_thres, RNA < y_thres) %>% nrow()\n",
    "        print(cr_rna_match %>% ggplot(aes(x = H3K4me2, y = RNA, color = H3K27me3)) +\n",
    "        geom_point(size = 0.5) +\n",
    "        geom_point(aes(x = H3K4me2, y = RNA), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 2, color = \"#ff1c1c\") +\n",
    "        geom_text_repel(aes(x = H3K4me2, y = RNA, label = gene_name), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 5, color = \"#ff1c1c\") +\n",
    "        theme_bw(base_size = 25) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "        scale_color_viridis() +\n",
    "        geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "        geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\") +\n",
    "        annotate(\"text\", x = c(0, 10, 0, 10), y = c(20, 20, -1, -1), label = paste0(c(panel1, panel2, panel3, panel4), \" genes\")) +\n",
    "        ggtitle(rna_select_cond)) \n",
    "\n",
    "\n",
    "        x_thres = 2.5\n",
    "        y_thres = 2.5\n",
    "        panel1 = cr_rna_match %>% dplyr::filter(H3K27me3 <= x_thres, RNA >= y_thres) %>% nrow()\n",
    "        panel2 = cr_rna_match %>% dplyr::filter(H3K27me3 > x_thres, RNA >= y_thres) %>% nrow()\n",
    "        panel3 = cr_rna_match %>% dplyr::filter(H3K27me3 <= x_thres, RNA < y_thres) %>% nrow()\n",
    "        panel4 = cr_rna_match %>% dplyr::filter(H3K27me3 > x_thres, RNA < y_thres) %>% nrow()\n",
    "        print(cr_rna_match %>% ggplot(aes(x = H3K27me3, y = RNA, color = H3K4me2)) +\n",
    "        geom_point(size = 0.5) +\n",
    "        geom_point(aes(x = H3K27me3, y = RNA), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 2, color = \"#ff1c1c\") +\n",
    "        geom_text_repel(aes(x = H3K27me3, y = RNA, label = gene_name), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 5, color = \"#ff1c1c\") +\n",
    "        theme_bw(base_size = 25) +\n",
    "        xlab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "        scale_color_viridis() +\n",
    "        geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "        geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\") +\n",
    "        annotate(\"text\", x = c(0, 10, 0, 10), y = c(20, 20, -1, -1), label = paste0(c(panel1, panel2, panel3, panel4), \" genes\")) +\n",
    "        ggtitle(rna_select_cond))\n",
    "    }\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## mean across healthy donors summarize gene number per category\n",
    "gene_num_summary = c()\n",
    "for(gene_group in unique(small_list$Group)){\n",
    "    # pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_scatterplot_SmallCuratedGeneList_\", gsub(\" \", \"_\", gene_group), \".pdf\"), width = 13, height =  11)\n",
    "    for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "        rna_select_cond = paste0(\"Input_\", cT)\n",
    "        cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "        rna_select_data = rna_expCount\n",
    "        cr_select_data = normMat$H3K27me3\n",
    "\n",
    "        rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "        cr_gene = rownames(cr_select_data) %>% tolower\n",
    "        rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "        rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "        cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "        rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "        cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "        \n",
    "        target_gene = small_list %>% dplyr::filter(Group == gene_group) %$% Gene %>% tolower\n",
    "\n",
    "        cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "        rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "        cr_rna_match = data.frame(\n",
    "            H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "            H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "            RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans,\n",
    "            gene_name = rna_gene[rna_row_index] %>% toupper\n",
    "        )\n",
    "        saveRDS(cr_rna_match, file = paste0(outPath, \"/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "\n",
    "        cr_rna_match_subset = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ]\n",
    "        x_thres = 4.5\n",
    "        y_thres = 2.5\n",
    "        panel1 = cr_rna_match_subset %>% dplyr::filter(H3K4me2 <= x_thres, RNA >= y_thres) %>% nrow()\n",
    "        panel2 = cr_rna_match_subset %>% dplyr::filter(H3K4me2 > x_thres, RNA >= y_thres) %>% nrow()\n",
    "        panel3 = cr_rna_match_subset %>% dplyr::filter(H3K4me2 <= x_thres, RNA < y_thres) %>% nrow()\n",
    "        panel4 = cr_rna_match_subset %>% dplyr::filter(H3K4me2 > x_thres, RNA < y_thres) %>% nrow()\n",
    "        \n",
    "        gene_num_summary = data.frame(\n",
    "            gene_group = gene_group, \n",
    "            cell_type = cT, \n",
    "            histone = \"H3K4me2\", \n",
    "            panel = paste0(\"panel\", 1:4), \n",
    "            gene_num = c(panel1, panel2, panel3, panel4), \n",
    "            total_gene_num = cr_rna_match_subset %>% nrow\n",
    "        ) %>% rbind(gene_num_summary, .)\n",
    "\n",
    "        x_thres = 2.5\n",
    "        y_thres = 2.5\n",
    "        panel1 = cr_rna_match_subset %>% dplyr::filter(H3K27me3 <= x_thres, RNA >= y_thres) %>% nrow()\n",
    "        panel2 = cr_rna_match_subset %>% dplyr::filter(H3K27me3 > x_thres, RNA >= y_thres) %>% nrow()\n",
    "        panel3 = cr_rna_match_subset %>% dplyr::filter(H3K27me3 <= x_thres, RNA < y_thres) %>% nrow()\n",
    "        panel4 = cr_rna_match_subset %>% dplyr::filter(H3K27me3 > x_thres, RNA < y_thres) %>% nrow()\n",
    "        \n",
    "        gene_num_summary = data.frame(\n",
    "            gene_group = gene_group, \n",
    "            cell_type = cT, \n",
    "            histone = \"H3K27me3\", \n",
    "            panel = paste0(\"panel\", 1:4), \n",
    "            gene_num = c(panel1, panel2, panel3, panel4), \n",
    "            total_gene_num = cr_rna_match_subset %>% nrow\n",
    "        ) %>% rbind(gene_num_summary, .)\n",
    "  \n",
    "    }\n",
    "    \n",
    "}\n",
    "gene_num_summary\n",
    "\n",
    "gene_num_summary %>% write.csv(., file = paste0(fig_path, \"/CSV/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_gene_group_panel_summary.csv\"), row.names = F)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#  for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "#         rna_select_cond = paste0(\"Input_\", cT)\n",
    "#         cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "#         rna_select_data = rna_expCount\n",
    "#         cr_select_data = normMat$H3K27me3\n",
    "\n",
    "#         rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "#         cr_gene = rownames(cr_select_data) %>% tolower\n",
    "#         rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "#         rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "#         cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "#         rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "#         cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "        \n",
    "#         target_gene = small_list %>% dplyr::filter(Group == gene_group) %$% Gene %>% tolower\n",
    "\n",
    "#         cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "#         rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "#         cr_rna_match = data.frame(\n",
    "#             H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "#             H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "#             RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans,\n",
    "#             gene_name = rna_gene[rna_row_index] %>% toupper\n",
    "#         )\n",
    "#         saveRDS(cr_rna_match, file = paste0(outPath, \"/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "#  }\n",
    "\n",
    "## add hd value\n",
    " for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "        rna_select_cond = paste0(\"Input_\", cT)\n",
    "        cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "        rna_select_data = rna_expCount\n",
    "        colnames(rna_select_data) = paste0(\"RNA_\", colnames(rna_select_data))\n",
    "        cr_select_data = normMat$H3K27me3\n",
    "\n",
    "        rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "        cr_gene = rownames(cr_select_data) %>% tolower\n",
    "        rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "        rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "        cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "        rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "        cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "        if(cT == \"EM\"){\n",
    "            cr_col_index_rm = which(cr_select_data %>% colnames %>% stringr::str_detect(\"EMRA_Input\"))\n",
    "            rna_col_index_rm = which(rna_select_data %>% colnames %>% stringr::str_detect(\"Input_EMRA\"))\n",
    "            cr_col_index = setdiff(cr_col_index, cr_col_index_rm)\n",
    "            rna_col_index = setdiff(rna_col_index, rna_col_index_rm)\n",
    "        }\n",
    "        #target_gene = small_list %>% dplyr::filter(Group == gene_group) %$% Gene %>% tolower\n",
    "\n",
    "        cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "        rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "        cr_rna_match = data.frame(\n",
    "                gene_name = rna_gene[rna_row_index] %>% toupper,\n",
    "            H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "            H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "            RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans\n",
    "        ) %>% cbind(., log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1)) %>%\n",
    "        cbind(., log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1)) %>% \n",
    "        cbind(., log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1))\n",
    "\n",
    "        k4_thres = 4.5\n",
    "        k27_thres = 2.5\n",
    "        bivalent_cr_rna = cr_rna_match %>% dplyr::filter(H3K27me3 >= k27_thres, H3K4me2 >= k4_thres)\n",
    "        write.csv(bivalent_cr_rna, file = paste0(\"CART_CUTRUN_Project/results/paper_figure/RNA_vs_CUTRUN/Bivalent_Genes_\", cT, \"_withHDcounts.csv\"))\n",
    "        saveRDS(cr_rna_match, file = paste0(outPath, \"/RDS/cr_rna_match_\", cT, \"_withHDcounts.rds\"))\n",
    " }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cT = \"N\"\n",
    "cr_rna_match = readRDS(paste0(\"CART_CUTRUN_Project/results/CUTANDRUN/analysis/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "x_thres = 4.5\n",
    "y_thres = 2.5\n",
    "cr_rna_match %>% ggplot(aes(x = H3K4me2, y = H3K27me3, color = RNA)) +\n",
    "        geom_point(size = 0.5) +\n",
    "        # geom_point(aes(x = H3K27me3, y = RNA), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 2, color = \"#ff1c1c\") +\n",
    "        # geom_text_repel(aes(x = H3K27me3, y = RNA, label = gene_name), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 5, color = \"#ff1c1c\") +\n",
    "        theme_bw(base_size = 25) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_viridis() +\n",
    "        geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "        geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\") +\n",
    "        # annotate(\"text\", x = c(0, 10, 0, 10), y = c(20, 20, -1, -1), label = paste0(c(panel1, panel2, panel3, panel4), \" genes\")) +\n",
    "        ggtitle(cT)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=18, repr.plot.height=8)\n",
    "\n",
    "library(plotGMM)\n",
    "library(magrittr)\n",
    "library(ggplot2)\n",
    "library(mixtools)\n",
    "x_thres = 4.25\n",
    "y_thres = 2.5\n",
    "# Fit a GMM using EM\n",
    "set.seed(576)\n",
    "\n",
    "cT = \"N\"\n",
    "# for(cT in c(\"N\")){ #, \"CM\", \"EM\"\n",
    "\n",
    "    pdf(paste0(\"CART_CUTRUN_Project/results/paper_figure_take3/gaussianMixture_histone_signal_TSS_\", cT, \".pdf\"), width = 14, height = 6)\n",
    "    cr_rna_match = readRDS(paste0(\"CART_CUTRUN_Project/results/CUTANDRUN/analysis/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "    mixmdl <- normalmixEM(cr_rna_match$H3K4me2, k = 2)\n",
    "\n",
    "    print(data.frame(x = cr_rna_match$H3K4me2) %>%\n",
    "    ggplot() +\n",
    "    geom_histogram(aes(x, ..density..), binwidth = 0.3, colour = \"black\",\n",
    "                    fill = \"white\") +\n",
    "    stat_function(geom = \"line\", fun = plot_mix_comps,\n",
    "                    args = list(mixmdl$mu[1] + 0.1, mixmdl$sigma[1], lam = mixmdl$lambda[1]),\n",
    "                    colour = \"red\", lwd = 1.5) +\n",
    "    stat_function(geom = \"line\", fun = plot_mix_comps,\n",
    "                    args = list(mixmdl$mu[2] + 0.2, mixmdl$sigma[2], lam = mixmdl$lambda[2]),\n",
    "                    colour = \"blue\", lwd = 1.5) +\n",
    "    theme_bw(base_size = 35) +\n",
    "    xlab(\"H3K4me2 Enrichment Signal around TSS\") +\n",
    "    ylab(\"Density\") +\n",
    "    geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "    ggtitle(cT))\n",
    "\n",
    "    mixmdl <- normalmixEM(cr_rna_match$H3K27me3, k = 2)\n",
    "    print(data.frame(x = cr_rna_match$H3K27me3) %>%\n",
    "    ggplot() +\n",
    "    geom_histogram(aes(x, ..density..), binwidth = 0.3, colour = \"black\",\n",
    "                    fill = \"white\") +\n",
    "    stat_function(geom = \"line\", fun = plot_mix_comps,\n",
    "                    args = list(mixmdl$mu[1] + 0.2, mixmdl$sigma[1] + 0.2, lam = mixmdl$lambda[1]),\n",
    "                    colour = \"red\", lwd = 1.5) +\n",
    "    stat_function(geom = \"line\", fun = plot_mix_comps,\n",
    "                    args = list(mixmdl$mu[2] + 0.4, mixmdl$sigma[2] - 0.2, lam = mixmdl$lambda[2]),\n",
    "                    colour = \"blue\", lwd = 1.5) +\n",
    "    theme_bw(base_size = 35) +\n",
    "    xlab(\"H3K27me3 Enrichment Signal around TSS\") +\n",
    "    ylab(\"Density\") +\n",
    "    geom_vline(xintercept = y_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "    ggtitle(cT))\n",
    "# }\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cT = \"CM\"\n",
    "\n",
    "pdf(paste0(\"CART_CUTRUN_Project/results/paper_figure_take3/gaussianMixture_histone_signal_TSS_\", cT, \".pdf\"), width = 14, height = 6)\n",
    "cr_rna_match = readRDS(paste0(\"CART_CUTRUN_Project/results/CUTANDRUN/analysis/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "mixmdl <- normalmixEM(cr_rna_match$H3K4me2, k = 2)\n",
    "\n",
    "print(data.frame(x = cr_rna_match$H3K4me2) %>%\n",
    "ggplot() +\n",
    "geom_histogram(aes(x, ..density..), binwidth = 0.3, colour = \"black\",\n",
    "                fill = \"white\") +\n",
    "stat_function(geom = \"line\", fun = plot_mix_comps,\n",
    "                args = list(mixmdl$mu[1], mixmdl$sigma[1], lam = mixmdl$lambda[1]),\n",
    "                colour = \"red\", lwd = 1.5) +\n",
    "stat_function(geom = \"line\", fun = plot_mix_comps,\n",
    "                args = list(mixmdl$mu[2] + 0.2, mixmdl$sigma[2], lam = mixmdl$lambda[2]),\n",
    "                colour = \"blue\", lwd = 1.5) +\n",
    "theme_bw(base_size = 35) +\n",
    "xlab(\"H3K4me2 Enrichment Signal around TSS\") +\n",
    "ylab(\"Density\") +\n",
    "geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "ggtitle(cT))\n",
    "\n",
    "mixmdl <- normalmixEM(cr_rna_match$H3K27me3, k = 2)\n",
    "print(data.frame(x = cr_rna_match$H3K27me3) %>%\n",
    "ggplot() +\n",
    "geom_histogram(aes(x, ..density..), binwidth = 0.3, colour = \"black\",\n",
    "                fill = \"white\") +\n",
    "stat_function(geom = \"line\", fun = plot_mix_comps,\n",
    "                args = list(mixmdl$mu[1], mixmdl$sigma[1], lam = mixmdl$lambda[1]),\n",
    "                colour = \"red\", lwd = 1.5) +\n",
    "stat_function(geom = \"line\", fun = plot_mix_comps,\n",
    "                args = list(mixmdl$mu[2] + 0.3, mixmdl$sigma[2] - 0.1, lam = mixmdl$lambda[2]),\n",
    "                colour = \"blue\", lwd = 1.5) +\n",
    "theme_bw(base_size = 35) +\n",
    "xlab(\"H3K27me3 Enrichment Signal around TSS\") +\n",
    "ylab(\"Density\") +\n",
    "geom_vline(xintercept = y_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "ggtitle(cT))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "cT = \"EM\"\n",
    "\n",
    "pdf(paste0(\"CART_CUTRUN_Project/results/paper_figure_take3/gaussianMixture_histone_signal_TSS_\", cT, \".pdf\"), width = 14, height = 6)\n",
    "cr_rna_match = readRDS(paste0(\"CART_CUTRUN_Project/results/CUTANDRUN/analysis/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "mixmdl <- normalmixEM(cr_rna_match$H3K4me2, k = 2)\n",
    "\n",
    "print(data.frame(x = cr_rna_match$H3K4me2) %>%\n",
    "ggplot() +\n",
    "geom_histogram(aes(x, ..density..), binwidth = 0.3, colour = \"black\",\n",
    "                fill = \"white\") +\n",
    "stat_function(geom = \"line\", fun = plot_mix_comps,\n",
    "                args = list(mixmdl$mu[1], mixmdl$sigma[1], lam = mixmdl$lambda[1]),\n",
    "                colour = \"red\", lwd = 1.5) +\n",
    "stat_function(geom = \"line\", fun = plot_mix_comps,\n",
    "                args = list(mixmdl$mu[2], mixmdl$sigma[2], lam = mixmdl$lambda[2]),\n",
    "                colour = \"blue\", lwd = 1.5) +\n",
    "theme_bw(base_size = 35) +\n",
    "xlab(\"H3K4me2 Enrichment Signal around TSS\") +\n",
    "ylab(\"Density\") +\n",
    "geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "ggtitle(cT))\n",
    "\n",
    "mixmdl <- normalmixEM(cr_rna_match$H3K27me3, k = 2)\n",
    "print(data.frame(x = cr_rna_match$H3K27me3) %>%\n",
    "ggplot() +\n",
    "geom_histogram(aes(x, ..density..), binwidth = 0.3, colour = \"black\",\n",
    "                fill = \"white\") +\n",
    "stat_function(geom = \"line\", fun = plot_mix_comps,\n",
    "                args = list(mixmdl$mu[1], mixmdl$sigma[1], lam = mixmdl$lambda[1]),\n",
    "                colour = \"red\", lwd = 1.5) +\n",
    "stat_function(geom = \"line\", fun = plot_mix_comps,\n",
    "                args = list(mixmdl$mu[2] + 0.1, mixmdl$sigma[2] - 0.1, lam = mixmdl$lambda[2]),\n",
    "                colour = \"blue\", lwd = 1.5) +\n",
    "theme_bw(base_size = 35) +\n",
    "xlab(\"H3K27me3 Enrichment Signal around TSS\") +\n",
    "ylab(\"Density\") +\n",
    "geom_vline(xintercept = y_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "ggtitle(cT))\n",
    "\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "x_thres = 4.5\n",
    "y_thres = 2.5\n",
    "\n",
    "bivProp = c()\n",
    "for(cT in c(\"N\", \"CM\", \"EM\")){\n",
    "    cr_rna_match = readRDS(paste0(\"CART_CUTRUN_Project/results/CUTANDRUN/analysis/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "    bivProp = data.frame(prop = (cr_rna_match %>% filter(H3K4me2 > 4.5, H3K27me3 > 2.5) %>% nrow)/nrow(cr_rna_match) * 100, num = cr_rna_match %>% filter(H3K4me2 > 4.5, H3K27me3 > 2.5) %>% nrow, celltype = cT, type = \"Bivalent\") %>% rbind(bivProp)\n",
    "    bivProp = data.frame(prop = 100 - (cr_rna_match %>% filter(H3K4me2 > 4.5, H3K27me3 > 2.5) %>% nrow)/nrow(cr_rna_match) * 100, num = cr_rna_match %>% filter(H3K4me2 > 4.5, H3K27me3 > 2.5) %>% nrow, celltype = cT, type = \"Non-bivalent\") %>% rbind(bivProp)\n",
    "}\n",
    "bivProp$celltype = factor(bivProp$celltype, levels = c(\"N\", \"CM\", \"EM\"))\n",
    "bivProp\n",
    "\n",
    "bivProp %>% ggplot(aes(x = celltype, y = prop, fill = type, label = round(prop, 2))) + \n",
    "geom_bar(stat = \"identity\") +\n",
    "theme_bw(base_size = 25) +\n",
    "scale_fill_brewer(palette = \"Set1\") +\n",
    "xlab(\"\")+\n",
    "ylab(\"Percentage of Genes\") +\n",
    "rremove(\"legend.title\") +\n",
    "theme(legend.position = \"top\")\n",
    "\n",
    "ggsave(\"CART_CUTRUN_Project/results/paper_figure_take3/Percentage_gene_bivalency.pdf\", width = 8, height = 8)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## get the gene region falling in K4+ K27+ and K4+K27+ category \n",
    "## generate the computeMatrix heatmap for three category for each cell type for +/-1kb or +/-5kb\n",
    "head(cr_rna_match)\n",
    "colnames(gtf_df)\n",
    "gene_full_list = gtf_df %>% data.frame %>% dplyr::filter(type == \"gene\") %>% dplyr::select(seqnames, start, end, gene_name, score, strand)\n",
    "\n",
    "for(cT in c(\"N\", \"CM\", \"EM\")){\n",
    "    cr_rna_match = readRDS(paste0(\"CART_CUTRUN_Project/results/CUTANDRUN/analysis/RDS/cr_rna_match_\", cT, \".rds\"))\n",
    "\n",
    "    k4p_k27n = cr_rna_match %>% filter(H3K4me2 > 4.5, H3K27me3 < 2.5) %$% gene_name  \n",
    "    k4p_k27p = cr_rna_match %>% filter(H3K4me2 > 4.5, H3K27me3 > 2.5) %$% gene_name\n",
    "    k4n_k27p = cr_rna_match %>% filter(H3K4me2 < 4, H3K27me3 > 2.5) %$% gene_name\n",
    "\n",
    "    gene_full_list %>% filter(gene_name %in% k4p_k27n) %>% write.table(paste0(\"CART_CUTRUN_Project/results/paper_figure_take3/K4pos_K27neg_gene_list_\", cT, \".bed\"), sep = \"\\t\", col.names = FALSE, row.names = FALSE, quote = FALSE)\n",
    "\n",
    "    gene_full_list %>% filter(gene_name %in% k4p_k27p) %>% write.table(paste0(\"CART_CUTRUN_Project/results/paper_figure_take3/K4pos_K27pos_gene_list_\", cT, \".bed\"), sep = \"\\t\", col.names = FALSE, row.names = FALSE, quote = FALSE)\n",
    "\n",
    "    gene_full_list %>% filter(gene_name %in% k4n_k27p) %>% write.table(paste0(\"CART_CUTRUN_Project/results/paper_figure_take3/K4neg_K27pos_gene_list_\", cT, \".bed\"), sep = \"\\t\", col.names = FALSE, row.names = FALSE, quote = FALSE)\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## for healthy donors\n",
    "\n",
    "for(gene_group in unique(small_list$Group)){\n",
    "    for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    \n",
    "        pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_scatterplot_SmallCuratedGeneList_\", gsub(\" \", \"_\", gene_group), \"_perDonor_celltype\", cT, \".pdf\"), width = 13, height =  11)\n",
    "    \n",
    "        rna_select_data = rna_expCount\n",
    "        cr_select_data = normMat$H3K27me3\n",
    "\n",
    "        rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "        cr_gene = rownames(cr_select_data) %>% tolower\n",
    "        rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "        rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "        cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "        rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "        cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "        \n",
    "        target_gene = small_list %>% dplyr::filter(Group == gene_group) %$% Gene %>% tolower\n",
    "\n",
    "        for(hd in paste0(\"HD\", 5:7)){\n",
    "            rna_select_cond = paste0(\"Input_\", cT, \"_\", hd)\n",
    "            cr_select_cond = paste0(cT, \"_Input_\", hd)\n",
    "            cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "            rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "    \n",
    "            cr_rna_match = data.frame(\n",
    "                H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1), # %>% rowMeans,\n",
    "                H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1), # %>% rowMeans,\n",
    "                RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1), # %>% rowMeans,\n",
    "                gene_name = rna_gene[rna_row_index] %>% toupper\n",
    "            )\n",
    "\n",
    "            x_thres = 4.5\n",
    "            y_thres = 2.5\n",
    "            panel1 = cr_rna_match %>% dplyr::filter(H3K4me2 <= x_thres, RNA >= y_thres) %>% nrow()\n",
    "            panel2 = cr_rna_match %>% dplyr::filter(H3K4me2 > x_thres, RNA >= y_thres) %>% nrow()\n",
    "            panel3 = cr_rna_match %>% dplyr::filter(H3K4me2 <= x_thres, RNA < y_thres) %>% nrow()\n",
    "            panel4 = cr_rna_match %>% dplyr::filter(H3K4me2 > x_thres, RNA < y_thres) %>% nrow()\n",
    "        \n",
    "            print(cr_rna_match %>% ggplot(aes(x = H3K4me2, y = RNA, color = H3K27me3)) +\n",
    "            geom_point(size = 0.5) +\n",
    "            geom_point(aes(x = H3K4me2, y = RNA), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 2, color = \"#ff1c1c\") +\n",
    "            geom_text_repel(aes(x = H3K4me2, y = RNA, label = gene_name), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 5, color = \"#ff1c1c\") +\n",
    "            theme_bw(base_size = 25) +\n",
    "            xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "            ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "            scale_color_viridis() +\n",
    "            geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "            geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\") +\n",
    "            annotate(\"text\", x = c(0, 10, 0, 10), y = c(20, 20, -1, -1), label = paste0(c(panel1, panel2, panel3, panel4), \" genes\")) +\n",
    "            ggtitle(rna_select_cond)) \n",
    "\n",
    "\n",
    "            x_thres = 2.5\n",
    "            y_thres = 2.5\n",
    "            panel1 = cr_rna_match %>% dplyr::filter(H3K27me3 <= x_thres, RNA >= y_thres) %>% nrow()\n",
    "            panel2 = cr_rna_match %>% dplyr::filter(H3K27me3 > x_thres, RNA >= y_thres) %>% nrow()\n",
    "            panel3 = cr_rna_match %>% dplyr::filter(H3K27me3 <= x_thres, RNA < y_thres) %>% nrow()\n",
    "            panel4 = cr_rna_match %>% dplyr::filter(H3K27me3 > x_thres, RNA < y_thres) %>% nrow()\n",
    "            print(cr_rna_match %>% ggplot(aes(x = H3K27me3, y = RNA, color = H3K4me2)) +\n",
    "            geom_point(size = 0.5) +\n",
    "            geom_point(aes(x = H3K27me3, y = RNA), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 2, color = \"#ff1c1c\") +\n",
    "            geom_text_repel(aes(x = H3K27me3, y = RNA, label = gene_name), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 5, color = \"#ff1c1c\") +\n",
    "            theme_bw(base_size = 25) +\n",
    "            xlab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "            ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "            scale_color_viridis() +\n",
    "            geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "            geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\") +\n",
    "            annotate(\"text\", x = c(0, 10, 0, 10), y = c(20, 20, -1, -1), label = paste0(c(panel1, panel2, panel3, panel4), \" genes\")) +\n",
    "            ggtitle(rna_select_cond))\n",
    "        }\n",
    "        dev.off()\n",
    "    }\n",
    "  \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## for healthy donors summary gene number in each panel\n",
    "gene_num_panel_summary = c()\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "        rna_select_data = rna_expCount\n",
    "        cr_select_data = normMat$H3K27me3\n",
    "\n",
    "        rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "        cr_gene = rownames(cr_select_data) %>% tolower\n",
    "        rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "        rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "        cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "        rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "        cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "        \n",
    "        target_gene = small_list %>% dplyr::filter(Group == gene_group) %$% Gene %>% tolower\n",
    "\n",
    "        for(hd in paste0(\"HD\", 5:7)){\n",
    "            rna_select_cond = paste0(\"Input_\", cT, \"_\", hd)\n",
    "            cr_select_cond = paste0(cT, \"_Input_\", hd)\n",
    "            cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "            rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "            \n",
    "            \n",
    "            cr_rna_match = data.frame(\n",
    "                H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1), # %>% rowMeans,\n",
    "                H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1), # %>% rowMeans,\n",
    "                RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1), # %>% rowMeans,\n",
    "                gene_name = rna_gene[rna_row_index] %>% toupper\n",
    "            )\n",
    "\n",
    "            x_thres = 4.5\n",
    "            y_thres = 2.5\n",
    " \n",
    "            panel1 = cr_rna_match %>% dplyr::filter(H3K4me2 <= x_thres, RNA >= y_thres) %>% nrow()\n",
    "            panel2 = cr_rna_match %>% dplyr::filter(H3K4me2 > x_thres, RNA >= y_thres) %>% nrow()\n",
    "            panel3 = cr_rna_match %>% dplyr::filter(H3K4me2 <= x_thres, RNA < y_thres) %>% nrow()\n",
    "            panel4 = cr_rna_match %>% dplyr::filter(H3K4me2 > x_thres, RNA < y_thres) %>% nrow()\n",
    "            gene_num_panel_summary = data.frame(gene_num = c(panel1, panel2, panel3, panel4), panel = paste0(\"Panel\", 1:4), celltype = cT, histone = \"H3K4me2\", hd = hd) %>% rbind(gene_num_panel_summary, .)\n",
    "\n",
    "            x_thres = 2.5\n",
    "            y_thres = 2.5\n",
    "\n",
    "            panel1 = cr_rna_match %>% dplyr::filter(H3K27me3 <= x_thres, RNA >= y_thres) %>% nrow()\n",
    "            panel2 = cr_rna_match %>% dplyr::filter(H3K27me3 > x_thres, RNA >= y_thres) %>% nrow()\n",
    "            panel3 = cr_rna_match %>% dplyr::filter(H3K27me3 <= x_thres, RNA < y_thres) %>% nrow()\n",
    "            panel4 = cr_rna_match %>% dplyr::filter(H3K27me3 > x_thres, RNA < y_thres) %>% nrow()\n",
    "            gene_num_panel_summary = data.frame(gene_num = c(panel1, panel2, panel3, panel4), panel = paste0(\"Panel\", 1:4), celltype = cT, histone = \"H3K27me3\", hd = hd) %>% rbind(gene_num_panel_summary, .)\n",
    "\n",
    "        }\n",
    "        \n",
    "  \n",
    "}\n",
    "pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_scatterplot_geneNum_per_quadrant_per_donor.pdf\"), width = 13, height =  11)\n",
    "gene_num_panel_summary$celltype = factor(gene_num_panel_summary$celltype, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\"))\n",
    "gene_num_panel_summary %>% ggplot(aes(hd, y = gene_num, fill = panel)) +\n",
    "geom_bar(stat = \"identity\", position =  \"stack\") +\n",
    "facet_grid(histone~celltype) +\n",
    "scale_fill_brewer(palette = \"Set1\") +\n",
    "theme_bw(base_size = 25) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Genes\")\n",
    "dev.off()\n",
    "\n",
    "gene_num_panel_summary$celltype = factor(gene_num_panel_summary$celltype, levels = c(\"N\", \"CM\", \"EM\", \"EMRA\"))\n",
    "gene_num_panel_summary %>% ggplot(aes(hd, y = gene_num, fill = panel)) +\n",
    "geom_bar(stat = \"identity\", position =  \"stack\") +\n",
    "facet_grid(histone~celltype) +\n",
    "scale_fill_brewer(palette = \"Set1\") +\n",
    "theme_bw(base_size = 25) +\n",
    "xlab(\"\") +\n",
    "ylab(\"# of Genes\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## get the gene name in each panel\n",
    "gene_num_panel_summary = c()\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "\n",
    "        rna_select_data = rna_expCount\n",
    "        cr_select_data = normMat$H3K27me3\n",
    "\n",
    "        rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "        cr_gene = rownames(cr_select_data) %>% tolower\n",
    "        rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "        rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "        cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "        rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "        cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "        \n",
    "        target_gene = small_list %>% dplyr::filter(Group == gene_group) %$% Gene %>% tolower\n",
    "\n",
    "        rna_select_cond = paste0(\"Input_\", cT)\n",
    "        cr_select_cond = paste0(cT, \"_Input\")\n",
    "        cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "        rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "            \n",
    "            \n",
    "        cr_rna_match = data.frame(\n",
    "                H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "                H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "                RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans,\n",
    "                gene_name = rna_gene[rna_row_index] %>% toupper\n",
    "        )\n",
    "\n",
    "        cr_rna_match_full = cr_rna_match %>% dplyr::select(gene_name, RNA, H3K4me2, H3K27me3) %>%\n",
    "        cbind(., data.frame(rna_select_data)[rna_row_index, rna_col_index]) %>% \n",
    "        cbind(., normMat$H3K4me2[cr_row_index, cr_col_index]) %>% \n",
    "        cbind(., normMat$H3K27me3[cr_row_index, cr_col_index])\n",
    "        \n",
    "        x_thres = 4.5\n",
    "        y_thres = 2.5\n",
    "        panel1_gene = cr_rna_match_full %>% dplyr::filter(H3K4me2 <= x_thres, RNA >= y_thres) \n",
    "        panel2_gene = cr_rna_match_full %>% dplyr::filter(H3K4me2 > x_thres, RNA >= y_thres) \n",
    "        panel3_gene = cr_rna_match_full %>% dplyr::filter(H3K4me2 <= x_thres, RNA < y_thres) \n",
    "        panel4_gene = cr_rna_match_full %>% dplyr::filter(H3K4me2 > x_thres, RNA < y_thres)\n",
    "        write.csv(panel1_gene, file = paste0(fig_path, \"/CSV/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_geneInfo_H3K4me2_celltype\", cT, \"_panel1.csv\"), row.names = F)\n",
    "        write.csv(panel2_gene, file = paste0(fig_path, \"/CSV/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_geneInfo_H3K4me2_celltype\", cT, \"_panel2.csv\"), row.names = F)\n",
    "        write.csv(panel3_gene, file = paste0(fig_path, \"/CSV/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_geneInfo_H3K4me2_celltype\", cT, \"_panel3.csv\"), row.names = F)\n",
    "        write.csv(panel4_gene, file = paste0(fig_path, \"/CSV/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_geneInfo_H3K4me2_celltype\", cT, \"_panel4.csv\"), row.names = F)\n",
    "      \n",
    "        x_thres = 2.5\n",
    "        y_thres = 2.5\n",
    "        panel1_gene = cr_rna_match_full %>% dplyr::filter(H3K27me3 <= x_thres, RNA >= y_thres)\n",
    "        panel2_gene = cr_rna_match_full %>% dplyr::filter(H3K27me3 > x_thres, RNA >= y_thres)\n",
    "        panel3_gene = cr_rna_match_full %>% dplyr::filter(H3K27me3 <= x_thres, RNA < y_thres)\n",
    "        panel4_gene = cr_rna_match_full %>% dplyr::filter(H3K27me3 > x_thres, RNA < y_thres)\n",
    "        write.csv(panel1_gene, file = paste0(fig_path, \"/CSV/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_geneInfo_H3K27me3_celltype\", cT, \"_panel1.csv\"), row.names = F)\n",
    "        write.csv(panel2_gene, file = paste0(fig_path, \"/CSV/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_geneInfo_H3K27me3_celltype\", cT, \"_panel2.csv\"), row.names = F)\n",
    "        write.csv(panel3_gene, file = paste0(fig_path, \"/CSV/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_geneInfo_H3K27me3_celltype\", cT, \"_panel3.csv\"), row.names = F)\n",
    "        write.csv(panel4_gene, file = paste0(fig_path, \"/CSV/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_geneInfo_H3K27me3_celltype\", cT, \"_panel4.csv\"), row.names = F)\n",
    "  \n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## bivalent gene categories\n",
    "gene_hist_group = c()\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "        rna_select_cond = paste0(\"Input_\", cT)\n",
    "        cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "        rna_select_data = rna_expCount\n",
    "        colnames(rna_select_data) = paste0(\"RNA_\", colnames(rna_select_data))\n",
    "        cr_select_data = normMat$H3K27me3\n",
    "\n",
    "        rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "        cr_gene = rownames(cr_select_data) %>% tolower\n",
    "        rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "        rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "        cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "        rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "        cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "        \n",
    "        cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "        rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "        if(cT == \"EM\"){\n",
    "            cr_col_index_rm = which(cr_select_data %>% colnames %>% stringr::str_detect(\"EMRA_Input\"))\n",
    "            rna_col_index_rm = which(rna_select_data %>% colnames %>% stringr::str_detect(\"Input_EMRA\"))\n",
    "            cr_col_index = setdiff(cr_col_index, cr_col_index_rm)\n",
    "            rna_col_index = setdiff(rna_col_index, rna_col_index_rm)\n",
    "        }\n",
    "        # cr_rna_match = data.frame(\n",
    "        #     gene_name = rna_gene[rna_row_index] %>% toupper,\n",
    "        #     H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "        #     H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "        #     RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans\n",
    "        # ) %>% mutate(H3K27me3_group = 0, H3K4me2_group = 0)\n",
    "\n",
    "        cr_rna_match = data.frame(\n",
    "            gene_name = rna_gene[rna_row_index] %>% toupper,\n",
    "            H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "            H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "            RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans\n",
    "        ) %>% mutate(H3K27me3_group = 0, H3K4me2_group = 0) %>% \n",
    "        cbind(., log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1)) %>%\n",
    "        cbind(., log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1)) %>% \n",
    "        cbind(., log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1)) \n",
    "\n",
    "\n",
    "        cr_rna_match$H3K27me3_group[which(cr_rna_match$H3K27me3 > 2.5)] = 1\n",
    "        cr_rna_match$H3K4me2_group[which(cr_rna_match$H3K4me2 > 4.5)] = 1\n",
    "        colnames(cr_rna_match)[2:6] = paste0(colnames(cr_rna_match)[2:6], \"_\", cT)\n",
    "        if(length(gene_hist_group) == 0){\n",
    "            gene_hist_group = cr_rna_match\n",
    "        }else{\n",
    "            gene_hist_group = cbind(gene_hist_group, cr_rna_match[, 2:ncol(cr_rna_match)])\n",
    "        }\n",
    "        print(colnames(gene_hist_group))\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "bivalent_active_gene = gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum != 0, H3K27me3_group_sum != 4, H3K4me2_group_sum != 0, H3K4me2_group_sum != 4) %>%\n",
    "dplyr::filter(H3K27me3_group_N >= H3K27me3_group_CM, H3K27me3_group_CM >= H3K27me3_group_EM, H3K27me3_group_EM >= H3K27me3_group_EMRA) %>%\n",
    "dplyr::filter(H3K4me2_group_N <= H3K4me2_group_CM, H3K4me2_group_CM <= H3K4me2_group_EM, H3K4me2_group_EM <= H3K4me2_group_EMRA) %>% dplyr::select(gene_name, H3K27me3_group_N, H3K27me3_group_CM, H3K27me3_group_EM, H3K27me3_group_EMRA, H3K4me2_group_N, H3K4me2_group_CM, H3K4me2_group_EM, H3K4me2_group_EMRA)\n",
    "\n",
    "gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum != 0, H3K27me3_group_sum != 4, H3K4me2_group_sum != 0, H3K4me2_group_sum != 4) %>%\n",
    "dplyr::filter(H3K27me3_group_N >= H3K27me3_group_CM, H3K27me3_group_CM >= H3K27me3_group_EM, H3K27me3_group_EM >= H3K27me3_group_EMRA) %>%\n",
    "dplyr::filter(H3K4me2_group_N <= H3K4me2_group_CM, H3K4me2_group_CM <= H3K4me2_group_EM, H3K4me2_group_EM <= H3K4me2_group_EMRA) %>% \n",
    "write.csv(file = \"CART_CUTRUN_Project/results/paper_figure/RNA_vs_CUTRUN/Bivalent_Genes_activation_from_N_to_EMRA_withHDcounts.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "bivalent_repress_gene = gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum != 0, H3K27me3_group_sum != 4, H3K4me2_group_sum != 0, H3K4me2_group_sum != 4) %>%\n",
    "dplyr::filter(H3K27me3_group_N <= H3K27me3_group_CM, H3K27me3_group_CM <= H3K27me3_group_EM, H3K27me3_group_EM <= H3K27me3_group_EMRA) %>%\n",
    "dplyr::filter(H3K4me2_group_N >= H3K4me2_group_CM, H3K4me2_group_CM >= H3K4me2_group_EM, H3K4me2_group_EM >= H3K4me2_group_EMRA) %>%\n",
    "dplyr::select(gene_name, H3K27me3_group_N, H3K27me3_group_CM, H3K27me3_group_EM, H3K27me3_group_EMRA, H3K4me2_group_N, H3K4me2_group_CM, H3K4me2_group_EM, H3K4me2_group_EMRA)\n",
    "\n",
    "gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum != 0, H3K27me3_group_sum != 4, H3K4me2_group_sum != 0, H3K4me2_group_sum != 4) %>%\n",
    "dplyr::filter(H3K27me3_group_N <= H3K27me3_group_CM, H3K27me3_group_CM <= H3K27me3_group_EM, H3K27me3_group_EM <= H3K27me3_group_EMRA) %>%\n",
    "dplyr::filter(H3K4me2_group_N >= H3K4me2_group_CM, H3K4me2_group_CM >= H3K4me2_group_EM, H3K4me2_group_EM >= H3K4me2_group_EMRA) %>%\n",
    "write.csv(file = \"CART_CUTRUN_Project/results/paper_figure/RNA_vs_CUTRUN/Bivalent_Genes_repression_from_N_to_EMRA_withHDcounts.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Poised to activated\n",
    "bivalent_posie_active_gene = gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum != 0, H3K27me3_group_sum != 4, H3K4me2_group_sum == 4) %>%\n",
    "dplyr::filter(H3K27me3_group_N >= H3K27me3_group_CM, H3K27me3_group_CM >= H3K27me3_group_EM, H3K27me3_group_EM >= H3K27me3_group_EMRA) %>%\n",
    "dplyr::select(gene_name, H3K27me3_group_N, H3K27me3_group_CM, H3K27me3_group_EM, H3K27me3_group_EMRA, H3K4me2_group_N, H3K4me2_group_CM, H3K4me2_group_EM, H3K4me2_group_EMRA)\n",
    "bivalent_posie_active_gene %>% head\n",
    "\n",
    "\n",
    "gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum != 0, H3K27me3_group_sum != 4, H3K4me2_group_sum == 4) %>%\n",
    "dplyr::filter(H3K27me3_group_N >= H3K27me3_group_CM, H3K27me3_group_CM >= H3K27me3_group_EM, H3K27me3_group_EM >= H3K27me3_group_EMRA) %>%\n",
    "write.csv(file = \"CART_CUTRUN_Project/results/paper_figure/RNA_vs_CUTRUN/Bivalent_Genes_poised_to_activated_from_N_to_EMRA_withHDcounts.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Poised to further repressed\n",
    "bivalent_posie_repress_gene = gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum == 4, H3K4me2_group_sum != 0, H3K4me2_group_sum != 4) %>%\n",
    "dplyr::filter(H3K27me3_group_N >= H3K27me3_group_CM, H3K27me3_group_CM >= H3K27me3_group_EM, H3K27me3_group_EM >= H3K27me3_group_EMRA) %>%\n",
    "dplyr::filter(H3K4me2_group_N >= H3K4me2_group_CM, H3K4me2_group_CM >= H3K4me2_group_EM, H3K4me2_group_EM >= H3K4me2_group_EMRA) %>%\n",
    "dplyr::select(gene_name, H3K27me3_group_N, H3K27me3_group_CM, H3K27me3_group_EM, H3K27me3_group_EMRA, H3K4me2_group_N, H3K4me2_group_CM, H3K4me2_group_EM, H3K4me2_group_EMRA)\n",
    "\n",
    "\n",
    "gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum == 4, H3K4me2_group_sum != 0, H3K4me2_group_sum != 4) %>%\n",
    "dplyr::filter(H3K27me3_group_N >= H3K27me3_group_CM, H3K27me3_group_CM >= H3K27me3_group_EM, H3K27me3_group_EM >= H3K27me3_group_EMRA) %>%\n",
    "dplyr::filter(H3K4me2_group_N >= H3K4me2_group_CM, H3K4me2_group_CM >= H3K4me2_group_EM, H3K4me2_group_EM >= H3K4me2_group_EMRA) %>%\n",
    "write.csv(file = \"CART_CUTRUN_Project/results/paper_figure/RNA_vs_CUTRUN/Bivalent_Genes_poised_to_further_repression_from_N_to_EMRA_withHDcounts.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## No mark to a single active mark\n",
    "bivalent_none_active_gene = gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum == 0, H3K4me2_group_sum != 0, H3K4me2_group_sum != 4) %>%\n",
    "dplyr::filter(H3K4me2_group_N <= H3K4me2_group_CM, H3K4me2_group_CM <= H3K4me2_group_EM, H3K4me2_group_EM <= H3K4me2_group_EMRA) %>%\n",
    "dplyr::select(gene_name, H3K27me3_group_N, H3K27me3_group_CM, H3K27me3_group_EM, H3K27me3_group_EMRA, H3K4me2_group_N, H3K4me2_group_CM, H3K4me2_group_EM, H3K4me2_group_EMRA)\n",
    "bivalent_none_active_gene %>% dim\n",
    "bivalent_none_active_gene %>% head\n",
    "\n",
    "gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum == 0, H3K4me2_group_sum != 0, H3K4me2_group_sum != 4) %>%\n",
    "dplyr::filter(H3K4me2_group_N <= H3K4me2_group_CM, H3K4me2_group_CM <= H3K4me2_group_EM, H3K4me2_group_EM <= H3K4me2_group_EMRA) %>% \n",
    "write.csv(file = \"CART_CUTRUN_Project/results/paper_figure/RNA_vs_CUTRUN/Bivalent_Genes_none_to_activation_from_N_to_EMRA_withHDcounts.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## None to only repress marker\n",
    "bivalent_none_repress_gene = gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum != 0, H3K27me3_group_sum != 4, H3K4me2_group_sum == 0) %>%\n",
    "dplyr::filter(H3K27me3_group_N <= H3K27me3_group_CM, H3K27me3_group_CM <= H3K27me3_group_EM, H3K27me3_group_EM <= H3K27me3_group_EMRA) %>%\n",
    "dplyr::select(gene_name, H3K27me3_group_N, H3K27me3_group_CM, H3K27me3_group_EM, H3K27me3_group_EMRA, H3K4me2_group_N, H3K4me2_group_CM, H3K4me2_group_EM, H3K4me2_group_EMRA)\n",
    "bivalent_none_repress_gene %>% dim\n",
    "bivalent_none_repress_gene %>% head()\n",
    "\n",
    "gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum != 0, H3K27me3_group_sum != 4, H3K4me2_group_sum == 0) %>%\n",
    "dplyr::filter(H3K27me3_group_N <= H3K27me3_group_CM, H3K27me3_group_CM <= H3K27me3_group_EM, H3K27me3_group_EM <= H3K27me3_group_EMRA) %>%\n",
    "write.csv(file = \"CART_CUTRUN_Project/results/paper_figure/RNA_vs_CUTRUN/Bivalent_Genes_none_to_repression_from_N_to_EMRA_withHDcounts.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## None to both active and repress marker\n",
    "bivalent_none_both_gene = gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum != 0, H3K27me3_group_sum != 4, H3K4me2_group_sum != 0, H3K4me2_group_sum != 4) %>%\n",
    "dplyr::filter(H3K27me3_group_N <= H3K27me3_group_CM, H3K27me3_group_CM <= H3K27me3_group_EM, H3K27me3_group_EM <= H3K27me3_group_EMRA) %>%\n",
    "dplyr::filter(H3K4me2_group_N <= H3K4me2_group_CM, H3K4me2_group_CM <= H3K4me2_group_EM, H3K4me2_group_EM <= H3K4me2_group_EMRA) %>%\n",
    "dplyr::select(gene_name, H3K27me3_group_N, H3K27me3_group_CM, H3K27me3_group_EM, H3K27me3_group_EMRA, H3K4me2_group_N, H3K4me2_group_CM, H3K4me2_group_EM, H3K4me2_group_EMRA)\n",
    "bivalent_none_both_gene %>% dim\n",
    "bivalent_none_both_gene %>% head()\n",
    "\n",
    "gene_hist_group %>% mutate(H3K27me3_group_sum = H3K27me3_group_N + H3K27me3_group_CM + H3K27me3_group_EM + H3K27me3_group_EMRA, H3K4me2_group_sum = H3K4me2_group_N + H3K4me2_group_CM + H3K4me2_group_EM + H3K4me2_group_EMRA) %>%\n",
    "dplyr::filter(H3K27me3_group_sum != 0, H3K27me3_group_sum != 4, H3K4me2_group_sum != 0, H3K4me2_group_sum != 4) %>%\n",
    "dplyr::filter(H3K27me3_group_N <= H3K27me3_group_CM, H3K27me3_group_CM <= H3K27me3_group_EM, H3K27me3_group_EM <= H3K27me3_group_EMRA) %>%\n",
    "dplyr::filter(H3K4me2_group_N <= H3K4me2_group_CM, H3K4me2_group_CM <= H3K4me2_group_EM, H3K4me2_group_EM <= H3K4me2_group_EMRA) %>%\n",
    "write.csv(file = \"CART_CUTRUN_Project/results/paper_figure/RNA_vs_CUTRUN/Bivalent_Genes_none_to_both_from_N_to_EMRA_withHDcounts.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## match the gene name -- original figure\n",
    "pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_scatterplot_expectedCounts.pdf\"), width = 13, height =  11)\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    rna_select_cond = paste0(\"Input_\", cT)\n",
    "    cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "    rna_select_data = rna_expCount\n",
    "    cr_select_data = normMat$H3K27me3\n",
    "\n",
    "    rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower %>% unique\n",
    "    cr_gene = rownames(cr_select_data) %>% tolower\n",
    "    rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "    rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "    cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "    rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "    cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "\n",
    "    cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "    rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "    cr_rna_match = data.frame(\n",
    "        H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "        H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "        RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans,\n",
    "        gene_length = log(common_gene$width_uniq)\n",
    "    )\n",
    "\n",
    "    print(cr_rna_match %>% ggplot(aes(x = H3K4me2, y = RNA, color = H3K27me3)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    # geom_hex(bins = 300) +\n",
    "    theme_bw(base_size = 25) +\n",
    "    xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "    ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(rna_select_cond))\n",
    "\n",
    "    print(cr_rna_match %>% ggplot(aes(x = H3K27me3, y = RNA, color = H3K4me2)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    # geom_hex(bins = 300) +\n",
    "    theme_bw(base_size = 25) +\n",
    "    xlab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "    ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(rna_select_cond))\n",
    "\n",
    "    print(cr_rna_match %>% ggplot(aes(x = H3K4me2, y = RNA, color = gene_length)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    # geom_hex(bins = 300) +\n",
    "    theme_bw(base_size = 25) +\n",
    "    xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "    ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(rna_select_cond))\n",
    "\n",
    "    print(cr_rna_match %>% ggplot(aes(x = H3K27me3, y = RNA, color = gene_length)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    # geom_hex(bins = 300) +\n",
    "    theme_bw(base_size = 25) +\n",
    "    xlab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "    ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(rna_select_cond))\n",
    "\n",
    "}\n",
    "dev.off()\n",
    "\n",
    "options(repr.plot.width=12, repr.plot.height=9)\n",
    "cr_rna_match %>% ggplot(aes(x = H3K4me2, y = RNA, color = H3K27me3)) +\n",
    "geom_point(size = 0.1) +\n",
    "# geom_hex(bins = 300) +\n",
    "theme_bw(base_size = 25) +\n",
    "xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "scale_color_viridis() +\n",
    "ggtitle(rna_select_cond)\n",
    "\n",
    "cr_rna_match %>% ggplot(aes(x = H3K27me3, y = RNA, color = H3K4me2)) +\n",
    "geom_point(size = 0.1) +\n",
    "# geom_hex(bins = 300) +\n",
    "theme_bw(base_size = 25) +\n",
    "xlab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "scale_color_viridis() +\n",
    "ggtitle(rna_select_cond)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## match the gene name -- original figure --TPM version\n",
    "pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_scatterplot_TPMversion.pdf\"), width = 13, height =  11)\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    rna_select_cond = paste0(\"Input_\", cT)\n",
    "    cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "    rna_select_data = rna_TPM\n",
    "    cr_select_data = normMat$H3K27me3\n",
    "\n",
    "    rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "    cr_gene = rownames(cr_select_data) %>% tolower\n",
    "    rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "    rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "    cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "    rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "    cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "\n",
    "    cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "    rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "    cr_rna_match = data.frame(\n",
    "        H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "        H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "        RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans\n",
    "    )\n",
    "\n",
    "    print(cr_rna_match %>% ggplot(aes(x = H3K4me2, y = RNA, color = H3K27me3)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    # geom_hex(bins = 300) +\n",
    "    theme_bw(base_size = 25) +\n",
    "    xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "    ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(rna_select_cond))\n",
    "\n",
    "    print(cr_rna_match %>% ggplot(aes(x = H3K27me3, y = RNA, color = H3K4me2)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    # geom_hex(bins = 300) +\n",
    "    theme_bw(base_size = 25) +\n",
    "    xlab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "    ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(rna_select_cond))\n",
    "\n",
    "}\n",
    "dev.off()\n",
    "\n",
    "options(repr.plot.width=12, repr.plot.height=9)\n",
    "cr_rna_match %>% ggplot(aes(x = H3K4me2, y = RNA, color = H3K27me3)) +\n",
    "geom_point(size = 0.1) +\n",
    "# geom_hex(bins = 300) +\n",
    "theme_bw(base_size = 25) +\n",
    "xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "scale_color_viridis() +\n",
    "ggtitle(rna_select_cond)\n",
    "\n",
    "cr_rna_match %>% ggplot(aes(x = H3K27me3, y = RNA, color = H3K4me2)) +\n",
    "geom_point(size = 0.1) +\n",
    "# geom_hex(bins = 300) +\n",
    "theme_bw(base_size = 25) +\n",
    "xlab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "scale_color_viridis() +\n",
    "ggtitle(rna_select_cond)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## match the gene name -- original figure --FPKM version\n",
    "pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_scatterplot_FPKMversion.pdf\"), width = 13, height =  11)\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    rna_select_cond = paste0(\"Input_\", cT)\n",
    "    cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "    rna_select_data = rna_FPKM\n",
    "    cr_select_data = normMat$H3K27me3\n",
    "\n",
    "    rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "    cr_gene = rownames(cr_select_data) %>% tolower\n",
    "    rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "    rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "    cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "    rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "    cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "\n",
    "    cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "    rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "    cr_rna_match = data.frame(\n",
    "        H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "        H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "        RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans\n",
    "    )\n",
    "\n",
    "    print(cr_rna_match %>% ggplot(aes(x = H3K4me2, y = RNA, color = H3K27me3)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    # geom_hex(bins = 300) +\n",
    "    theme_bw(base_size = 25) +\n",
    "    xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "    ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(rna_select_cond))\n",
    "\n",
    "    print(cr_rna_match %>% ggplot(aes(x = H3K27me3, y = RNA, color = H3K4me2)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    # geom_hex(bins = 300) +\n",
    "    theme_bw(base_size = 25) +\n",
    "    xlab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "    ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(rna_select_cond))\n",
    "\n",
    "}\n",
    "dev.off()\n",
    "\n",
    "options(repr.plot.width=12, repr.plot.height=9)\n",
    "cr_rna_match %>% ggplot(aes(x = H3K4me2, y = RNA, color = H3K27me3)) +\n",
    "geom_point(size = 0.1) +\n",
    "# geom_hex(bins = 300) +\n",
    "theme_bw(base_size = 25) +\n",
    "xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "scale_color_viridis() +\n",
    "ggtitle(rna_select_cond)\n",
    "\n",
    "cr_rna_match %>% ggplot(aes(x = H3K27me3, y = RNA, color = H3K4me2)) +\n",
    "geom_point(size = 0.1) +\n",
    "# geom_hex(bins = 300) +\n",
    "theme_bw(base_size = 25) +\n",
    "xlab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "scale_color_viridis() +\n",
    "ggtitle(rna_select_cond)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "options(repr.plot.width=18, repr.plot.height=11)\n",
    "\n",
    "## check the bivalent peaks\n",
    "for(gene_group in unique(small_list$Group)){\n",
    "## mean across healthy donors\n",
    "    pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_TSSneighborHistoneSignal_scatterplot_SmallCuratedGeneList_\", gsub(\" \", \"_\", gene_group), \"_BivalentDecipher.pdf\"), width = 18, height =  16)\n",
    "    for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "        rna_select_cond = paste0(\"Input_\", cT)\n",
    "        cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "        rna_select_data = rna_expCount\n",
    "        cr_select_data = normMat$H3K27me3\n",
    "\n",
    "        rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "        cr_gene = rownames(cr_select_data) %>% tolower\n",
    "        rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "        rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "        cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "        rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "        cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "        \n",
    "        target_gene = small_list %>% dplyr::filter(Group == gene_group) %$% Gene %>% tolower\n",
    "\n",
    "        cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "        rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "        cr_rna_match = data.frame(\n",
    "            H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "            H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "            RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans,\n",
    "            gene_name = rna_gene[rna_row_index] %>% toupper\n",
    "        )\n",
    "\n",
    "        x_thres = 4.5\n",
    "        y_thres = 2.5\n",
    "        cr_rna_match$H3K4me2_panel = \"I\"\n",
    "        cr_rna_match$H3K4me2_panel[which(cr_rna_match$H3K4me2 > x_thres & cr_rna_match$RNA >= y_thres)] = \"II\"\n",
    "        cr_rna_match$H3K4me2_panel[which(cr_rna_match$H3K4me2 <= x_thres & cr_rna_match$RNA < y_thres)] = \"III\"\n",
    "        cr_rna_match$H3K4me2_panel[which(cr_rna_match$H3K4me2 > x_thres & cr_rna_match$RNA < y_thres)] = \"IV\"\n",
    "\n",
    "        print(cr_rna_match %>% ggplot(aes(x = H3K4me2, y = H3K27me3, color = H3K4me2_panel)) +\n",
    "        geom_point(size = 0.5) +\n",
    "        geom_point(aes(x = H3K4me2, y = H3K27me3), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 2, color = \"#ff1c1c\") +\n",
    "        geom_text_repel(aes(x = H3K4me2, y = H3K27me3, label = gene_name), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 5, color = \"#ff1c1c\") +\n",
    "        theme_bw(base_size = 25) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_brewer(palette = \"Set1\") +\n",
    "        geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "        geom_hline(yintercept = 2.5, linetype = \"dashed\", color = \"grey10\") +\n",
    "        guides(colour = guide_legend(override.aes = list(size=10))) +\n",
    "        ggtitle(rna_select_cond)) +\n",
    "        coord_cartesian(ylim = c(0, 8), xlim = c(0, 10))\n",
    "        \n",
    "        print(cr_rna_match %>% ggplot(aes(x = H3K4me2, y = H3K27me3, color = H3K4me2_panel)) +\n",
    "        geom_point(size = 0.5) +\n",
    "        geom_point(aes(x = H3K4me2, y = H3K27me3), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 2, color = \"#ff1c1c\") +\n",
    "        geom_text_repel(aes(x = H3K4me2, y = H3K27me3, label = gene_name), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 5, color = \"#ff1c1c\") +\n",
    "        theme_bw(base_size = 25) +\n",
    "        facet_wrap(~H3K4me2_panel) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_brewer(palette = \"Set1\") +\n",
    "        geom_vline(xintercept = x_thres, linetype = \"dashed\", color = 'grey10') +\n",
    "        geom_hline(yintercept = 2.5, linetype = \"dashed\", color = \"grey10\") +\n",
    "        guides(colour = guide_legend(override.aes = list(size=10))) +\n",
    "        ggtitle(rna_select_cond)) +\n",
    "        coord_cartesian(ylim = c(0, 8), xlim = c(0, 10))\n",
    "\n",
    "        x_thres = 2.5\n",
    "        y_thres = 2.5\n",
    "        cr_rna_match$H3K27me3_panel = \"I\"\n",
    "        cr_rna_match$H3K27me3_panel[which(cr_rna_match$H3K27me3 > x_thres & cr_rna_match$RNA >= y_thres)] = \"II\"\n",
    "        cr_rna_match$H3K27me3_panel[which(cr_rna_match$H3K27me3 <= x_thres & cr_rna_match$RNA < y_thres)] = \"III\"\n",
    "        cr_rna_match$H3K27me3_panel[which(cr_rna_match$H3K27me3 > x_thres & cr_rna_match$RNA < y_thres)] = \"IV\"\n",
    "\n",
    "        print(cr_rna_match %>% ggplot(aes(x = H3K4me2, y = H3K27me3, color = H3K27me3_panel)) +\n",
    "        geom_point(size = 0.5) +\n",
    "        geom_point(aes(x = H3K4me2, y = H3K27me3), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 2, color = \"#ff1c1c\") +\n",
    "        geom_text_repel(aes(x = H3K4me2, y = H3K27me3, label = gene_name), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 5, color = \"#ff1c1c\") +\n",
    "        theme_bw(base_size = 25) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_brewer(palette = \"Set1\") +\n",
    "        geom_vline(xintercept = 4.5, linetype = \"dashed\", color = 'grey10') +\n",
    "        geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\") +\n",
    "        guides(colour = guide_legend(override.aes = list(size=10))) +\n",
    "        ggtitle(rna_select_cond)) +\n",
    "        coord_cartesian(ylim = c(0, 8), xlim = c(0, 10))\n",
    "\n",
    "        print(cr_rna_match %>% ggplot(aes(x = H3K4me2, y = H3K27me3, color = H3K27me3_panel)) +\n",
    "        geom_point(size = 0.5) +\n",
    "        geom_point(aes(x = H3K4me2, y = H3K27me3), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 2, color = \"#ff1c1c\") +\n",
    "        geom_text_repel(aes(x = H3K4me2, y = H3K27me3, label = gene_name), data = cr_rna_match[which(rna_gene[rna_row_index] %in% target_gene), ], size = 5, color = \"#ff1c1c\") +\n",
    "        theme_bw(base_size = 25) +\n",
    "        facet_wrap(~H3K27me3_panel) +\n",
    "        xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "        ylab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "        scale_color_brewer(palette = \"Set1\") +\n",
    "        geom_vline(xintercept = 4.5, linetype = \"dashed\", color = 'grey10') +\n",
    "        geom_hline(yintercept = y_thres, linetype = \"dashed\", color = \"grey10\") +\n",
    "        guides(colour = guide_legend(override.aes = list(size=10))) +\n",
    "        ggtitle(rna_select_cond)) +\n",
    "        coord_cartesian(ylim = c(0, 8), xlim = c(0, 10))\n",
    "    }\n",
    "    dev.off()\n",
    "}"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### METHOD 2 Consider all peaks that are near TSS and Normalize peak length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## load peak information and match with nearest genes\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "mPeak$H3K4me2 = mPeak$H3K4me2[which(width(mPeak$H3K4me2) < quantile(width(mPeak$H3K4me2), 0.75))]\n",
    "mPeak$H3K27me3 = mPeak$H3K27me3[which(width(mPeak$H3K27me3) < quantile(width(mPeak$H3K27me3), 0.75))]\n",
    "\n",
    "max_distance = 5000\n",
    "peak_match_gene = list()\n",
    "for(hist in c(\"H3K4me2\", \"H3K27me3\")){\n",
    "    dp_match_gene = distanceToNearest(mPeak[[hist]], gene_gr, select = \"all\") %>% data.frame %>% dplyr::filter(distance <= max_distance)\n",
    "    peak_match_gene[[hist]] = data.frame(peak_id = dp_match_gene$queryHits, gene_name = gene_gr$gene_name[dp_match_gene$subjectHits]) #%>% group_by(peak_id) %>% summarize(nearest_genes = paste(gene_name, collapse = \";\"))\n",
    "\n",
    "}\n",
    "\n",
    "## normalize the peak count by sequencing depth\n",
    "normMat = countMat\n",
    "for(hist in c(\"H3K4me2\", \"H3K27me3\")){\n",
    "    for(sample_each in colnames(countMat[[hist]])){\n",
    "        normMat[[hist]][, sample_each] = log1p(countMat[[hist]][, sample_each]/designInfo$depth[which(designInfo$exps == sample_each)] * max(designInfo$depth))\n",
    "    }\n",
    "}\n",
    "\n",
    "head(countMat$H3K4me2)\n",
    "head(normMat$H3K4me2)\n",
    "\n",
    "## Method I.\n",
    "## normalize peak count scaled by peak length \n",
    "for(hist in c(\"H3K4me2\", \"H3K27me3\")){\n",
    "    peak_width = data.frame(mPeak[[hist]])$width\n",
    "    for(i in 1:nrow(normMat[[hist]])){\n",
    "         normMat[[hist]][i,] =  normMat[[hist]][i, ]/peak_width[i] * median(peak_width)\n",
    "    }\n",
    "}\n",
    "head(normMat$H3K4me2)\n",
    "\n",
    "## Method II.\n",
    "## normalize peak count standardized by peak length \n",
    "# for(hist in c(\"H3K4me2\", \"H3K27me3\")){\n",
    "    \n",
    "#     for(i in 1:nrow(normMat[[hist]])){\n",
    "#         tmp = normMat[[hist]][i, ]\n",
    "#         normMat[[hist]][i,] =  (tmp - mean(tmp))/sd(tmp)\n",
    "#     }\n",
    "# }\n",
    "# head(normMat$H3K4me2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## get the matching genes\n",
    "hist = \"H3K4me2\"\n",
    "pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_HistonePeakScalePeakLength_scatterplot.pdf\"), width = 13, height =  11)\n",
    "# pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_HistonePeakStandardizePeakLength_scatterplot.pdf\"), width = 13, height =  11)\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "\n",
    "\n",
    "        tmp = match(peak_match_gene[[hist]]$gene_name %>% tolower, rownames(rna_expCount) %>% gsub(\".*_\", \"\", .) %>% tolower)\n",
    "        cr_row_index = which(!is.na(tmp))\n",
    "        rna_row_index = tmp[which(!is.na(tmp))]\n",
    "\n",
    "        rna_select_cond = paste0(\"Input_\", cT)\n",
    "        cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "        rna_select_data = rna_expCount\n",
    "        cr_select_data = normMat[[hist]]\n",
    "\n",
    "        cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "        rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "        ## for scalePeakLength method\n",
    "        cr_rna_match = data.frame(\n",
    "                H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "                H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "                RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans\n",
    "        )\n",
    "\n",
    "        ## for standardizePeakLength method\n",
    "        # cr_rna_match = data.frame(\n",
    "        #         H3K27me3 = normMat$H3K27me3[cr_row_index, cr_col_index]  %>% rowMeans,\n",
    "        #         H3K4me2 = normMat$H3K4me2[cr_row_index, cr_col_index] %>% rowMeans,\n",
    "        #         RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans\n",
    "        # )\n",
    "\n",
    "    print(cr_rna_match %>% ggplot(aes(x = H3K4me2, y = RNA, color = H3K27me3)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    # geom_hex(bins = 300) +\n",
    "    theme_bw(base_size = 25) +\n",
    "    xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "    ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(rna_select_cond))\n",
    "\n",
    "    print(cr_rna_match %>% ggplot(aes(x = H3K27me3, y = RNA, color = H3K4me2)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    # geom_hex(bins = 300) +\n",
    "    theme_bw(base_size = 25) +\n",
    "    xlab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "    ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(rna_select_cond))\n",
    "}\n",
    "dev.off()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "pdf(paste0(fig_path, \"/RNAseq_CUTRUN_GEX_HistonePeakScalePeakLength_scatterplot.pdf\"), width = 13, height =  11)\n",
    "for(cT in c(\"N\", \"CM\", \"EM\", \"EMRA\")){\n",
    "    rna_select_cond = paste0(\"Input_\", cT)\n",
    "    cr_select_cond = paste0(cT, \"_Input\")\n",
    "\n",
    "    rna_select_data = rna_expCount\n",
    "    cr_select_data = normMat$H3K27me3\n",
    "\n",
    "    rna_gene = rownames(rna_select_data) %>% gsub(\".*_\", \"\", .) %>% tolower\n",
    "    cr_gene = rownames(cr_select_data) %>% tolower\n",
    "    rna_gene_in_cr = match(rna_gene, cr_gene)\n",
    "    rna_gene[which(!is.na(rna_gene_in_cr))] %>% head\n",
    "    cr_gene[rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]] %>% head\n",
    "\n",
    "    rna_row_index = which(!is.na(rna_gene_in_cr))\n",
    "    cr_row_index = rna_gene_in_cr[which(!is.na(rna_gene_in_cr))]\n",
    "\n",
    "    cr_col_index = which(cr_select_data %>% colnames %>% stringr::str_detect(cr_select_cond))\n",
    "    rna_col_index = which(rna_select_data %>% colnames %>% stringr::str_detect(rna_select_cond))\n",
    "\n",
    "    cr_rna_match = data.frame(\n",
    "        H3K27me3 = log2(normMat$H3K27me3[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "        H3K4me2 = log2(normMat$H3K4me2[cr_row_index, cr_col_index] + 1) %>% rowMeans,\n",
    "        RNA = log2(data.frame(rna_select_data)[rna_row_index, rna_col_index] + 1) %>% rowMeans\n",
    "    )\n",
    "\n",
    "    print(cr_rna_match %>% ggplot(aes(x = H3K4me2, y = RNA, color = H3K27me3)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    # geom_hex(bins = 300) +\n",
    "    theme_bw(base_size = 25) +\n",
    "    xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "    ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(rna_select_cond))\n",
    "\n",
    "    print(cr_rna_match %>% ggplot(aes(x = H3K27me3, y = RNA, color = H3K4me2)) +\n",
    "    geom_point(size = 0.5) +\n",
    "    # geom_hex(bins = 300) +\n",
    "    theme_bw(base_size = 25) +\n",
    "    xlab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "    ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "    scale_color_viridis() +\n",
    "    ggtitle(rna_select_cond))\n",
    "\n",
    "}\n",
    "dev.off()\n",
    "\n",
    "options(repr.plot.width=12, repr.plot.height=9)\n",
    "cr_rna_match %>% ggplot(aes(x = H3K4me2, y = RNA, color = H3K27me3)) +\n",
    "geom_point(size = 0.1) +\n",
    "# geom_hex(bins = 300) +\n",
    "theme_bw(base_size = 25) +\n",
    "xlab(\"H3K4me2: log2(scaled count + 1)\") +\n",
    "ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "scale_color_viridis() +\n",
    "ggtitle(rna_select_cond)\n",
    "\n",
    "cr_rna_match %>% ggplot(aes(x = H3K27me3, y = RNA, color = H3K4me2)) +\n",
    "geom_point(size = 0.1) +\n",
    "# geom_hex(bins = 300) +\n",
    "theme_bw(base_size = 25) +\n",
    "xlab(\"H3K27me3: log2(scaled count + 1)\") +\n",
    "ylab(\"RNA-seq: log2(expected count + 1)\") +\n",
    "scale_color_viridis() +\n",
    "ggtitle(rna_select_cond)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
