{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## CUTRUN DE and assign to nearest genes\n",
    "\n",
    "library(pacman)\n",
    "p_load(data.table, dplyr, ggplot2, viridis, magrittr, VennDiagram, ggpubr, limma, edgeR, tidyr, GenomicRanges, RColorBrewer, pheatmap, Seurat, fgsea, GSEABase, limma, TxDb.Hsapiens.UCSC.hg38.knownGene, org.Hs.eg.db)\n",
    "\n",
    "cell_list <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hd_list <- paste0(\"HD\", c(1:3, 5:7))\n",
    "expr_list = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "cell_comp_list = c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "hist_list <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "in_path <- \"CART_CUTRUN_Project/results/RNAseq/process/RSEM/\"\n",
    "out_path <- \"CART_CUTRUN_Project/results/RNAseq/analysis/RSEM/\"\n",
    "inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## output DP with nearest genes\n",
    "load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/IDR_master_peak_list_chromVar_count_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "load(file = paste0(outPath, \"/RData/results_histList_hd1-7_adjustPeak_noChrXYM.RData\")) ## voomDDS results\n",
    "gene_gr <- genes(TxDb.Hsapiens.UCSC.hg38.knownGene, columns = \"gene_id\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "fig_path <- \"CART_CUTRUN_Project/results/paper_figure/CUTRUN_DP_nearestGene/\"\n",
    "\n",
    "\n",
    "for(hist in c(\"H3K27me3\", \"H3K4me2\")){\n",
    "    dp_match_gene <- distanceToNearest(mPeak[[hist]], gene_gr, select = \"all\")\n",
    "    dp_match_gene %>% data.frame %$% distance %>% summary\n",
    "    peak_match_gene <- data.frame(peak_id = dp_match_gene@from, gene_name = mapIds(org.Hs.eg.db, gene_gr$gene_id[dp_match_gene@to], \"SYMBOL\", \"ENTREZID\") %>% as.vector) %>% group_by(peak_id) %>% summarize(nearest_genes = paste(gene_name, collapse = \";\"))\n",
    "\n",
    "    for(sample in names(results[[hist]])){\n",
    "        info = sample %>% strsplit(\"_\") %>% unlist\n",
    "        if(info[1] == \"Input\"){\n",
    "            hd_list = paste0(\"HD\", 4:7)\n",
    "        }else{\n",
    "            hd_list = paste0(\"HD\", 1:7)\n",
    "        }\n",
    "        limma_res <- results[[hist]][[sample]]\n",
    "        limma_res <- cbind(data.frame(mPeak[[hist]])[limma_res$GeneIndex, c(\"seqnames\", \"start\", \"end\")], limma_res)\n",
    "        dp_res <- right_join(peak_match_gene, limma_res, by = c(\"peak_id\" = \"GeneIndex\")) %>% arrange(adj.P.Val)\n",
    "        \n",
    "        ## raw\n",
    "        data_raw = countMat[[hist]][\n",
    "            dp_res$peak_id, \n",
    "                c(\n",
    "                    paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", hd_list), \n",
    "                    paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", hd_list)\n",
    "                )]\n",
    "        colnames(data_raw) = c(paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", hd_list, \"_raw\"), paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", hd_list, \"_raw\"))\n",
    "        tmp = cbind(dp_res, data_raw)\n",
    "        \n",
    "        ## scaled\n",
    "        data_norm <- countMat[[hist]]/colSums(countMat[[hist]]) * 1000000\n",
    "        data_norm_scaled = data_norm[\n",
    "                dp_res$peak_id, \n",
    "                c(\n",
    "                    paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", hd_list), \n",
    "                    paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", hd_list)\n",
    "                )]\n",
    "        colnames(data_norm_scaled) = c(paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", hd_list, \"_scaled\"), paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", hd_list, \"_scaled\"))\n",
    "        tmp = cbind(tmp, data_norm_scaled) \n",
    "        \n",
    "        ## log2\n",
    "        data_norm_log2 = log2(data_norm[\n",
    "                dp_res$peak_id, \n",
    "                c(\n",
    "                    paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", hd_list), \n",
    "                    paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", hd_list)\n",
    "                )] + 1)\n",
    "        colnames(data_norm_log2) = c(paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", hd_list, \"_log2\"), paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", hd_list, \"_log2\"))\n",
    "        tmp = cbind(tmp, data_norm_log2) %>% data.frame\n",
    "        \n",
    "        ## log2 FC\n",
    "        if(info[1] == \"Input\"){\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD4_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD4_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD4_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD5_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD5_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD5_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD6_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD6_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD6_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD7_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD7_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD7_scaled\")])\n",
    "\n",
    "        }else{\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD1_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD1_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD1_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD2_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD2_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD2_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD3_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD3_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD3_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD4_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD4_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD4_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD5_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD5_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD5_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD6_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD6_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD6_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD7_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD7_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD7_scaled\")])\n",
    "\n",
    "        }\n",
    "        \n",
    "        write.table(tmp, file = paste0(fig_path, \"CUTRUN_DP_\", hist, \"_\", sample, \"_pairwise_withNearestGenes.csv\"), sep = \",\", quote = F, row.names = F, col.names = T)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "## Gene TSS neighborhood DP\n",
    "\n",
    "library(chromVAR)\n",
    "library(SummarizedExperiment)\n",
    "library(Matrix)\n",
    "outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "\n",
    "countMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_countMat.rds\"))\n",
    "normMat = readRDS(file = paste0(outPath, \"/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_normMat.rds\")) ##  countMat[[hist]][, k]/seqDepth[[hist]][k] * 16294327 ## divide by the seqDepth and multiply by the largest seqDepth\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "# CUTRUN pairwise DP\n",
    "# 1. pairwise\n",
    "# inPath = \"CART_CUTRUN_Project/results/CUTANDRUN/process/\"\n",
    "# outPath = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/\"\n",
    "# load(file = paste0(outPath, \"/RData/masterPeak_peakAnno_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "# load(file = paste0(outPath, \"/RData/IDR_master_peak_list_chromVar_count_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "# load(file = paste0(outPath, \"/RData/countMat_designInfo_histList_hd1-7_SEACRcontrolTop10_noChrXYM.RData\"))\n",
    "\n",
    "outPath = \"CART_CUTRUN_Project/results/paper_figure/CUTRUN_DP_TSS/\"\n",
    "\n",
    "cellList <- c(\"N\", \"CM\", \"EM\", \"EMRA\")\n",
    "hdList <- paste0(\"HD\", 1:7)\n",
    "exprList = c(\"Input\", \"Product\", \"Stim1\", \"Stim2\", \"Stim3\")\n",
    "cellCompList = c(\"N_CM\", \"N_EM\", \"N_EMRA\", \"CM_EM\", \"CM_EMRA\", \"EM_EMRA\")\n",
    "histList <- c(\"H3K27me3\", \"H3K4me2\")\n",
    "\n",
    "checkRep = function(hist, cell, expr, hd){\n",
    "  rep = \"rep1\"\n",
    "  if(paste(hist, cell, expr, hd, sep = \"_\") == \"H3K27me3_CM_Input_HD1\"){\n",
    "    rep = \"rep2\"\n",
    "  }\n",
    "  if(paste(hist, cell, expr, hd, sep = \"_\") == \"H3K27me3_N_Input_HD2\"){\n",
    "    rep = \"rep2\"\n",
    "  }\n",
    "\n",
    "  return(rep)\n",
    "\n",
    "}\n",
    "\n",
    "checkDup = function(expr, hd){\n",
    "\n",
    "  dupType = \"normLibSize\" #\"norm\"\n",
    "  if(expr == \"Input\" && hd %in% paste0(\"HD\", 1:3)){\n",
    "    dupType = \"stringent\"\n",
    "  }\n",
    "  return(dupType)\n",
    "\n",
    "}\n",
    "\n",
    "checkBam = function(expr, hd){\n",
    "\n",
    "  bamType = \"bowtie2_align.bam\"\n",
    "  if(expr == \"Input\" && hd %in% paste0(\"HD\", 1:3)){\n",
    "    bamType = \"bowtie2_align.sorted.rmDup.sortName.bam\"\n",
    "  }\n",
    "  return(bamType)\n",
    "\n",
    "}\n",
    "\n",
    "checkHD = function(expr){\n",
    "\n",
    "  if(expr == \"Input\"){\n",
    "    return(paste0(\"HD\", 4:7))\n",
    "  }else{\n",
    "    return(paste0(\"HD\", 1:7))\n",
    "  }\n",
    "}\n",
    "\n",
    "target = c()\n",
    "for(cell in cellList){\n",
    "  for(expr in exprList){\n",
    "    hdL = checkHD(expr)\n",
    "    for(hd in hdL){\n",
    "      if( !((cell == \"EMRA\") & (expr == \"Input\") && (hd == \"HD2\"))){\n",
    "        target = rbind(target, data.frame(cell = cell, expr = expr, humanDonor = hd))\n",
    "      }\n",
    "\n",
    "    }\n",
    "  }\n",
    "}\n",
    "target\n",
    "\n",
    "## Consider human donor as random variable.\n",
    "## Experimental design\n",
    "treat <- factor(paste(target$cell, target$expr, sep=\".\"))\n",
    "design <- model.matrix(~0 + treat)\n",
    "colnames(design) <- levels(treat)\n",
    "contrast <- makeContrasts(Input_N_CM = CM.Input - N.Input,\n",
    "                    Input_N_EM = EM.Input - N.Input,\n",
    "                    Input_N_EMRA = EMRA.Input - N.Input,\n",
    "                    Input_CM_EM = EM.Input - CM.Input,\n",
    "                    Input_CM_EMRA = EMRA.Input - CM.Input,\n",
    "                    Input_EM_EMRA = EMRA.Input - EM.Input,\n",
    "                    Product_N_CM = CM.Product - N.Product,\n",
    "                    Product_N_EM = EM.Product - N.Product,\n",
    "                    Product_N_EMRA = EMRA.Product - N.Product,\n",
    "                    Product_CM_EM = EM.Product - CM.Product,\n",
    "                    Product_CM_EMRA = EMRA.Product - CM.Product,\n",
    "                    Product_EM_EMRA = EMRA.Product - EM.Product,\n",
    "                    Stim1_N_CM = CM.Stim1 - N.Stim1,\n",
    "                    Stim1_N_EM = EM.Stim1 - N.Stim1,\n",
    "                    Stim1_N_EMRA = EMRA.Stim1 - N.Stim1,\n",
    "                    Stim1_CM_EM = EM.Stim1 - CM.Stim1,\n",
    "                    Stim1_CM_EMRA = EMRA.Stim1 - CM.Stim1,\n",
    "                    Stim1_EM_EMRA = EMRA.Stim1 - EM.Stim1,\n",
    "                    Stim2_N_CM = CM.Stim2 - N.Stim2,\n",
    "                    Stim2_N_EM = EM.Stim2 - N.Stim2,\n",
    "                    Stim2_N_EMRA = EMRA.Stim2 - N.Stim2,\n",
    "                    Stim2_CM_EM = EM.Stim2 - CM.Stim2,\n",
    "                    Stim2_CM_EMRA = EMRA.Stim2 - CM.Stim2,\n",
    "                    Stim2_EM_EMRA = EMRA.Stim2 - EM.Stim2,\n",
    "                    Stim3_N_CM = CM.Stim3 - N.Stim3,\n",
    "                    Stim3_N_EM = EM.Stim3 - N.Stim3,\n",
    "                    Stim3_N_EMRA = EMRA.Stim3 - N.Stim3,\n",
    "                    Stim3_CM_EM = EM.Stim3 - CM.Stim3,\n",
    "                    Stim3_CM_EMRA = EMRA.Stim3 - CM.Stim3,\n",
    "                    Stim3_EM_EMRA = EMRA.Stim3 - EM.Stim3,\n",
    "                    levels = design)\n",
    "\n",
    "results = vector(\"list\", length(histList))\n",
    "voomDDS = vector(\"list\", length(histList))\n",
    "for(hist in histList){\n",
    "  ## Filter and delete low expressed genes\n",
    "    selectR <- which(rowSums(countMat[[hist]]) > 5) ## remove low count genes\n",
    "    dataS <- countMat[[hist]][selectR,]\n",
    "    geneNameList = rownames(data)[selectR]\n",
    "\n",
    "    voomDDS[[hist]] <- voom(counts = dataS, design = design, normalize.method = \"cyclicloess\", plot = FALSE)\n",
    "    ## option 1 using voomDDS option2 using normDDS as normalized input.\n",
    "    inputDDS <- voomDDS[[hist]]\n",
    "    corfit <- duplicateCorrelation(inputDDS, design, block = target$humanDonor)\n",
    "    ## corfit$consensus\n",
    "    fit <- lmFit(inputDDS, design, block = target$humanDonor, correlation = corfit$consensus)\n",
    "    fitContrast <- contrasts.fit(fit, contrast)\n",
    "    fitBayes <- eBayes(fitContrast, robust = TRUE)\n",
    "\n",
    "\n",
    "    results[[hist]] = list()\n",
    "    for(i in 1:ncol(contrast)){\n",
    "        ## Results\n",
    "        res <- topTable(fit = fitBayes, adjust.method = 'fdr', coef = i, number = nrow(inputDDS), sort = 'P')\n",
    "\n",
    "        res <- data.table(GeneName = geneNameList[as.numeric(rownames(res))], GeneIndex = rownames(res) %>% as.numeric, res)\n",
    "        res[, Significance := ifelse((adj.P.Val <= 0.1 & sign(logFC) == 1 & abs(logFC) >= 1), 'Up',\n",
    "                              ifelse((adj.P.Val <= 0.1 & sign(logFC) == -1 & abs(logFC) >= 1), 'Down', 'notDE'))]\n",
    "        results[[hist]][[i]] = res\n",
    "\n",
    "        ## Output\n",
    "        write.table(res, file = paste0(outPath, '/CUTRUN_limma_tables/TSS_DE_', colnames(contrast)[i], '_HD1-7_adj0.1_logFC1.csv'), quote = FALSE, row.names = FALSE, sep = \",\")\n",
    "    }\n",
    "    names(results[[hist]]) <- colnames(contrast)\n",
    "\n",
    "}\n",
    "saveRDS(list(results, voomDDS), file = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_limmaDEresults.rds\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "fig_path <- \"CART_CUTRUN_Project/results/paper_figure/CUTRUN_DP_TSS/\"\n",
    "results = readRDS( file = \"CART_CUTRUN_Project/results/CUTANDRUN/analysis/RData/TSS_neighbors_chromVar_count_histList_hd1-7_noChrXYM_limmaDEresults.rds\")[[1]]\n",
    "\n",
    "\n",
    "for(hist in c(\"H3K27me3\", \"H3K4me2\")){\n",
    "    \n",
    "    for(sample in names(results[[hist]])){\n",
    "        info = sample %>% strsplit(\"_\") %>% unlist\n",
    "        if(info[1] == \"Input\"){\n",
    "            hd_list = paste0(\"HD\", 4:7)\n",
    "        }else{\n",
    "            hd_list = paste0(\"HD\", 1:7)\n",
    "        }\n",
    "        limma_res <- results[[hist]][[sample]]\n",
    "        data_norm <- countMat[[hist]]/colSums(countMat[[hist]]) * 1000000\n",
    "\n",
    "        data_raw = countMat[[hist]][\n",
    "                limma_res$ID, \n",
    "                c(\n",
    "                    paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", hd_list), \n",
    "                    paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", hd_list)\n",
    "                )]\n",
    "        colnames(data_raw) = c(\n",
    "                    paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", hd_list, \"_raw\"), \n",
    "                    paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", hd_list, \"_raw\")\n",
    "                )\n",
    "        tmp = cbind(limma_res, data_raw)\n",
    "\n",
    "        data_norm_scaled =  data_norm[\n",
    "                limma_res$ID, \n",
    "                c(\n",
    "                    paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", hd_list), \n",
    "                    paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", hd_list)\n",
    "                )]\n",
    "        colnames(data_norm_scaled) = c(\n",
    "            paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", hd_list, \"_scaled\"), \n",
    "            paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", hd_list, \"_scaled\")\n",
    "        )\n",
    "        tmp = cbind(tmp, data_norm_scaled)\n",
    "\n",
    "        data_norm_log2 = log2(data_norm[\n",
    "                limma_res$ID, \n",
    "                c(\n",
    "                    paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", hd_list), \n",
    "                    paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", hd_list)\n",
    "                )] + 1)\n",
    "        colnames(data_norm_log2) = c(\n",
    "            paste0(hist, \"_\", info[2], \"_\", info[1], \"_\", hd_list, \"_log2\"), \n",
    "            paste0(hist, \"_\", info[3], \"_\", info[1], \"_\", hd_list, \"_log2\")\n",
    "        )\n",
    "        tmp = cbind(tmp, data_norm_log2) %>% data.frame\n",
    "\n",
    "\n",
    "        if(info[1] == \"Input\"){\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD4_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD4_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD4_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD5_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD5_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD5_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD6_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD6_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD6_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD7_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD7_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD7_scaled\")])\n",
    "\n",
    "        }else{\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD1_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD1_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD1_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD2_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD2_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD2_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD3_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD3_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD3_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD4_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD4_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD4_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD5_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD5_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD5_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD6_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD6_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD6_scaled\")])\n",
    "            tmp[, paste0(hist, \"_log2(\", info[3], \"/\", info[2], \")_\", info[1], \"_HD7_scaled\")] = log2(tmp[, paste0(hist, \"_\", info[3], \"_\", info[1], \"_HD7_scaled\")] / tmp[, paste0(hist, \"_\", info[2], \"_\", info[1], \"_HD7_scaled\")])\n",
    "\n",
    "        }\n",
    "        \n",
    "        write.table(tmp, file = paste0(fig_path, \"CUTRUN_DP_\", hist, \"_\", sample, \"_pairwise_TSSneighborhood.csv\"), sep = \",\", quote = F, row.names = F, col.names = T)\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.2"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
